{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eba806f6",
   "metadata": {},
   "source": [
    "# orthoMCL_GSM_geneComparisons.\n",
    "\n",
    "This notebook seeks to reconcile genes and reactions for the *Lipomyces starkeyi* GSM. Gene orthologs were obtained via orthoMCL (provided by Dr. Yichao Han for four species). \n",
    "\n",
    "\n",
    "GSMs were originally built using only bbh between the scaffold species and *L. starkeyi*. As an example: bbh was performed between *Rhodosporidium toruloides* and *L. starkeyi*. Reactions with suspected *L. starkeyi* genes were ported from the rto scaffold to the lst GSM. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2cb25c9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6d5c2160",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "This model seems to have confidenceScores instead of rxnConfidenceScores field. Will use confidenceScores for what rxnConfidenceScores represents.\n",
      "This model seems to have metCharge instead of metCharges field. Will use metCharge for what metCharges represents.\n",
      "No defined compartments in model model. Compartments will be deduced heuristically using regular expressions.\n",
      "Using regular expression found the following compartments:c, e, g, m, n, r, v, x\n"
     ]
    }
   ],
   "source": [
    "# load GSMs.\n",
    "rto = cobra.io.load_json_model('../models/Rt_IFO0880.json')\n",
    "sce = cobra.io.load_matlab_model('../models/yeast8_modifiedwBIGGnames.mat')\n",
    "ylip = cobra.io.load_matlab_model('../models/twoModels/iYLI647_corr.mat')\n",
    "\n",
    "# lst_from_sce = cobra.io.load_matlab_model(\"../models/Lst_v_0.0_modelRemovedYeast8Genes.matlab\")\n",
    "# lst_from_ylip = cobra.io.load_json_model('../models/OLD_models_prior16Dec22/Lst_v0.1_model_yarrowiaBased.json')\n",
    "\n",
    "# lst_from_rto = cobra.io.load_json_model(\"../models/Lst_v_0.1_modelRemovedGenes.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "751b5959",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjust iYLI647 parameters for consistent annotation. \n",
    "for m in ylip.metabolites:\n",
    "    if ('[' in m.id):\n",
    "        ylip.metabolites.get_by_id(m.id).id = m.id.replace('[','_').replace(']','')\n",
    "    if ('_L_' in m.id):\n",
    "        ylip.metabolites.get_by_id(m.id).id = m.id.replace('_L_','__L_')\n",
    "    if ('_D_' in m.id):\n",
    "        ylip.metabolites.get_by_id(m.id).id = m.id.replace('_D_','__D_')\n",
    "for r in ylip.reactions:\n",
    "    if ('(e)' in r.id):\n",
    "        ylip.reactions.get_by_id(r.id).id = r.id.replace('(e)','_e')\n",
    "    if(('_L_') in r.id):\n",
    "        ylip.reactions.get_by_id(r.id).id = r.id.replace('_L_','__L_')\n",
    "\n",
    "    if(('_D_') in r.id):\n",
    "        ylip.reactions.get_by_id(r.id).id = r.id.replace('_D_','__D_')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd576b95",
   "metadata": {},
   "source": [
    "load the *L. starkeyi* framework model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0feb1a44",
   "metadata": {},
   "outputs": [],
   "source": [
    "## below is the original model used. \n",
    "# model = cobra.io.load_json_model(\"../models/Lst_v_0.1_modelRemovedGenes.json\")\n",
    "## this is a model to verify. \n",
    "model = cobra.io.load_json_model(\"../models/Lst_v_0.0_removedRTOGenes_forPUB.json\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad41a3aa",
   "metadata": {},
   "source": [
    "### load orthoMCL data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "577455ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Lst</th>\n",
       "      <th>Rto</th>\n",
       "      <th>Sce</th>\n",
       "      <th>Yli</th>\n",
       "      <th>s2</th>\n",
       "      <th>s3</th>\n",
       "      <th>s4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lst|G1IA6-22289-MONOMER</td>\n",
       "      <td>Lipst1_1_113024</td>\n",
       "      <td>10878</td>\n",
       "      <td>YML069W</td>\n",
       "      <td>YALI0D25058g</td>\n",
       "      <td>1.447</td>\n",
       "      <td>1.776</td>\n",
       "      <td>1.588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Lst|G1IA6-16751-MONOMER</td>\n",
       "      <td>Lipst1_1_441</td>\n",
       "      <td>14212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YALI0F19734g</td>\n",
       "      <td>1.055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lst|G1IA6-23047-MONOMER</td>\n",
       "      <td>Lipst1_1_175925</td>\n",
       "      <td>12668</td>\n",
       "      <td>YKL046C</td>\n",
       "      <td>YALI0D02563g</td>\n",
       "      <td>0.159</td>\n",
       "      <td>1.579</td>\n",
       "      <td>1.284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lst|G1IA6-17901-MONOMER</td>\n",
       "      <td>Lipst1_1_91016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YALI0E00396g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.895</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Unnamed: 0              Lst    Rto      Sce           Yli  \\\n",
       "0  Lst|G1IA6-22289-MONOMER  Lipst1_1_113024  10878  YML069W  YALI0D25058g   \n",
       "1  Lst|G1IA6-16751-MONOMER     Lipst1_1_441  14212      NaN  YALI0F19734g   \n",
       "2  Lst|G1IA6-23047-MONOMER  Lipst1_1_175925  12668  YKL046C  YALI0D02563g   \n",
       "3  Lst|G1IA6-17901-MONOMER   Lipst1_1_91016    NaN      NaN  YALI0E00396g   \n",
       "\n",
       "      s2     s3     s4  \n",
       "0  1.447  1.776  1.588  \n",
       "1  1.055    NaN  1.272  \n",
       "2  0.159  1.579  1.284  \n",
       "3    NaN    NaN  0.895  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orthoMCL = pd.read_csv('../orthoMCL/orthoMCL_orthologs_2023.csv')\n",
    "orthoMCL.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "01f04089",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain the genes present in each GSM. \n",
    "\n",
    "g = ylip.genes\n",
    "ylip_GSM_genes = [temp.id for temp in g]\n",
    "\n",
    "g = rto.genes\n",
    "rto_GSM_genes = [temp.id for temp in g]\n",
    "\n",
    "g = sce.genes\n",
    "sce_GSM_genes = [temp.id for temp in g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3b3eaabf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "810\n",
      "790\n",
      "512\n"
     ]
    }
   ],
   "source": [
    "# the number of genes in the GSM that have Lst orthologs. \n",
    "\n",
    "# rto to Lst\n",
    "print(len(set(orthoMCL.Rto.to_list()).intersection(set(rto_GSM_genes))))\n",
    "\n",
    "# sce to Lst\n",
    "print(len(set(orthoMCL.Sce.to_list()).intersection(set(sce_GSM_genes))))\n",
    "\n",
    "# yli to Lst\n",
    "print(len(set(orthoMCL.Yli.to_list()).intersection(set(ylip_GSM_genes))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f30f090c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# more genes are present in the direct bbh blasting scaffolds for GSMs. \n",
    "# print(len(lst_from_rto.genes))\n",
    "# print(len(lst_from_sce.genes))\n",
    "# print(len(lst_from_ylip.genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0aaeb5ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dictionary mapping the genes for each species. \n",
    "lst_rto_dict = dict(zip(orthoMCL.Rto,orthoMCL.Lst))\n",
    "lst_sce_dict = dict(zip(orthoMCL.Sce,orthoMCL.Lst))\n",
    "lst_ylip_dict = dict(zip(orthoMCL.Yli,orthoMCL.Lst))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a002e33b",
   "metadata": {},
   "source": [
    "### next to do.\n",
    "\n",
    "comparisons between GSMs using python. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7d59483e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# determine reactions annotations in all three GSMs. \n",
    "# and either ylip:rto or sce:rto GSMs.\n",
    "\n",
    "ylip_rto_overlap = []\n",
    "ylip_sce_rto_overlap = []\n",
    "rto_sce_overlap = []\n",
    "\n",
    "# start with Rhodo GSM reaction names. \n",
    "for r in rto.reactions:\n",
    "    \n",
    "    # search for ID in yarrowia. \n",
    "    if r.id in ylip.reactions:\n",
    "        \n",
    "        # now search for ID in saccharomyces.\n",
    "        if r.id in sce.reactions:\n",
    "            ylip_sce_rto_overlap.append(r.id)\n",
    "        \n",
    "        # reaction not in saccharomyces. \n",
    "        # ylip and rto overlapping reactions. \n",
    "        else:\n",
    "            ylip_rto_overlap.append(r.id)\n",
    "    \n",
    "    # reaction not in yarrowia.\n",
    "    # sce and rto overlapping reactions.\n",
    "    elif r.id in sce.reactions:\n",
    "        rto_sce_overlap.append(r.id)\n",
    "\n",
    "\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6f381e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain the ylip:sce reaction overlap. \n",
    "# obtain the sce and ylip only genes. \n",
    "\n",
    "ylip_sce_overlap = []\n",
    "ylip_rxn_only = []\n",
    "ylip_gene_only = []\n",
    "sce_rxn_only = []\n",
    "sce_gene_only = []\n",
    "\n",
    "# now loop through the yarrowia reactions. \n",
    "for r in ylip.reactions:\n",
    "    \n",
    "    # check if the id is not already in one of the previously determined lists. \n",
    "    if (r.id not in ylip_sce_rto_overlap) and (r.id not in rto_sce_overlap) and (r.id not in ylip_rto_overlap):\n",
    "        \n",
    "        # check if the reaction is in saccharomyces. \n",
    "        if r.id in sce.reactions:\n",
    "            ylip_sce_overlap.append(r.id)\n",
    "        \n",
    "        # reaction is only present in yarrowia. \n",
    "        else:\n",
    "            ylip_genes = r.gene_reaction_rule\n",
    "            \n",
    "            # obtain the gpr for the reaction if it is also present in the orthoMCL groups. \n",
    "            for g in [ylip_genes]:\n",
    "                if g in set(orthoMCL.Yli.to_list()).intersection(set(ylip_GSM_genes)):\n",
    "                    ylip_rxn_only.append(r.id)\n",
    "                    ylip_gene_only.append(g)\n",
    "\n",
    "# now loop through saccharomyces reactions. \n",
    "for r in sce.reactions:\n",
    "    \n",
    "    # check if the reaction has already been inlcuded in one of the previously determined lists. \n",
    "    if (r.id not in ylip_sce_rto_overlap) and (r.id not in rto_sce_overlap) and (r.id not in ylip_rto_overlap) and (r.id not in ylip_sce_overlap):\n",
    "        \n",
    "        # reaction is only in saccharomyces. \n",
    "        sce_genes = r.gene_reaction_rule\n",
    "        \n",
    "        # obtain the gpr for the reaction if it is also present in the orthoMCL groups. \n",
    "        for g in [sce_genes]:\n",
    "            if g in set(orthoMCL.Yli.to_list()).intersection(set(ylip_GSM_genes)):\n",
    "                sce_rxn_only.append(r.id)\n",
    "                sce_gene_only.append(g)\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc2f1ba4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7055f899",
   "metadata": {},
   "source": [
    "## Grab the L. starkeyi ortholog for reactions that have genes present in the orthoMCL groups.\n",
    "\n",
    "#### note, this section does not modify the GSM. It just obtains orthologs and new GPRs.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab05eb4",
   "metadata": {},
   "source": [
    "compare all three sets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f7fa0f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# definition to compare the genes in common reactions. \n",
    "def compare_three_lists(list1, list2, list3):\n",
    "\n",
    "    \n",
    "    # Convert lists to sets.\n",
    "    set1, set2, set3 = set(list1), set(list2), set(list3)\n",
    "    \n",
    "    # Find all intersections and differences.\n",
    "    common_all = set1 & set2 & set3\n",
    "    \n",
    "    only_12 = (set1 & set2) - set3\n",
    "    only_13 = (set1 & set3) - set2\n",
    "    only_23 = (set2 & set3) - set1\n",
    "    \n",
    "    only_1 = set1 - set2 - set3\n",
    "    only_2 = set2 - set1 - set3\n",
    "    only_3 = set3 - set1 - set2\n",
    "    \n",
    "    all_three_diff = (set1 | set2 | set3) - common_all - only_12 - only_13 - only_23 - only_1 - only_2 - only_3\n",
    "    \n",
    "    # Construct verbose explanation.\n",
    "    explanation = {}\n",
    "    \n",
    "#     explanation['common_all'] = f\"Items common in all three lists: {list(common_all)}\"\n",
    "#     explanation['only_12'] = f\"Items found only in list 1 and list 2: {list(only_12)}\"\n",
    "#     explanation['only_13'] = f\"Items found only in list 1 and list 3: {list(only_13)}\"\n",
    "#     explanation['only_23'] = f\"Items found only in list 2 and list 3: {list(only_23)}\"\n",
    "#     explanation['only_1'] = f\"Items unique to list 1: {list(only_1)}\"\n",
    "#     explanation['only_2'] = f\"Items unique to list 2: {list(only_2)}\"\n",
    "#     explanation['only_3'] = f\"Items unique to list 3: {list(only_3)}\"\n",
    "#     explanation['all_three_diff'] = f\"Items present in all three lists but not shared by all of them simultaneously: {list(all_three_diff)}\"\n",
    "    \n",
    "    explanation['common_all'] = list(common_all)\n",
    "    explanation['only_12'] = list(only_12)\n",
    "    explanation['only_13'] = list(only_13)\n",
    "    explanation['only_23'] = list(only_23)\n",
    "    explanation['only_1'] = list(only_1)\n",
    "    explanation['only_2'] = list(only_2)\n",
    "    explanation['only_3'] = list(only_3)\n",
    "    explanation['all_three_diff'] = list(all_three_diff)\n",
    "    \n",
    "    \n",
    "    return explanation\n",
    "\n",
    "\n",
    "# gene overlap in all three.\n",
    "\n",
    "_names = []\n",
    "_gpr_sce = []\n",
    "_gpr_ylip = []\n",
    "_gpr_rto = []\n",
    "\n",
    "_lst_rto_gene = []\n",
    "_lst_sce_gene = []\n",
    "_lst_ylip_gene = []\n",
    "\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "_lst_gene = []\n",
    "_ylip_gene = []\n",
    "\n",
    "\n",
    "consensus = []\n",
    "consensus_count = []\n",
    "consensus_source = []\n",
    "_output = []\n",
    "\n",
    "\n",
    "_common_all = []\n",
    "_only_12 = []\n",
    "_only_13 = []\n",
    "_only_23 = []\n",
    "_only_1 = []\n",
    "_only_2 = []\n",
    "_only_3 = []\n",
    "_all_three_diff = []\n",
    "\n",
    "# loop through the reactions present in all three GSMs and determine the L. starkeyi ortholog. \n",
    "for rxn in ylip_sce_rto_overlap:\n",
    "# test cases: \n",
    "# for rxn in ['NNDPRm','CSp','MTHFCm','PYRt2','PDE4','CSm']:\n",
    "    r = rto.reactions.get_by_id(rxn)\n",
    "    y = ylip.reactions.get_by_id(rxn)\n",
    "    s = sce.reactions.get_by_id(rxn)\n",
    "    \n",
    "    _names.append(r.name)\n",
    "    _gpr_rto.append(r.gene_reaction_rule)\n",
    "    _gpr_sce.append(s.gene_reaction_rule)\n",
    "    _gpr_ylip.append(y.gene_reaction_rule)\n",
    "\n",
    "    _id.append(r.id)\n",
    "    _stoich.append(r.reaction)\n",
    "    _subsystem.append(r.subsystem)\n",
    "    \n",
    "    \n",
    "    # rto genes.\n",
    "    temp1 = []\n",
    "    for g in r.genes:\n",
    "        try:\n",
    "            temp1.append(lst_rto_dict[g.id])\n",
    "        except:\n",
    "            temp1.append('NA')\n",
    "            donothing=0\n",
    "    _lst_rto_gene.append(temp1)\n",
    "    \n",
    "    # sce genes.\n",
    "    temp2 = []\n",
    "    for g in s.genes:\n",
    "        try:\n",
    "            temp2.append(lst_sce_dict[g.id])\n",
    "        except:\n",
    "            temp2.append('NA')\n",
    "            donothing=0\n",
    "    _lst_sce_gene.append(temp2)\n",
    "    \n",
    "    # ylip genes. \n",
    "    temp3 = []\n",
    "    for g in y.genes:\n",
    "        try:\n",
    "            temp3.append(lst_ylip_dict[g.id])\n",
    "        except:\n",
    "            temp3.append('NA')\n",
    "            donothing=0\n",
    "    _lst_ylip_gene.append(temp3)\n",
    "\n",
    "    \n",
    "\n",
    "    ### comparisions below - general gene info is above.\n",
    "\n",
    "    # remove NA for comparisions. \n",
    "    if 'NA' in temp1:\n",
    "        temp1.remove('NA')\n",
    "        temp1 = [e for e in temp1 if e not in ('NA')]\n",
    "    if 'NA' in temp2:\n",
    "        temp2.remove('NA')\n",
    "        temp2 = [e for e in temp2 if e not in ('NA')]\n",
    "\n",
    "    if 'NA' in temp3:\n",
    "        temp3.remove('NA')\n",
    "        temp3 = [e for e in temp3 if e not in ('NA')]\n",
    "    \n",
    "    result = compare_three_lists(temp1, temp2, temp3)\n",
    "    _output.append(result)\n",
    "    \n",
    "    \n",
    "    _common_all.append(result['common_all'])\n",
    "    _only_12.append(result['only_12'])\n",
    "    _only_13.append(result['only_13'])\n",
    "    _only_23.append(result['only_23'])\n",
    "    _only_1.append(result['only_1'])\n",
    "    _only_2.append(result['only_2'])\n",
    "    _only_3.append(result['only_3'])\n",
    "    _all_three_diff.append(result['all_three_diff'])\n",
    "#     for key, value in result.items():\n",
    "        \n",
    "#         print(f\"{key}: {value}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2506eca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3 = pd.DataFrame([_names,_id,_stoich,_subsystem,_gpr_rto,_gpr_sce,_gpr_ylip,_lst_rto_gene,_lst_sce_gene,_lst_ylip_gene,_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3,_all_three_diff]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e5115131",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9b1e3887",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3 = pd.DataFrame([_names,_id,_stoich,_subsystem,_gpr_rto,_gpr_sce,_gpr_ylip,_lst_rto_gene,_lst_sce_gene,_lst_ylip_gene,_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3,_all_three_diff]).T\n",
    "# df3.to_csv('allcomparisions_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d1cd7b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6f87416e",
   "metadata": {},
   "source": [
    "compare the ylip:sce overlap. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e1e27681",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rxn overlap for ylip and sce.    \n",
    "# _names = []\n",
    "# _gpr_sce = []\n",
    "# _gpr_ylip = []\n",
    "# _gpr_rto = []\n",
    "\n",
    "# _lst_rto_gene = []\n",
    "# _lst_sce_gene = []\n",
    "# _lst_ylip_gene = []\n",
    "\n",
    "# _id = []\n",
    "# _stoich = []\n",
    "# _subsystem = []\n",
    "# _lst_gene = []\n",
    "# _ylip_gene = []\n",
    "\n",
    "# _common_all = []\n",
    "# _only_12 = []\n",
    "# _only_13 = []\n",
    "# _only_23 = []\n",
    "# _only_1 = []\n",
    "# _only_2 = []\n",
    "# _only_3 = []\n",
    "# _all_three_diff = []\n",
    "\n",
    "for rxn in ylip_sce_overlap:\n",
    "#     r = rto.reactions.get_by_id(rxn)\n",
    "    y = ylip.reactions.get_by_id(rxn)\n",
    "    s = sce.reactions.get_by_id(rxn)\n",
    "    \n",
    "    _names.append(s.name)\n",
    "#     _gpr_rto.append(r.gene_reaction_rule)\n",
    "    _gpr_rto.append('None')\n",
    "    _gpr_sce.append(s.gene_reaction_rule)\n",
    "    _gpr_ylip.append(y.gene_reaction_rule)\n",
    "\n",
    "    _id.append(s.id)\n",
    "    _stoich.append(s.reaction)\n",
    "    _subsystem.append(s.subsystem)\n",
    "    \n",
    "    temp1 = []\n",
    "#     for g in [r.gene_reaction_rule]:\n",
    "#         try:\n",
    "#             temp1.append(lst_rto_dict[g])\n",
    "#         except:\n",
    "    temp1.append('NA')\n",
    "    _lst_rto_gene.append(temp1)\n",
    "    \n",
    "    temp2 = []\n",
    "    for g in s.genes:\n",
    "        try:\n",
    "            temp2.append(lst_sce_dict[g.id])\n",
    "        except:\n",
    "            temp2.append('NA')\n",
    "            donothing=0\n",
    "    _lst_sce_gene.append(temp2)\n",
    "    \n",
    "    temp3 = []\n",
    "    for g in y.genes:\n",
    "        try:\n",
    "            temp3.append(lst_ylip_dict[g.id])\n",
    "        except:\n",
    "            temp3.append('NA')\n",
    "            donothing=0\n",
    "\n",
    "    _lst_ylip_gene.append(temp3)\n",
    "    \n",
    "    # remove NA for comparisions. \n",
    "    if 'NA' in temp1:\n",
    "        temp1.remove('NA')\n",
    "        temp1 = [e for e in temp1 if e not in ('NA')]\n",
    "\n",
    "    if 'NA' in temp2:\n",
    "        temp2.remove('NA')\n",
    "        temp2 = [e for e in temp2 if e not in ('NA')]\n",
    "\n",
    "    if 'NA' in temp3:\n",
    "        temp3.remove('NA')\n",
    "        temp3 = [e for e in temp3 if e not in ('NA')]\n",
    "    \n",
    "    result = compare_three_lists(temp1, temp2, temp3)\n",
    "    \n",
    "    \n",
    "    _common_all.append(result['common_all'])\n",
    "    _only_12.append(result['only_12'])\n",
    "    _only_13.append(result['only_13'])\n",
    "    _only_23.append(result['only_23'])\n",
    "    _only_1.append(result['only_1'])\n",
    "    _only_2.append(result['only_2'])\n",
    "    _only_3.append(result['only_3'])\n",
    "    _all_three_diff.append(result['all_three_diff'])\n",
    "#     for key, value in result.items():\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ff9c13be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3 = pd.DataFrame([_names,_id,_stoich,_subsystem,_gpr_sce,_gpr_ylip,_lst_sce_gene,_lst_ylip_gene,_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3,_all_three_diff]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5128bb39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7a89c797",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3.to_csv('ylip_sce_rxn_overlap_genes.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d52c785",
   "metadata": {},
   "source": [
    "compare the ylip:rto overlap. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b9529dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# overlap ylip and rto.  \n",
    "# _names = []\n",
    "# _gpr_sce = []\n",
    "# _gpr_ylip = []\n",
    "# _gpr_rto = []\n",
    "\n",
    "# _lst_rto_gene = []\n",
    "# _lst_sce_gene = []\n",
    "# _lst_ylip_gene = []\n",
    "\n",
    "# _id = []\n",
    "# _stoich = []\n",
    "# _subsystem = []\n",
    "# _lst_gene = []\n",
    "# _ylip_gene = []\n",
    "\n",
    "# _common_all = []\n",
    "# _only_12 = []\n",
    "# _only_13 = []\n",
    "# _only_23 = []\n",
    "# _only_1 = []\n",
    "# _only_2 = []\n",
    "# _only_3 = []\n",
    "# _all_three_diff = []\n",
    "\n",
    "for rxn in ylip_rto_overlap:\n",
    "    r = rto.reactions.get_by_id(rxn)\n",
    "    y = ylip.reactions.get_by_id(rxn)\n",
    "#     s = sce.reactions.get_by_id(rxn)\n",
    "    \n",
    "    _names.append(r.name)\n",
    "    _gpr_rto.append(r.gene_reaction_rule)\n",
    "#     _gpr_sce.append(s.gene_reaction_rule)\n",
    "    _gpr_sce.append('None')\n",
    "    \n",
    "    \n",
    "    _gpr_ylip.append(y.gene_reaction_rule)\n",
    "\n",
    "    _id.append(r.id)\n",
    "    _stoich.append(r.reaction)\n",
    "    _subsystem.append(r.subsystem)\n",
    "    \n",
    "    temp1 = []\n",
    "    for g in r.genes:\n",
    "        try:\n",
    "            temp1.append(lst_rto_dict[g.id])\n",
    "        except:\n",
    "            temp1.append('NA')\n",
    "            donothing=0\n",
    "\n",
    "    _lst_rto_gene.append(temp1)\n",
    "    \n",
    "    temp2 = []\n",
    "    temp2.append('NA')\n",
    "#     for g in [s.gene_reaction_rule]:\n",
    "#         try:\n",
    "#             temp2.append(lst_sce_dict[g])\n",
    "#         except:\n",
    "#             temp2.append('NA')\n",
    "    _lst_sce_gene.append(temp2)\n",
    "    \n",
    "    temp3 = []\n",
    "    for g in y.genes:\n",
    "        try:\n",
    "            temp3.append(lst_ylip_dict[g.id])\n",
    "        except:\n",
    "            temp3.append('NA')\n",
    "            donothing=0\n",
    "\n",
    "    _lst_ylip_gene.append(temp3)\n",
    "      \n",
    "        \n",
    "    \n",
    "    # remove NA for comparisions. \n",
    "    # remove NA for comparisions. \n",
    "    if 'NA' in temp1:\n",
    "        temp1.remove('NA')\n",
    "        temp1 = [e for e in temp1 if e not in ('NA')]\n",
    "    \n",
    "    if 'NA' in temp2:\n",
    "        temp2.remove('NA')\n",
    "        temp2 = [e for e in temp2 if e not in ('NA')]\n",
    "\n",
    "    if 'NA' in temp3:\n",
    "        temp3.remove('NA')\n",
    "        temp3 = [e for e in temp3 if e not in ('NA')]\n",
    "    \n",
    "    result = compare_three_lists(temp1, temp2, temp3)\n",
    "    \n",
    "    _common_all.append(result['common_all'])\n",
    "    _only_12.append(result['only_12'])\n",
    "    _only_13.append(result['only_13'])\n",
    "    _only_23.append(result['only_23'])\n",
    "    _only_1.append(result['only_1'])\n",
    "    _only_2.append(result['only_2'])\n",
    "    _only_3.append(result['only_3'])\n",
    "    _all_three_diff.append(result['all_three_diff'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8909a1c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3 = pd.DataFrame([_names,_id,_stoich,_subsystem,_gpr_rto,_gpr_ylip,_lst_rto_gene,_lst_ylip_gene,_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3,_all_three_diff]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a0c76ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3.to_csv('ylip_rto_overlap_genes.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7037914",
   "metadata": {},
   "source": [
    "compare the sce:rto overlap. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b756fb36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NA', 'NA', 'NA', 'Lipst1_1_6303']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_74900']\n",
      "['Lipst1_1_314464']\n",
      "['Lipst1_1_5453']\n",
      "['Lipst1_1_265510']\n",
      "['Lipst1_1_265510', 'Lipst1_1_127634']\n",
      "['Lipst1_1_289621']\n",
      "['Lipst1_1_70370']\n",
      "['Lipst1_1_76289']\n",
      "['Lipst1_1_105640']\n",
      "['Lipst1_1_4472']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_6303']\n",
      "['Lipst1_1_102591']\n",
      "['Lipst1_1_314768']\n",
      "['Lipst1_1_423']\n",
      "['Lipst1_1_1263']\n",
      "['Lipst1_1_2889']\n",
      "['Lipst1_1_107753']\n",
      "['Lipst1_1_73838']\n",
      "['Lipst1_1_3878']\n",
      "['Lipst1_1_53536']\n",
      "['Lipst1_1_72751']\n",
      "['NA', 'Lipst1_1_2256']\n",
      "['NA', 'Lipst1_1_2256', 'Lipst1_1_61026']\n",
      "['Lipst1_1_76289']\n",
      "['Lipst1_1_338549']\n",
      "['Lipst1_1_48101']\n",
      "['Lipst1_1_27755']\n",
      "['NA', 'Lipst1_1_282']\n",
      "['Lipst1_1_81155']\n",
      "['Lipst1_1_46126']\n",
      "['Lipst1_1_142956']\n",
      "['Lipst1_1_3974']\n",
      "['Lipst1_1_3974']\n",
      "['Lipst1_1_246321']\n",
      "['Lipst1_1_2195']\n",
      "['Lipst1_1_314464']\n",
      "['Lipst1_1_46126']\n",
      "['Lipst1_1_3313']\n",
      "['NA', 'Lipst1_1_102207']\n",
      "['Lipst1_1_71146']\n",
      "['Lipst1_1_1110']\n",
      "['Lipst1_1_931']\n",
      "['Lipst1_1_76359']\n",
      "['NA', 'NA', 'Lipst1_1_69867']\n",
      "['NA', 'NA', 'Lipst1_1_69867', 'NA', 'Lipst1_1_70253']\n",
      "['NA', 'Lipst1_1_65274']\n",
      "['NA', 'Lipst1_1_65274', 'Lipst1_1_261247']\n",
      "['Lipst1_1_99746']\n",
      "['Lipst1_1_99746', 'NA', 'Lipst1_1_2772']\n",
      "['NA', 'Lipst1_1_2256']\n",
      "['NA', 'Lipst1_1_2256', 'Lipst1_1_61026']\n",
      "['Lipst1_1_70370']\n",
      "['Lipst1_1_99746']\n",
      "['Lipst1_1_99746', 'NA', 'Lipst1_1_2772']\n",
      "['Lipst1_1_61657']\n",
      "['Lipst1_1_75890']\n",
      "['Lipst1_1_76289']\n",
      "['Lipst1_1_314464']\n",
      "['Lipst1_1_50618']\n",
      "['Lipst1_1_50618', 'NA', 'NA', 'Lipst1_1_7158']\n",
      "['NA', 'Lipst1_1_3540']\n",
      "['NA', 'NA', 'Lipst1_1_74823']\n",
      "['Lipst1_1_3974']\n",
      "['Lipst1_1_75163']\n",
      "['Lipst1_1_69028']\n",
      "['Lipst1_1_2436']\n",
      "['Lipst1_1_338549']\n",
      "['Lipst1_1_3860']\n",
      "['Lipst1_1_1263']\n",
      "['Lipst1_1_3532']\n",
      "['Lipst1_1_1559']\n",
      "['Lipst1_1_75335']\n",
      "['Lipst1_1_69288']\n",
      "['Lipst1_1_931']\n",
      "['Lipst1_1_423']\n",
      "['Lipst1_1_4566']\n",
      "['Lipst1_1_121954']\n",
      "['Lipst1_1_121954', 'NA', 'Lipst1_1_177423']\n",
      "['Lipst1_1_46126']\n",
      "['Lipst1_1_5184']\n",
      "['Lipst1_1_42480']\n",
      "['Lipst1_1_56835']\n",
      "['Lipst1_1_1857']\n",
      "['NA', 'Lipst1_1_201837']\n",
      "['Lipst1_1_99746']\n",
      "['Lipst1_1_99746', 'NA', 'Lipst1_1_2772']\n",
      "['Lipst1_1_6465']\n",
      "['Lipst1_1_69288']\n",
      "['NA', 'Lipst1_1_101999']\n",
      "['Lipst1_1_1356']\n",
      "['Lipst1_1_1356', 'Lipst1_1_5189']\n",
      "['Lipst1_1_76289']\n",
      "['Lipst1_1_972']\n",
      "['Lipst1_1_2766']\n",
      "['Lipst1_1_594']\n",
      "['Lipst1_1_54959']\n",
      "['Lipst1_1_54959', 'Lipst1_1_2352']\n",
      "['Lipst1_1_54959', 'Lipst1_1_2352', 'NA', 'Lipst1_1_74347']\n",
      "['Lipst1_1_46126']\n",
      "['Lipst1_1_67618']\n",
      "['Lipst1_1_42480']\n",
      "['Lipst1_1_338549']\n",
      "['Lipst1_1_70181']\n",
      "['NA', 'Lipst1_1_76061']\n",
      "['Lipst1_1_3974']\n",
      "['Lipst1_1_46126']\n",
      "['Lipst1_1_99746']\n",
      "['Lipst1_1_99746', 'NA', 'Lipst1_1_2772']\n",
      "['Lipst1_1_61657']\n",
      "['Lipst1_1_3923']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_44049']\n",
      "['Lipst1_1_73367']\n",
      "['Lipst1_1_5784']\n",
      "['Lipst1_1_70370']\n",
      "['Lipst1_1_103302']\n",
      "['Lipst1_1_310928']\n",
      "['Lipst1_1_870']\n",
      "['Lipst1_1_314768']\n",
      "['Lipst1_1_314768']\n",
      "['Lipst1_1_265510']\n",
      "['Lipst1_1_265510', 'Lipst1_1_127634']\n",
      "['Lipst1_1_92025']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_121954']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_121954', 'Lipst1_1_6303']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_121954', 'Lipst1_1_6303', 'NA', 'Lipst1_1_177423']\n",
      "['Lipst1_1_67618']\n",
      "['Lipst1_1_70370']\n",
      "['NA', 'Lipst1_1_276115']\n",
      "['Lipst1_1_5895']\n",
      "['Lipst1_1_2102']\n",
      "['Lipst1_1_284310']\n",
      "['Lipst1_1_4155']\n",
      "['NA', 'Lipst1_1_2256']\n",
      "['NA', 'Lipst1_1_2256', 'Lipst1_1_61026']\n",
      "['NA', 'Lipst1_1_67991']\n",
      "['Lipst1_1_68284']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_6303']\n",
      "['Lipst1_1_3985']\n",
      "['Lipst1_1_265510']\n",
      "['Lipst1_1_265510', 'Lipst1_1_127634']\n",
      "['Lipst1_1_5784']\n",
      "['Lipst1_1_3974']\n",
      "['Lipst1_1_33829']\n",
      "['NA', 'Lipst1_1_3540']\n",
      "['Lipst1_1_76289']\n",
      "['Lipst1_1_57798']\n",
      "['NA', 'NA', 'NA', 'Lipst1_1_74862']\n",
      "['Lipst1_1_72834']\n",
      "['Lipst1_1_72834', 'Lipst1_1_71351']\n",
      "['Lipst1_1_5576']\n",
      "['Lipst1_1_72360']\n",
      "['Lipst1_1_70370']\n",
      "['Lipst1_1_265510']\n",
      "['Lipst1_1_265510', 'Lipst1_1_127634']\n",
      "['Lipst1_1_70181']\n",
      "['Lipst1_1_275']\n",
      "['Lipst1_1_99746']\n",
      "['Lipst1_1_99746', 'NA', 'Lipst1_1_2772']\n",
      "['NA', 'Lipst1_1_3431']\n",
      "['Lipst1_1_70253']\n",
      "['Lipst1_1_70253', 'Lipst1_1_69038']\n",
      "['Lipst1_1_73838']\n",
      "['Lipst1_1_75138']\n",
      "['Lipst1_1_1110']\n",
      "['Lipst1_1_1110', 'Lipst1_1_72701']\n",
      "['Lipst1_1_50618']\n",
      "['Lipst1_1_50618', 'NA', 'NA', 'Lipst1_1_7158']\n",
      "['Lipst1_1_176686']\n",
      "['Lipst1_1_73351']\n",
      "['Lipst1_1_97460']\n",
      "['Lipst1_1_4054']\n",
      "['Lipst1_1_49747']\n",
      "['Lipst1_1_5472']\n",
      "['Lipst1_1_925']\n",
      "['Lipst1_1_76716']\n",
      "['Lipst1_1_6779']\n",
      "['Lipst1_1_3290']\n",
      "['Lipst1_1_53798']\n",
      "['Lipst1_1_4390']\n",
      "['Lipst1_1_50110']\n",
      "['Lipst1_1_73932']\n",
      "['Lipst1_1_76716']\n",
      "['Lipst1_1_3617']\n",
      "['Lipst1_1_314464']\n",
      "['Lipst1_1_67687']\n",
      "['Lipst1_1_67687']\n",
      "['Lipst1_1_72108']\n",
      "['Lipst1_1_47259']\n",
      "['Lipst1_1_75992']\n",
      "['Lipst1_1_6459']\n",
      "['Lipst1_1_1356']\n",
      "['Lipst1_1_71319']\n"
     ]
    }
   ],
   "source": [
    "# overlap sce and rto.    \n",
    "# _names = []\n",
    "# _gpr_sce = []\n",
    "# _gpr_ylip = []\n",
    "# _gpr_rto = []\n",
    "\n",
    "# _lst_rto_gene = []\n",
    "# _lst_sce_gene = []\n",
    "# _lst_ylip_gene = []\n",
    "\n",
    "# _id = []\n",
    "# _stoich = []\n",
    "# _subsystem = []\n",
    "# _lst_gene = []\n",
    "# _ylip_gene = []\n",
    "\n",
    "# _common_all = []\n",
    "# _only_12 = []\n",
    "# _only_13 = []\n",
    "# _only_23 = []\n",
    "# _only_1 = []\n",
    "# _only_2 = []\n",
    "# _only_3 = []\n",
    "# _all_three_diff = []\n",
    "\n",
    "for rxn in rto_sce_overlap:\n",
    "    r = rto.reactions.get_by_id(rxn)\n",
    "#     y = ylip.reactions.get_by_id(rxn)\n",
    "    s = sce.reactions.get_by_id(rxn)\n",
    "    \n",
    "    _names.append(r.name)\n",
    "    _gpr_rto.append(r.gene_reaction_rule)\n",
    "    _gpr_sce.append(s.gene_reaction_rule)\n",
    "#     _gpr_ylip.append(y.gene_reaction_rule)\n",
    "    _gpr_ylip.append('None')\n",
    "\n",
    "\n",
    "    _id.append(r.id)\n",
    "    _stoich.append(r.reaction)\n",
    "    _subsystem.append(r.subsystem)\n",
    "    \n",
    "    temp1 = []\n",
    "    for g in r.genes:\n",
    "        try:\n",
    "            temp1.append(lst_rto_dict[g.id])\n",
    "            print(temp1)\n",
    "        except:\n",
    "            temp1.append('NA')\n",
    "            donothing=0\n",
    "\n",
    "    _lst_rto_gene.append(temp1)\n",
    "    \n",
    "    temp2 = []\n",
    "    for g in s.genes:\n",
    "        try:\n",
    "            temp2.append(lst_sce_dict[g.id])\n",
    "        except:\n",
    "#             temp2.append('NA')\n",
    "            donothing=0\n",
    "            \n",
    "    _lst_sce_gene.append(temp2)\n",
    "    \n",
    "    temp3 = []\n",
    "    temp3.append('NA')\n",
    "    \n",
    "#     for g in [y.gene_reaction_rule]:\n",
    "#         try:\n",
    "#             temp.append(lst_ylip_dict[g])\n",
    "#         except:\n",
    "#             temp.append('NA')\n",
    "    _lst_ylip_gene.append(temp3)\n",
    "    \n",
    "    # remove NA for comparisions. \n",
    "    if 'NA' in temp1:\n",
    "        temp1.remove('NA')\n",
    "        temp1 = [e for e in temp1 if e not in ('NA')]\n",
    "    if 'NA' in temp2:\n",
    "        temp2.remove('NA')\n",
    "        temp2 = [e for e in temp2 if e not in ('NA')]\n",
    "\n",
    "    if 'NA' in temp3:\n",
    "        temp3.remove('NA')\n",
    "        temp3 = [e for e in temp3 if e not in ('NA')]\n",
    "        \n",
    "        \n",
    "    result = compare_three_lists(temp1, temp2, temp3)\n",
    "    \n",
    "    \n",
    "    _common_all.append(result['common_all'])\n",
    "    _only_12.append(result['only_12'])\n",
    "    _only_13.append(result['only_13'])\n",
    "    _only_23.append(result['only_23'])\n",
    "    _only_1.append(result['only_1'])\n",
    "    _only_2.append(result['only_2'])\n",
    "    _only_3.append(result['only_3'])\n",
    "    _all_three_diff.append(result['all_three_diff'])           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5ea29c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3 = pd.DataFrame([_names,_id,_stoich,_subsystem,_gpr_rto,_gpr_sce,_lst_rto_gene,_lst_sce_gene,_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3,_all_three_diff]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3425b694",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df3.to_csv('rto_sce_overlap_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9fd91a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "51c65ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save output. \n",
    "# df3 = pd.DataFrame([_names,_id,_stoich,_subsystem,_gpr_rto,_gpr_sce,_gpr_ylip,_lst_rto_gene,_lst_sce_gene,_lst_ylip_gene,_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3,_all_three_diff]).T\n",
    "# df3.to_csv('allcomparisions_genes.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c95d2338",
   "metadata": {},
   "source": [
    "#### next \n",
    "\n",
    "Automate the GPR for each reaction and model. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0403ef81",
   "metadata": {},
   "outputs": [],
   "source": [
    "def concatenate_elements(*lists):\n",
    "    # Ensure that all lists have the same length by filling shorter lists with empty strings\n",
    "    labels_list = []\n",
    "\n",
    "        \n",
    "    max_len = max(len(lst) for lst in lists)\n",
    "    for lst in lists:\n",
    "        while len(lst) < max_len:\n",
    "            lst.append([''])\n",
    "    \n",
    "    # Concatenate lists\n",
    "    result = []\n",
    "    \n",
    "    \n",
    "\n",
    "    for i in range(max_len):\n",
    "        combined_string = \" or \".join(\n",
    "            \" or; \".join(sub_item.replace(\"None\", \"\").replace(\"NA\",\" \") for sub_item in lst[i]) for lst in lists \n",
    "            if len(lst) > i and lst[i]\n",
    "            )\n",
    "        # potential alternative.\n",
    "#         combined_string = ' or '.join(item.replace(\"None\", \"\").replace(\"NA\",\" \") for sublist in nested_lists for subsublist in sublist for item in subsublist)\n",
    "\n",
    "        result.append([combined_string])\n",
    "        \n",
    "    \n",
    "        # Generate label for this concatenated string\n",
    "        if combined_string:\n",
    "            current_labels = []\n",
    "            for j, lst in enumerate(lists):\n",
    "                if len(lst) > 0 and lst[i]:\n",
    "                    if j == 0:\n",
    "                        current_labels.append(\"rto,sce,ylip\")\n",
    "                    elif j == 1:\n",
    "                        current_labels.append(\"rto,sce\")\n",
    "                    elif j == 2:\n",
    "                        current_labels.append(\"rto,ylip\")\n",
    "                    elif j == 3:\n",
    "                        current_labels.append(\"sce,ylip\")\n",
    "                    elif j == 4:\n",
    "                        current_labels.append(\"rto\")\n",
    "                    elif j == 5:\n",
    "                        current_labels.append(\"sce\")\n",
    "                    elif j == 6:\n",
    "                        current_labels.append(\"ylip\")\n",
    "            labels_list.append(\" ; \".join(current_labels))\n",
    "        else:\n",
    "            labels_list.append('')    \n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "#             current_labels = [f\"list_{j+1}\" for j, lst in enumerate(lists) if len(lst) > i and lst[i]]\n",
    "#             labels_list.append(\" or \".join(current_labels))\n",
    "    \n",
    "    return result,labels_list\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "151755ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene annotation for each ortholog and the species source of the ortholog. \n",
    "result,result_labels = concatenate_elements(_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3effea7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a3a5d1c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0869f1f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lipst1_1_73838 (From: rto,ylip)\n",
      "Lipst1_1_103302 (From: rto,sce)\n",
      "Lipst1_1_258919 (From: rto,sce,ylip)\n",
      "Lipst1_1_63385 (From: rto,sce,ylip)\n",
      "Lipst1_1_60832 or Lipst1_1_43727 (From: rto,sce,ylip ; rto,ylip)\n",
      "Lipst1_1_68609 or Lipst1_1_70253 (From: rto,sce,ylip ; rto,sce)\n",
      "Lipst1_1_3835 or Lipst1_1_57709 (From: sce ; ylip)\n",
      "Lipst1_1_4166 or Lipst1_1_6739 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_47324 (From: rto,sce,ylip)\n",
      "Lipst1_1_2102 or Lipst1_1_70967 (From: rto ; sce)\n",
      "Lipst1_1_54023 (From: rto,sce,ylip)\n",
      "Lipst1_1_68065 or Lipst1_1_47067 (From: rto ; ylip)\n",
      "Lipst1_1_302882 (From: rto,sce,ylip)\n",
      "Lipst1_1_102877 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_102877 (From: rto,sce,ylip)\n",
      "Lipst1_1_92627 or Lipst1_1_47067 (From: rto,sce,ylip ; sce,ylip)\n",
      "Lipst1_1_4094 or Lipst1_1_2516 (From: rto,sce ; ylip)\n",
      "Lipst1_1_69224 or Lipst1_1_1857 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_72774 (From: rto,sce,ylip)\n",
      "Lipst1_1_73707 (From: rto,sce,ylip)\n",
      "Lipst1_1_275 (From: rto,sce,ylip)\n",
      "Lipst1_1_35483 (From: rto,sce,ylip)\n",
      "Lipst1_1_1834 (From: rto,sce,ylip)\n",
      "Lipst1_1_75883 (From: rto,sce)\n",
      "Lipst1_1_68121 (From: rto,sce,ylip)\n",
      "Lipst1_1_89607 (From: rto,sce,ylip)\n",
      "Lipst1_1_103580 or Lipst1_1_164851 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_70511 (From: rto,sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_69043 (From: ylip)\n",
      "Lipst1_1_196787 or Lipst1_1_201989 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_47324 (From: rto,sce,ylip)\n",
      "Lipst1_1_72400 (From: rto,sce,ylip)\n",
      "Lipst1_1_72898 or Lipst1_1_4939 or; Lipst1_1_71658 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_241647 (From: ylip)\n",
      "Lipst1_1_241647 (From: ylip)\n",
      "Lipst1_1_157499 (From: rto,sce,ylip)\n",
      "Lipst1_1_71969 or Lipst1_1_3835 or Lipst1_1_57709 (From: rto ; sce ; ylip)\n",
      "Lipst1_1_5409 or Lipst1_1_5857 (From: rto ; ylip)\n",
      "Lipst1_1_6430 (From: rto,sce,ylip)\n",
      "Lipst1_1_76556 or Lipst1_1_73409 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_3600 (From: rto)\n",
      "Lipst1_1_70762 or Lipst1_1_59512 or Lipst1_1_147168 (From: rto,ylip ; sce ; ylip)\n",
      "Lipst1_1_61809 (From: rto,sce,ylip)\n",
      "Lipst1_1_70474 (From: rto,sce,ylip)\n",
      "Lipst1_1_73409 or Lipst1_1_76556 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_3313 (From: rto,sce,ylip)\n",
      "Lipst1_1_72902 (From: rto,sce,ylip)\n",
      "Lipst1_1_72597 or Lipst1_1_7296 or Lipst1_1_54130 or; Lipst1_1_32541 (From: rto,sce ; sce ; ylip)\n",
      "Lipst1_1_283125 (From: rto,sce,ylip)\n",
      "Lipst1_1_75567 (From: rto,sce,ylip)\n",
      "Lipst1_1_1998 (From: rto,ylip)\n",
      "Lipst1_1_72597 (From: rto,sce)\n",
      "Lipst1_1_338549 (From: rto)\n",
      "Lipst1_1_75887 (From: rto,sce,ylip)\n",
      "Lipst1_1_5880 or Lipst1_1_4996 (From: rto,sce ; ylip)\n",
      "Lipst1_1_132 (From: rto,sce,ylip)\n",
      "Lipst1_1_1228 (From: rto,ylip)\n",
      "Lipst1_1_68121 (From: rto,sce,ylip)\n",
      "Lipst1_1_73845 or Lipst1_1_75188 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_36529 or Lipst1_1_39586 or Lipst1_1_4390 or; Lipst1_1_48116 (From: rto,ylip ; sce ; ylip)\n",
      "Lipst1_1_75883 or Lipst1_1_58573 (From: rto,sce ; ylip)\n",
      "Lipst1_1_71969 or Lipst1_1_3835 or Lipst1_1_57709 (From: rto ; sce ; ylip)\n",
      "Lipst1_1_157098 (From: rto,ylip)\n",
      " (From: )\n",
      "Lipst1_1_6444 (From: rto,sce,ylip)\n",
      "Lipst1_1_70877 (From: rto,sce,ylip)\n",
      "Lipst1_1_3290 (From: rto,sce,ylip)\n",
      "Lipst1_1_228 or Lipst1_1_69723 (From: rto,sce,ylip ; sce)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 (From: rto,sce,ylip)\n",
      "Lipst1_1_72093 (From: rto,sce,ylip)\n",
      "Lipst1_1_168 (From: rto,sce,ylip)\n",
      "Lipst1_1_99746 (From: rto)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_51135 (From: rto,sce,ylip)\n",
      "Lipst1_1_4946 (From: rto,sce)\n",
      "Lipst1_1_72563 (From: rto,sce,ylip)\n",
      "Lipst1_1_69933 or Lipst1_1_73260 (From: rto,sce,ylip ; rto,ylip)\n",
      "Lipst1_1_73737 (From: rto,sce,ylip)\n",
      "Lipst1_1_4939 or; Lipst1_1_71658 or Lipst1_1_72898 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_239921 (From: rto,sce,ylip)\n",
      "Lipst1_1_52642 (From: rto,sce,ylip)\n",
      "Lipst1_1_105290 (From: rto,sce,ylip)\n",
      "Lipst1_1_52812 or Lipst1_1_4095 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_105569 (From: rto,sce,ylip)\n",
      "Lipst1_1_3975 (From: rto,sce,ylip)\n",
      "Lipst1_1_67222 (From: rto,sce)\n",
      "Lipst1_1_1741 (From: rto,sce,ylip)\n",
      "Lipst1_1_338321 (From: rto,sce,ylip)\n",
      "Lipst1_1_67513 (From: rto,sce,ylip)\n",
      "Lipst1_1_70511 (From: rto,sce,ylip)\n",
      "Lipst1_1_60832 or Lipst1_1_43727 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_87988 (From: rto,sce,ylip)\n",
      "Lipst1_1_3313 (From: rto,sce,ylip)\n",
      "Lipst1_1_1263 (From: rto,sce,ylip)\n",
      "Lipst1_1_67949 (From: rto,sce,ylip)\n",
      "Lipst1_1_241647 (From: rto,sce,ylip)\n",
      "Lipst1_1_241647 (From: ylip)\n",
      "Lipst1_1_41207 (From: rto,sce,ylip)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_215 (From: rto,sce,ylip)\n",
      "Lipst1_1_404 or Lipst1_1_931 (From: rto,ylip ; sce)\n",
      "Lipst1_1_110167 (From: rto,sce,ylip)\n",
      "Lipst1_1_4054 (From: rto,sce,ylip)\n",
      "Lipst1_1_2767 (From: rto,sce,ylip)\n",
      "Lipst1_1_37388 (From: rto,sce,ylip)\n",
      "Lipst1_1_256229 (From: rto,sce)\n",
      "Lipst1_1_6367 (From: rto,sce,ylip)\n",
      "Lipst1_1_2213 (From: rto,sce,ylip)\n",
      "Lipst1_1_162168 (From: rto,sce,ylip)\n",
      "Lipst1_1_46503 or Lipst1_1_322121 or Lipst1_1_69043 (From: rto,sce,ylip ; rto ; ylip)\n",
      "Lipst1_1_51135 (From: rto,sce)\n",
      "Lipst1_1_5453 (From: sce,ylip)\n",
      "Lipst1_1_72320 or Lipst1_1_94934 (From: rto,sce,ylip ; rto,ylip)\n",
      "Lipst1_1_71964 or; Lipst1_1_69736 or; Lipst1_1_67537 (From: rto,sce,ylip)\n",
      "Lipst1_1_2516 or Lipst1_1_327549 (From: rto ; sce)\n",
      "Lipst1_1_193402 (From: rto,sce,ylip)\n",
      "Lipst1_1_1998 (From: rto,sce,ylip)\n",
      "Lipst1_1_265581 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_60832 or; Lipst1_1_43727 (From: rto,sce,ylip)\n",
      "Lipst1_1_3900 or Lipst1_1_73932 (From: rto,sce,ylip ; sce)\n",
      "Lipst1_1_74823 or Lipst1_1_5931 (From: rto ; sce)\n",
      "Lipst1_1_2955 or Lipst1_1_309296 (From: rto,sce ; rto)\n",
      "Lipst1_1_1586 or Lipst1_1_256229 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_2516 or Lipst1_1_327549 (From: rto,sce,ylip ; sce)\n",
      "Lipst1_1_63385 (From: rto,sce,ylip)\n",
      "Lipst1_1_74203 (From: rto,sce,ylip)\n",
      "Lipst1_1_73838 (From: rto,ylip)\n",
      "Lipst1_1_84 or Lipst1_1_90973 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_50182 (From: rto,sce,ylip)\n",
      "Lipst1_1_132 (From: rto,sce,ylip)\n",
      "Lipst1_1_6459 (From: rto,sce)\n",
      "Lipst1_1_119165 (From: rto,sce,ylip)\n",
      "Lipst1_1_5184 (From: rto,sce,ylip)\n",
      "Lipst1_1_338321 (From: rto,sce,ylip)\n",
      "Lipst1_1_25577 (From: rto,sce,ylip)\n",
      "Lipst1_1_6251 (From: rto,sce,ylip)\n",
      "Lipst1_1_85639 (From: rto,sce,ylip)\n",
      "Lipst1_1_3128 (From: rto,sce,ylip)\n",
      "Lipst1_1_73838 (From: rto,sce,ylip)\n",
      "Lipst1_1_182606 (From: rto,sce,ylip)\n",
      "Lipst1_1_313744 or Lipst1_1_4956 or Lipst1_1_542 (From: rto,sce,ylip ; rto ; ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 or; Lipst1_1_57088 (From: rto ; sce)\n",
      "Lipst1_1_168 (From: rto,sce,ylip)\n",
      "Lipst1_1_1228 (From: rto,sce,ylip)\n",
      "Lipst1_1_110210 (From: rto,sce,ylip)\n",
      "Lipst1_1_104780 (From: rto,sce,ylip)\n",
      "Lipst1_1_6459 (From: rto,sce)\n",
      "Lipst1_1_72802 or Lipst1_1_265200 or Lipst1_1_4166 (From: rto,sce,ylip ; rto,ylip ; rto)\n",
      "Lipst1_1_7017 or Lipst1_1_2089 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_69224 or Lipst1_1_1857 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_4095 or Lipst1_1_52812 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_241647 (From: ylip)\n",
      "Lipst1_1_4946 or Lipst1_1_1110 (From: rto,sce ; rto)\n",
      "Lipst1_1_5931 or Lipst1_1_74823 (From: sce,ylip ; rto)\n",
      "Lipst1_1_316 (From: rto,sce,ylip)\n",
      "Lipst1_1_70370 (From: rto,sce,ylip)\n",
      "Lipst1_1_258919 (From: rto,sce,ylip)\n",
      "Lipst1_1_72149 (From: rto,sce)\n",
      "Lipst1_1_3573 (From: rto,sce,ylip)\n",
      "Lipst1_1_288751 (From: rto,sce,ylip)\n",
      "Lipst1_1_75188 or Lipst1_1_73845 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_68590 (From: rto,ylip)\n",
      "Lipst1_1_2790 (From: rto,sce,ylip)\n",
      "Lipst1_1_925 (From: rto,sce,ylip)\n",
      "Lipst1_1_3458 (From: rto,sce,ylip)\n",
      "Lipst1_1_5229 (From: rto,sce,ylip)\n",
      "Lipst1_1_73407 or Lipst1_1_101674 (From: rto,sce,ylip ; sce,ylip)\n",
      "Lipst1_1_1896 (From: rto,sce,ylip)\n",
      "Lipst1_1_68472 (From: rto,sce,ylip)\n",
      "Lipst1_1_73737 (From: rto,sce,ylip)\n",
      "Lipst1_1_75768 (From: rto,sce,ylip)\n",
      "Lipst1_1_75883 (From: rto,sce)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_102853 (From: ylip)\n",
      "Lipst1_1_238934 (From: rto,sce,ylip)\n",
      "Lipst1_1_3931 (From: rto,sce,ylip)\n",
      "Lipst1_1_6067 (From: rto,sce,ylip)\n",
      "Lipst1_1_117393 (From: rto,sce,ylip)\n",
      "Lipst1_1_7080 (From: rto,sce)\n",
      "Lipst1_1_404 or Lipst1_1_70253 (From: rto,sce,ylip ; sce)\n",
      "Lipst1_1_33932 (From: rto,sce,ylip)\n",
      "Lipst1_1_71969 or Lipst1_1_3835 or Lipst1_1_57709 (From: rto ; sce ; ylip)\n",
      "Lipst1_1_102877 (From: rto,sce,ylip)\n",
      "Lipst1_1_47324 (From: rto,sce,ylip)\n",
      "Lipst1_1_74872 (From: rto,sce,ylip)\n",
      "Lipst1_1_67613 (From: rto,sce,ylip)\n",
      "Lipst1_1_5978 (From: rto,sce,ylip)\n",
      "Lipst1_1_71443 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_76037 (From: rto,sce,ylip)\n",
      "Lipst1_1_74867 or; Lipst1_1_1724 or Lipst1_1_79742 or; Lipst1_1_67480 or; Lipst1_1_55686 (From: rto,sce,ylip ; rto,sce)\n",
      "Lipst1_1_117342 or Lipst1_1_102591 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_94412 (From: rto,sce,ylip)\n",
      "Lipst1_1_102287 (From: rto,sce,ylip)\n",
      "Lipst1_1_6557 (From: rto,sce,ylip)\n",
      "Lipst1_1_104780 (From: rto,sce,ylip)\n",
      "Lipst1_1_2110 (From: rto,sce,ylip)\n",
      "Lipst1_1_46980 (From: rto,sce,ylip)\n",
      "Lipst1_1_69113 (From: rto,sce,ylip)\n",
      "Lipst1_1_73341 (From: rto,sce,ylip)\n",
      "Lipst1_1_54569 or Lipst1_1_3600 or Lipst1_1_111440 (From: rto,sce,ylip ; rto ; sce)\n",
      "Lipst1_1_105712 or Lipst1_1_58334 or Lipst1_1_48449 (From: sce,ylip ; rto ; sce)\n",
      "Lipst1_1_70723 (From: rto,sce,ylip)\n",
      "Lipst1_1_3128 (From: rto,sce,ylip)\n",
      "Lipst1_1_72319 (From: rto,sce,ylip)\n",
      "Lipst1_1_7085 (From: rto,sce)\n",
      "Lipst1_1_1263 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 or Lipst1_1_54569 (From: rto ; sce ; ylip)\n",
      "Lipst1_1_75883 or Lipst1_1_58573 (From: rto,sce ; ylip)\n",
      "Lipst1_1_68186 (From: rto,sce,ylip)\n",
      "Lipst1_1_75883 (From: rto,sce)\n",
      "Lipst1_1_7391 (From: rto,sce,ylip)\n",
      "Lipst1_1_1034 (From: rto,sce,ylip)\n",
      "Lipst1_1_316583 (From: rto,sce,ylip)\n",
      "Lipst1_1_72186 (From: rto,sce,ylip)\n",
      "Lipst1_1_102585 (From: rto,sce,ylip)\n",
      "Lipst1_1_68855 or Lipst1_1_72137 or Lipst1_1_61809 (From: rto,ylip ; rto ; sce)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_88614 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_4257 (From: rto,ylip)\n",
      "Lipst1_1_84 or Lipst1_1_90973 (From: rto,ylip ; sce,ylip)\n",
      "Lipst1_1_332345 (From: rto,sce,ylip)\n",
      "Lipst1_1_25577 (From: rto,sce,ylip)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_54569 or Lipst1_1_3600 or Lipst1_1_111440 (From: rto,sce,ylip ; rto ; sce)\n",
      "Lipst1_1_69113 (From: rto,sce,ylip)\n",
      "Lipst1_1_72935 or; Lipst1_1_4904 or; Lipst1_1_32736 or; Lipst1_1_90916 or; Lipst1_1_86782 or; Lipst1_1_75804 or; Lipst1_1_52074 or; Lipst1_1_25267 or; Lipst1_1_101415 or; Lipst1_1_448 or Lipst1_1_72399 or; Lipst1_1_46841 or; Lipst1_1_71305 or Lipst1_1_2175 (From: rto,sce ; rto ; sce)\n",
      "Lipst1_1_6444 (From: rto,sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_55185 (From: rto,sce)\n",
      "Lipst1_1_103097 (From: rto,sce)\n",
      "Lipst1_1_51135 (From: rto,sce,ylip)\n",
      "Lipst1_1_71969 or Lipst1_1_3835 or Lipst1_1_57709 (From: rto ; sce ; ylip)\n",
      "Lipst1_1_275 (From: rto,sce,ylip)\n",
      "Lipst1_1_2223 (From: rto,ylip)\n",
      "Lipst1_1_3431 (From: rto,sce,ylip)\n",
      "Lipst1_1_72319 (From: rto,sce,ylip)\n",
      "Lipst1_1_234091 (From: rto,sce,ylip)\n",
      "Lipst1_1_156565 (From: rto,sce)\n",
      "Lipst1_1_3938 (From: rto,sce,ylip)\n",
      "Lipst1_1_241647 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_47324 (From: rto,sce,ylip)\n",
      "Lipst1_1_61809 (From: rto,sce,ylip)\n",
      "Lipst1_1_72810 or Lipst1_1_31624 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_870 (From: rto,sce,ylip)\n",
      "Lipst1_1_67945 or Lipst1_1_74569 or; Lipst1_1_244790 (From: sce,ylip ; rto)\n",
      "Lipst1_1_70900 or Lipst1_1_4340 (From: sce,ylip ; rto)\n",
      "Lipst1_1_72797 or Lipst1_1_47259 (From: sce,ylip ; rto)\n",
      "Lipst1_1_72215 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_104169 or Lipst1_1_4018 or Lipst1_1_255632 (From: rto,sce,ylip ; sce,ylip ; sce)\n",
      "Lipst1_1_1069 (From: rto,sce)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_177423 or; Lipst1_1_121954 or Lipst1_1_5651 or Lipst1_1_6983 (From: rto,ylip ; sce ; ylip)\n",
      "Lipst1_1_205437 or Lipst1_1_201837 or; Lipst1_1_206087 (From: rto,sce ; ylip)\n",
      "Lipst1_1_75169 (From: rto,sce,ylip)\n",
      "Lipst1_1_1998 (From: rto,ylip)\n",
      "Lipst1_1_274 (From: rto,sce,ylip)\n",
      "Lipst1_1_62277 or Lipst1_1_244782 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_70723 (From: rto,sce,ylip)\n",
      "Lipst1_1_336589 (From: rto,sce,ylip)\n",
      "Lipst1_1_70305 (From: rto,sce,ylip)\n",
      "Lipst1_1_72797 or Lipst1_1_47259 (From: rto,ylip ; rto)\n",
      "Lipst1_1_105823 (From: rto,sce,ylip)\n",
      "Lipst1_1_4166 or Lipst1_1_6739 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_870 (From: rto,sce,ylip)\n",
      "Lipst1_1_69651 (From: rto,sce)\n",
      "Lipst1_1_205437 or Lipst1_1_201837 or; Lipst1_1_206087 (From: rto,sce ; ylip)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 or Lipst1_1_70253 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_1344 or Lipst1_1_68607 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_7080 (From: rto,sce)\n",
      "Lipst1_1_594 (From: rto,sce,ylip)\n",
      "Lipst1_1_4490 or Lipst1_1_69192 (From: rto,sce,ylip ; rto,sce)\n",
      "Lipst1_1_155998 (From: rto,sce)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_72186 (From: rto,sce,ylip)\n",
      "Lipst1_1_76545 or; Lipst1_1_2524 (From: rto)\n",
      "Lipst1_1_1792 (From: rto,sce,ylip)\n",
      "Lipst1_1_102477 or; Lipst1_1_3555 (From: rto,sce,ylip)\n",
      "Lipst1_1_71543 (From: rto,sce,ylip)\n",
      "Lipst1_1_72597 (From: rto,sce)\n",
      "Lipst1_1_161705 (From: rto,sce,ylip)\n",
      "Lipst1_1_117734 or Lipst1_1_154486 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_60832 or Lipst1_1_43727 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_69113 (From: rto,sce,ylip)\n",
      "Lipst1_1_101656 (From: rto,sce,ylip)\n",
      "Lipst1_1_5128 (From: rto,sce,ylip)\n",
      "Lipst1_1_75322 (From: rto,sce,ylip)\n",
      "Lipst1_1_2223 (From: rto,sce,ylip)\n",
      "Lipst1_1_194216 (From: rto,sce,ylip)\n",
      "Lipst1_1_51135 (From: rto,sce)\n",
      "Lipst1_1_1356 or Lipst1_1_71673 or Lipst1_1_6750 (From: rto ; sce ; ylip)\n",
      "Lipst1_1_68186 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_59512 (From: rto,sce,ylip)\n",
      "Lipst1_1_72215 or Lipst1_1_7298 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_72400 (From: rto,sce,ylip)\n",
      "Lipst1_1_5341 (From: rto,sce,ylip)\n",
      "Lipst1_1_5184 (From: rto,sce,ylip)\n",
      "Lipst1_1_59861 (From: rto,sce,ylip)\n",
      "Lipst1_1_2790 (From: rto,sce,ylip)\n",
      "Lipst1_1_324941 (From: rto,sce,ylip)\n",
      "Lipst1_1_53798 or Lipst1_1_71326 (From: rto,sce,ylip ; sce)\n",
      "Lipst1_1_139115 (From: rto,sce,ylip)\n",
      "Lipst1_1_588 (From: rto,sce,ylip)\n",
      "Lipst1_1_1741 (From: rto,ylip)\n",
      "Lipst1_1_73838 (From: rto,sce,ylip)\n",
      "Lipst1_1_51135 (From: rto,sce,ylip)\n",
      "Lipst1_1_205437 or Lipst1_1_201837 or; Lipst1_1_206087 (From: rto ; ylip)\n",
      "Lipst1_1_5235 (From: rto,sce,ylip)\n",
      "Lipst1_1_117633 (From: rto,sce,ylip)\n",
      "Lipst1_1_149190 (From: rto,ylip)\n",
      "Lipst1_1_105823 (From: rto,sce,ylip)\n",
      "Lipst1_1_69850 (From: sce,ylip)\n",
      "Lipst1_1_3267 (From: rto,sce,ylip)\n",
      "Lipst1_1_70534 or Lipst1_1_1110 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_67294 (From: rto,sce)\n",
      "Lipst1_1_54569 or Lipst1_1_3600 or Lipst1_1_111440 (From: rto,sce,ylip ; rto ; sce)\n",
      "Lipst1_1_68607 (From: rto,sce,ylip)\n",
      "Lipst1_1_934 (From: rto,sce,ylip)\n",
      "Lipst1_1_1263 (From: rto,sce,ylip)\n",
      "Lipst1_1_4523 or Lipst1_1_73977 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_75883 (From: rto,sce)\n",
      "Lipst1_1_314464 (From: rto,sce,ylip)\n",
      "Lipst1_1_157098 or Lipst1_1_4003 (From: rto,ylip ; rto)\n",
      "Lipst1_1_70713 (From: rto,sce,ylip)\n",
      "Lipst1_1_47565 (From: rto,sce,ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_6813 (From: rto,sce,ylip)\n",
      "Lipst1_1_54569 or Lipst1_1_3600 or Lipst1_1_111440 (From: rto,sce,ylip ; rto ; sce)\n",
      "Lipst1_1_74675 (From: rto,sce,ylip)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_2413 (From: rto,sce,ylip)\n",
      "Lipst1_1_3313 or Lipst1_1_89420 (From: rto,sce,ylip ; sce,ylip)\n",
      "Lipst1_1_54569 or Lipst1_1_3600 or Lipst1_1_111440 (From: rto,sce,ylip ; rto ; sce)\n",
      "Lipst1_1_67972 (From: rto,sce,ylip)\n",
      "Lipst1_1_1274 (From: rto,sce,ylip)\n",
      "Lipst1_1_2436 or Lipst1_1_285732 (From: rto,ylip ; sce,ylip)\n",
      "Lipst1_1_189449 (From: rto,sce,ylip)\n",
      "Lipst1_1_47324 (From: rto,sce,ylip)\n",
      "Lipst1_1_59861 (From: rto,sce,ylip)\n",
      "Lipst1_1_54184 or Lipst1_1_3514 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_2790 (From: rto,sce,ylip)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 (From: rto,sce,ylip)\n",
      "Lipst1_1_72797 or Lipst1_1_47259 or Lipst1_1_72533 or; Lipst1_1_70288 or; Lipst1_1_2905 (From: sce,ylip ; rto ; ylip)\n",
      "Lipst1_1_60710 or; Lipst1_1_2214 (From: rto,sce,ylip)\n",
      "Lipst1_1_5692 (From: rto,sce,ylip)\n",
      "Lipst1_1_51135 (From: rto,sce,ylip)\n",
      "Lipst1_1_1228 (From: rto,ylip)\n",
      "Lipst1_1_4003 or Lipst1_1_157098 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_4780 (From: sce,ylip)\n",
      "Lipst1_1_2790 (From: rto,sce,ylip)\n",
      "Lipst1_1_68107 (From: rto,sce)\n",
      "Lipst1_1_4257 (From: rto,sce,ylip)\n",
      "Lipst1_1_299000 (From: rto,sce)\n",
      "Lipst1_1_149190 (From: rto,sce,ylip)\n",
      "Lipst1_1_157098 (From: rto,sce,ylip)\n",
      "Lipst1_1_75883 or Lipst1_1_58573 (From: rto,sce ; ylip)\n",
      "Lipst1_1_147292 (From: rto,sce,ylip)\n",
      "Lipst1_1_72751 (From: rto,sce,ylip)\n",
      "Lipst1_1_74014 or Lipst1_1_239921 (From: rto ; sce)\n",
      "Lipst1_1_72319 (From: rto,sce,ylip)\n",
      "Lipst1_1_5128 (From: rto,sce,ylip)\n",
      "Lipst1_1_1741 (From: rto,ylip)\n",
      "Lipst1_1_75116 (From: rto,sce,ylip)\n",
      "Lipst1_1_58737 (From: rto,sce,ylip)\n",
      "Lipst1_1_5288 (From: rto,sce,ylip)\n",
      "Lipst1_1_67179 or Lipst1_1_76570 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_54569 or Lipst1_1_3600 or Lipst1_1_111440 or; Lipst1_1_1540 (From: rto,sce,ylip ; rto ; sce)\n",
      "Lipst1_1_3625 or Lipst1_1_154956 (From: sce,ylip ; rto)\n",
      "Lipst1_1_70001 (From: rto,sce)\n",
      "Lipst1_1_5292 (From: rto,sce,ylip)\n",
      "Lipst1_1_6813 (From: rto,sce)\n",
      "Lipst1_1_72193 (From: rto,sce,ylip)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 or Lipst1_1_70253 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_99817 (From: rto)\n",
      "Lipst1_1_275 or Lipst1_1_3985 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_48617 (From: rto,sce,ylip)\n",
      "Lipst1_1_72774 (From: rto,sce,ylip)\n",
      "Lipst1_1_2790 (From: rto,sce,ylip)\n",
      "Lipst1_1_164851 or Lipst1_1_103580 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_241647 (From: ylip)\n",
      "Lipst1_1_255632 or; Lipst1_1_104169 or Lipst1_1_4018 (From: rto,sce,ylip ; sce,ylip)\n",
      "Lipst1_1_1591 (From: rto,sce,ylip)\n",
      "Lipst1_1_2195 or Lipst1_1_103460 (From: rto,sce,ylip ; sce,ylip)\n",
      "Lipst1_1_209648 (From: rto,sce,ylip)\n",
      "Lipst1_1_72106 or Lipst1_1_265581 (From: rto,sce ; ylip)\n",
      "Lipst1_1_118171 (From: rto,sce,ylip)\n",
      "Lipst1_1_315712 or Lipst1_1_5184 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_117734 or Lipst1_1_154486 (From: rto,ylip ; rto)\n",
      "Lipst1_1_72902 (From: rto,sce,ylip)\n",
      "Lipst1_1_67652 (From: rto,sce,ylip)\n",
      "Lipst1_1_1741 (From: rto,ylip)\n",
      "Lipst1_1_4166 or Lipst1_1_6739 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_70496 or; Lipst1_1_4980 (From: rto,sce,ylip)\n",
      "Lipst1_1_1998 (From: rto,ylip)\n",
      "Lipst1_1_72381 or; Lipst1_1_67690 or Lipst1_1_41960 (From: rto,ylip ; sce,ylip)\n",
      "Lipst1_1_33725 or Lipst1_1_73351 (From: rto,sce,ylip ; rto,sce)\n",
      " (From: )\n",
      "Lipst1_1_54569 or Lipst1_1_3600 or Lipst1_1_111440 (From: rto,sce,ylip ; rto ; sce)\n",
      "Lipst1_1_1064 (From: rto,sce,ylip)\n",
      "Lipst1_1_70973 or Lipst1_1_61315 or; Lipst1_1_48518 or; Lipst1_1_67480 (From: rto,sce,ylip ; rto,sce)\n",
      "Lipst1_1_206 or Lipst1_1_80397 or Lipst1_1_30795 or Lipst1_1_4715 or Lipst1_1_5784 (From: rto,sce,ylip ; rto,sce ; sce,ylip ; rto ; ylip)\n",
      "Lipst1_1_3965 (From: rto,sce,ylip)\n",
      "Lipst1_1_148853 (From: rto,sce,ylip)\n",
      "Lipst1_1_3041 or Lipst1_1_71443 or Lipst1_1_6367 (From: rto,sce,ylip ; rto,sce ; rto)\n",
      "Lipst1_1_2524 (From: rto,sce,ylip)\n",
      "Lipst1_1_3809 (From: rto,sce,ylip)\n",
      "Lipst1_1_48449 or Lipst1_1_105712 (From: rto,sce,ylip ; sce,ylip)\n",
      "Lipst1_1_2774 (From: rto,sce,ylip)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 (From: rto,sce,ylip)\n",
      "Lipst1_1_238934 or; Lipst1_1_4540 (From: rto,sce,ylip)\n",
      "Lipst1_1_50963 (From: rto)\n",
      "Lipst1_1_60832 or Lipst1_1_43727 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_2102 or Lipst1_1_70967 (From: rto ; sce)\n",
      "Lipst1_1_326473 (From: rto,sce,ylip)\n",
      "Lipst1_1_2791 (From: rto,sce,ylip)\n",
      "Lipst1_1_73436 (From: rto,sce,ylip)\n",
      "Lipst1_1_314464 (From: rto,sce,ylip)\n",
      "Lipst1_1_338321 (From: rto,sce,ylip)\n",
      "Lipst1_1_5558 (From: rto,sce,ylip)\n",
      "Lipst1_1_2348 (From: rto,sce,ylip)\n",
      "Lipst1_1_5675 (From: rto,sce,ylip)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 or Lipst1_1_70253 (From: rto,sce,ylip ; rto)\n",
      "Lipst1_1_103302 (From: rto,sce)\n",
      "Lipst1_1_70370 or Lipst1_1_72782 (From: rto,sce ; ylip)\n",
      "Lipst1_1_147292 (From: rto,sce,ylip)\n",
      "Lipst1_1_643 (From: rto,sce,ylip)\n",
      "Lipst1_1_314464 (From: ylip)\n",
      "Lipst1_1_60314 (From: rto,sce,ylip)\n",
      "Lipst1_1_75883 or Lipst1_1_58573 (From: rto,sce ; ylip)\n",
      "Lipst1_1_7004 (From: rto)\n",
      "Lipst1_1_72728 (From: rto,sce,ylip)\n",
      "Lipst1_1_72381 or Lipst1_1_67690 (From: rto,sce,ylip ; rto,sce)\n",
      "Lipst1_1_85 or Lipst1_1_333987 or; Lipst1_1_4296 (From: rto,sce,ylip ; sce,ylip)\n",
      "Lipst1_1_205437 or Lipst1_1_201837 or Lipst1_1_50110 or; Lipst1_1_323814 or Lipst1_1_206087 (From: rto,sce ; sce,ylip ; rto ; ylip)\n",
      "Lipst1_1_72113 or Lipst1_1_234091 (From: rto,sce,ylip ; ylip)\n",
      "Lipst1_1_3600 or Lipst1_1_111440 (From: rto ; sce)\n",
      "Lipst1_1_72320 or Lipst1_1_94934 (From: rto,sce,ylip ; rto,sce)\n",
      "Lipst1_1_47667 (From: rto,sce,ylip)\n",
      "Lipst1_1_73882 (From: rto,sce,ylip)\n",
      "Lipst1_1_67910 or Lipst1_1_70223 (From: rto,sce,ylip ; rto,ylip)\n",
      "Lipst1_1_6367 (From: rto,sce,ylip)\n",
      "Lipst1_1_314464 (From: ylip)\n",
      "Lipst1_1_69966 (From: rto,sce,ylip)\n",
      "Lipst1_1_69891 (From: rto,sce,ylip)\n",
      "Lipst1_1_260066 (From: rto,sce,ylip)\n",
      "Lipst1_1_1145 (From: rto,sce,ylip)\n",
      "Lipst1_1_74347 or Lipst1_1_70234 (From: rto,sce,ylip ; rto,ylip)\n",
      "Lipst1_1_76147 (From: rto)\n",
      "Lipst1_1_115 or; Lipst1_1_4467 (From: rto,sce,ylip)\n",
      "Lipst1_1_3355 (From: rto,sce,ylip)\n",
      "Lipst1_1_328593 or Lipst1_1_4739 (From: rto,sce,ylip ; rto)\n",
      " (From: )\n",
      "Lipst1_1_6746 (From: rto,sce,ylip)\n",
      "Lipst1_1_4523 or Lipst1_1_73977 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_6058 (From: sce)\n",
      "Lipst1_1_108550 or Lipst1_1_67513 (From: sce,ylip ; rto)\n",
      "Lipst1_1_50624 or Lipst1_1_57415 or Lipst1_1_1339 or; Lipst1_1_74374 (From: sce,ylip ; rto ; sce)\n",
      "Lipst1_1_89278 (From: sce)\n",
      "Lipst1_1_36529 or Lipst1_1_39586 or Lipst1_1_4390 or; Lipst1_1_48116 (From: rto,ylip ; sce ; ylip)\n",
      "Lipst1_1_6795 or Lipst1_1_75116 (From: sce,ylip ; rto)\n",
      "Lipst1_1_3988 (From: rto,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_117633 or Lipst1_1_72673 (From: rto,sce ; ylip)\n",
      " (From: )\n",
      "Lipst1_1_72834 (From: sce)\n",
      "Lipst1_1_2357 (From: rto)\n",
      "Lipst1_1_50090 or Lipst1_1_238934 (From: sce,ylip ; rto)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_6058 (From: sce)\n",
      " (From: )\n",
      "Lipst1_1_6058 (From: sce)\n",
      " (From: )\n",
      "Lipst1_1_6058 (From: sce)\n",
      "Lipst1_1_101656 (From: rto)\n",
      "Lipst1_1_72342 (From: rto,sce)\n",
      "Lipst1_1_121954 or Lipst1_1_5651 or Lipst1_1_76630 (From: rto ; sce ; ylip)\n",
      "Lipst1_1_274802 or Lipst1_1_101656 (From: sce,ylip ; rto)\n",
      "Lipst1_1_50110 (From: rto)\n",
      "Lipst1_1_315712 or Lipst1_1_5184 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_74823 or Lipst1_1_5931 or; Lipst1_1_71227 (From: rto ; sce)\n",
      "Lipst1_1_72342 (From: rto,sce)\n",
      "Lipst1_1_74591 or Lipst1_1_2633 (From: rto ; sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_156565 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_69872 (From: rto)\n",
      "Lipst1_1_73446 or; Lipst1_1_103731 (From: rto,sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_74247 or Lipst1_1_73414 (From: sce,ylip ; rto)\n",
      "Lipst1_1_71411 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_72774 (From: rto)\n",
      "Lipst1_1_4390 (From: rto)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_69723 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_177423 or; Lipst1_1_121954 or; Lipst1_1_6303 or Lipst1_1_5651 or Lipst1_1_76630 (From: rto ; sce ; ylip)\n",
      " (From: )\n",
      "Lipst1_1_6205 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_6205 (From: rto)\n",
      "Lipst1_1_2357 or Lipst1_1_55653 (From: rto ; sce)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_154956 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_205437 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_205437 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_71543 or; Lipst1_1_222534 (From: ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_5692 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_314464 (From: ylip)\n",
      "Lipst1_1_2955 or Lipst1_1_6225 (From: sce ; ylip)\n",
      "Lipst1_1_72571 (From: ylip)\n",
      " (From: )\n",
      "Lipst1_1_4094 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_246003 or; Lipst1_1_3320 or Lipst1_1_67480 (From: sce,ylip ; sce)\n",
      "Lipst1_1_246003 or; Lipst1_1_3320 or; Lipst1_1_67480 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_2594 (From: sce,ylip)\n",
      "Lipst1_1_72193 (From: ylip)\n",
      "Lipst1_1_314464 (From: ylip)\n",
      "Lipst1_1_314464 (From: ylip)\n",
      "Lipst1_1_72557 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_107475 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_211414 (From: sce)\n",
      " (From: )\n",
      "Lipst1_1_85639 (From: ylip)\n",
      "Lipst1_1_182606 (From: sce,ylip)\n",
      "Lipst1_1_67262 (From: sce,ylip)\n",
      "Lipst1_1_73715 (From: sce,ylip)\n",
      "Lipst1_1_5903 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_238934 or; Lipst1_1_4540 (From: ylip)\n",
      " (From: )\n",
      "Lipst1_1_73737 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_72342 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_48518 or Lipst1_1_61315 or; Lipst1_1_70973 or; Lipst1_1_67480 (From: sce,ylip ; sce)\n",
      " (From: )\n",
      "Lipst1_1_70973 or Lipst1_1_61315 or; Lipst1_1_48518 or; Lipst1_1_67480 (From: sce,ylip ; sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_1740 (From: sce,ylip)\n",
      "Lipst1_1_55005 (From: ylip)\n",
      "Lipst1_1_70001 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_65274 or Lipst1_1_261247 (From: sce ; ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_5184 or Lipst1_1_315712 (From: sce,ylip ; ylip)\n",
      "Lipst1_1_5184 or Lipst1_1_315712 (From: sce,ylip ; ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_2772 (From: sce,ylip)\n",
      "Lipst1_1_74886 (From: sce,ylip)\n",
      "Lipst1_1_170681 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_2633 or Lipst1_1_74591 (From: sce ; ylip)\n",
      "Lipst1_1_50182 (From: ylip)\n",
      "Lipst1_1_72144 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_60314 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_5229 or; Lipst1_1_3988 (From: ylip)\n",
      "Lipst1_1_67480 or Lipst1_1_48518 or; Lipst1_1_61315 or; Lipst1_1_70973 or; Lipst1_1_246003 or; Lipst1_1_3320 (From: sce,ylip ; sce)\n",
      "Lipst1_1_48518 or Lipst1_1_61315 or; Lipst1_1_70973 or; Lipst1_1_67480 (From: sce,ylip ; sce)\n",
      "Lipst1_1_70973 or Lipst1_1_61315 or; Lipst1_1_48518 or; Lipst1_1_67480 (From: sce,ylip ; sce)\n",
      "Lipst1_1_67480 or Lipst1_1_61315 or; Lipst1_1_70973 or; Lipst1_1_48518 (From: sce,ylip ; sce)\n",
      " (From: )\n",
      "Lipst1_1_110167 (From: ylip)\n",
      "Lipst1_1_73435 or Lipst1_1_6913 or; Lipst1_1_111304 or; Lipst1_1_114537 or; Lipst1_1_180820 or; Lipst1_1_54800 or; Lipst1_1_4310 or; Lipst1_1_2954 (From: sce ; ylip)\n",
      "Lipst1_1_50110 (From: sce)\n",
      " (From: )\n",
      "Lipst1_1_68590 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_54073 or Lipst1_1_4566 or; Lipst1_1_3884 (From: sce,ylip ; ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_112377 or Lipst1_1_75138 (From: sce,ylip ; ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_60710 or; Lipst1_1_2214 (From: sce,ylip)\n",
      "Lipst1_1_68056 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_72728 (From: sce,ylip)\n",
      "Lipst1_1_70762 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_54023 (From: sce,ylip)\n",
      "Lipst1_1_54023 (From: sce,ylip)\n",
      "Lipst1_1_70234 or; Lipst1_1_73846 (From: sce)\n",
      "Lipst1_1_3931 (From: sce,ylip)\n",
      "Lipst1_1_335824 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_3010 (From: sce)\n",
      "Lipst1_1_54023 (From: sce,ylip)\n",
      "Lipst1_1_314464 (From: ylip)\n",
      "Lipst1_1_332244 (From: sce,ylip)\n",
      "Lipst1_1_114537 or; Lipst1_1_180820 or; Lipst1_1_54800 or; Lipst1_1_6913 or; Lipst1_1_4310 or; Lipst1_1_2954 or; Lipst1_1_111304 (From: ylip)\n",
      "Lipst1_1_117393 (From: sce,ylip)\n",
      "Lipst1_1_73932 or Lipst1_1_3458 (From: sce ; ylip)\n",
      "Lipst1_1_3458 (From: ylip)\n",
      "Lipst1_1_48150 (From: sce,ylip)\n",
      "Lipst1_1_73932 or Lipst1_1_124163 or; Lipst1_1_5321 (From: sce,ylip ; sce)\n",
      "Lipst1_1_73932 (From: ylip)\n",
      "Lipst1_1_4013 (From: sce,ylip)\n",
      "Lipst1_1_976 (From: sce,ylip)\n",
      " (From: )\n",
      "Lipst1_1_310928 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_180684 (From: sce)\n",
      " (From: )\n",
      "Lipst1_1_4188 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_310928 (From: sce,ylip)\n",
      "Lipst1_1_976 (From: sce)\n",
      "Lipst1_1_68196 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_102853 (From: ylip)\n",
      "Lipst1_1_870 (From: sce,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_54828 (From: sce,ylip)\n",
      "Lipst1_1_205437 (From: sce)\n",
      "Lipst1_1_147168 or; Lipst1_1_70762 (From: ylip)\n",
      "Lipst1_1_28592 or; Lipst1_1_73864 or; Lipst1_1_58405 or Lipst1_1_151 (From: sce,ylip ; ylip)\n",
      "Lipst1_1_70973 or Lipst1_1_61315 or; Lipst1_1_48518 or; Lipst1_1_67480 (From: sce,ylip ; sce)\n",
      "Lipst1_1_39586 or Lipst1_1_4390 or; Lipst1_1_36529 or; Lipst1_1_48116 (From: sce ; ylip)\n",
      " (From: )\n",
      "Lipst1_1_73732 (From: ylip)\n",
      " (From: )\n",
      "Lipst1_1_75768 (From: sce,ylip)\n",
      "Lipst1_1_4583 or Lipst1_1_73341 (From: sce ; ylip)\n",
      " (From: )\n",
      "Lipst1_1_2633 (From: sce)\n",
      "Lipst1_1_46980 (From: rto,ylip)\n",
      "Lipst1_1_2790 (From: rto,ylip)\n",
      "Lipst1_1_50945 (From: ylip)\n",
      "Lipst1_1_57709 (From: rto,ylip)\n",
      "Lipst1_1_71969 or Lipst1_1_57709 (From: rto ; ylip)\n",
      " (From: )\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      "Lipst1_1_76544 (From: rto,ylip)\n",
      " (From: )\n",
      "Lipst1_1_47324 or; Lipst1_1_2790 or; Lipst1_1_75883 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_3498 (From: rto,ylip)\n",
      "Lipst1_1_2790 or; Lipst1_1_75883 or; Lipst1_1_160699 or; Lipst1_1_47324 or; Lipst1_1_158150 (From: rto)\n",
      "Lipst1_1_73286 or; Lipst1_1_287747 (From: rto,ylip)\n",
      "Lipst1_1_57709 (From: rto,ylip)\n",
      "Lipst1_1_47324 or; Lipst1_1_2790 or; Lipst1_1_75883 (From: rto)\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      "Lipst1_1_73286 or; Lipst1_1_287747 (From: rto,ylip)\n",
      "Lipst1_1_2790 or; Lipst1_1_75883 or; Lipst1_1_160699 or; Lipst1_1_47324 or; Lipst1_1_158150 (From: rto)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 (From: rto,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_284310 (From: ylip)\n",
      "Lipst1_1_6224 or; Lipst1_1_6225 or Lipst1_1_72701 or; Lipst1_1_75352 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_47324 (From: rto,ylip)\n",
      "Lipst1_1_6224 or; Lipst1_1_6225 or Lipst1_1_72701 or; Lipst1_1_75352 (From: rto,ylip ; ylip)\n",
      " (From: )\n",
      "Lipst1_1_4956 or; Lipst1_1_313744 (From: rto)\n",
      "Lipst1_1_338549 (From: rto)\n",
      "Lipst1_1_67994 or; Lipst1_1_104545 or Lipst1_1_3871 (From: rto ; ylip)\n",
      "Lipst1_1_59861 (From: rto,ylip)\n",
      "Lipst1_1_2767 (From: rto,ylip)\n",
      "Lipst1_1_71969 (From: rto)\n",
      "Lipst1_1_4472 (From: rto,ylip)\n",
      "Lipst1_1_57709 (From: ylip)\n",
      "Lipst1_1_105970 (From: rto,ylip)\n",
      " (From: )\n",
      "Lipst1_1_2907 or; Lipst1_1_74933 or; Lipst1_1_49131 or Lipst1_1_69250 (From: rto,ylip ; rto)\n",
      " (From: )\n",
      "Lipst1_1_6224 or; Lipst1_1_6225 or Lipst1_1_72701 or; Lipst1_1_75352 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_115 or; Lipst1_1_4467 (From: rto,ylip)\n",
      "Lipst1_1_47324 or; Lipst1_1_2790 or; Lipst1_1_75883 (From: rto)\n",
      "Lipst1_1_6750 (From: rto,ylip)\n",
      "Lipst1_1_70877 (From: rto,ylip)\n",
      "Lipst1_1_6224 or; Lipst1_1_6225 or Lipst1_1_72701 or; Lipst1_1_75352 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_74417 (From: rto,ylip)\n",
      "Lipst1_1_67994 or; Lipst1_1_104545 or Lipst1_1_3871 (From: rto ; ylip)\n",
      "Lipst1_1_4883 (From: rto,ylip)\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      "Lipst1_1_2790 or; Lipst1_1_75883 or; Lipst1_1_160699 or; Lipst1_1_47324 or; Lipst1_1_158150 (From: rto)\n",
      "Lipst1_1_74417 (From: rto,ylip)\n",
      "Lipst1_1_73286 or; Lipst1_1_287747 (From: rto,ylip)\n",
      "Lipst1_1_103920 or; Lipst1_1_37982 (From: rto,ylip)\n",
      "Lipst1_1_6224 or; Lipst1_1_6225 or Lipst1_1_72701 or; Lipst1_1_75352 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      "Lipst1_1_57709 (From: rto,ylip)\n",
      "Lipst1_1_46921 (From: rto,ylip)\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      "Lipst1_1_57709 (From: rto,ylip)\n",
      "Lipst1_1_2363 or; Lipst1_1_29458 or Lipst1_1_70253 (From: rto,ylip ; rto)\n",
      "Lipst1_1_71400 or Lipst1_1_69441 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      " (From: )\n",
      "Lipst1_1_57709 (From: rto,ylip)\n",
      "Lipst1_1_47324 or; Lipst1_1_2790 or; Lipst1_1_75883 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_75883 (From: rto)\n",
      "Lipst1_1_75883 (From: rto)\n",
      "Lipst1_1_2348 (From: rto,ylip)\n",
      "Lipst1_1_53798 (From: rto,ylip)\n",
      " (From: )\n",
      "Lipst1_1_71969 (From: rto)\n",
      "Lipst1_1_53798 (From: rto,ylip)\n",
      "Lipst1_1_70877 or Lipst1_1_74886 (From: rto ; ylip)\n",
      "Lipst1_1_6224 or; Lipst1_1_6225 or Lipst1_1_72701 or; Lipst1_1_75352 (From: rto,ylip ; ylip)\n",
      "Lipst1_1_75883 or Lipst1_1_58573 (From: rto ; ylip)\n",
      "Lipst1_1_70691 (From: rto,ylip)\n",
      "Lipst1_1_141437 (From: rto,ylip)\n",
      "Lipst1_1_57709 (From: rto,ylip)\n",
      "Lipst1_1_53798 (From: rto,ylip)\n",
      "Lipst1_1_75883 or Lipst1_1_58573 (From: rto ; ylip)\n",
      "Lipst1_1_4094 (From: rto,ylip)\n",
      "Lipst1_1_275 (From: rto,ylip)\n",
      " (From: )\n",
      "Lipst1_1_57709 (From: rto,ylip)\n",
      "Lipst1_1_117342 (From: rto)\n",
      "Lipst1_1_117895 (From: rto)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_314464 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_70194 (From: rto,ylip)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_71147 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_5601 or; Lipst1_1_72166 (From: rto)\n",
      "Lipst1_1_6303 or Lipst1_1_4197 (From: rto ; sce)\n",
      " (From: )\n",
      "Lipst1_1_74900 (From: rto,sce)\n",
      "Lipst1_1_314464 (From: rto)\n",
      "Lipst1_1_5453 or Lipst1_1_5116 or; Lipst1_1_6972 (From: rto ; sce)\n",
      "Lipst1_1_127634 or Lipst1_1_265510 (From: rto,sce ; rto)\n",
      "Lipst1_1_289621 or Lipst1_1_3546 (From: rto,sce ; sce)\n",
      "Lipst1_1_70370 (From: rto,sce)\n",
      "Lipst1_1_76289 (From: rto,sce)\n",
      "Lipst1_1_105640 (From: rto,sce)\n",
      "Lipst1_1_4472 (From: rto,sce)\n",
      "Lipst1_1_6303 or Lipst1_1_4197 (From: rto ; sce)\n",
      "Lipst1_1_102591 (From: rto)\n",
      "Lipst1_1_92627 (From: sce)\n",
      "Lipst1_1_314768 (From: rto,sce)\n",
      "Lipst1_1_423 (From: rto,sce)\n",
      "Lipst1_1_1263 (From: rto,sce)\n",
      "Lipst1_1_2889 (From: rto,sce)\n",
      "Lipst1_1_107753 (From: rto,sce)\n",
      "Lipst1_1_73838 (From: rto,sce)\n",
      "Lipst1_1_3878 (From: rto,sce)\n",
      "Lipst1_1_53536 (From: rto,sce)\n",
      "Lipst1_1_72751 (From: rto,sce)\n",
      "Lipst1_1_2256 or; Lipst1_1_61026 or Lipst1_1_6033 or; Lipst1_1_97209 (From: rto ; sce)\n",
      "Lipst1_1_76289 (From: rto,sce)\n",
      "Lipst1_1_338549 (From: rto)\n",
      "Lipst1_1_48101 (From: rto,sce)\n",
      "Lipst1_1_27755 (From: rto,sce)\n",
      "Lipst1_1_282 or Lipst1_1_5651 (From: rto ; sce)\n",
      "Lipst1_1_81155 (From: rto,sce)\n",
      "Lipst1_1_46126 or Lipst1_1_48239 (From: rto,sce ; sce)\n",
      "Lipst1_1_142956 (From: rto,sce)\n",
      "Lipst1_1_3974 (From: rto,sce)\n",
      "Lipst1_1_3974 (From: rto,sce)\n",
      "Lipst1_1_246321 (From: rto,sce)\n",
      "Lipst1_1_2195 (From: rto)\n",
      "Lipst1_1_314464 (From: rto)\n",
      "Lipst1_1_46126 (From: rto,sce)\n",
      "Lipst1_1_3313 (From: rto)\n",
      "Lipst1_1_102207 (From: rto,sce)\n",
      "Lipst1_1_71146 (From: rto,sce)\n",
      " (From: )\n",
      "Lipst1_1_1110 (From: rto,sce)\n",
      "Lipst1_1_931 or Lipst1_1_6152 or; Lipst1_1_5188 (From: rto,sce ; sce)\n",
      "Lipst1_1_76359 (From: rto)\n",
      "Lipst1_1_69867 or; Lipst1_1_70253 or Lipst1_1_104402 (From: rto ; sce)\n",
      "Lipst1_1_261247 or; Lipst1_1_65274 or Lipst1_1_72041 (From: rto,sce ; sce)\n",
      " (From: )\n",
      "Lipst1_1_2772 or Lipst1_1_99746 (From: rto,sce ; rto)\n",
      "Lipst1_1_2256 or; Lipst1_1_61026 or Lipst1_1_111440 or; Lipst1_1_6033 or; Lipst1_1_119165 or; Lipst1_1_97209 (From: rto ; sce)\n",
      "Lipst1_1_70370 (From: rto,sce)\n",
      "Lipst1_1_2772 or Lipst1_1_99746 (From: rto,sce ; rto)\n",
      "Lipst1_1_61657 (From: rto,sce)\n",
      "Lipst1_1_75890 (From: rto)\n",
      "Lipst1_1_76289 (From: rto,sce)\n",
      "Lipst1_1_314464 (From: rto)\n",
      "Lipst1_1_7158 or; Lipst1_1_50618 or Lipst1_1_6034 or; Lipst1_1_48239 (From: rto ; sce)\n",
      "Lipst1_1_3540 (From: rto,sce)\n",
      "Lipst1_1_74823 or Lipst1_1_5931 (From: rto ; sce)\n",
      "Lipst1_1_3974 (From: rto,sce)\n",
      "Lipst1_1_75163 (From: rto,sce)\n",
      "Lipst1_1_69028 (From: rto,sce)\n",
      "Lipst1_1_2436 or Lipst1_1_285732 (From: rto ; sce)\n",
      "Lipst1_1_338549 (From: rto)\n",
      "Lipst1_1_3860 (From: rto,sce)\n",
      "Lipst1_1_1263 (From: rto,sce)\n",
      " (From: )\n",
      "Lipst1_1_3532 (From: rto)\n",
      "Lipst1_1_1559 (From: rto,sce)\n",
      "Lipst1_1_75335 (From: rto,sce)\n",
      "Lipst1_1_69288 (From: rto,sce)\n",
      "Lipst1_1_931 or Lipst1_1_5596 (From: rto,sce ; sce)\n",
      "Lipst1_1_423 (From: rto,sce)\n",
      "Lipst1_1_4566 (From: rto,sce)\n",
      "Lipst1_1_177423 or Lipst1_1_121954 (From: rto,sce ; rto)\n",
      "Lipst1_1_46126 (From: rto,sce)\n",
      "Lipst1_1_5184 (From: rto,sce)\n",
      "Lipst1_1_42480 (From: rto,sce)\n",
      "Lipst1_1_56835 (From: rto,sce)\n",
      "Lipst1_1_1857 (From: rto,sce)\n",
      "Lipst1_1_201837 or Lipst1_1_72253 (From: rto,sce ; sce)\n",
      "Lipst1_1_2772 or Lipst1_1_99746 (From: rto,sce ; rto)\n",
      "Lipst1_1_6465 or Lipst1_1_73761 or; Lipst1_1_5188 (From: rto,sce ; sce)\n",
      "Lipst1_1_69288 (From: rto,sce)\n",
      "Lipst1_1_101999 (From: rto,sce)\n",
      "Lipst1_1_5189 or Lipst1_1_1356 (From: rto,sce ; rto)\n",
      "Lipst1_1_76289 (From: rto,sce)\n",
      "Lipst1_1_972 (From: rto,sce)\n",
      "Lipst1_1_67195 (From: sce)\n",
      "Lipst1_1_2766 (From: rto,sce)\n",
      "Lipst1_1_594 (From: rto,sce)\n",
      "Lipst1_1_74347 or Lipst1_1_2352 or; Lipst1_1_54959 (From: rto,sce ; rto)\n",
      "Lipst1_1_46126 (From: rto,sce)\n",
      "Lipst1_1_67618 (From: rto,sce)\n",
      "Lipst1_1_42480 (From: rto,sce)\n",
      "Lipst1_1_338549 (From: rto)\n",
      "Lipst1_1_70181 (From: rto,sce)\n",
      "Lipst1_1_76061 (From: rto,sce)\n",
      "Lipst1_1_3974 (From: rto,sce)\n",
      "Lipst1_1_46126 (From: rto,sce)\n",
      "Lipst1_1_2772 or Lipst1_1_99746 (From: rto,sce ; rto)\n",
      "Lipst1_1_61657 (From: rto,sce)\n",
      "Lipst1_1_3923 (From: rto,sce)\n",
      "Lipst1_1_44049 or Lipst1_1_70253 (From: rto,sce ; sce)\n",
      " (From: )\n",
      "Lipst1_1_73367 (From: rto,sce)\n",
      "Lipst1_1_5784 (From: rto,sce)\n",
      "Lipst1_1_70370 (From: rto)\n",
      "Lipst1_1_103302 (From: rto,sce)\n",
      "Lipst1_1_310928 (From: rto,sce)\n",
      "Lipst1_1_870 (From: rto,sce)\n",
      "Lipst1_1_314768 (From: rto,sce)\n",
      "Lipst1_1_314768 (From: rto,sce)\n",
      "Lipst1_1_127634 or Lipst1_1_265510 (From: rto,sce ; rto)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_92025 (From: rto,sce)\n",
      "Lipst1_1_177423 or; Lipst1_1_121954 or; Lipst1_1_6303 or Lipst1_1_4197 (From: rto ; sce)\n",
      "Lipst1_1_67618 (From: rto,sce)\n",
      "Lipst1_1_70370 (From: rto,sce)\n",
      "Lipst1_1_276115 or Lipst1_1_114186 (From: rto ; sce)\n",
      "Lipst1_1_5895 (From: rto,sce)\n",
      "Lipst1_1_52540 (From: sce)\n",
      "Lipst1_1_2102 or Lipst1_1_70967 (From: rto ; sce)\n",
      "Lipst1_1_284310 (From: rto,sce)\n",
      "Lipst1_1_4155 (From: rto,sce)\n",
      "Lipst1_1_2256 or; Lipst1_1_61026 or Lipst1_1_6033 or; Lipst1_1_52551 or; Lipst1_1_97209 (From: rto ; sce)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_67991 (From: rto,sce)\n",
      "Lipst1_1_68284 (From: rto,sce)\n",
      "Lipst1_1_6303 or Lipst1_1_4197 (From: rto ; sce)\n",
      "Lipst1_1_3985 (From: rto,sce)\n",
      "Lipst1_1_127634 or Lipst1_1_265510 (From: rto,sce ; rto)\n",
      "Lipst1_1_5784 (From: rto,sce)\n",
      "Lipst1_1_3974 (From: rto,sce)\n",
      "Lipst1_1_33829 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_3540 (From: rto,sce)\n",
      "Lipst1_1_76289 (From: rto,sce)\n",
      "Lipst1_1_57798 (From: rto,sce)\n",
      "Lipst1_1_74862 or Lipst1_1_70253 (From: rto ; sce)\n",
      " (From: )\n",
      "Lipst1_1_72834 or; Lipst1_1_71351 (From: rto)\n",
      "Lipst1_1_5576 (From: rto,sce)\n",
      "Lipst1_1_72360 (From: rto,sce)\n",
      "Lipst1_1_70370 (From: rto)\n",
      "Lipst1_1_127634 or Lipst1_1_265510 (From: rto,sce ; rto)\n",
      "Lipst1_1_70181 (From: rto,sce)\n",
      "Lipst1_1_275 (From: rto,sce)\n",
      "Lipst1_1_2772 or Lipst1_1_99746 (From: rto,sce ; rto)\n",
      "Lipst1_1_3431 (From: rto,sce)\n",
      "Lipst1_1_70253 or; Lipst1_1_69038 (From: rto,sce)\n",
      "Lipst1_1_73838 (From: rto)\n",
      "Lipst1_1_75138 (From: rto,sce)\n",
      "Lipst1_1_1110 or; Lipst1_1_72701 (From: rto,sce)\n",
      "Lipst1_1_7158 or; Lipst1_1_50618 or Lipst1_1_6034 or; Lipst1_1_48239 (From: rto ; sce)\n",
      "Lipst1_1_176686 (From: rto)\n",
      "Lipst1_1_73351 (From: rto,sce)\n",
      "Lipst1_1_97460 or Lipst1_1_48807 (From: rto ; sce)\n",
      "Lipst1_1_4054 or Lipst1_1_73352 (From: rto ; sce)\n",
      "Lipst1_1_49747 or Lipst1_1_71533 (From: rto,sce ; sce)\n",
      "Lipst1_1_5472 (From: rto)\n",
      "Lipst1_1_146577 (From: sce)\n",
      "Lipst1_1_925 (From: rto)\n",
      "Lipst1_1_71411 (From: sce)\n",
      "Lipst1_1_76716 or Lipst1_1_5571 (From: rto ; sce)\n",
      "Lipst1_1_6779 or Lipst1_1_73351 (From: rto ; sce)\n",
      "Lipst1_1_3290 (From: rto)\n",
      " (From: )\n",
      "Lipst1_1_53798 (From: rto,sce)\n",
      "Lipst1_1_4390 (From: rto)\n",
      "Lipst1_1_50110 (From: rto)\n",
      "Lipst1_1_73932 or Lipst1_1_3458 (From: rto ; sce)\n",
      " (From: )\n",
      "Lipst1_1_76716 (From: rto,sce)\n",
      "Lipst1_1_3617 (From: rto,sce)\n",
      "Lipst1_1_314464 (From: rto)\n",
      "Lipst1_1_67687 (From: rto,sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_170681 or; Lipst1_1_103302 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_67687 (From: rto,sce)\n",
      "Lipst1_1_72108 (From: rto,sce)\n",
      "Lipst1_1_6006 (From: sce)\n",
      " (From: )\n",
      "Lipst1_1_170681 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_47259 (From: rto)\n",
      "Lipst1_1_75992 (From: rto)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_6459 (From: rto)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_1356 (From: rto)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_71543 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_170681 (From: sce)\n",
      " (From: )\n",
      "Lipst1_1_170681 (From: sce)\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      " (From: )\n",
      "Lipst1_1_71319 (From: rto,sce)\n",
      " (From: )\n"
     ]
    }
   ],
   "source": [
    "for r, label in zip(result, result_labels):\n",
    "    print(f\"{r[0]} (From: {label})\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8719ea6b",
   "metadata": {},
   "source": [
    "### Note - section above did not add to model - just obtained orthologs and new GPRs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "42fb53ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save output. \n",
    "# df3 = pd.DataFrame([_names,_id,_stoich,_subsystem,_gpr_rto,_gpr_sce,_gpr_ylip,_lst_rto_gene,_lst_sce_gene,_lst_ylip_gene,_common_all,_only_12,_only_13,_only_23,_only_1,_only_2,_only_3,_all_three_diff,result,result_labels]).T\n",
    "# df3.to_csv('allcomparisions_genes_withGPR_.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "97e10d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rxnID to new GPR.\n",
    "__dict = dict(zip(_id,result))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b40691ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rxnID to organism source.\n",
    "__dict2 = dict(zip(_id,result_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e3b8adb",
   "metadata": {},
   "source": [
    "verify model and add a needed reaction for growth on lipid compounds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5e1e3bdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reaction needed to enable growth for lipid compounds. \n",
    "model.add_reactions([rto.reactions.GAPP.copy()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "77e87bdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# examine the model genes for non-Lst genes.\n",
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f7e207f",
   "metadata": {},
   "source": [
    "check that the genes are all associated with a reaction. remove the ones that are not. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6b64d02c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71\n",
      "21\n"
     ]
    }
   ],
   "source": [
    "wrong_g = []\n",
    "wrong_rx = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g.append(g)\n",
    "            wrong_rx.append(rx)\n",
    "print(len(set(wrong_g)))\n",
    "print(len(set(wrong_rx)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "745d125b",
   "metadata": {},
   "source": [
    "71 genes are associated with 21 reactions that currently do not have Lst genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "46d26cc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<Reaction ACRNtp at 0x2a1bc3a10>,\n",
       " <Reaction ALDD2ym at 0x15b811ad0>,\n",
       " <Reaction ATPS3m at 0x15bcd1890>,\n",
       " <Reaction C30CPT1 at 0x2a1bc32d0>,\n",
       " <Reaction C3STKR1er at 0x2a1bc1e10>,\n",
       " <Reaction C3STKR2er at 0x2a17cab10>,\n",
       " <Reaction CRNtp at 0x2a1bc8150>,\n",
       " <Reaction CYOO6m at 0x2a18c2f10>,\n",
       " <Reaction CYOR_u9m at 0x15b829810>,\n",
       " <Reaction DNTPPA at 0x2a1841d90>,\n",
       " <Reaction DRTPPD at 0x2a1b92f90>,\n",
       " <Reaction FA160COAabcp at 0x2a1ba3990>,\n",
       " <Reaction FA183COAabcp at 0x2a1920c90>,\n",
       " <Reaction FA240tp at 0x2a1a134d0>,\n",
       " <Reaction GAPP at 0x16a228f90>,\n",
       " <Reaction H2Otm at 0x2a1782c90>,\n",
       " <Reaction OIVD2m at 0x2a1bbb6d0>,\n",
       " <Reaction PMDPHT at 0x2a1bc2350>,\n",
       " <Reaction PRPPS at 0x2a1bba210>,\n",
       " <Reaction SERPTer at 0x2a13c8f50>,\n",
       " <Reaction TRDR at 0x2a1b93390>}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(wrong_rx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "aac20f27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/kn/zzns_smn1q79xdnf1tgqk__r0000gn/T/tmp_184yos8.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1898 rows, 4166 columns, 16794 nonzeros\n"
     ]
    }
   ],
   "source": [
    "model_v01 = model.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d81a45ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ICLp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Isocitrate lyase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a1395150</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>icit_x --> glx_x + succ_x</p>\n",
       "                    <p style='text-align:right'>Isocitrate --> Glyoxylate + Succinate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_112377</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ICLp at 0x2a1395150>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('ICLp')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab8bb241",
   "metadata": {},
   "source": [
    "## Cell block below replaces the Lst model reaction GPR with orthoMCL && adds reactions that were not in the model but are in rto and have orthologs.\n",
    "\n",
    "also keeps gapfilled reactions that enabled growth. \n",
    "replace genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "60ac3beb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2083"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bce48be7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.reactions.GAPP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6e7a1e30",
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'GAPP' in list(__dict.keys()):\n",
    "    print('yes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "702263bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list(__dict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "bb7a0009",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "old H2Otm 9014\n",
      "old DNTPPA 15385\n",
      "old PMDPHT 11513\n",
      "old ACRNtp 10149\n",
      "old CRNtp 10149\n",
      "old FA240tp 9912\n",
      "old FA160COAabcp 13167 and 9637\n",
      "old DRTPPD 15506\n"
     ]
    }
   ],
   "source": [
    "# all the reactions added below come from the rto model. The sources of the Lst genes can come from rto, sce or ylip. \n",
    "\n",
    "and_checks = []\n",
    "maintained_v01_gpr = []\n",
    "gpr_replaced = []\n",
    "no_gpr_bothMethods = []\n",
    "added_to_model = []\n",
    "not_in_model = []\n",
    "added_to_model_source = []\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# iterate through dict (reaction.id:gpr)\n",
    "for r in list(__dict.keys()):\n",
    "    \n",
    "    # if the r is included in the model. \n",
    "    if r in model.reactions:\n",
    "        \n",
    "        # obtain the reaction.\n",
    "        model_rxn = model.reactions.get_by_id(r)\n",
    "#         print(model_rxn.gene_reaction_rule)\n",
    "        old_gpr = model_rxn.gene_reaction_rule\n",
    "\n",
    "        # obtain the new gpr. \n",
    "        new_gpr = __dict[r][0]\n",
    "        \n",
    "        # a semi-colon represents a potential \"and\" reaction for further validation.\n",
    "        if ';' in new_gpr:\n",
    "            new_gpr = new_gpr.replace(';','')\n",
    "            \n",
    "            # further validation.\n",
    "            and_checks.append(r)\n",
    "    \n",
    "    # reaction not in model.\n",
    "    else:\n",
    "        value = __dict[r][0]\n",
    "        if (value) and (r in rto.reactions):\n",
    "            copy_rxn = rto.reactions.get_by_id(r).copy()\n",
    "            model.add_reactions([copy_rxn])\n",
    "            \n",
    "            \n",
    "            if ';' in value:\n",
    "                value = value.replace(';','')\n",
    "                new_gpr = value\n",
    "                # further validation\n",
    "                and_checks.append(r)\n",
    "            model.reactions.get_by_id(r).gene_reaction_rule = value\n",
    "            \n",
    "            \n",
    "            added_to_model.append(r)\n",
    "            added_to_model_source.append(__dict2[r])\n",
    "        \n",
    "        else:\n",
    "            not_in_model.append(r)\n",
    "    \n",
    "#     if (\"Lipst1\" not in new_gpr) and (new_gpr):\n",
    "#         print(\"new\",r, new_gpr)\n",
    "#     if (\"Lipst1\" not in old_gpr) and (old_gpr):\n",
    "#         print(\"old\", r, old_gpr)\n",
    "        \n",
    "    if (r=='GAPP'):\n",
    "        print('new',new_gpr,'old',old_gpr)\n",
    "    # if there is a gene value from orthoMCL.     \n",
    "    if (old_gpr) and (new_gpr) and (r in model.reactions):\n",
    "        gpr_replaced.append(r)\n",
    "        model_rxn.gene_reaction_rule = new_gpr\n",
    "        if (\"Lipst1\" not in new_gpr):\n",
    "            print(\"new\",r, new_gpr)\n",
    "    elif (new_gpr) and (r in model.reactions):\n",
    "        gpr_replaced.append(r)\n",
    "        model_rxn.gene_reaction_rule = new_gpr\n",
    "    \n",
    "        if (\"Lipst1\" not in new_gpr):\n",
    "            print(\"new2\",r, new_gpr)\n",
    "\n",
    "    # no new gpr -- remove reaction?\n",
    "    # keep old reaction?\n",
    "    elif (old_gpr) and (r in model.reactions):\n",
    "        maintained_v01_gpr.append(r)\n",
    "        if (\"Lipst1\" not in old_gpr) and (old_gpr):\n",
    "            print(\"old\", r, old_gpr)\n",
    "\n",
    "    elif ((r in model.reactions)):\n",
    "        no_gpr_bothMethods.append(r)\n",
    "    else: \n",
    "        not_in_model.append(r)\n",
    "        \n",
    "            \n",
    "#             print(f'{old_gpr} and {new_gpr}:')\n",
    "\n",
    "        #model_rxn.gene_reaction_rule = new_gpr\n",
    "\n",
    "#     print(r,new_gpr,model_rxn.gene_reaction_rule)\n",
    "        \n",
    "        # assign new gpr rule.\n",
    "#         model_rxn.gene_reaction_rule = new_gpr\n",
    "#         print(model_rxn.gene_reaction_rule)\n",
    "    \n",
    "    # if it is in the rto model (so was removed because lack of GPR)\n",
    "#     elif r in rto.reactions:\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "6f4be130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['H2Otm',\n",
       " 'DNTPPA',\n",
       " 'PMDPHT',\n",
       " 'ACRNtp',\n",
       " 'CRNtp',\n",
       " 'FA240tp',\n",
       " 'FA160COAabcp',\n",
       " 'DRTPPD']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maintained_v01_gpr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "4b1c48e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "105"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "eba77214",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12737 has no associated reaction\n",
      "9688 has no associated reaction\n",
      "12730 has no associated reaction\n",
      "10848 has no associated reaction\n",
      "9940 has no associated reaction\n",
      "16359 has no associated reaction\n",
      "11958 has no associated reaction\n",
      "ATP8 has no associated reaction\n",
      "14786 has no associated reaction\n",
      "13842 has no associated reaction\n",
      "9619 has no associated reaction\n",
      "13759 has no associated reaction\n",
      "11967 has no associated reaction\n",
      "ATP9 has no associated reaction\n",
      "15287 has no associated reaction\n",
      "13424 has no associated reaction\n",
      "15880 has no associated reaction\n",
      "14912 has no associated reaction\n",
      "13053 has no associated reaction\n",
      "ATP6 has no associated reaction\n",
      "13252 has no associated reaction\n",
      "15589 has no associated reaction\n",
      "10674 has no associated reaction\n",
      "9080 has no associated reaction\n",
      "13426 has no associated reaction\n",
      "11145 has no associated reaction\n",
      "16592 has no associated reaction\n",
      "15900 has no associated reaction\n",
      "9910 has no associated reaction\n",
      "14014 has no associated reaction\n",
      "11290 has no associated reaction\n",
      "11172 has no associated reaction\n",
      "14237 has no associated reaction\n",
      "16019 has no associated reaction\n",
      "9838 has no associated reaction\n",
      "10937 has no associated reaction\n",
      "16648 has no associated reaction\n",
      "10043 has no associated reaction\n",
      "13627 has no associated reaction\n",
      "15762 has no associated reaction\n",
      "8454 has no associated reaction\n",
      "14623 has no associated reaction\n",
      "12742 has no associated reaction\n",
      "9700 has no associated reaction\n",
      "10293 has no associated reaction\n",
      "13228 has no associated reaction\n",
      "11362 has no associated reaction\n",
      "12752 has no associated reaction\n",
      "11907 has no associated reaction\n",
      "13813 has no associated reaction\n",
      "13048 has no associated reaction\n",
      "11328 has no associated reaction\n",
      "10704 has no associated reaction\n",
      "10705 has no associated reaction\n",
      "10200 has no associated reaction\n",
      "13413 has no associated reaction\n",
      "15037 has no associated reaction\n",
      "14866 has no associated reaction\n",
      "12715 has no associated reaction\n",
      "12761 has no associated reaction\n",
      "8434 has no associated reaction\n",
      "71\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g)))\n",
    "print(len(set(wrong_rx2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "30e6b681",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<Reaction ACRNtp at 0x2a1bc3a10>,\n",
       " <Reaction C30CPT1 at 0x2a1bc32d0>,\n",
       " <Reaction C3STKR1er at 0x2a1bc1e10>,\n",
       " <Reaction C3STKR2er at 0x2a17cab10>,\n",
       " <Reaction CRNtp at 0x2a1bc8150>,\n",
       " <Reaction CYOO6m at 0x2a18c2f10>,\n",
       " <Reaction CYOR_u9m at 0x15b829810>,\n",
       " <Reaction DNTPPA at 0x2a1841d90>,\n",
       " <Reaction DRTPPD at 0x2a1b92f90>,\n",
       " <Reaction FA160COAabcp at 0x2a1ba3990>,\n",
       " <Reaction FA183COAabcp at 0x2a1920c90>,\n",
       " <Reaction FA240tp at 0x2a1a134d0>,\n",
       " <Reaction GAPP at 0x16a228f90>,\n",
       " <Reaction H2Otm at 0x2a1782c90>,\n",
       " <Reaction OIVD2m at 0x2a1bbb6d0>,\n",
       " <Reaction PMDPHT at 0x2a1bc2350>,\n",
       " <Reaction SERPTer at 0x2a13c8f50>}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(wrong_rx2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cb538bd",
   "metadata": {},
   "source": [
    "{<Reaction ACRNtp at 0x2a19ea1d0>,\n",
    " <Reaction C30CPT1 at 0x2a19e9a90>,\n",
    " <Reaction C3STKR1er at 0x2a08e0a50>,\n",
    " <Reaction C3STKR2er at 0x2a19e8b10>,\n",
    " <Reaction CRNtp at 0x2a19ea8d0>,\n",
    " <Reaction CYOO6m at 0x2a10856d0>,\n",
    " <Reaction CYOR_u9m at 0x2a19bbad0>,\n",
    " <Reaction DNTPPA at 0x2a19df810>,\n",
    " <Reaction DRTPPD at 0x2a19c5710>,\n",
    " <Reaction FA160COAabcp at 0x2a1092590>,\n",
    " <Reaction FA183COAabcp at 0x2a176b410>,\n",
    " <Reaction FA240tp at 0x2a1769c90>,\n",
    " <Reaction GAPP at 0x2a19ead90>,\n",
    " <Reaction H2Otm at 0x2a090d450>,\n",
    " <Reaction OIVD2m at 0x2a19a6d90>,\n",
    " <Reaction PMDPHT at 0x2a19e92d0>,\n",
    " <Reaction SERPTer at 0x2a0bc76d0>}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c773beb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "43d95035",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g)))\n",
    "print(len(set(wrong_rx2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb0ae663",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1e3e514d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "39"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)-2083"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7c1ba35",
   "metadata": {},
   "source": [
    "39 reactions added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "be892257",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ICLp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Isocitrate lyase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a1395150</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>icit_x --> glx_x + succ_x</p>\n",
       "                    <p style='text-align:right'>Isocitrate --> Glyoxylate + Succinate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_112377</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ICLp at 0x2a1395150>"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('ICLp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "ebc40401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['H2Otm',\n",
       " 'DNTPPA',\n",
       " 'PMDPHT',\n",
       " 'ACRNtp',\n",
       " 'CRNtp',\n",
       " 'FA240tp',\n",
       " 'FA160COAabcp',\n",
       " 'DRTPPD']"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maintained_v01_gpr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "8cb3d038",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "39"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(added_to_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4085c270",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.06130800269480152</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>5.101E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>3.219E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0003782</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.86%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.03591</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.003784</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>3.721E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4047</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.898</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.01632</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.127E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001482</td>\n",
       "      <td>9</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0001905</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.03541</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.008992</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.006619</td>\n",
       "      <td>10</td>\n",
       "      <td>1.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.007228</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01279</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.03146</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.00375</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.009722</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.01963</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.01057</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.003498</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.005834</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.01446</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.02065</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.01337</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.00143</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.003823</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.01687</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3aap_c</td>\n",
       "      <td>DM_3aap_c</td>\n",
       "      <td>-0.04342</td>\n",
       "      <td>5</td>\n",
       "      <td>6.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-8.54E-05</td>\n",
       "      <td>5</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001482</td>\n",
       "      <td>15</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001482</td>\n",
       "      <td>10</td>\n",
       "      <td>0.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-3.126</td>\n",
       "      <td>1</td>\n",
       "      <td>87.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.0002626</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.514</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1403</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001536</td>\n",
       "      <td>1</td>\n",
       "      <td>0.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.008273</td>\n",
       "      <td>27</td>\n",
       "      <td>6.25%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a4114d50>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_v01.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b32caf35",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.06233019485004844</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>5.186E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>3.272E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0003845</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.84%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.03651</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.003847</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>3.783E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.001627</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4114</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.846</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.01659</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.006248</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.179E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001507</td>\n",
       "      <td>9</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0001937</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.009142</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.00673</td>\n",
       "      <td>10</td>\n",
       "      <td>1.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002136</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.007349</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01301</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.03199</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.003813</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.009884</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.01996</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.01074</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.003556</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.005932</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.0147</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.0136</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.001453</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.003887</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.01715</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3aap_c</td>\n",
       "      <td>DM_3aap_c</td>\n",
       "      <td>-0.04415</td>\n",
       "      <td>5</td>\n",
       "      <td>6.25%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-8.683E-05</td>\n",
       "      <td>5</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001507</td>\n",
       "      <td>15</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001507</td>\n",
       "      <td>10</td>\n",
       "      <td>0.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-3.078</td>\n",
       "      <td>1</td>\n",
       "      <td>87.16%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.000267</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.489</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1426</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001562</td>\n",
       "      <td>1</td>\n",
       "      <td>0.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.008411</td>\n",
       "      <td>27</td>\n",
       "      <td>6.43%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a43415d0>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4d45dd00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/kn/zzns_smn1q79xdnf1tgqk__r0000gn/T/tmpaoqfjyvs.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1904 rows, 4244 columns, 17194 nonzeros\n"
     ]
    }
   ],
   "source": [
    "model_remove_original_gpr = model.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da65d74b",
   "metadata": {},
   "source": [
    "##### removing the maintained reactions that were gapfilled leads to no growth. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "296a24f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with model_remove_original_gpr:\n",
    "for r in maintained_v01_gpr:\n",
    "    rxn = model_remove_original_gpr.reactions.get_by_id(r)\n",
    "    model_remove_original_gpr.remove_reactions([rxn])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5ec93cf4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.0</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>0.488</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>0.122</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e</td>\n",
       "      <td>-0.732</td>\n",
       "      <td>2</td>\n",
       "      <td>50.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>btd_RR_e</td>\n",
       "      <td>EX_btd_RR_e</td>\n",
       "      <td>-0.244</td>\n",
       "      <td>4</td>\n",
       "      <td>33.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-0.488</td>\n",
       "      <td>1</td>\n",
       "      <td>16.67%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-0.244</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.732</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a65cac10>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_remove_original_gpr.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "8599d0ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2114"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model_remove_original_gpr.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "694b0f8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2122"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "722a9704",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2083"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model_v01.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0f6bcb94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(maintained_v01_gpr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "bfd3d1de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H2Otm 9014\n",
      "DNTPPA 15385\n",
      "PMDPHT 11513\n",
      "ACRNtp 10149\n",
      "CRNtp 10149\n",
      "FA240tp 9912\n",
      "FA160COAabcp 13167 and 9637\n",
      "DRTPPD 15506\n"
     ]
    }
   ],
   "source": [
    "for r in (maintained_v01_gpr):\n",
    "    print(r, model.reactions.get_by_id(r).gene_reaction_rule)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e1a8be9",
   "metadata": {},
   "source": [
    "summary of above findings.\n",
    "\n",
    "8 reactions all were gap filled in order to function orginally. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84104014",
   "metadata": {},
   "source": [
    "# look for other orthologs from individual rto genes potentially left out. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "28144c3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "277\n"
     ]
    }
   ],
   "source": [
    "# this cell block looks at the rto model and determines if there are other orthologs that can be added to the lst GSM. \n",
    "# previous code only looked at reactions that were in multiple models (i.e., in the rto and ylip model). \n",
    "\n",
    "# counter. \n",
    "count=0\n",
    "\n",
    "# storage variables. \n",
    "ortholog_available = []\n",
    "no_ortholog = []\n",
    "no_rxn = []\n",
    "genes = []\n",
    "\n",
    "\n",
    "# obtain list of rto, model reactions.\n",
    "rto_rxns = list(rto.reactions)\n",
    "rto_rxns = [x.id for x in rto_rxns]\n",
    "model_rxns = list(model.reactions)\n",
    "model_rxns = [x.id for x in model_rxns]\n",
    "\n",
    "# dictionary with keys of rto genes, values of Lst genes. \n",
    "lst_dict_keys = list(lst_rto_dict.keys())\n",
    "\n",
    "# determine how many reactions were not added to the model. \n",
    "diff_rxns = set(rto_rxns) - set(model_rxns)\n",
    "print(len(diff_rxns))\n",
    "\n",
    "# check the gpr of the reactions. \n",
    "for r in diff_rxns:\n",
    "    \n",
    "    # rxn is the rto object.\n",
    "    rxn = rto.reactions.get_by_id(r)\n",
    "\n",
    "    # gpr is the rto gene reaction rule. \n",
    "    gpr = rxn.genes\n",
    "    \n",
    "    \n",
    "    # temporary storage for new orthologs. \n",
    "    temp_gpr = []\n",
    "\n",
    "    # tracking orthologs. \n",
    "    count_ortho=0\n",
    "    \n",
    "    \n",
    "    # determine the number of orthologs for the GPR.\n",
    "    for i in range(len(gpr)):\n",
    "        \n",
    "        # obtain the individual genes. \n",
    "        g=list(gpr)[i]\n",
    "\n",
    "        # evaluate if there are any orthologs lst_orthologs. \n",
    "        if g.id in lst_dict_keys:\n",
    "\n",
    "            # obtain the lst gene ortholog.\n",
    "            value = lst_rto_dict[g.id]\n",
    "\n",
    "            # check if the ortholog is blank. \n",
    "            if value:\n",
    "                temp_gpr.append(value)\n",
    "                count_ortho+=1\n",
    "                    \n",
    "    # if there were no orthologs. \n",
    "    if count_ortho==0:\n",
    "        no_ortholog.append(r)\n",
    "    \n",
    "    # there was at least one ortholog for the reaction. \n",
    "    elif count_ortho>0:\n",
    "        # append the reaction to this list. \n",
    "        ortholog_available.append(r)\n",
    "        # append the gpr to this list. \n",
    "        genes.append(temp_gpr)\n",
    "        \n",
    "        # make a GPR rule for the orthologs. \n",
    "        temp = \" or \".join(sub_item.replace(\"None\", \"\").replace(\"NA\",\" \") for sub_item in temp_gpr)\n",
    "        \n",
    "#         print(temp, '\\n',temp_gpr)\n",
    "        \n",
    "        # add the reaction and GPR to the Lst model. \n",
    "        model.add_reactions([rxn.copy()])\n",
    "        model.reactions.get_by_id(r).gene_reaction_rule = temp\n",
    "\n",
    "        # track the number of reactions added. \n",
    "        count+=1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "5a07e92d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "0574f1bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "dea0e0a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15762 has no associated reaction\n",
      "13348 has no associated reaction\n",
      "14170 has no associated reaction\n",
      "12742 has no associated reaction\n",
      "9700 has no associated reaction\n",
      "10293 has no associated reaction\n",
      "13228 has no associated reaction\n",
      "11362 has no associated reaction\n",
      "12752 has no associated reaction\n",
      "11907 has no associated reaction\n",
      "13813 has no associated reaction\n",
      "12570 has no associated reaction\n",
      "14805 has no associated reaction\n",
      "11203 has no associated reaction\n",
      "8885 has no associated reaction\n",
      "12116 has no associated reaction\n",
      "9274 has no associated reaction\n",
      "10704 has no associated reaction\n",
      "10705 has no associated reaction\n",
      "14070 has no associated reaction\n",
      "12867 has no associated reaction\n",
      "15694 has no associated reaction\n",
      "16404 has no associated reaction\n",
      "9113 has no associated reaction\n",
      "11702 has no associated reaction\n",
      "14078 has no associated reaction\n",
      "10607 has no associated reaction\n",
      "13017 has no associated reaction\n",
      "10017 has no associated reaction\n",
      "8446 has no associated reaction\n",
      "8444 has no associated reaction\n",
      "NAD6 has no associated reaction\n",
      "9144 has no associated reaction\n",
      "15864 has no associated reaction\n",
      "15998 has no associated reaction\n",
      "NAD2 has no associated reaction\n",
      "10541 has no associated reaction\n",
      "12497 has no associated reaction\n",
      "15947 has no associated reaction\n",
      "8817 has no associated reaction\n",
      "8909 has no associated reaction\n",
      "NAD3 has no associated reaction\n",
      "12653 has no associated reaction\n",
      "14717 has no associated reaction\n",
      "14418 has no associated reaction\n",
      "13656 has no associated reaction\n",
      "13411 has no associated reaction\n",
      "8699 has no associated reaction\n",
      "11857 has no associated reaction\n",
      "14900 has no associated reaction\n",
      "15973 has no associated reaction\n",
      "12543 has no associated reaction\n",
      "9348 has no associated reaction\n",
      "10010 has no associated reaction\n",
      "NAD4 has no associated reaction\n",
      "11151 has no associated reaction\n",
      "12482 has no associated reaction\n",
      "11411 has no associated reaction\n",
      "14261 has no associated reaction\n",
      "NAD1 has no associated reaction\n",
      "15179 has no associated reaction\n",
      "13268 has no associated reaction\n",
      "10318 has no associated reaction\n",
      "8530 has no associated reaction\n",
      "13925 has no associated reaction\n",
      "NAD4L has no associated reaction\n",
      "13214 has no associated reaction\n",
      "NAD5 has no associated reaction\n",
      "71\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g)))\n",
    "print(len(set(wrong_rx2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "bcb33d9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACRNtp',\n",
       " 'C30CPT1',\n",
       " 'C3STKR1er',\n",
       " 'C3STKR2er',\n",
       " 'CRNtp',\n",
       " 'CYOO6m',\n",
       " 'CYOR_u9m',\n",
       " 'DNTPPA',\n",
       " 'DRTPPD',\n",
       " 'FA160COAabcp',\n",
       " 'FA183COAabcp',\n",
       " 'FA240tp',\n",
       " 'GAPP',\n",
       " 'H2Otm',\n",
       " 'OIVD2m',\n",
       " 'PMDPHT',\n",
       " 'SERPTer'}"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2 = [x.id for x in wrong_rx2]\n",
    "(set(wrong_rx2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc8c542d",
   "metadata": {},
   "source": [
    "## above cell added 26 reactions that had orthologs but were not previously included. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "ac97c172",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "251"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(no_ortholog)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "e67363bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ortholog_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "65f953c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2148"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "bea60e21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2083"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model_v01.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "47478a16",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)-2083"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "815f030a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.07253593170088728</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>6.035E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>3.808E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0004475</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.04248</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.004477</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>4.403E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.001893</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4788</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.329</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.01931</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.007271</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.699E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001754</td>\n",
       "      <td>9</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0002254</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.04189</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.01064</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.007832</td>\n",
       "      <td>10</td>\n",
       "      <td>1.29%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.01431</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002486</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.008552</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01514</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.03723</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.004437</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.0115</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.02323</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.004138</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.01711</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.02444</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.01582</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.001691</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.004524</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.01996</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3aap_c</td>\n",
       "      <td>DM_3aap_c</td>\n",
       "      <td>-0.05137</td>\n",
       "      <td>5</td>\n",
       "      <td>8.21%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-0.000101</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001754</td>\n",
       "      <td>15</td>\n",
       "      <td>0.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001754</td>\n",
       "      <td>10</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-2.599</td>\n",
       "      <td>1</td>\n",
       "      <td>83.12%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.0003107</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.242</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1659</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001818</td>\n",
       "      <td>1</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.009788</td>\n",
       "      <td>27</td>\n",
       "      <td>8.45%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a66a7c10>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0e6c64c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.06130800269480152</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>5.101E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>3.219E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0003782</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.86%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.03591</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.003784</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>3.721E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4047</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.898</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.01632</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.127E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001482</td>\n",
       "      <td>9</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0001905</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.03541</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.008992</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.006619</td>\n",
       "      <td>10</td>\n",
       "      <td>1.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.0121</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.007228</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01279</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.03146</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.00375</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.009722</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.01963</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.01057</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.003498</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.005834</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.01446</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.02065</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.01337</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.00143</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.003823</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.01687</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3aap_c</td>\n",
       "      <td>DM_3aap_c</td>\n",
       "      <td>-0.04342</td>\n",
       "      <td>5</td>\n",
       "      <td>6.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-8.54E-05</td>\n",
       "      <td>5</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001482</td>\n",
       "      <td>15</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001482</td>\n",
       "      <td>10</td>\n",
       "      <td>0.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-3.126</td>\n",
       "      <td>1</td>\n",
       "      <td>87.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.0002626</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.514</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1403</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001536</td>\n",
       "      <td>1</td>\n",
       "      <td>0.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.008273</td>\n",
       "      <td>27</td>\n",
       "      <td>6.25%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a68181d0>"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_v01.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6c223e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "14188356",
   "metadata": {},
   "source": [
    "# look at sce and ylip to determine individual reactions that could be added. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc66e2ec",
   "metadata": {},
   "source": [
    "following cell does not add reactions - finds sce individual reactions that have orthologs for Lst."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "7180d130",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3234\n"
     ]
    }
   ],
   "source": [
    "# this cell now looks at the sce model to find individual reactions that have an Lst gene ortholog. \n",
    "# this cell does not modify the model but rather tracks the reactions that have potential for additions. \n",
    "\n",
    "# track objects. \n",
    "sce_ortholog_available = []\n",
    "no_ortholog = []\n",
    "no_rxn = []\n",
    "genes = []\n",
    "\n",
    "\n",
    "# obtain list of sce and model reactions. \n",
    "sce_rxns = list(sce.reactions)\n",
    "sce_rxns = [x.id for x in sce_rxns]\n",
    "model_rxns = list(model.reactions)\n",
    "model_rxns = [x.id for x in model_rxns]\n",
    "\n",
    "# obtain keys of the sce:lst dictionary. Keys are the sce genes. \n",
    "sce_dict_keys = list(lst_sce_dict.keys())\n",
    "\n",
    "# determine the reactions that have not yet been added to the model.\n",
    "diff_rxns = set(sce_rxns) - set(model_rxns)\n",
    "\n",
    "print(len(diff_rxns))\n",
    "\n",
    "# iterate through the not added reactions.\n",
    "for r in diff_rxns:\n",
    "    \n",
    "    # rxn is the sce reaction IDs.\n",
    "    rxn = sce.reactions.get_by_id(r)\n",
    "\n",
    "    # gpr are the sce reaction genes. \n",
    "    gpr = rxn.genes\n",
    "    \n",
    "    # holder for the Lst orthologs. \n",
    "    temp_gpr = []\n",
    "\n",
    "\n",
    "    count_ortho=0\n",
    "        \n",
    "    # determine the number of orthologs for the GPR.\n",
    "    for i in range(len(gpr)):\n",
    "        \n",
    "        # g is the individual genes in the rxns. \n",
    "        g=list(gpr)[i]\n",
    "\n",
    "        # examine lst_orthologs. \n",
    "        if g.id in sce_dict_keys:\n",
    "\n",
    "            # obtain the lst gene ortholog.\n",
    "            value = lst_sce_dict[g.id]\n",
    "\n",
    "            # check if the ortholog is blank. \n",
    "            if value:\n",
    "                temp_gpr.append(value)\n",
    "                count_ortho+=1\n",
    "                    \n",
    "    # if there were no orthologs. \n",
    "    if count_ortho==0:\n",
    "        no_ortholog.append(r)\n",
    "    \n",
    "    # if one ortholog exist for a reaction, append it sce_ortholog_available list. \n",
    "    elif count_ortho>0:\n",
    "        sce_ortholog_available.append(r)\n",
    "        genes.append(temp_gpr)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "02e7fb7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1668"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sce_ortholog_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "b1a12ae8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2148"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "651b91e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2148-2083"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b6878f",
   "metadata": {},
   "source": [
    "potential to add 1668 more reactions. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0c29e8d",
   "metadata": {},
   "source": [
    "#### ylip orthologs. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73879644",
   "metadata": {},
   "source": [
    "following cell does not add reactions - finds ylip reactions that have orthologs to Lst.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1571d7df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "633\n"
     ]
    }
   ],
   "source": [
    "# this cell now looks at the ylip model to find individual reactions that have an Lst gene ortholog. \n",
    "# this cell does not modify the model but rather tracks the reactions that have potential for additions. \n",
    "\n",
    "# track objects.\n",
    "ylip_ortholog_available = []\n",
    "no_ortholog = []\n",
    "no_rxn = []\n",
    "genes = []\n",
    "\n",
    "# obtain the ylip and model reaction lists. \n",
    "ylip_rxns = list(ylip.reactions)\n",
    "ylip_rxns = [x.id for x in ylip_rxns]\n",
    "model_rxns = list(model.reactions)\n",
    "model_rxns = [x.id for x in model_rxns]\n",
    "\n",
    "# obtain keys of the ylip:lst dictionary. Keys are the ylip genes.\n",
    "ylip_dict_keys = list(lst_ylip_dict.keys())\n",
    "\n",
    "# determine the reactions that are in the ylip GSM but not in the Lst GSM model. \n",
    "diff_rxns = set(ylip_rxns) - set(model_rxns)\n",
    "print(len(diff_rxns))\n",
    "\n",
    "# iterate through the reactions not included. \n",
    "for r in diff_rxns:\n",
    "    \n",
    "    # rxn holds the ylip reactions. \n",
    "    rxn = ylip.reactions.get_by_id(r)\n",
    "    \n",
    "    # gpr holds the ylip reaction genes. \n",
    "    gpr = rxn.genes\n",
    "    \n",
    "    # list object to hold the Lst orthologs. \n",
    "    temp_gpr = []\n",
    "\n",
    "    count_ortho=0\n",
    "        \n",
    "    # determine the number of orthologs for the GPR.\n",
    "    for i in range(len(gpr)):\n",
    "        \n",
    "        # g is the individual ylip genes in the reaction. \n",
    "        g=list(gpr)[i]\n",
    "\n",
    "        # examine lst_orthologs. \n",
    "        if g.id in ylip_dict_keys:\n",
    "\n",
    "            # obtain the lst gene ortholog.\n",
    "            value = lst_ylip_dict[g.id]\n",
    "\n",
    "            # check if the ortholog is blank. \n",
    "            if value:\n",
    "                temp_gpr.append(value)\n",
    "                count_ortho+=1\n",
    "                    \n",
    "    # if there were no orthologs. \n",
    "    if count_ortho==0:\n",
    "        no_ortholog.append(r)\n",
    "    \n",
    "    # append reactions for which there is one Lst ortholog to the ylip_ortholog_available list. \n",
    "    elif count_ortho>0:\n",
    "        ylip_ortholog_available.append(r)\n",
    "        genes.append(temp_gpr)\n",
    "        \n",
    "#         temp = \" or \".join(sub_item.replace(\"None\", \"\").replace(\"NA\",\" \") for sub_item in temp_gpr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "0682a52e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)-2083"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "70d7416a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "260"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ylip_ortholog_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "de2b71af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1668"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sce_ortholog_available)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "524c1ed0",
   "metadata": {},
   "source": [
    "There are 260 potential ylip reactions that can be added. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ebab549",
   "metadata": {},
   "source": [
    "## following section seeks to determine if there are reactions in the individual sce or ylip GSMs that are in the Lst draft GSM, but may have had seperate ID conventions. \n",
    "\n",
    "i.e.\n",
    "\n",
    "- ReacR1 = a+b --> c\n",
    "- ReacR_1 = a+b --> c\n",
    "\n",
    "The two reactions are the same but would not have turned up the same in the previous iteration. \n",
    "\n",
    "Now we try to match the naming conventions instead of IDs to identify these reactions. \n",
    "\n",
    "i.e., \n",
    "- ReacR1 = a+b --> c : name = Reaction 1\n",
    "- ReacR_1 = a+b --> c : name = Reaction 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "d891665e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell gathers the reactions that have orthologs from the sce model that will be used for making a dictionary. \n",
    "\n",
    "# list holders for reaction details. \n",
    "_names = []\n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "\n",
    "\n",
    "# gather the reaction details from the possible ortholog reactions. \n",
    "for rx in sce_ortholog_available:\n",
    "    r = sce.reactions.get_by_id(rx)\n",
    "    _names.append(r.name)\n",
    "    _gpr.append(r.gene_reaction_rule)\n",
    "    _id.append(r.id)\n",
    "    _stoich.append(r.reaction)\n",
    "    \n",
    "# store the names here. \n",
    "sce_rxns_name = _names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba0f61bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "14347ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here we obtain the model reactions and names. \n",
    "model_rxns_name = model.reactions\n",
    "model_rxns_id = [x.id for x in model_rxns_name]\n",
    "model_rxns_name = [x.name for x in model_rxns_name] \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "909e8332",
   "metadata": {},
   "source": [
    "now we check how many names match the sce and model reactions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "21f31621",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "source": [
    "temp = []\n",
    "for r in sce_rxns_name:\n",
    "    if r in model_rxns_name:\n",
    "        temp.append(r)\n",
    "print(len(temp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "fd1345f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C-3 sterol dehydrogenase (4-methylzymosterol)', '3-mercaptopyruvate sulfurtransferase', '3-methyl-2-oxobutanoate hydroxymethyltransferase', '2-keto-4-methylthiobutyrate transamination', 'C-14 sterol reductase', 'Phytosphingosine phosphate lyase', 'UDP-N-acetylglucosamine diphosphorylase', 'C-4 sterol methyl oxidase (4,4-dimethylzymosterol)', 'NAD kinase', 'GTP cyclohydrolase II', 'fatty-acid--CoA ligase (decanoate)', 'NAD diphosphatase']\n"
     ]
    }
   ],
   "source": [
    "print(temp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f48cac35",
   "metadata": {},
   "source": [
    "## match reaction names so we can add orthologs even if reaction IDs do not match.\n",
    "\n",
    "Many reaction names have different writing conventions (i.e., spaces, dashes, captilization). We will use the python 're' library to replace many of those names to try and match reactions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "aaeb4de5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "786d7d83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace the model reaction special characters and make all characters lower case. \n",
    "model_rxns_name = [x.replace(\" \",\"\").replace('-','').replace(':',\"\").replace(',',\"\").replace(')',\"\").replace(')',\"\").replace('\\'',\"\").replace('mitochondrialmembrane',\"\").lower() for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'[^a-zA-Z0-9]', '', x) for x in model_rxns_name]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e4693792",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1668"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sce_rxns_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "eae755b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4131"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sce.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "33f092e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace the sce reaction name special characters and make all characters lower case. \n",
    "sce_rxns_name = [re.sub(r' c$',\"\",x) for x in sce_rxns_name]\n",
    "sce_rxns_name = [re.sub(r' v$',\"\",x) for x in sce_rxns_name]\n",
    "\n",
    "sce_rxns_name = [x.replace(\" \",\"\").replace('-','').replace(':',\"\").replace(',',\"\").replace(')',\"\").replace(')',\"\").replace('\\'',\"\").replace('+',\"\").replace('mitochondrialmembrane',\"\").lower() for x in sce_rxns_name]\n",
    "\n",
    "sce_rxns_name = [re.sub(r'[^a-zA-Z0-9]', '', x) for x in sce_rxns_name]\n",
    "\n",
    "# the yeast-8 scaffold model has many membrane/envelope compartments. \n",
    "# Here we will replace those names for initial matching purposes.\n",
    "sce_rxns_name = [x.replace('mitochondrialmembrane',\"\").replace('golgimembrane',\"\").replace('ermembrane',\"\") for x in sce_rxns_name]\n",
    "sce_rxns_name = [x.replace('cellenvelopecytoplasm',\"\").replace('vacuolarmembrane',\"\") for x in sce_rxns_name]\n",
    "sce_rxns_name = [re.sub(r'mitochondrial$',\"\",x) for x in sce_rxns_name]\n",
    "sce_rxns_name = [re.sub(r'mitochondria$',\"\",x) for x in sce_rxns_name]\n",
    "sce_rxns_name = [re.sub(r'cellenvelope$',\"\",x) for x in sce_rxns_name]\n",
    "sce_rxns_name = [re.sub(r'cytoplasm$',\"\",x) for x in sce_rxns_name]\n",
    "sce_rxns_name = [re.sub(r'cytoplasmic$',\"\",x) for x in sce_rxns_name]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "cf6e006b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mitochondrialalcoholdehydrogenase\n",
      "glutamyltrnaglnamidotransferasesubunitbmitochondrialgluadtsubunitbec635cytochromecoxidaseassemblyfactorpet112\n",
      "magnesiumtransportermrs2mitochondrialrnasplicingproteinmrs2\n",
      "ironsulfurclusterassemblyprotein2mitochondrialironsulfurclusterscaffoldprotein2\n",
      "putativelipoateproteinligaseaec63120alteredinheritancerateofmitochondriaprotein22\n"
     ]
    }
   ],
   "source": [
    "for r in sce_rxns_name:\n",
    "    if 'mitochondria' in r:\n",
    "        print(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "19c6c6e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for r in sce_rxns_name:\n",
    "    if 'cytoplasm' in r:\n",
    "        print(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "8232d6f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1668"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sce_rxns_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52ad5c78",
   "metadata": {},
   "source": [
    "potentially 1386 reactions can be added from sce. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "40987b7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1386"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "len(set(sce_rxns_name))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f3c133f",
   "metadata": {},
   "source": [
    "however, many reactions have the same name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "67e2f450",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'ipsphospholipasec' found at positions [54, 66, 84, 119, 135, 156, 166, 235, 300, 322, 325, 421, 429, 462, 534, 536, 553, 587, 593, 595, 617, 649, 682, 690, 716, 781, 815, 827, 947, 978, 980, 982, 1177, 1186, 1189, 1194, 1198, 1200, 1202, 1203, 1212, 1220, 1245, 1255, 1264, 1299, 1324, 1328, 1339, 1369, 1410, 1425, 1435, 1470, 1512, 1520, 1552, 1580, 1587, 1654] in the list.\n"
     ]
    }
   ],
   "source": [
    "phrase = 'ipsphospholipasec'\n",
    "try:\n",
    "    position = sce_rxns_name.index(phrase)\n",
    "    positions = [i for i, x in enumerate(sce_rxns_name) if x == phrase]\n",
    "    if positions:\n",
    "        print(f\"'{phrase}' found at positions {positions} in the list.\")\n",
    "#     print(f\"'{phrase}' found at position {position} in the list.\")\n",
    "except ValueError:\n",
    "    print(f\"'{phrase}' not found in the list.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "584361e5",
   "metadata": {},
   "source": [
    "## but first, lets check how many reactions are already in the model. \n",
    "\n",
    "88 reactions are already in model and need updated GPRs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "d17160e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88\n"
     ]
    }
   ],
   "source": [
    "temp = []\n",
    "for r in sce_rxns_name:\n",
    "    if r in model_rxns_name:\n",
    "        temp.append(r)\n",
    "print(len(temp))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "556be29d",
   "metadata": {},
   "source": [
    "88 reactions are already in model and need updated GPRs for the sce genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "4c179b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "## this cell removes more of the unique identifiers from the model reaction names. \n",
    "model_rxns_name = [re.sub(r' c$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r' v$',\"\",x) for x in model_rxns_name]\n",
    "\n",
    "model_rxns_name = [x.replace(\" \",\"\").replace('-','').replace(':',\"\").replace(',',\"\").replace(')',\"\").replace(')',\"\").replace('\\'',\"\").replace('+',\"\").replace('mitochondrialmembrane',\"\").lower() for x in model_rxns_name]\n",
    "\n",
    "model_rxns_name = [re.sub(r'[^a-zA-Z0-9]', '', x) for x in model_rxns_name]\n",
    "\n",
    "# model_rxns_name = [x.replace('mitochondrialmembrane',\"\").replace('golgimembrane',\"\").replace('ermembrane',\"\") for x in model_rxns_name]\n",
    "model_rxns_name = [x.replace('vacuolar',\"\").replace('vacuolarmembrane',\"\") for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'mitochondrial$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'mitochondria$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cellenvelope$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cytoplasm$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cytoplasmic$',\"\",x) for x in model_rxns_name]\n",
    "\n",
    "model_rxns_name = [re.sub(r'peroxisomal$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cytoplasmic$',\"\",x) for x in model_rxns_name]\n",
    "\n",
    "# sce_rxns_name = [x.replace('mitochondrialmembrane',\"\").replace('golgimembrane',\"\").replace('ermembrane',\"\") for x in sce_rxns_name]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "110dfd83",
   "metadata": {},
   "source": [
    "correction. 102 reactions need updated GPRs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "4a415be1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110\n",
      "102\n"
     ]
    }
   ],
   "source": [
    "temp = []\n",
    "for r in sce_rxns_name:\n",
    "    if r in model_rxns_name:\n",
    "        temp.append(r)\n",
    "print(len(temp))\n",
    "print(len(set(temp)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "9e77ae96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary of the model rxn names to the original ids. \n",
    "model_dict_sce = dict(zip(model_rxns_name,model_rxns_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "8c84c84a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aldehydedehydrogenase2phenylethanolnadp': 'ALCD25yi',\n",
       " 'methenyltetrahydrifikatecyclohydrolase': 'MTHFCm',\n",
       " 'ampnucleosidase': 'AMPN',\n",
       " 'diacylglycerolcholinephosphotransferaseyeastspecific': 'DAGCPTer_RT',\n",
       " 'pyruvatetransportinviaprotonsymport': 'PYRt2',\n",
       " 'nicotinatenucleotidediphosphorylasecarboxylating': 'NNDPR',\n",
       " 'hydroxymethylglutarylcoasynthase': 'HMGCOAS',\n",
       " '35cyclicnucleotidephosphodiesterase': 'PDE3',\n",
       " 'phosphoadenylylsulfatereductasethioredoxin': 'PAPSR',\n",
       " 'acylcoaoxidasedecanoylcoa': 'ACOAO4p',\n",
       " '3isopropylmalate3methyltransferase': '3IPM3MT',\n",
       " 'udpglucose6dehydrogenase': 'UDPGD',\n",
       " 'cytidinekinasegtp': 'CYTDK2',\n",
       " 'fattyacidcoaligasetetradecenoate': 'FACOAL141p',\n",
       " 'acetylcoaacyltransferasetetradecanoylcoa': 'ACACT7p',\n",
       " 'polyamineoxidase': 'POLYAO',\n",
       " 'mevalonatekinaseatp': 'MEVK1',\n",
       " 'datpuridine5phosphotransferase': 'DATUP',\n",
       " 'taurinedioxygenase': 'TAUDO',\n",
       " 'citratesynthase': 'CSm',\n",
       " 'phosphoserinetransaminase': 'PSERT',\n",
       " 'guanylatekinasegmpdatp': 'GK2',\n",
       " 'lcysteinereversibletransportviaprotonsymport': 'CYSt2r',\n",
       " 'hatpase': 'ATPS2v',\n",
       " 'sphinganinephosphatelyase': 'SPHPLer',\n",
       " 'phospholipasebphosphatidylcholineyeastspecificextracellular': 'PLBPCe_RT',\n",
       " 'aldehydedehydrogenasephenylacetaldehydenad': 'ALDD19xr',\n",
       " 'glutathionehydralaseperiplasmic': 'GTHRDH_syn',\n",
       " 'nucleosidediphosphatekinaseatpdcdp': 'NDPK7',\n",
       " 'propane12diolnadp1oxidoreductase': 'PPDOy',\n",
       " 'glyoxylatereductasenadp': 'GLYCLTDy',\n",
       " 'deoxyguanylatekinasedgmpatp': 'DGK1',\n",
       " 'sphingolipidlongchainbasekinasephytosphingosine': 'SLCBK2er',\n",
       " 'ceramide1synthase26cendoplasmicreticular': 'CERS126er',\n",
       " 'inositol145triphosphate6kinasenucleus': 'MI145P6Kn',\n",
       " 'adentylatekinasegtp': 'ADK3m',\n",
       " 'dimethylallyltranstransferase': 'DMATT',\n",
       " 'phosphomethylpyrimidinekinase': 'PMPK',\n",
       " 'lactoylglutathionelyase': 'LGTHL',\n",
       " 'purinenucleosidephosphorylaseguanosine': 'PUNP3',\n",
       " 'prephenatedehydratase': 'PPNDH',\n",
       " 'glycyltrnasynthetase': 'GLYTRS',\n",
       " '3hydroxyacylcoadehydrogenase3oxodecanoylcoa': 'HACD5p',\n",
       " 'glycerol3phosphateacyltransferaseglyceronephosphateyeastspecific': 'GAT2er_RT',\n",
       " 'cdpdiacylglycerolsynthetaseyeastspecific': 'DASYNm_RT',\n",
       " 's3hydroxybutanoylcoahydrolyase': 'ECOAH1p',\n",
       " 'kynureninase': 'KYN',\n",
       " 'histidinolphosphatetransaminase': 'HSTPT',\n",
       " 'ironiitransport': 'FE2tm',\n",
       " '3methyl2oxopentanoatedecarboxylase': '3MOPDC',\n",
       " 'peroxisomalacylcoathioesterase': 'PTE12x',\n",
       " 'glutamyltrnasynthetase': 'GLUTRSm',\n",
       " 'phosphoribosylaminoimidazolecarboxamideformyltransferase': 'AICART',\n",
       " '42aminoethyl12benzenedioloxygenoxidoreductasedeaminatingflavincontaining': '42A12BOOX',\n",
       " 'triosephosphateisomerase': 'TPI',\n",
       " 'coatransportermitochondrialirreversible': 'COAtim',\n",
       " 'nicotinatemononucleotideadenylyltransferase': 'NADS2',\n",
       " 'acetylcoacacetyltransferaseoctanoylcoa': 'ACACT4p',\n",
       " 'histidyltrnasynthetase': 'HISTRSm',\n",
       " 'phenylalanyltrnasynthetase': 'PHETRS',\n",
       " '2oxo4methyl3carboxypentanoatedecarboxylation': 'OMCDCm',\n",
       " 'dtmpkinase': 'DTMPK',\n",
       " 'aldehydedehydrogenaseisobutylalcoholnadp': 'ALCD23yi',\n",
       " 'fattyacidcoaligasehexadecanoate': 'FACOAL160',\n",
       " '6phosphogluconolactonase': 'PGLp',\n",
       " 'lthreoninedeaminase': 'THRD_L',\n",
       " 'betaglucosidase': 'BG_MBDG',\n",
       " 'tryptophanyltrnasynthetase': 'TRPTRSm',\n",
       " 'sadenosyllmethioninetransportinviaprotonsymport': 'AMETt2',\n",
       " 'osuccinylhomoserinelyaselcysteine': 'SHSL1',\n",
       " 'oacetylhomoserinethiollyase': 'AHSERL2',\n",
       " 'acetatetransporter': 'ACtr',\n",
       " 'cytochromep450lanosterol14alphademethylasenad': 'LNS14DMrx',\n",
       " '2methylacylcoadehydrogenasetrans2methylbut2enoylcoaforming': 'ACOAD10m',\n",
       " 'thymidinephosphorylase': 'TMDPP',\n",
       " 'inositoltransportinviaprotonsymport': 'INSTt2',\n",
       " 'inositolphosphorylceramideceramide224cphospholipasec': 'IPC224PLCg_RT',\n",
       " 'dctpcytidine5phosphotransferase': 'DCTCP',\n",
       " 'hexokinasedfructoseatp': 'HEX7',\n",
       " 'homoisocitratedehydrogenase': 'HICITDm',\n",
       " 'malatetransport': 'MALtm',\n",
       " 'amidase': 'AMID3',\n",
       " 'pantothenatekinase': 'PNTK',\n",
       " 'fattyacidoxidationoctanoylcoa': 'FAO80p',\n",
       " 'adenosinedeaminase': 'ADA',\n",
       " 'glutamatedehydrogenasenad': 'GLUDxi',\n",
       " 'uraciltransportinviaprotonsymport': 'URAt2',\n",
       " 'ergosterolreversibletransport': 'ERGSTt',\n",
       " 'carnitineoacetyltransferasereversedirection': 'CSNATp',\n",
       " 'udpglucose4epimerase': 'UDPG4E',\n",
       " 'shikimatedehydrogenase': 'SHK3Dr',\n",
       " 'succinatesemialdehydedehydrogenasenadp': 'SSALy',\n",
       " 'cysteinedesulfhydrase': 'CYSDS',\n",
       " 'ribokinase': 'RBK',\n",
       " '3methyl2oxobutanoatehydroxymethyltransferase': 'MOHMT',\n",
       " 'inositolphosphorylceramideceramide324cphospholipasec': 'IPC324PLCg_RT',\n",
       " 'phosphoribosylglycinamideformyltransferase': 'GARFT',\n",
       " 'pyridoxaminekinase': 'PYDAMK',\n",
       " '3hydroxylkynureninehydrolase': 'HKYNH',\n",
       " 'datpcytidine5phosphotransferase': 'DATCY',\n",
       " 'atpnicotinamidenucleotideadenylyltransferasenucleus': 'NNATn',\n",
       " 'methionyltrnasynthetase': 'METTRS',\n",
       " 'sformylglutathionehydralase': 'SFGTHi',\n",
       " 'inositolphosphoryceramidesynthaseceramide226cyeastspecific': 'IPCS226g_RT',\n",
       " 'acidphosphataseextracellularsecreted': 'ACP1e',\n",
       " 'formatetetrahydrofolateligase': 'FTHFLi',\n",
       " 'aspartatesemialdehydedehydrogenase': 'ASAD',\n",
       " '3hydroxyacylcoadehydratase3hydroxyhexadecanoylcoa': 'ECOAH7p',\n",
       " 'fattyacidcoaligasetetradecanoate': 'FACOAL140',\n",
       " 'enoylcoahydratase3hydroxy2methylbutyrylcoaforming': 'ECOAH9m',\n",
       " 'hypoxanthinephosphoribosyltransferasehypoxanthine': 'HXPRT',\n",
       " 'c22steroldesaturasenad': 'C22STDSrx',\n",
       " 'nitrilase': 'NTRLASE',\n",
       " 'dihydroxyaciddehydratase23dihydroxy3methylpentanoate': 'DHAD2m',\n",
       " 'udpglucoseergosterolglucosyltransferase': 'ERGSTGLCT',\n",
       " 'fattyacidcoaligaseoctadecynoate': 'FACOAL182p',\n",
       " 'pyrroline5carboxylatereductase': 'P5CR',\n",
       " 'phytosphingosinephosphatelyase': 'PSPHPLer',\n",
       " 'myoinositol1phosphatase': 'MI1PP',\n",
       " 'sodiumprotonantiporterhnais11': 'NAt3_1g',\n",
       " 'nucleosidediphosphatekinaseatpdidp': 'NDPK10',\n",
       " '3phospho5adenylylsulfate3phosphohydrolase': 'BPNT2',\n",
       " 'coproporphyrinogenoxidaseo2required': 'CPPPGO',\n",
       " 'putrescinetransportinviaprotonantiportirreversible': 'PTRCt3i',\n",
       " 'naprtaserev': 'NAPRT',\n",
       " 'mannoseinositolp2ceramideceramide324cphospholipasec': 'MIP2C324PLCg_RT',\n",
       " 'alcoholdehydrogenaseethanol': 'ALCD2x',\n",
       " 'nucleosidediphosphatekinaseatpcdp': 'NDPK3',\n",
       " 'lanosterolreversibletransport': 'LANOSTt',\n",
       " 'fattyacidcoaligasedodecanoate': 'FACOAL120p',\n",
       " 'enoylcoahydratasec120': 'ECOAH5m',\n",
       " 'c4sterolmethyloxidase4methylzymosterol': 'C4STMO2er',\n",
       " 'uridylatekinasedump': 'URIDK2r',\n",
       " 'nucleosidetriphosphatepyrophosphorylaseitp': 'NTPP9',\n",
       " 'ceramide2psynthase24c': 'CERS2p24er',\n",
       " 'allophanatehydrolase': 'ALPHNH',\n",
       " 'deoxyuridinephosphorylase': 'DURIPP',\n",
       " 'glutamatedehydrogenasenadp': 'GLUDy',\n",
       " 'homoserineotransacetylase': 'HSERTA',\n",
       " 'citratetransport': 'CITtcm',\n",
       " 'fmnadenylyltransferase': 'FMNATm',\n",
       " 'tyrosyltrnasynthetase': 'TYRTRSm',\n",
       " 'isoleucyltrnasynthetase': 'ILETRS',\n",
       " 'aspartatekinase': 'ASPK',\n",
       " 'diphthinesynthase': 'DIPS',\n",
       " 'udpgalactosetransportgolgiapparatus': 'UDPGALt2g',\n",
       " 'llactaldehydenadp1oxidoreductase': 'LALDO3',\n",
       " 'homoserinekinase': 'HSK',\n",
       " 'gtpgdptranslocasemitochondrialelectroneutral': 'GTPt2m',\n",
       " 'phosphopantothenoylcysteinedecarboxylase': 'PPCDC',\n",
       " 'atpadenylyltransferase': 'ATPATF3',\n",
       " 'phosphoribosylatppyrophosphatase': 'PRATPP',\n",
       " 'aspartyltrnasynthetase': 'ASPTRSm',\n",
       " 'ltyrosinetransportinviaprotonantiport': 'TYRt6',\n",
       " 'phosphatidylserinedecarboxylaseyeastspecific': 'PSERDer_RT',\n",
       " 'impcyclohydrolase': 'IMPC',\n",
       " 'dgtpuridine5phosphotransferase': 'DGTUP',\n",
       " 'nucleosidediphosphatekinaseatpidp': 'NDPK9',\n",
       " 'sucrosehydrolyzingenxymeextracellular': 'SUCRe',\n",
       " 'ltyrosinetransportoutviaprotonsymport': 'TYRt7',\n",
       " 'ribulose5phosphate3epimerase': 'RPE',\n",
       " 'hexokinasedmannoseatp': 'HEX4',\n",
       " 'lisoleucinetransportoutviaprotonsymport': 'ILEt7',\n",
       " '3dehydroquinatedehydrataseirreversible': 'DHQTi',\n",
       " 'ornithinemitochondrialtransportviaprotonantiport': 'ORNLYStm',\n",
       " 'leucinetransaminase': 'LEUTA',\n",
       " 'enoylcoahydratasec140': 'ECOAH6m',\n",
       " 'nadhdehydrogenasecytosolic': 'NADH2_u9cm',\n",
       " 'cholinephosphatecytididyltransferase': 'CHLPCTD',\n",
       " 'nucleosidediphosphatekinaseatpudp': 'NDPK2',\n",
       " 'ceramide2psynthase26c': 'CERS2p26er',\n",
       " 'pyruvatekinase': 'PYK',\n",
       " 'fattyacylcoasynthasenc180coa': 'FAS180COA',\n",
       " 'propanoylcoaacceptor23oxidoreductase': 'PPCOAOm',\n",
       " 'glycinehydroxymethyltransferasereversible': 'GHMT2r',\n",
       " 'acetylcoacacyltransferasedecanoylcoa': 'ACACT5p',\n",
       " 'allantoateamidinohydrolasereversible': 'ALLTAHr',\n",
       " 'ceramide3synthase24c': 'CERS324er',\n",
       " 'phosphoglucosaminemutase': 'PGAMT',\n",
       " 'phosphomevalonatekinase': 'PMEVK',\n",
       " 'fattyacylcoasynthasenc80coalumpedreaction': 'FAS80COA_L',\n",
       " 'dihydrofolatesynthase': 'DHFS',\n",
       " 'acetylglutamatekinase': 'ACGKm',\n",
       " 'phosphoglyceratedehydrogenase': 'PGCD',\n",
       " 'porphobilinogensynthase': 'PPBNGS',\n",
       " '15phosphoribosyl55phosphoribosylaminomethylideneaminoimidazole4carboxamideisomerase': 'PRMICI',\n",
       " '4phosphopantothenatelcysteineligase': 'PPNCL3',\n",
       " 'phosphoglyceratemutase': 'PGM',\n",
       " 'cytidylatekinasedcmp': 'CYTK2',\n",
       " 'mannoseinositolp2ceramideceramide224cphospholipasec': 'MIP2C224PLCg_RT',\n",
       " 'aspartate1decarboxylase': 'ASP1DC',\n",
       " 'ethanolaminephosphotransferaseyeastspecific': 'ETHAPTer_RT',\n",
       " 'phosphoglucomutase': 'PGMT',\n",
       " 'reducedglutathioneviaabcsystem': 'GTHRDabcv',\n",
       " 'dolichylphosphatemannoseproteinmannosyltransferaseendoplasmicreticular': 'DOLPMMer',\n",
       " 'deoxyadenylatekinase': 'DADK',\n",
       " 'dttpcytidine5phosphotransferase': 'DTTGY',\n",
       " 'zymosterolreversibletransport': 'ZYMSTt',\n",
       " 'adenylatecyclase': 'ADNCYC',\n",
       " 'uridinetransportinviaprotonsymport': 'URIt2',\n",
       " 'aspartatecarbamoyltransferase': 'ASPCT',\n",
       " 'cysteinyltrnasynthetase': 'CYSTRS',\n",
       " 'larganinereversibletransportviaprotonsymport': 'ARGt2r',\n",
       " 'atpcitratelyase': 'ACITL',\n",
       " 'glycolipid12alphadmannosyltransferasegolgiapparatus': 'G12MT1g',\n",
       " 'phosphofructokinase': 'PFK',\n",
       " 'aminobutyraldehydedehydrogenase': 'ABUTD',\n",
       " 'phenethylamineoxidase': 'PEAMNO',\n",
       " 'aldehydedehydrogenaseacetaldehydenad': 'ALDD2x',\n",
       " 'inositolphosphorylceramidesynthaseceramide224cyeastspecific': 'IPCS224g_RT',\n",
       " 'alkalinephosphatase': 'ALKP',\n",
       " '3hydroxyisobutyrylcoahydrolase': '3HBCOAHLm',\n",
       " 'lasparaginetransportinviaprotonantiport': 'ASNt6',\n",
       " 'dgtpcytidine5phosphotransferase': 'DGTCY',\n",
       " 'atpasecytosolic': 'ATPS',\n",
       " 'phosphoacetylglucosaminemutase': 'ACGAMPM',\n",
       " 'phenylalaninetransaminase': 'PHETA1',\n",
       " 'hexokinasedglucosamineatp': 'HEX10',\n",
       " '16betaglucansynthase': '16GS',\n",
       " 'malonylcoaacptransacylase': 'MCOATAm',\n",
       " 'nadsynthasenh3nuclear': 'NADS1n',\n",
       " 'alanylaminopeptidasecysgly': 'AMPTASECG',\n",
       " 'tetrahydrofolatelglutamategammaligaseadpforming': 'THFGLUS',\n",
       " 'adenylatekinase': 'CYTK2_1',\n",
       " 'rrbutanedioldehydrogenase': 'BTDD_RR',\n",
       " 'gtpcyclohydrolasei': 'GTPCI',\n",
       " 'methylenetetrahydrofolatedehydrogenasenadp': 'MTHFD',\n",
       " 'leucyltrnasynthetase': 'LEUTRSm',\n",
       " 'phosphatidylnmethylethanolaminenmethyltransferaseyeastspecific': 'PMETMer_RT',\n",
       " 'glycerol3phosphatedehydrogenasenad': 'G3PD1ir',\n",
       " 'glycylpeptidentetradecanoyltransferase': 'GLPT',\n",
       " 'deoxyribokinase': 'DRBK',\n",
       " 'adenosinetransportinviaprotonsymport': 'ADNt2',\n",
       " 'ureareversibletransportviaprotonsymport2h': 'UREA2t2',\n",
       " 'aspartatetransaminase': 'ASPTAm',\n",
       " 'mannoseinositolphosphorylceramidesynthaseceramide326cyeastspecific': 'MIPCS326g_RT',\n",
       " 'phosphoribosylampcyclohydrolase': 'PRAMPC',\n",
       " 'adenylsuccinatelyase': 'ADSL1r',\n",
       " 'arginase': 'ARGN',\n",
       " 'isocitratedehydrogenasenadp': 'ICDHyp',\n",
       " 'transketolase': 'TKT1',\n",
       " 'sadenosylmethioninedelta24sterolcmethyltransferase': 'SAM24MTer',\n",
       " 'biotinacetylcoacarboxylaseligase': 'BACCLm',\n",
       " 'dutpuridine5phosphotransferase': 'DUTUP',\n",
       " 'glutathioneoxidoreductase': 'GTHOm',\n",
       " 'ampatptransporter': 'ATP2tp_H',\n",
       " 'glutamatedecarboxylase': 'GLUDC',\n",
       " 'lmethioninereversibletransportviaprotonsymport': 'METt2r',\n",
       " '3hydroxybutanoylcoanadpoxidoreductase': 'HBCO_nadp',\n",
       " 'valinetransaminasemitochondiral': 'VALTAm',\n",
       " 'glutathionetransport': 'GTHRDt2',\n",
       " 'lmethioninesoxidereductase': 'METSOXR1',\n",
       " 'phospholipasedyeastspecific': 'PLDn_RT',\n",
       " 'naprtase': 'NAPRTm',\n",
       " 'fattyacidcoaligasenc260': 'FACOAL260',\n",
       " 'pyridoxinekinase': 'PYDXNK',\n",
       " 'alcoholdehydrogenaseethanolnadp': 'ALCD2y',\n",
       " 'imidazoleglycerol3phosphatesynthase': 'IG3PS',\n",
       " '1acylglycerol3phosphateacyltransferaseyeastspecific': 'AGATer_RT',\n",
       " 'dctpdeaminase': 'DCTPD',\n",
       " 'cytidinetransportinviaprotonsymport': 'CYTDt2',\n",
       " 'uracilphosphoribosyltransferase': 'UPPRT',\n",
       " 'enoylcoahydratase3hydroxyisobutyrylcoaforming': 'ECOAH12m',\n",
       " '3sulfoalaninecarboxylyase': 'LCYSTCBOXL',\n",
       " 'dihydroxyacetonekinase': 'DHAK',\n",
       " 'geranyltranstransferase': 'GRTT',\n",
       " 'inositolphosphorylceramidesynthaseceramide124cyeastspecific': 'IPCS124g_RT',\n",
       " 'ureacarboxylase': 'UREASE',\n",
       " 'acylcoathioesterase': 'PTE10m',\n",
       " 'tyrosinetransaminase': 'TYRTA',\n",
       " 'glycogenphosphorylase': 'GLCP',\n",
       " 'pyruvatedecarboxylase': 'PYRDC',\n",
       " 'phosphatidylinositolsynthaseyeastspecific': 'PINOSer_RT',\n",
       " 'mannoseinositolphosphorylceramideceramide124cphospholipasec': 'MIPC124PLCg_RT',\n",
       " 'phosphatetransporter': 'PIt2m',\n",
       " 'pyridoxamineoxygenoxidoreductasedeaminating': 'PYDXO_1',\n",
       " 'riboflavinsynthase': 'RBFSb',\n",
       " 'phosphomannomutase': 'PMANM',\n",
       " 'nadsynthasenh3': 'NADS1',\n",
       " 'phosphatidylserinedecarboxylaseyeastspecificgolgi': 'PSERDg_RT',\n",
       " 'hydroxyacylglutathionehydrolase': 'GLYOX',\n",
       " 'dlactatedehydrogenasecytosolic': 'D_LACDcm',\n",
       " 'acylcoaoxidasedodecanoylcoa': 'ACOAO5p',\n",
       " '5aminolevulinatesynthase': 'ALASm',\n",
       " 'aldehydedehydrogenaseisoamylalcoholnad': 'ALCD24xi',\n",
       " 'nucleosidediphosphatasegdpgolgi': 'NDP3g',\n",
       " 'diacylglycerolpyrophosphatephosphataseyeastspecific': 'DAGPYPer_RT',\n",
       " 'spermidinetransportinviaprotonantiport': 'SPMDt3i',\n",
       " '3methyl2oxobutanoatedecarboxylase': '3MOBDC',\n",
       " 'ctp12diacylsnglycerol3phosphotransferase': 'DAGKer_RT',\n",
       " 'malatedehydrogenase': 'MDHp',\n",
       " 'seryltrnasynthetase': 'SERTRSm',\n",
       " 'transaldolase': 'TALA',\n",
       " 'aldehydedehydrogenase2phenylethanolnad': 'ALCD25xi',\n",
       " 'glyceraldehyde3phosphatedehydrogenase': 'GAPD',\n",
       " 'atppantothenoyllcysteine4phosphotransferase': 'APCPT',\n",
       " 'ornithinecarbamoyltransferaseirreversible': 'OCBTm',\n",
       " 'c5steroldesaturase': 'C5STDSer',\n",
       " '13betaglucansynthase': '13GS',\n",
       " 'sirohydrochlorindehydrogenasenad': 'SHCHD2',\n",
       " 'galactokinase': 'GALKr',\n",
       " 'thiaminediphosphokinase': 'TMDPK',\n",
       " 'ceramide1hydroxylase26c': 'CERH126er',\n",
       " 'citratehydrolyasecisaconitateforming': 'ACONTam',\n",
       " 'dctpuridine5phosphotransferase': 'DCTUP',\n",
       " 'mannoseinositolphosphorylceramideceramide324cphospholipasec': 'MIPC324PLCg_RT',\n",
       " 's2hydroxyacidoxidase': 'GLYCTO1p',\n",
       " 'fecosterolesterhydrolaseyeastspecific': 'FECOSTESTH_RT',\n",
       " '3hydroxyacylcoadehydrogenase3oxohexadecanoylcoa': 'HACD7p',\n",
       " 'lthreoninereversibletransportviaprotonsymport': 'THRt2r',\n",
       " 'electrontransferflavoprotein': 'ETF',\n",
       " 'nucleotidephosphatase': 'NP1',\n",
       " 'pantetheinephosphateadenylyltransferase': 'PTPATi',\n",
       " 'indole3glycerolphosphatesynthase': 'IGPS',\n",
       " 'aldehydedehydrogenaseindole3acetaldehydenad': 'ALDD20x',\n",
       " 'llysinetransportinviaprotonantiport': 'LYSt6',\n",
       " 'methioninesynthase': 'METS',\n",
       " 'cystathionineblyase': 'CYSTL',\n",
       " '1phosphatidylinositol35bisphosphate5phosphataseyeastspecific': 'PIN3Per_RT',\n",
       " 'nadkinase': 'NADK',\n",
       " '5nucleotidaseump': 'NTD2',\n",
       " 'mannoseinositolphosphorylceramidesynthaseceramide226cyeastspecific': 'MIPCS226g_RT',\n",
       " 'cdpdiacylglycerolserineophosphatidyltransferaeyeastspecific': 'CDPDGPm_RT',\n",
       " 'histidinolphosphatase': 'HISTP',\n",
       " 'fattyacidcoaligasehexadecenoate': 'FACOAL161',\n",
       " 'guanylatekinasegmpatp': 'GK1',\n",
       " 'acylcoaepisterolacyltransferaseyeastspecific': 'EPISTATer_RT',\n",
       " 'isobutylacetatehydrolyzingesterase': 'ACHLE2',\n",
       " 'acetylcoacacetyltransferasedodecanoyl': 'ACACT6p',\n",
       " 'gmpsynthase': 'GMPS',\n",
       " '4amino4deoxychorismatesynthase': 'ADCS',\n",
       " 'impdehydrogenase': 'IMPDm',\n",
       " 'tryptophantransaminase': 'TRPTA',\n",
       " 'coahydrolase': 'COAHLp',\n",
       " 'lanosterolesterhydrolaseyeastspecific': 'LANOSTESTH_RT',\n",
       " 'ornithinetransacetylaseirreversible': 'ORNTACim',\n",
       " 'lvalinereversibletransportviaprotonsymport': 'VALt2r',\n",
       " 'cs24sterolreductaseendoplamicreticular': 'C24STRer',\n",
       " 'lasparaginetransportoutviaprotonsymport': 'ASNt7',\n",
       " 'yumpsynthetase': 'YUMPS',\n",
       " 'phosphatidylcholinediacylglycerolacyltransferase': 'PCDAGATer_RT',\n",
       " 'inositolphosphorylceramideceramide226cphospholipasec': 'IPC226PLCg_RT',\n",
       " 'fattyacylcoasynthasenc140coa': 'FAS140COA',\n",
       " 'acetolactatesynthase': 'ACLSm',\n",
       " 'glucose6phosphatedehydrogenase': 'G6PDH2rp',\n",
       " '1phosphatidylinositol3kinaseyeastspecific': 'PIN3Kn_RT',\n",
       " 'sphinganinenadphoxygenoxidoreductase4hydroxylating': 'PSPHSer',\n",
       " 'pyruvatedehydrogenase': 'PDHm',\n",
       " 'ammoniareversibletransport': 'NH4t',\n",
       " 'gammaglutamylcysteinesynthetase': 'GLUCYS',\n",
       " 'alanyltrnasynthetase': 'ALATRS',\n",
       " 'protoporphyrinogenoxidase': 'PPPGOm',\n",
       " 'dcmpdeaminase': 'DCMPDA',\n",
       " 'glutaminyltrnasynthetase': 'GLNTRS',\n",
       " 'adenosinemonophosphatedeaminase': 'AMPDA',\n",
       " 'ornithinecarbamoyltransferase': 'OCBT',\n",
       " 'acetylcoacacetyltransferase': 'ACACT1m',\n",
       " 'adeninetransportinviaprotonsymport': 'ADEt2',\n",
       " 'alphaalphatrehalase': 'TREHe',\n",
       " 'lglutamatetransportviaprotonsymportreversible': 'GLUt2r',\n",
       " 'fecosterolreversibletransport': 'FECOSTt',\n",
       " 'betafructofuranosidase': 'BFFS',\n",
       " 'triglycerolsynthesis': 'TRIGSer_RT',\n",
       " '2deoxydarabinoheptulosonate7phosphatesynthetase': 'DDPAm',\n",
       " '14alphaglucanbranchingenzyme': 'GBEZ',\n",
       " 'lhydroxyprolinereductasenadp': 'HPROb',\n",
       " 'thiaminephosphatediphosphorylase': 'TMPPP',\n",
       " 'fructosebisphosphatealdolase': 'FBA',\n",
       " '34dihydroxy2butanone4phosphatesynthase': 'DB4PS',\n",
       " 'argininosuccinatesynthase': 'ARGSS',\n",
       " '3dehydroquinatesynthase': 'DHQS',\n",
       " 'lleucinereversibletransportviaprotonsymport': 'LEUt2r',\n",
       " '3phosphoshikimate1carboxyvinyltransferase': 'PSCVT',\n",
       " '3hydroxyacylcoadehydratase3hydroxytetradecanoylcoa': 'ECOAH6p',\n",
       " 'ctpsynthasenh3': 'CTPS1',\n",
       " 'acylcoalanosterolacyltransferaseyeastspecific': 'LANOSTATer_RT',\n",
       " 'mannoseinositolp2ceramideceramide226cphospholipasec': 'MIP2C226PLCg_RT',\n",
       " 'nacetylglucosamine6phosphatesynthase': 'ACGAM6PS',\n",
       " 'glutathionesynthetase': 'GTHS',\n",
       " 'guanylatecyclase': 'GUACYC',\n",
       " 'glutamatetransportuniporter': 'LYSt7m',\n",
       " 'phosphoribosylaminoimidazolesuccinocarboxamidesynthase': 'PRASCSi',\n",
       " 'utpuridine5phosphotransferase': 'UTUP',\n",
       " 'prephenatedehydrogenasenadp': 'PPND2',\n",
       " 'phosphoribosylglycinamidesynthase': 'PRAGSr',\n",
       " 'fattyacylcoasynthasenc120coa': 'FAS120COA',\n",
       " 'glutamate5semialdehydedehydrogenase': 'G5SD',\n",
       " '5methylthioribose1phosphateisomerase': 'MTRI',\n",
       " 'cysteinesynthase': 'CYSS',\n",
       " 'glycogensynthaseudpglc': 'GLCS2',\n",
       " 'lhistidinereversibletransportviaprotonsymport': 'HISt2r',\n",
       " 'lallothreoninedehydrogenase': 'ATHRDHr',\n",
       " 'dolichylphosphatedmannosyltransferase': 'DOLPMTcer',\n",
       " 'peptidealphanacetyltransferase': 'PEPAT',\n",
       " 'cysteinetransaminase': 'CYSTA',\n",
       " 'lprolinereversibletransportviaprotonsymport': 'PROt2r',\n",
       " '2methylacylcoadehydrogenase2methylprop2enoylcoaforming': 'ACOAD9m',\n",
       " '25diamino6ribosylamino43hpyrimidinone5phosphatereductasenadh': 'DROPPRx',\n",
       " 'pyridoxine5phosphatephosphatase': 'PDXPP',\n",
       " 'inorganicdiphosphatase': 'PPA',\n",
       " 'phosphoribosylaminoimidazolecarboxylase': 'AIRCr',\n",
       " 'glycerolkinase': 'GLYKm',\n",
       " 'carnitineoaceyltransferase': 'CSNAT3x',\n",
       " 'nucleosidetriphosphatepyrophosphorylaseditp': 'NTPP10',\n",
       " 'purinenucleosidephosphorylasedeoxyadenosine': 'PUNP2',\n",
       " 'hemeosynthase': 'HEMEOSm',\n",
       " 'phosphatidylinositol4kinasenuclearyeastspecifc': 'PIN4Kn_RT',\n",
       " 'adenylosuccinatelyase': 'ADSL2r',\n",
       " 'nucleosidediphosphatekinaseatpdgdp': 'NDPK5',\n",
       " 'lalaninereversibletransportviaprotonsymport': 'ALAt2r',\n",
       " 'cytosinetransportinviaprotonsymport': 'CSNt2',\n",
       " 'glyceroldehydrogenasenadpdependent': 'GLYCDy',\n",
       " 'lglutaminetransportinviaprotonantiport': 'GLNt6',\n",
       " 'dihydroxyaciddehydratase23dihydroxy3methylbutanoate': 'DHAD1m',\n",
       " 'hco3equilibrationreaction': 'HCO3E',\n",
       " 'nformyllkynurenineamidohydrolase': 'FKYNH',\n",
       " 'umpkinasenuclear': 'UMPKn',\n",
       " 'inositol1345triphosphate6kinasenucleus': 'MI1345PKn',\n",
       " 'purinenucleosidephosphorylaseinosine': 'PUNP5',\n",
       " 'arginyltrnasynthetase': 'ARGTRSm',\n",
       " 'homoserinedehydrogenasenadhirreversible': 'HSDxi',\n",
       " 'tyrosinetransaminaseirreversible': 'TYRTAip',\n",
       " 'glycogenstarchsynthase': 'GLYGS',\n",
       " 'dfructose1phosphatedglyceraldehyde3phosphatelyase': 'FBA2',\n",
       " 'asparaginyltrnasynthetasemiotchondrial': 'ASNTRSm',\n",
       " 'h2otransportviadiffusion': 'H2Ot',\n",
       " 'laspartatetransportoutviaprotonsymport': 'ASPt7',\n",
       " 'adpribosediphosphatase': 'ADPRDP',\n",
       " 'phosphoribosylformylglycinamidinesynthase': 'PRFGS',\n",
       " 'adenosylmethionine8amino7oxononanoatetransaminase': 'AMAOTrm',\n",
       " 'isoleucinetransaminase': 'ILETA',\n",
       " 'phosphatidylserinesynthaseyeastspecific': 'PSERSm_RT',\n",
       " 'lphenylalaninereversibletransportviaprotonsymport': 'PHEt2r',\n",
       " 'udpnacetylglucosaminediphosphorylase': 'UDPACGLP',\n",
       " 'acetylcoaacyltransferasetetracosanoylcoa': 'ACACT9p',\n",
       " 'mannoseinositolphosphorylceramidesynthaseceramide324cyeastspecific': 'MIPCS324g_RT',\n",
       " 'asparaginesynthaseglutaminehydrolysing': 'ASNS1',\n",
       " 'episterolesterhydrolaseyeastspecific': 'EPISTESTH_RT',\n",
       " '510methylenetetrahydrofolatereductasenadh': 'MTHFR2',\n",
       " 'alanineglyoxylateaminotransferaseirreversible': 'AGTi',\n",
       " 'sphinganinenadp3oxidoreductase': '3DSPHRer',\n",
       " 'phosphatidylinositol4kinaseyeastspecific': 'PIN4Ker_RT',\n",
       " 'pyridoxineoxidase': 'PYDXNO',\n",
       " 'farnesyltranstransferase': 'FRTT',\n",
       " 'pantothenatereversibletransportviaprotonsymport': 'PNTOt2',\n",
       " '2dehydropantoate2reductase': 'DPR',\n",
       " 'dsorbitoldehydrogenasedfructoseproducing': 'SBTD_D2',\n",
       " 'lserinedeaminase': 'SERD_L',\n",
       " 'formaldehydedehydrogenase': 'FALDH2',\n",
       " 'llysinereversibletransportviaprotonsymport': 'LYSt2r',\n",
       " 'trehalosephosphatase': 'TRE6PP',\n",
       " 'adenosylmethioninedecarboxylase': 'ADMDC',\n",
       " 'nucleosidediphosphatekinaseatpdadp': 'NDPK8',\n",
       " 'ltyrosinereversibletransportviaprotonsymport': 'TYRt2r',\n",
       " 'alcoholdehydrogenaseglycerolnadp': 'ALCD19y',\n",
       " 'mannoseinositolphosphorylceramideceramide226cphospholipasec': 'MIPC226PLCg_RT',\n",
       " 'dfructosetransportinviaprotonsymport': 'FRUt2',\n",
       " 'aldehydedehydrogenase2methylbutanolnad': 'ALCD22xi',\n",
       " 'adenylosuccinatesynthase': 'ADSS',\n",
       " 'darabinono14lactoneoxidase': 'ARAB14LO',\n",
       " 'phosphatidylethanolaminenmethyltransferaseyeastspecific': 'PETOHMer_RT',\n",
       " 'glutathioneperidoxase': 'GTHPi',\n",
       " 'thiaminetransportinviaprotonsymport': 'THMt2',\n",
       " 'uroporphyrinogenmethyltransferase': 'UPP3MT',\n",
       " 'ribose5phosphateisomerase': 'RPI',\n",
       " 'mannoseinositolp2ceramideceramide126cphospholipasec': 'MIP2C126PLCg_RT',\n",
       " 'lserinedehydrogenase': 'LSERDHr',\n",
       " 'hydroxyethylthiazolekinase': 'HETZK',\n",
       " 'uroporphyrinogeniiisynthase': 'UPP3S',\n",
       " 'glucan14alphaglucosidase': 'GLCGSD',\n",
       " 'thymidylatesynthase': 'TMDS',\n",
       " 'xylitoldehydrogenasedxyuloseforming': 'XYLTD_D',\n",
       " 'valyltrnasynthetase': 'VALTRSm',\n",
       " 'uridinekinaseatpuridine': 'URIK1',\n",
       " 'acylcoaergosterolacyltransferaseyeastspecific': 'ERGSTATer_RT',\n",
       " 'pyridoxamine5phosphateoxidase': 'PYAM5PO',\n",
       " 'nadhdehydrogenase': 'NADH2_u9m2',\n",
       " '3isopropylmalatedehydrogenase': 'IPMD',\n",
       " 'inositolphosphorylceramideceramide326cphospholipasec': 'IPC326PLCg_RT',\n",
       " 'chitinsynthase': 'CHTNS',\n",
       " 'mannoseinositolphosphorylceramidesynthaseceramide224cyeastspecific': 'MIPCS224g_RT',\n",
       " 'dmannosetransportinviaprotonsymport': 'MANt2',\n",
       " 'ribonucleosidediphosphatereductaseadp': 'RNDR1',\n",
       " 'acetylcoacacyltransferase': 'ACACT10m',\n",
       " 'sulfateadenylyltransferase': 'SADT',\n",
       " '5nucleotidasecmp': 'NTD4',\n",
       " 'lipidphosphatephosphataseyeastspecific': 'LPPer_RT',\n",
       " '5methylthio5deoxydribulose1phosphatedehydratase': 'MDRPD',\n",
       " 'cardiolipinsynthaseyeastspecific': 'CLPNSm_RT',\n",
       " 'threoninealdolase': 'THRA',\n",
       " 'fattyacylcoasynthasenc160coa': 'FAS160COA',\n",
       " 'laminoadipatesemialdehydedehydrogenasenadph': 'AASAD1',\n",
       " 'homoacontinatehydratase': 'HACONTbm',\n",
       " 'utpcytidine5phosphotransferase': 'UTCY',\n",
       " '4aminobenzoatesynthase': 'ADCL',\n",
       " 'methylmalonatesemialdehydedehydrogenaseacylating': 'MMSAD3m',\n",
       " 'gluconokinase': 'GNK',\n",
       " 'lallothreoninealdolase': 'THRA2',\n",
       " 'squaleneepoxidaseendoplasmicreticularnad': 'SQLErx',\n",
       " 'nucleosidediphosphatekinaseatpdtdp': 'NDPK4',\n",
       " 'phosphoribosylaminoimidazolesynthase': 'PRAIS',\n",
       " 'acetylcoasynthetase': 'ACS',\n",
       " 'phosphogluconatedehydrogenase': 'GND',\n",
       " 'sulfitereductasenadph2': 'SULR',\n",
       " 'cholinetransportviaprotonsymport': 'CHLt2',\n",
       " 'ceramide3synthase26c': 'CERS326er',\n",
       " 'purinenucleosidephosphorylaseadenosine': 'PUNP1',\n",
       " 'fattyacidcoaligaseoctadecenoate': 'FACOAL181p',\n",
       " 'methionyltrnaformyltransferase': 'FMETTRSm',\n",
       " 'catalasea': 'CATp',\n",
       " 'lleucinetransportinviaprotonantiport': 'LEUt6',\n",
       " '1dmyoinositol4phosphatephosphohydrolase': 'MI4PP',\n",
       " 'fadfmnantiport': 'FADFMNtm',\n",
       " 'sphingoidbasephosphatephosphatasesphinganine1phosphataseendoplasmicreticulum': 'SBPP1er',\n",
       " 'ethylacetatehydrolyzingesterase': 'ACHLE3',\n",
       " 'guaninereversibletransportviaprotonsymport': 'GUAt2r',\n",
       " 'lglutamatetransportoutviaprotonsymport': 'GLUt7',\n",
       " 'episterolreversibletransport': 'EPISTt',\n",
       " 'saccharopinedehydrogenasenadplglutamateforming': 'SACCD1',\n",
       " '2isopropylmalatehydratase': 'IPPMIb',\n",
       " 'c8sterolisomerase': 'C8STIer',\n",
       " 'inositolphosphorylceramideceramide126cphospholipasec': 'IPC126PLCg_RT',\n",
       " 'phospholipasebphosphatidylinositolyeastspecificextracellular': 'PLBP1Ie_RT',\n",
       " 'atppantetheine4phosphotransferase': 'APNPT',\n",
       " 'phosphoglyceratekinase': 'PGK',\n",
       " 'ctpcmpantiport': 'CTPtm2',\n",
       " 'lornithine2oxoacidaminotransferase': 'ORNTArm',\n",
       " 'purinenucleosidephosphorylasexanthosine': 'PUNP7',\n",
       " 'adenosinekinase': 'ADNK1',\n",
       " 'cytidylatekinasecmp': 'CYTK1',\n",
       " 'ribonucleosidediphosphatereductaseudp': 'RNDR4',\n",
       " 'alphaglucosidase': 'MALT',\n",
       " 'acetylcoasynthasepropionate': 'ACS2',\n",
       " 'ctpsynthaseglutamine': 'CTPS2',\n",
       " 'orntithinereversibletransportinviaprotonsymport': 'ORNt2r',\n",
       " 'isocitratehydrolyasecisaconitateforming': 'ACONTbm',\n",
       " 'cystathionineglyase': 'CYSTGL',\n",
       " 'enoylcoahydratasec100': 'ECOAH4m',\n",
       " '5oxoprolineamidohydrolaseatphydrolysingir': 'OPAH',\n",
       " 'phospholipasebphosphatidylcholineyeastspecific': 'PLBPC_RT',\n",
       " 'lleucinetransportoutviaprotonsymport': 'LEUt7',\n",
       " 'acyldihydroxyacetonephosphatereductaseyeastspecific': 'ADHAPRer_RT',\n",
       " 'c22steroldesaturasenadp': 'C22STDSer',\n",
       " 'saccharopinedehydrogenasenadllysineforming': 'SACCD2',\n",
       " 'phosphofructokinases7p': 'PFK_3',\n",
       " 'lisoleucinetransportinviaprotonantiport': 'ILEt6',\n",
       " 'prolyltrnasynthetase': 'PROTRS',\n",
       " 'acylcoaoxidaseoctadecanoylcoa': 'ACOAO8p',\n",
       " 'fructosebisphosphatase': 'FBP',\n",
       " 'nicotinamidenucleotideadenylyltransferase': 'NMNATn',\n",
       " 'ltryptophanoxygen23oxidoreductasedecyclizing': 'TRPO2',\n",
       " '35bisphosphatenucleotidase': 'BPNT',\n",
       " 'inositolphosphorylceramidesynthaseceramide126cyeastspecific': 'IPCS126g_RT',\n",
       " 'aldehydedehydrogenasetryptopholnad': 'ALCD26xi',\n",
       " 'sirohydrochlorinferrochelatase': 'SHCHF',\n",
       " 'purinenucleosidephosphorylasedeoxyguanosine': 'PUNP4',\n",
       " 'methenyltetrahydrofolatecyclohydrolase': 'MTHFCx',\n",
       " 'phosphatidylinositol4phosphate5kinaseyeastspecfic': 'PI4P5Kn_RT',\n",
       " 'uridylatekinasedumpnuclear': 'URIDK2rn',\n",
       " 'dgalactosetransportinviaprotonsymport': 'GALt2',\n",
       " '2keto4methylthiobutyratetransamination': 'UNK3',\n",
       " 'ornithinedecarboxylase': 'ORNDC',\n",
       " 'cholinekinase': 'CHOLK',\n",
       " 'mannoseinositolphosphorylceramideceramide126cphospholipasec': 'MIPC126PLCg_RT',\n",
       " '2isopropylmalatesynthase': 'IPPS',\n",
       " 'allantoinase': 'ALLTN',\n",
       " 'nucleosidetriphosphatepyrophosphorylasextp': 'NTPP11',\n",
       " 'pyridoxal5phosphatephosphatase': 'PYDXPP',\n",
       " 'succinatefumaratetransport': 'SUCFUMtm',\n",
       " 'methylenetetrahydrofolatedehydrogenasenad': 'MTHFD2i',\n",
       " 'glycero3phospho1inositoltransportextracellulartocytosol': 'G3PIt',\n",
       " 'phenylpyruvatedecarboxylase': 'PPYRDC',\n",
       " 'aldosereductasemethylglyoxal': 'ALR2',\n",
       " 'pyruvatecarboxylase': 'PC',\n",
       " 'dlactatetransportviaprotonsymport': 'D_LACt2',\n",
       " 'glyceroltransportviachannel': 'GLYCt',\n",
       " 'glycinereversibletransportviaprotonsymport': 'GLYt2r',\n",
       " 'adenylylsulfatekinase': 'ADSK',\n",
       " 'phosphoenolpyruvatecarboxykinase': 'PPCK',\n",
       " 'taurcholateviaabcsystem': 'TCHOLAabcv',\n",
       " 'shikimatekinase': 'SHKK',\n",
       " 'lysyltrnasynthetase': 'LYSTRSm',\n",
       " 'triacylglycerollipaseyeastspecific': 'TAGL_RT',\n",
       " 'fumaratereductasecytosolic': 'FRDcm',\n",
       " '3hydroxyacylcoadehydrogenase3oxotetradecanoylcoa': 'HACD6p',\n",
       " 'l4hydroxyglutamatesemialdehydedehydrogenase': '4HGLSDm',\n",
       " 'methylenefattyacylphospholipidsynthaseyeastspecific': 'MFAPSer_RT',\n",
       " 'aldehydedehydrogenaseindole3acetaldehydenadp': 'ALDD20y',\n",
       " 'pyridoxine5phosphateoxidase': 'PDX5POi',\n",
       " 'guaninephosphoribosyltransferase': 'GUAPRT',\n",
       " 'glutamate5kinase': 'GLU5K',\n",
       " 'mannoseinositolphosphorylceramideceramide224cphospholipasec': 'MIPC224PLCg_RT',\n",
       " 'chorismatemutase': 'CHORM',\n",
       " 'fattyacidcoaligaseoctanoate': 'FACOAL80p',\n",
       " 'lcysteine2oxoglutarateaminotransferase': 'CYSTAm',\n",
       " 'lisoleucinereversibletransportviaprotonsymport': 'ILEt2r',\n",
       " 'udpglucosehexose1phosphateuridylyltransferase': 'UGLT',\n",
       " 'itpcytidine5phosphotransferase': 'ITCY',\n",
       " 'lglutaminereversibletransportviaprotonsymport': 'GLNt2r',\n",
       " 'adeninedeaminase': 'ADD',\n",
       " 'glutaminefructose6phosphatetransaminase': 'GF6PTA',\n",
       " 'deoxyadenosinedeaminase': 'DADA',\n",
       " 'nucleosidediphosphatekinaseatpgdp': 'NDPK1',\n",
       " 'mannoseinositolphosphorylceramideceramide326cphospholipasec': 'MIPC326PLCg_RT',\n",
       " 'ceramide2synthase24cendoplasmicreticular': 'CERS224er',\n",
       " 'mannoseinositolp2ceramideceramide124cphospholipasec': 'MIP2C124PLCg_RT',\n",
       " 'enolase': 'ENO',\n",
       " 'transaconitate3methyltransferase': 'ACON3MT',\n",
       " '3hydroxyacylcoadehydrogenase3oxooctadecanoylcoa': 'HACD8p',\n",
       " 'hexokinasedglucoseatp': 'HEX1',\n",
       " '3mercaptopyruvatesulfurtransferase': 'MCPST',\n",
       " 'lasparaginereversibletransportviaprotonsymport': 'ASNt2r',\n",
       " 'adenosylhomocysteinase': 'ADSHm',\n",
       " 'adeninephosphoribosyltransferase': 'ADPT',\n",
       " 'acteylornithinetransaminaseirreversible': 'ACOTAim',\n",
       " 'inositolphosphorylceramidesynthaseceramide324cyeastspecific': 'IPCS324g_RT',\n",
       " 'nicotinamidenmethyltransferase': 'NNMT',\n",
       " '3hydroxyacylcoadehydrogenase3oxohexacosylcoa': 'HACD10p',\n",
       " 'dihydroorotase': 'DHORTS',\n",
       " 'acetylcoaacyltransferasehexadecanoylcoa': 'ACACT8p',\n",
       " 'mannose1phosphateguanylyltransferase': 'MAN1PT',\n",
       " 'inositolphosphorylceramideceramide124cphospholipasec': 'IPC124PLCg_RT',\n",
       " 'peroxidasemultiplesubstrates': 'PRDX',\n",
       " 'acylcpaoxidasetetradecanoylcoa': 'ACOAO6p',\n",
       " 'acetohydroxyacidisomeroreductase': 'KARA1im',\n",
       " 'lsorbitoldehydrogenaselsorboseproducing': 'SBTD_L',\n",
       " 'aldehydedehydrogenase2methylbutanolnadp': 'ALCD22yi',\n",
       " 'deaminonadnucleotidohydrolase': 'DNADDP',\n",
       " 'succinatecoaligaseadpforming': 'SUCOASm',\n",
       " 'threoninesynthase': 'THRS',\n",
       " 'sphingoidbasephosphatephosphatasephytosphingosine1phosphateendoplasmicreticulum': 'SBPP2er',\n",
       " 'umpkinase': 'UMPK',\n",
       " 'pyridoxalkinase': 'PYDXK',\n",
       " 'dihydroneopterinmonophosphatedephosphorylase': 'DNMPPA',\n",
       " 'mannoseinositolp2ceramideceramide326cphospholipasec': 'MIP2C326PLCg_RT',\n",
       " 'acylcoaoxidasehexadecanoylcoa': 'ACOAO7p',\n",
       " 'acylcoafecosterolacyltransferaseyeastspecific': 'FECOSTATer_RT',\n",
       " 'acetaldehydecondensation': 'ACALDCD',\n",
       " 'sadenosyllmethioninereversibletransport': 'AMETt2m',\n",
       " 'ceramide1hydroxylase24c': 'CERH124er',\n",
       " 'orotidine5phosphatedecarboxylase': 'OMPDC',\n",
       " 'lysophosphatidylcholineacyltransferaseacyltransferaseyeastspecific': 'LPCATer_RT',\n",
       " 'inositol1456tetrakisphosphate3kinasenucleus': 'MI1456PKn',\n",
       " 'inositolphosphorylceramidesynthaseceramide326cyeastspecific': 'IPCS326g_RT',\n",
       " 'xanthinephosphoribosyltransferase': 'XPPT',\n",
       " '3hydroxyacylcoadehydratase3hydroxydecanoylcoa': 'ECOAH4p',\n",
       " 'riboflavinkinase': 'RBFK',\n",
       " 'nucleosidediphosphatekinaseatpdudp': 'NDPK6',\n",
       " 'isocitratelyase': 'ICLp',\n",
       " 'nicotinicacidtransport': 'NACt',\n",
       " 'imidazoleacetaldeydedehydrogenase': 'IMACTD',\n",
       " '3hydroxybutanoylcoahydrolyase': 'ECOAH1m',\n",
       " 'sedoheptulose17bisphosphatedglyceraldehyde3phosphatelyase': 'FBA3',\n",
       " '3isopropylmalatedehydratase': 'IPPMIa',\n",
       " 'ribosylnicotinamidekinase': 'RNMK',\n",
       " 'sulfateadenylyltransferaseadp': 'SLFAT',\n",
       " 'glycerol3phosphateacyltransferaseglycerol3phosphateyeastspecific': 'GAT1er_RT',\n",
       " 'threonyltrnasynthetase': 'THRTRSm',\n",
       " 'hydroxypyruvatereductasenadp': 'HPYRRy',\n",
       " '3hydroxyanthranilate34dioxygenase': '3HAO',\n",
       " 'mannoseinositiolphophorylceramidesynthaseceramide124cyeastspecific': 'MIPCS124g_RT',\n",
       " 'nicotinatenucleotideadenylyltransferase': 'NNATm',\n",
       " 'tryptophansynthaseindoleglycerolphosphate': 'TRPS1',\n",
       " '1phosphatidylinositol45bisphosphatephosphodiesteraseyeastspecific': 'PI45BPPn_RT',\n",
       " 'cytosinedeaminase': 'CSND',\n",
       " 'lserinereversibletransportviaprotonsymport': 'SERt2r',\n",
       " 'maltosetransportinviaprotonsymport': 'MALTt2',\n",
       " 'dephosphocoakinase': 'DPCOAK',\n",
       " 'isopentenyldiphosphatedisomerase': 'IPDDI',\n",
       " 'phosphatidylinositol3phosphate5kinaseyeastspecfic': 'PI3P5Kn_RT',\n",
       " 'galactose1phosphateuridylyltransferase': 'GALT',\n",
       " 'poylamineoxidase': 'POLYAO2',\n",
       " 'dutpcytidine5phosphotransferase': 'DUTCP',\n",
       " 'lalaninetransaminase': 'ALATA_Lm',\n",
       " 'c3steroldehydrogenasezymosterol': 'C3STDH2er',\n",
       " 'acetylcoahydrolasemitochondrialirreversible': 'ACOAHim',\n",
       " 'gdpmannoseantiport': 'GDPMANNtg',\n",
       " 'ribonucleosidediphosphatereductasecdp': 'RNDR3',\n",
       " 'sperminetransportviaprotonantiportirreversible': 'SPRMt2i',\n",
       " 'hypotheticalenyme': 'HYPOE',\n",
       " 'atpcytidine5phosphotransferase': 'CYTDK1',\n",
       " 'imidazoleglycerolphosphatedehydratase': 'IGPDH',\n",
       " 'hydroxymethylpyrimidinekinaseatp': 'HMPK1',\n",
       " 'fattyacidcoaligasenc240': 'FACOAL240p',\n",
       " 'acylcoaoxidasehexacanoylcoa': 'ACOAO9p',\n",
       " 'sulfitetransporteffluxcytosoltoextracellular': 'SO3ti',\n",
       " 'alphaalphatrehalosephosphatesynthaseudpforming': 'TRE6PS',\n",
       " 'utpglucose1phosphateuridylyltransferaseirreversible': 'GALUi',\n",
       " 'diphosphomevalonatedecarboxylase': 'DPMVD',\n",
       " 'carbamoylphosphatesynthaseglutaminehydrolysing': 'CBPS',\n",
       " 'dttpuridine5phosphotransferase': 'DTTUP',\n",
       " 'fattyacidcoaligasedecanoate': 'FACOAL100p',\n",
       " 'glutarylcoadehydrogenase': 'GLUTCOADHm',\n",
       " 'zymosterolesterhydrolaseyeastspecific': 'ZYMSTESTH_RT',\n",
       " 'kynurenine3monooxygenase': 'KYN3OX',\n",
       " 'uridinekinaseitpuridine': 'URIK3',\n",
       " 'cystathioninebetasynthase': 'CYSTS',\n",
       " 'methylamineoxygenoxidoreductasedeaminatingcoppercontaining': 'MAOX',\n",
       " 'aldehydedehydrogenaseisoamylalcoholnadp': 'ALCD24yi',\n",
       " 'formatedehydrogenase': 'FDH',\n",
       " 'nicotinatedribonucleotidesynthaseatp': 'NICRNS',\n",
       " '5methylthioadenosinephosphorylase': 'MTAP',\n",
       " 'catalase': 'CATm',\n",
       " 'succinatetransport': 'SUCCtm',\n",
       " 'orotatephosphoribosyltransferase': 'ORPT',\n",
       " 'ceramide2synthase26cendoplasmicreticular': 'CERS226er',\n",
       " 'uridinekinasegtpuridine': 'URIK2',\n",
       " 'isocitratedehydrogenasenad': 'ICDHxm',\n",
       " 'lhydroxyprolinereductasenad': 'HPROa',\n",
       " 'allantoinirreversibleuniport': 'ALLTNti',\n",
       " 'sphingolipidlongchainbasekinasesphinganine': 'SLCBK1er',\n",
       " '6phosphofructo2kinase': 'PFK26',\n",
       " 'dutpdiphosphatase': 'DUTPDPm',\n",
       " 'fattyacylcoasynthasenc100coa': 'FAS100COA',\n",
       " '3hydroxyacylcoadehydratase3hydroxyoctadecanoylcoa': 'ECOAH8p',\n",
       " 'glycero3phosphocholinetransportextracellulartocytosol': 'G3PCt',\n",
       " 'lglutaminetransportoutviaprotonsymport': 'GLNt7',\n",
       " 'laspartatereversibletransportviaprotonsymport': 'ASPt2r',\n",
       " 'glucose6phosphateisomerase': 'PGI',\n",
       " 'glycinecleavagesystem': 'GLYCLm',\n",
       " 'phosphatereversibletransportviasymport': 'PIt2r',\n",
       " 'chorismatesynthase': 'CHORS',\n",
       " 'anthranilatephosphoribosyltransferase': 'ANPRT',\n",
       " 'glycerophosphodiesterphosphodiesteraseglycerophosphocholine': 'GPDDA1',\n",
       " 'nacteylglutamatesynthase': 'ACGSm',\n",
       " 'glutamatesynthasenadh2': 'GLUSx',\n",
       " '3deoxydarabinoheptulosonate7phosphatesynthetase': 'DDPA',\n",
       " 'mannose6phosphateisomerase': 'MAN6PI',\n",
       " 's3hydroxybutanoylcoanadoxidoreductase': 'HACD1p',\n",
       " 'isoamylacetatehydrolyzingesterase': 'ACHLE1',\n",
       " 'hemeomonooxygenase': 'HEMEASm',\n",
       " 'aconitase': 'ACONT3m',\n",
       " 'anthranilatesynthase': 'ANS',\n",
       " 'allantoateirreversibleuniport': 'ALLTTti',\n",
       " 'hnrnpargininenmethyltransferase': 'ARMT',\n",
       " 'gtpcyclohydrolaseii': 'GTPCII',\n",
       " '3hydroxypropionylcoahydrolyase': 'PRPNCOAHYDm',\n",
       " 'succinatedehydrogenaseubiquinone6': 'SUCD2_u9m',\n",
       " 'glutaminephosphoribosyldiphosphateamidotransferase': 'GLUPRT',\n",
       " 'malatesynthase': 'MALSp',\n",
       " 'aldehydedehydrogenaseacetaldehydenadp': 'ALDD2y',\n",
       " 'histidinoldehydrogenase': 'HISTD',\n",
       " 'pyridoxal5phosphatesynthase': 'PYDXS',\n",
       " 'ketolacidreductoisomerase2aceto2hydroxybutanoate': 'KARA2im',\n",
       " 'hydroxymethylglutarylcoareductase': 'HMGCOAR',\n",
       " 'inositol145trisphosphate3kinasenucleus': 'MI145PKn',\n",
       " 'gammalinolenicacidcoaligaseampforming': 'FACOAL1831p',\n",
       " 'ap4ahydrolase': 'AP4AH',\n",
       " 'ribonucleosidediphosphatereductasegdp': 'RNDR2',\n",
       " 'llactatereversibletransportviaprotonsymport': 'L_LACt2r',\n",
       " 'guaninedeaminase': 'GUAD',\n",
       " 'indole3pyruvatedecarboxylase': 'INDPYRD',\n",
       " '4methyl2oxopentanoatedecarboxylase': '4MOPDC',\n",
       " 'lanosterolsynthase': 'LNSTLSr',\n",
       " '1phosphatidylinositol45bisphosphate5phosphataseyeastspecific': 'PIN3Pn_RT',\n",
       " 'aspartateaminotransferasecysteateforming': 'LCYSTATm',\n",
       " 'argininosuccinatelyase': 'ARGSL',\n",
       " 'acylcoazymosterolacyltransferaseyeastspecific': 'ZYMSTATer_RT',\n",
       " 'fumarase': 'FUMm',\n",
       " 'mannoseinositolphosphorylceramidesynthaseceramide126cyeastspecific': 'MIPCS126g_RT',\n",
       " 'ceramide1synthase24cendoplasmicreticular': 'CERS124er',\n",
       " 'utpumpantiport': 'UTPtm2',\n",
       " 'purinenucleosidephosphorylase': 'PNP',\n",
       " '3hydroxyacylcoadehydratase3hydroxydodecanoylcoa': 'ECOAH5p',\n",
       " 'homogentisateoxygen12oxidoreductasedecyclizing': 'HGNTOR',\n",
       " 'aldehydedehydrogenaseisobutylalcoholnad': 'ALCD23xi',\n",
       " 'lglutamate5semialdehydenadoxidoreductasearginineandprolinemetabolismec15112': 'G5SADm',\n",
       " 'malicenzymenad': 'ME1m',\n",
       " 'fructose26bisphosphate2phosphatase': 'FBP26',\n",
       " '1dmyoinositol3phosphatephosphohydrolase': 'MI3PP',\n",
       " 'homoserinedehydrogenasenadph': 'HSDy',\n",
       " '3hydroxyacylcoadehydratase3hydroxyhexacosylcoa': 'ECOAH11p',\n",
       " 'glucosetransportuniport': 'GLCt1',\n",
       " 'enoylcoahydratasec160': 'ECOAH7m',\n",
       " 'asparaginyltrnasynthetase': 'ASNTRS',\n",
       " 'adentylatekinaseitp': 'ADK4m',\n",
       " 'ltryptophanreversibletransportviaprotonsymport': 'TRPt2r',\n",
       " 'phosphopentomutase': 'PPM',\n",
       " 'ergosterolesterhydrolaseyeastspecific': 'ERGSTESTH_RT',\n",
       " 'hydrogenperoxidereductasethioredoxin': 'THIORDXi',\n",
       " '510methylenetetrahydrofolatereductasenadph': 'MTHFR3',\n",
       " 'peptidetransportinviaprotonsymport': 'PEPDt3',\n",
       " 'ornithinetransaminase': 'ORNTA',\n",
       " 'purinenucleosidephosphorylasedeoxyinosine': 'PUNP6',\n",
       " 'sulfateirreversibleuniport': 'SO4ti',\n",
       " 'nacetylgglutamylphosphatereductaseirreversible': 'AGPRim',\n",
       " '3mercaptopyruvatecyanidesulfurtransferase': 'HMR_6515',\n",
       " 'glutaminesynthetase': 'GLNS',\n",
       " 'uroporphyrinogendecarboxylaseuroporphyrinogeniii': 'UPPDC1',\n",
       " 'formatetetrahydrofolateligasemitochondrialirreversible': 'FTHFLmi',\n",
       " 'methylisocitratelyase': 'MCITL2m',\n",
       " 'pantothenatesynthase': 'PANTS',\n",
       " 'fattyacidcoaligaseoctadecanoate': 'FACOAL180p',\n",
       " 'hydroxymethylbilanesynthase': 'HMBS',\n",
       " 'acetylcoacarboxylase': 'ACCOAC',\n",
       " 'squaleneepoxidaseendoplasmicreticularnadp': 'SQLEr',\n",
       " 'lhistidinetransportinviaprotonantiport': 'HISt6',\n",
       " 'fumarylacetoacetase': 'FUMAC',\n",
       " '2aceto2hydroxybutanoatesynthase': 'ACHBSm',\n",
       " 'atpphosphoribosyltransferase': 'ATPPRT',\n",
       " 'methionineadenosyltransferase': 'METAT',\n",
       " 'mannosyloligosaccharide12alphamannosidaseasnglcnacglcnacmanmanmanmanmanmanmanman': 'MM3',\n",
       " 'linoleoylcoahydrolase1829z12z': 'FACOAE1829Z12Z',\n",
       " 'lmethioninersulfoxidereductase': 'METSOXR2',\n",
       " 'fattyacidcoathioesterasedecanoate': 'FACOAE100',\n",
       " 'isc4fe4stransfer': 'I4FE4STm',\n",
       " 'udpnacetyldmannosamineoxidoreductase': 'UACMAMO',\n",
       " 'fattyacidcoathioesterasetetradecanoate': 'FACOAE140',\n",
       " 'succinyldiaminopimelatetransaminase': 'SDPTA',\n",
       " 'dfructose6phosphatephosphatase': 'F6PP',\n",
       " 'dolichylphosphatebetaglucosyltransferase': 'UDPDOLPT',\n",
       " 'trans2enoylcoareductasenadphc120': 'ACOAR5m',\n",
       " 'xylulokinase': 'XYLK',\n",
       " 'alphaamylase': 'AAMYLASE',\n",
       " 'adenosine35phosphate5nucleotidohydrolasenucleus': 'APNn',\n",
       " 'udpglucuronatec4decarboxylase': 'UDPGDC',\n",
       " 'lipoyloctanoyltransferase': 'LIPOCTm',\n",
       " 'nucleobasetransportthioguanineextracellular': 'TGUAt',\n",
       " 'alpha12mannosyltransferasedoldpglcnacglcnacmanmanmanmanmanmanman': 'G12MT3',\n",
       " 'glucose6phosphatephosphatase': 'G6PP',\n",
       " 'nadhxdehydratase': 'NADHXD2',\n",
       " 'voltagesensitivenachannel': 'NA1tm',\n",
       " '2oalphamannosyl6phosphatedglyceratehydrolase': 'MANPGH',\n",
       " 'udplara4nformyltransferase': 'ULA4NFT',\n",
       " 'betainelipidsynthase3amino3carboxypropyltransferase1819z1819z': 'ACPT1819Z1819Z',\n",
       " 'butanolsynthase': 'BTS',\n",
       " 'equilibrativenucleosidetransporter': 'ADNtm',\n",
       " 'malonicsemialdehydereductase': 'MSAR',\n",
       " 'malonylcoamethyltransferase': 'MALCOAMT',\n",
       " 'aldehydedehydrogenasepropanalnadp': 'ALDD3y',\n",
       " 'phosphogluconatephosphatase': 'GNP',\n",
       " 'maltodextrinphosphorylasemaltohexaose': 'MLTP2',\n",
       " 'spermidineprotonantiporterreversible': 'SPMDtmr',\n",
       " 'cyanidesulfurtransferase': 'CYANST',\n",
       " 'nucleosidetriphosphatepyrophosphorylasedgtp': 'NTPP1',\n",
       " '3phosphohydroxypyruvatephosphatase': 'HPYRP',\n",
       " 'acetatekinase': 'ACKr',\n",
       " 'fattyacidcoathioesteraseoctadecanoate': 'FACOAE180',\n",
       " 'bis5adenosyltriphosphatase': 'BTP',\n",
       " 'betaalaninedipeptidase': 'BADP',\n",
       " '7hydroxysteroiddehydrogenase': 'CHOLID',\n",
       " 'pyridoxinenadp4oxidoreductase': 'PYDXOR',\n",
       " 'ureidoglycolatehydrolase': 'UGLYCH',\n",
       " '56dihydro5fluorouracilamidohydrolase': '56DH5FLURAAMH',\n",
       " 'isc2fe2ssynthesisii': 'I2FE2SS2m',\n",
       " '25diketodgluconatereductase': 'DKGLCNR1',\n",
       " '34dihydroxyphenylacetateametomethyltransferase': '34DHPHAMT',\n",
       " 'aminophospholipidtranslocasepgp1819z160extracellular': 'PGP1819Z160t',\n",
       " 'exo13betaglucanglucohydrase': '13BGHe',\n",
       " 'lacaldehydereductasespropane12diolforming': 'LCARSyi',\n",
       " 'trehalose6phosphatehydrolase': 'TRE6PH',\n",
       " 'acetylcoaacetylcoacacetyltransferase': 'ACACT1p',\n",
       " '3hydroxycinnamatehydroxylase': '3HCINNMH',\n",
       " 'isc2fe2ssynthesis': 'I2FE2SSm',\n",
       " 'dihydropteroatesynthase': 'DHPS2',\n",
       " 'uroporphyrinogendecarboxylaseuroporphyrinogeni': 'UPPDC2',\n",
       " 'nadphdependentdihydrocurcuminreductase': 'DHCURR',\n",
       " 'fattyacidcoathioesterasehexadecenoate': 'FACOAE161',\n",
       " 'hemecligase': 'HEMELm',\n",
       " 'formamidase': 'FORA',\n",
       " 'uratepermease': 'URATEt_1',\n",
       " 'glyceratekinase': 'GLYCK',\n",
       " 'lipoyladenylateproteinligase': 'LIPAMPLm',\n",
       " 'udpglcnacdolichylphosphatenacetylglucosaminephosphotransferase': 'GLCNACPT',\n",
       " 'thiaminphosphataseextracellular': 'THMPe',\n",
       " 'cuexportingatpase': 'Cu1ATPase',\n",
       " 'formyltetrahydrofolatedeformylase': 'FTHFD',\n",
       " 'gammalinolenoylcoahydrolase1836z9z12z': 'FACOAE1836Z9Z12Z',\n",
       " 'glutaredoxinreductase': 'GRXRm',\n",
       " 'dtdp4dehydrorhamnosereductase': 'TDPDRR',\n",
       " 'glucan13alphaglucosidaseasnglcnacglcnacmanmanmanmanmanmanmanmanmanglca': 'MG3A',\n",
       " 'glycerophosphodiesterphosphodiesteraseglycerophosphoethanolamine': 'GPDDA2',\n",
       " 'nucleosidetriphosphatepyrophosphorylasegtp': 'NTPP2',\n",
       " 'acetylcoaoxalatecoatransferase': 'ACOXT',\n",
       " 'aminophospholipidtranslocasepe1819z1819zextracellular': 'PE1819Z1819Zt',\n",
       " 'aminophospholipidtranslocasepail1819z160extracellular': 'PAIL1819Z160t',\n",
       " 'phenylhydantoinase': 'PHDA',\n",
       " 'betainelipidsynthasetrimethylase1601819z': 'TM1601819Z',\n",
       " 'purinenucleosidephorphoramidase': 'PNSPA',\n",
       " 'acetamidase': 'AMID_1',\n",
       " 'trans2enoylcoareductasenadphc140': 'ACOAR6m',\n",
       " 'adenosine5phosphosulfatereductase': 'APSR',\n",
       " '6mercaptopurinephosphoribosyltransferase': '6MPURPRT',\n",
       " 'ethanolaminekinase': 'ETHAK',\n",
       " 'nadhhydratase': 'NADHHR',\n",
       " 'chitobiosyldiphosphodolicholbetamannosyltransferase': 'BDMT',\n",
       " '5dehydrodgluconatereductase': '5DGLCNR',\n",
       " 'acylcoadehydrogenasehexanoylcoa': 'ACOAD2p',\n",
       " 'mannosyloligosaccharide12alphamannosidaseasnglcnacglcnacmanmanmanmanman': 'MM1',\n",
       " 'fattyacidcoathioesteraseoctanoate': 'FACOAE80',\n",
       " 'fattyacidcoathioesteraseoctadecenoate': 'FACOAE181',\n",
       " 'isccysteinedesulfuration': 'ICYSDSm',\n",
       " 'isc2fe2sregeneration': 'I2FE2SRm',\n",
       " 'enoylcoahydratasec60': 'ECOAH2m',\n",
       " '1acylglycerolacylhydrolase160': 'MAGAH160',\n",
       " 'aminophospholipidtranslocasepg1819z160extracellular': 'PG1819Z160t',\n",
       " '6thioinosine5monophosphatenadoxidoreductase': '6TINS5MPOR',\n",
       " '3hydroxybutanoatenadoxidoreductase': 'BDHm',\n",
       " 'acylcoadehydrogenasebutanoylcoa': 'ACOAD1m',\n",
       " 'cyanateaminohydrolase': 'CYNTAH',\n",
       " 'trans2enoylcoareductasenadphc100': 'ACOAR4m',\n",
       " 'arabinose5phosphateisomerase': 'A5PISO',\n",
       " 'nucleobasetransportfluorouracilextracellular': '5FLURAt',\n",
       " 'glyoxylatereductasenad': 'GLYCLTDxm',\n",
       " 'glycerophosphodiesterphosphodiesteraseglycerophosphoserine': 'GPDDA3',\n",
       " 'ribose5phosphatephosphatase': 'R5PP',\n",
       " 'shydroxyoctanoylcoanadoxidoreductase': 'HACD3p',\n",
       " 'alphaalphatrehalaseglca': 'AATHA',\n",
       " 'dolichylphosphatebetadmannosyltransferase': 'DOLPMT',\n",
       " 'betainelipidsynthase3amino3carboxypropyltransferase1601819z': 'ACPT1601819Z',\n",
       " 'folatereductase': 'FOLR2_1',\n",
       " 'spermidineacetyltransferase': 'SPMDAT1',\n",
       " 'glycinecacetyltransferase': 'GLYATm',\n",
       " 'dihydrofolatereductase': 'DHFRim',\n",
       " 'carboxylicaciddissociation': 'H2CO3D',\n",
       " 'sulfatetransportinviaprotonsymport': 'SO4t2',\n",
       " 'ornithinetransacetylase': 'ORNTAC',\n",
       " 'phosphoribosylanthranilateisomeraseirreversible': 'PRAIi',\n",
       " 'lipoateatpadenylatetransferase': 'LIPATPTm',\n",
       " 'glycerol3phosphatedehydrogenasefad': 'G3PD_u9cm',\n",
       " 'udpdgalactosenad6oxidoreductase': 'UDPGALOR',\n",
       " 'glucose1phosphatase': 'G1PP',\n",
       " 'dserinedehydrogenase': 'DSERDHr',\n",
       " 'hydroxypyruvatereductasenad': 'HPYRRx',\n",
       " 'transtranscisgeranylgeranyldiphosphatesynthase': 'FT',\n",
       " 'hydrogencarbonatelglutamineamidoligase': 'CBPSm',\n",
       " 'trans2enoylcoareductasenadphc160': 'ACOAR7m',\n",
       " 'glucan13alphaglucosidaseasnglcnacglcnacmanmanmanmanmanmanmanmanmanglcb': 'MG3B',\n",
       " 'nacetylglutamatesynthase': 'ACGS',\n",
       " 'aldose1epimerase': 'A1E',\n",
       " 'nadhxepimerase': 'NADHXE',\n",
       " 'formamidaseglyoxysomal': 'FORAMDx',\n",
       " 'mannosyloligosaccharide1316alphamannosidaseasnglcnacglcnacmanmanman': 'MM2',\n",
       " 'mannosyloligosaccharideglucosidaseasnglcnacglcnacmanmanmanmanmanmanmanmanmanglcglcglcb': 'MG1B',\n",
       " 'betainelipidsynthasetrimethylase1819z1819z': 'TM1819Z1819Z',\n",
       " 'nadphdependentcurcuminreductase': 'CURR',\n",
       " 'erythrose4phosphatephosphatase': 'E4PP',\n",
       " 'gudbutnah': 'GUDBUTNAH',\n",
       " '1acylglycerolacylhydrolase180': 'MAGAH180',\n",
       " 'nucleobasetransport6mercaptopurineextracellular': '6MPURt',\n",
       " 'ribulose5phosphatephosphatase': 'RU5PP',\n",
       " 'ralphahydroxyglutaratedehydrogenase': 'GHBTRHDm',\n",
       " 'inositol13456pentakisphosphate2kinasenuclear': 'MI13456PKn',\n",
       " 'phosphatetransportnuclear': 'PItn',\n",
       " '3oxoadipateenollactonehydrolase': 'OXOAEL',\n",
       " '5ketodgluconate5reductase': '5DKGR',\n",
       " 'fattyacidcoathioesterasehexanoate': 'FACOAE60',\n",
       " 'nicotinicacidmononucleotidepyrophosphorylase': 'NAMNPP',\n",
       " 'glycerophosphodiesterphosphodiesteraseglycerophosphoglycerol': 'GPDDA4',\n",
       " 'shydroxyhexanoylcoahydrolyase': 'ECOAH2p',\n",
       " 'tryptophansynthaseindole': 'TRPS2',\n",
       " 'maltodextrinphosphorylasemaltopentaose': 'MLTP1',\n",
       " 'urateoxidase': 'URO',\n",
       " 'serinedecarboxylase': 'SERDC',\n",
       " 'ferrousironuptakegtpdependentplasmamembrane': 'FE2GTPabc',\n",
       " 'dolichyldiphosphooligosaccharideproteinglycotransferase': 'DOLASNT',\n",
       " '10formyltetrahydrofolatelglutamateligaseadpforming': '10FTHFGLULLm',\n",
       " 'ferricironuptakeplasmamembrane': 'FE3t',\n",
       " 'udplarabinose4epimerase': 'UA4E',\n",
       " '1acylglycerolacylhydrolase1819z': 'MAGAH1819Z',\n",
       " 'aminophospholipidtranslocasepe1801829z12zextracellular': 'PE1801829Z12Zt',\n",
       " 'lipoatesynthase': 'LIPOSm',\n",
       " 'nadphxdehydratase': 'NADPHXD2',\n",
       " 'alpha12mannosyltransferasedoldpglcnacglcnacmanmanmanmanmanmanmanmanman': 'G12MT4',\n",
       " 'h2otransport': 'H2Otm',\n",
       " 'dihydooroticacidubiquinone6irreversible': 'DHORD_u9m',\n",
       " '6thioxanthine5monophosphatelglutamineamidoligase': '6TXAN5MPAML',\n",
       " 'nacetylgglutamylphosphatereductase': 'AGPR',\n",
       " 'alpha16mannosyltransferasedoldpglcnacglcnacmanmanman': 'G16MT',\n",
       " 'shydroxyhexanoylcoanadoxidoreductase': 'HACD2p',\n",
       " 'thioguaninephosphoribosyltransferase': 'TGUAPRT',\n",
       " 'fattyacidcoathioesterasedodecanoate': 'FACOAE120',\n",
       " 'fattyacidcoathioesterasehexadecanoate': 'FACOAE160',\n",
       " 'acireductonedioxygenase': 'ACDO',\n",
       " 'isc4fe4sreduction': 'I4FE4SRm',\n",
       " 'trans2enoylcoareductasenadphc60': 'ACOAR2m',\n",
       " 'dfructose1phosphatephosphatase': 'F1PP',\n",
       " 'cholinedehydrogenase': 'CHOLD',\n",
       " 'superoxidedismutase': 'SPODMm',\n",
       " 'ammoniaperoxisomaltransport': 'NH4tp',\n",
       " 'aminophospholipidtranslocasepe1819z1829z12zextracellular': 'PE1819Z1829Z12Zt',\n",
       " 'acylcarrierproteinsynthase': 'ACPS1',\n",
       " 'enoylcoahydratasec80': 'ECOAH3m',\n",
       " 'glycinetrnaligase': 'GLYTLm',\n",
       " '5diphosphoinositol12346pentakisphosphatesynthase': 'PMI12346PSn',\n",
       " 'mannosyloligosaccharideglucosidaseasnglcnacglcnacmanmanmanmanmanmanmanmanmanglcglcglca': 'MG1A',\n",
       " 'lseryladenylatesynthase': 'SERASr',\n",
       " 'spermidineacetyltransferasen8': 'SPMDAT2',\n",
       " 'aminophospholipidtranslocasepe1801819zextracellular': 'PE1801819Zt',\n",
       " 'nucleobasetransporttegafurextracellular': 'TEGAt',\n",
       " 'nucleosidetriphosphatepyrophosphorylasedttp': 'NTPP7',\n",
       " 'hexanoylcoaacetylcoacacyltransferase': 'ACACT3p',\n",
       " 'alpha12mannosyltransferasedoldpglcnacglcnacmanmanmanmanman': 'G12MT2',\n",
       " 'octanoatenonlipoylatedapodomainligase': 'OCTNLLm',\n",
       " 'glucan13alphaglucosidaseasnglcnacglcnacmanmanmanmanmanmanmanmanmanglcglcb': 'MG2B',\n",
       " 'nadphxepimerase': 'NADPHXE',\n",
       " 'glycerol2phosphatephosphatase': 'G2PP',\n",
       " 'alcoholdehydrogenaseglycerol': 'ALCD19',\n",
       " 'biotinsynthase': 'BTS5m',\n",
       " 'butanolsynthasenadph': 'BTS_nadph',\n",
       " ...}"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_dict_sce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7937841c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c082945",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "49bb2e6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110 110\n",
      "102\n"
     ]
    }
   ],
   "source": [
    "# here we create two lists that have the model IDs from the sce scaffold and the modified reaction names. \n",
    "\n",
    "# holders for the names and IDs.\n",
    "temp = []\n",
    "temp2 = []\n",
    "model_temp_id_sce = []\n",
    "\n",
    "# iterate through the sce modified reaction names. \n",
    "for i,r in enumerate(sce_rxns_name):\n",
    "    \n",
    "    # check if the modified sce modified reaction name matches the model modified reaction name. \n",
    "    if r in model_rxns_name:\n",
    "        \n",
    "        # append the modified reaction name to temp.\n",
    "        temp.append(r)\n",
    "        \n",
    "        # append the reaction _id to temp 2. \n",
    "        temp2.append(_id[i])\n",
    "        \n",
    "        # append the Lst model reaction ID that corresponds to the reaction name\n",
    "        model_temp_id_sce.append(model_dict_sce[r])\n",
    "\n",
    "# output should be 110 and 102. \n",
    "print(len(temp),len(temp2))\n",
    "print(len(set(temp)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "d1b6a912",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary of the modified reaction names to the model reaction IDs.\n",
    "\n",
    "# dictionary of modified reaction names to the model reaction ID. \n",
    "sce_additional_rxns_toModel = dict(zip(temp,model_temp_id_sce))\n",
    "\n",
    "# dictionary of modified reaction names to the sce GSM reaction ID. \n",
    "sce_additional_rxns = dict(zip(temp,temp2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "065ce254",
   "metadata": {},
   "source": [
    "### Repeat this process for *Y. lipolytica* GSM. \n",
    "\n",
    "Determine reactins in ylip that share names with the model reactions, but have diferent IDs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "632ae11e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output the ortholog from Ylip to csv to validate reactions\n",
    "_names = []\n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "\n",
    "# gather the reaction details from the possible ortholog reactions. \n",
    "for rx in ylip_ortholog_available:\n",
    "    r = ylip.reactions.get_by_id(rx)\n",
    "    _names.append(r.name)\n",
    "    _gpr.append(r.gene_reaction_rule)\n",
    "    _id.append(r.id)\n",
    "    _stoich.append(r.reaction)\n",
    "\n",
    "# store the ylip GSM names here. \n",
    "ylip_rxns_name = _names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "a81b861a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gather the reaction names in the model.\n",
    "# note - this variable name is re-used from the sce section and will be overwritten by running this cell. \n",
    "model_rxns_name = model.reactions\n",
    "model_rxns_name = [x.name for x in model_rxns_name] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "e1079319",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace naming conventions that can be 'messy'. \n",
    "model_rxns_name = [re.sub(r' c$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r' v$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [x.replace(\" \",\"\").replace('-','').replace(':',\"\").replace(',',\"\").replace(')',\"\").replace(')',\"\").replace('\\'',\"\").replace('+',\"\").replace('mitochondrialmembrane',\"\").lower() for x in model_rxns_name]\n",
    "\n",
    "model_rxns_name = [re.sub(r'[^a-zA-Z0-9]', '', x) for x in model_rxns_name]\n",
    "\n",
    "model_rxns_name = [x.replace('vacuolar',\"\").replace('vacuolarmembrane',\"\") for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'mitochondrial$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'mitochondria$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cellenvelope$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cytoplasm$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cytoplasmic$',\"\",x) for x in model_rxns_name]\n",
    "\n",
    "model_rxns_name = [re.sub(r'peroxisomal$',\"\",x) for x in model_rxns_name]\n",
    "model_rxns_name = [re.sub(r'cytoplasmic$',\"\",x) for x in model_rxns_name]\n",
    "\n",
    "# sce_rxns_name = [x.replace('mitochondrialmembrane',\"\").replace('golgimembrane',\"\").replace('ermembrane',\"\") for x in sce_rxns_name]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62fb778f",
   "metadata": {},
   "source": [
    "replace the ylip naming conventions that can be 'messy'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "6ede0add",
   "metadata": {},
   "outputs": [],
   "source": [
    "ylip_rxns_name = [re.sub(r' c$',\"\",x) for x in ylip_rxns_name]\n",
    "ylip_rxns_name = [re.sub(r' v$',\"\",x) for x in ylip_rxns_name]\n",
    "\n",
    "ylip_rxns_name = [x.replace(\" \",\"\").replace('-','').replace(':',\"\").replace(',',\"\").replace(')',\"\").replace(')',\"\").replace('\\'',\"\").replace('+',\"\").replace('mitochondrialmembrane',\"\").lower() for x in ylip_rxns_name]\n",
    "\n",
    "ylip_rxns_name = [re.sub(r'[^a-zA-Z0-9]', '', x) for x in ylip_rxns_name]\n",
    "\n",
    "ylip_rxns_name = [x.replace('mitochondrialmembrane',\"\").replace('golgimembrane',\"\").replace('ermembrane',\"\") for x in ylip_rxns_name]\n",
    "ylip_rxns_name = [x.replace('cellenvelopecytoplasm',\"\").replace('vacuolarmembrane',\"\") for x in ylip_rxns_name]\n",
    "ylip_rxns_name = [re.sub(r'mitochondrial$',\"\",x) for x in ylip_rxns_name]\n",
    "ylip_rxns_name = [re.sub(r'mitochondria$',\"\",x) for x in ylip_rxns_name]\n",
    "ylip_rxns_name = [re.sub(r'cellenvelope$',\"\",x) for x in ylip_rxns_name]\n",
    "ylip_rxns_name = [re.sub(r'cytoplasm$',\"\",x) for x in ylip_rxns_name]\n",
    "ylip_rxns_name = [re.sub(r'cytoplasmic$',\"\",x) for x in ylip_rxns_name]\n",
    "\n",
    "# sce_rxns_name = [x.replace('mitochondrialmembrane',\"\").replace('golgimembrane',\"\").replace('ermembrane',\"\") for x in sce_rxns_name]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6298f90a",
   "metadata": {},
   "source": [
    "260 reactions have orthologs in the ylip model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "f61fe2d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "260"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# length of reactions with orthologs only in the ylip model.\n",
    "len(ylip_rxns_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "5c4d9235",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dictionary of the replaced model reaction names to the reaction IDs. \n",
    "model_dict = dict(zip(model_rxns_name,model_rxns_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "83cf4616",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "108 108\n",
      "101\n"
     ]
    }
   ],
   "source": [
    "# here we create two lists that have the model IDs from the sce scaffold and the modified reaction names. \n",
    "\n",
    "# holders for the names and IDs.\n",
    "temp = []\n",
    "temp2 = []\n",
    "model_temp_id = []\n",
    "\n",
    "# iterate through the ylip modified reaction names. \n",
    "for i,r in enumerate(ylip_rxns_name):\n",
    "\n",
    "    # check if the modified sce modified reaction name matches the model modified reaction name. \n",
    "    if r in model_rxns_name:\n",
    "\n",
    "        # append the reaction _id to temp 2. \n",
    "        temp.append(r)\n",
    "        \n",
    "        # append the reaction _id to temp 2. \n",
    "        temp2.append(_id[i])\n",
    "        \n",
    "        # append the Lst model reaction ID that corresponds to the reaction name\n",
    "        model_temp_id.append(model_dict[r])\n",
    "\n",
    "print(len(temp),len(temp2))\n",
    "print(len(set(temp)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "ce263789",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary of the modified reaction names to the model reaction IDs. \n",
    "\n",
    "# dictionary of modified reaction names to the model reaction ID. \n",
    "ylip_additional_rxns_toModel = dict(zip(temp,model_temp_id))\n",
    "\n",
    "# dictionary of modified reaction names to the sce GSM reaction ID. \n",
    "ylip_additional_rxns = dict(zip(temp,temp2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "5d3175ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary of the modified reaction names to the ylip GSM reaction IDs. \n",
    "\n",
    "ylip_additional_rxns = dict(zip(temp,temp2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "b00d24ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ICL'"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip_additional_rxns['isocitratelyase']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "aa2fa487",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this cell contains a modified function that was previously used to compare three lists. \n",
    "# this cell instead will take two lists of genes and determine if any of the genes overlap or are unique.\n",
    "# the function returns the unique genes. \n",
    "\n",
    "# definition to compare the genes in common reactions. \n",
    "def compare_two_lists(list1, list2):\n",
    "\n",
    "    # Convert lists to sets.\n",
    "    set1, set2 = set(list1), set(list2)\n",
    "    \n",
    "    # Find all intersections and differences.\n",
    "    common_all = set1 & set2\n",
    "        \n",
    "    only_1 = set1 - set2\n",
    "    only_2 = set2 - set1\n",
    "    \n",
    "    all_diff = (set1 | set2 ) - common_all - only_1 - only_2\n",
    "    \n",
    "    # Construct verbose explanation.\n",
    "    explanation = {}\n",
    "\n",
    "    \n",
    "    explanation['common_all'] = list(common_all)\n",
    "\n",
    "    explanation['only_1'] = list(only_1)\n",
    "    explanation['only_2'] = list(only_2)\n",
    "    explanation['all_diff'] = list(all_diff)\n",
    "    \n",
    "    \n",
    "    return explanation\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6948ec07",
   "metadata": {},
   "source": [
    "## This section uses the collected reaction names and modifies the Lst model GPR as needed. \n",
    "following cell updates the model GPR for overlapping reactions in ylip."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "85e35f2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Lipst1_1_112377']\n",
      "['Lipst1_1_112377', 'Lipst1_1_75138']\n",
      "{'common_all': ['Lipst1_1_112377'], 'only_1': [], 'only_2': ['Lipst1_1_75138'], 'all_diff': []}\n",
      "[['Lipst1_1_112377'], [], ['Lipst1_1_75138']]\n",
      "Lipst1_1_112377 or Lipst1_1_75138\n",
      "YALI0C16885g or YALI0F31999g\n"
     ]
    }
   ],
   "source": [
    "# this cell updates the model GPR with the overlapping ylip reactions. \n",
    "\n",
    "# storage lists. \n",
    "# gene overlap in the ylip orthologs and the model genes for each reaction. \n",
    "_names = []\n",
    "_gpr_sce = []\n",
    "_gpr_ylip = []\n",
    "_gpr_model = []\n",
    "\n",
    "_lst_model_gene = []\n",
    "_lst_sce_gene = []\n",
    "_lst_ylip_gene = []\n",
    "\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "_lst_gene = []\n",
    "_ylip_gene = []\n",
    "\n",
    "\n",
    "consensus = []\n",
    "consensus_count = []\n",
    "consensus_source = []\n",
    "_output = []\n",
    "\n",
    "\n",
    "_common_all = []\n",
    "_only_1 = []\n",
    "_only_2 = []\n",
    "_all_diff = []\n",
    "\n",
    "# iterate through the modified reaction names that are in ylip and the Lst model. \n",
    "for rx in list(ylip_additional_rxns.keys()):\n",
    "    \n",
    "    # get actual GSM IDs.\n",
    "    model_rx = ylip_additional_rxns_toModel[rx]\n",
    "    ylip_rx = ylip_additional_rxns[rx]\n",
    "    \n",
    "    # obtain the reaction objects.\n",
    "    m = model.reactions.get_by_id(model_rx)\n",
    "    y = ylip.reactions.get_by_id(ylip_rx)\n",
    "    \n",
    "    # obtain the reaction GPRs.\n",
    "    _gpr_model = (m.gene_reaction_rule)\n",
    "    _gpr_ylip = (y.gene_reaction_rule)\n",
    "    \n",
    "#     # check. \n",
    "#     if y.id == 'ICL':\n",
    "#         print(_gpr_ylip,_gpr_model)\n",
    "\n",
    "    # obtain the reaction meta data.\n",
    "    _names.append(m.name)\n",
    "    _id.append(m.id)\n",
    "    _stoich.append(m.reaction)\n",
    "    _subsystem.append(m.subsystem)\n",
    "    \n",
    "    ## to do - insert check to make sure not adding multiple GPRs. \n",
    "    \n",
    "    # Lst model current GPR.\n",
    "    temp1 = []\n",
    "    for g in m.genes:\n",
    "        temp1.append(g.id)\n",
    "    _lst_model_gene.append(temp1)\n",
    "    if y.id == 'ICL':\n",
    "        print(temp1)\n",
    "\n",
    "    \n",
    "    # ylip genes. \n",
    "    temp3 = []\n",
    "    gpr = y.genes\n",
    "    \n",
    "    # iterate over each individual gene. \n",
    "    for i in range(len(gpr)):\n",
    "        g=list(gpr)[i]\n",
    "\n",
    "        # examine lst_orthologs. \n",
    "        if g.id in ylip_dict_keys:\n",
    "\n",
    "            # obtain the lst gene ortholog.\n",
    "            value = lst_ylip_dict[g.id]\n",
    "\n",
    "            # check if the ortholog is blank. \n",
    "            if value:\n",
    "                temp3.append(value)\n",
    "#                 count_ortho+=1\n",
    "\n",
    "    if y.id == 'ICL':\n",
    "        print(temp3)\n",
    "    _lst_ylip_gene.append(temp3)\n",
    "    \n",
    "    # call the function to compare the genes in each GSM reaction, which are stored as list temp1 and temp3. \n",
    "    result = compare_two_lists(temp1, temp3)\n",
    "    \n",
    "    if y.id == 'ICL':\n",
    "        print(result)\n",
    "\n",
    "    # assign the results here. \n",
    "    _common_all = (result['common_all'])\n",
    "    _only_1 = (result['only_1'])\n",
    "    _only_2 = (result['only_2'])\n",
    "    _all_three_diff = (result['all_diff'])\n",
    "    \n",
    "    # append them into a single list.\n",
    "    trifecta = []\n",
    "    trifecta.append(_common_all)\n",
    "    trifecta.append(_only_1)\n",
    "    trifecta.append(_only_2)\n",
    "    \n",
    "    if y.id =='ICL':\n",
    "        print(trifecta)   \n",
    "\n",
    "    # if there is a new gene that needs to be added. \n",
    "    if result['only_2']:\n",
    "        \n",
    "        \n",
    "        max_len = max(len(lst) for lst in trifecta)\n",
    "        \n",
    "        # join them as 'or' gpr rules. \n",
    "        for i in range(max_len):\n",
    "            temp_gpr = ' or '.join(item for sublist in trifecta for item in sublist)\n",
    "    \n",
    "    # else keep the original gene reaction rule. \n",
    "    else:\n",
    "        temp_gpr = m.gene_reaction_rule\n",
    "\n",
    "    if y.id == 'ICL':\n",
    "        print(temp_gpr)\n",
    "        print(y.gene_reaction_rule)\n",
    "\n",
    "    # assign new GPR. \n",
    "    m.gene_reaction_rule = temp_gpr\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "ea7bd9d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>model</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>15a49dd50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1121</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1348</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>646</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>69</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*biomass_C - 1.0*biomass_C_reverse_c1d5c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>c, e, m, n, x, r, g, v</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model model at 0x15a49dd50>"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "4143605c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ICLp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Isocitrate lyase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a1395150</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>icit_x --> glx_x + succ_x</p>\n",
       "                    <p style='text-align:right'>Isocitrate --> Glyoxylate + Succinate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_112377 or Lipst1_1_75138</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ICLp at 0x2a1395150>"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('ICLp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "14db83c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)-2083"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "17eacf56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.07253593170088728</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>6.035E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>3.808E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0004475</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.04248</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.004477</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>4.403E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.001893</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4788</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.329</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.01931</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.007271</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.699E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001754</td>\n",
       "      <td>9</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0002254</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.04189</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.01064</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.007832</td>\n",
       "      <td>10</td>\n",
       "      <td>1.29%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.01431</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002486</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.008552</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01514</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.03723</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.004437</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.0115</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.02323</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.004138</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.01711</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.02444</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.01582</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.001691</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.004524</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.01996</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3aap_c</td>\n",
       "      <td>DM_3aap_c</td>\n",
       "      <td>-0.05137</td>\n",
       "      <td>5</td>\n",
       "      <td>8.21%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-0.000101</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001754</td>\n",
       "      <td>15</td>\n",
       "      <td>0.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001754</td>\n",
       "      <td>10</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-2.599</td>\n",
       "      <td>1</td>\n",
       "      <td>83.12%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.0003107</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.242</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1659</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001818</td>\n",
       "      <td>1</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.009788</td>\n",
       "      <td>27</td>\n",
       "      <td>8.45%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a6973e90>"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "f7fd76d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "45a55573",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "\n",
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g2)))\n",
    "print(len(set(wrong_rx2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "92eb6958",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06be1168",
   "metadata": {},
   "source": [
    "## This section uses the collected reaction names and modifies the Lst model GPR as needed. \n",
    "following cell updates the model GPR for overlapping reactions in sce.\n",
    "#### sce"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affb2491",
   "metadata": {},
   "source": [
    "following cell updates gpr from SCE reactions that have naming overlaps. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "e23660e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list: ['Lipst1_1_112377'] [] ['Lipst1_1_75138']\n",
      "trifecta [['Lipst1_1_112377'], ['Lipst1_1_75138'], []]\n",
      "Lipst1_1_112377 or Lipst1_1_75138\n"
     ]
    }
   ],
   "source": [
    "# this cell updates the model GPR with the overlapping ylip reactions. \n",
    "\n",
    "# storage lists. \n",
    "# gene overlap in the sce orthologs and the model genes for each reaction. \n",
    "_names = []\n",
    "_gpr_sce = []\n",
    "_gpr_ylip = []\n",
    "_gpr_model = []\n",
    "\n",
    "_lst_model_gene = []\n",
    "_lst_sce_gene = []\n",
    "_lst_ylip_gene = []\n",
    "\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "_lst_gene = []\n",
    "_ylip_gene = []\n",
    "\n",
    "\n",
    "consensus = []\n",
    "consensus_count = []\n",
    "consensus_source = []\n",
    "_output = []\n",
    "\n",
    "\n",
    "_common_all = []\n",
    "\n",
    "_only_1 = []\n",
    "_only_2 = []\n",
    "_all_diff = []\n",
    "\n",
    "# iterate through the sce reactions that have orthologs and are also in the model reactions (based on name matching).\n",
    "for rx in list(sce_additional_rxns.keys()):\n",
    "    \n",
    "    # get actual GSM IDs.\n",
    "    model_rx = sce_additional_rxns_toModel[rx]\n",
    "    sce_rx = sce_additional_rxns[rx]\n",
    "    \n",
    "    # obtain the reaction objects.\n",
    "    m = model.reactions.get_by_id(model_rx)\n",
    "    s = sce.reactions.get_by_id(sce_rx)\n",
    "    \n",
    "    # obtain the reaction GPRs.\n",
    "    _gpr_model.append(m.gene_reaction_rule)\n",
    "    _gpr_sce.append(s.gene_reaction_rule)\n",
    "\n",
    "    # obtain the reaction meta data.\n",
    "    _names.append(m.name)\n",
    "    _id.append(m.id)\n",
    "    _stoich.append(m.reaction)\n",
    "    _subsystem.append(m.subsystem)\n",
    "    \n",
    "    ## to do - insert check to make sure not adding multiple GPRs. \n",
    "    \n",
    "    # model current GPR.\n",
    "    temp1 = []\n",
    "    for g in m.genes:\n",
    "        temp1.append(g.id)\n",
    "    _lst_model_gene.append(temp1)\n",
    "    \n",
    "\n",
    "    # sce genes. \n",
    "    temp3 = []\n",
    "    gpr = s.genes\n",
    "    \n",
    "    # iterate through the genes in the sce GSM. \n",
    "    for i in range(len(gpr)):\n",
    "        g=list(gpr)[i]\n",
    "\n",
    "        # examine lst_orthologs. \n",
    "        if g.id in sce_dict_keys:\n",
    "\n",
    "            # obtain the lst gene ortholog.\n",
    "            value = lst_sce_dict[g.id]\n",
    "\n",
    "            # check if the ortholog is blank. \n",
    "            if value:\n",
    "                temp3.append(value)\n",
    "#                 count_ortho+=1\n",
    "\n",
    "    _lst_sce_gene.append(temp3)\n",
    "    \n",
    "\n",
    "    # call comparing function for two lists. \n",
    "    result = compare_two_lists(temp1, temp3)\n",
    "\n",
    "    # assign the results. \n",
    "    _common_all = (result['common_all'])\n",
    "    _only_1 = (result['only_1'])\n",
    "    _only_2 = (result['only_2'])\n",
    "    _all_three_diff = (result['all_diff'])\n",
    "    \n",
    "    # concate into a single list. \n",
    "    trifecta = []\n",
    "    trifecta.append(_common_all)\n",
    "    trifecta.append(_only_1)\n",
    "    trifecta.append(_only_2)\n",
    "    \n",
    "    # check. \n",
    "    if s.id == 'ICL':\n",
    "        print(\"list:\",_common_all,_only_2,_only_1)\n",
    "        print(\"trifecta\",trifecta)\n",
    "    \n",
    "    # if there is a new gene to add. \n",
    "    if result['only_2']:\n",
    "        max_len = max(len(lst) for lst in trifecta)\n",
    "        \n",
    "        # concate to new gpr with 'or' gene reaction rules. \n",
    "        for i in range(max_len):\n",
    "            temp_gpr = ' or '.join(item for sublist in trifecta for item in sublist)\n",
    "                \n",
    "    # no new genes - keep original reaction gene reaction rule. \n",
    "    else:\n",
    "        temp_gpr = m.gene_reaction_rule\n",
    "\n",
    "    # check. \n",
    "    if s.id =='ICL':\n",
    "        print(temp_gpr)\n",
    "\n",
    "    # assign new GPR. \n",
    "    m.gene_reaction_rule = temp_gpr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "a4fe25c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2148"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "fccff237",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2148-2083"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "07da3571",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44\n",
      "44\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "print(len(set(wrong)))\n",
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g2)))\n",
    "print(len(set(wrong_rx2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f799689f",
   "metadata": {},
   "source": [
    "## Progess status.\n",
    "\n",
    "We:\n",
    "- created a scaffold GSM originally built from rto bidirectional best hits.\n",
    "- compiled orthologs using orthoMCL.\n",
    "\n",
    "then we: \n",
    "- obtained reactions in multiple GSMs.\n",
    "- concated GPRs using the unique genes in the overlapping reaction IDs.\n",
    "- updated the Lst bbh model with the multi-species GPRs. \n",
    "- added rto reactions to the Lst model that were only in rto AND had orthologs. \n",
    "- matched ylip and sce reaction names to find other overlapping reactions (multi-species).\n",
    "- modifed Lst model with multi-species GPRs. \n",
    "\n",
    "Next we will:\n",
    "- deterimine ylip/sce reactions that have Lst orthologs that ARE not in the GSM (i.e., genes not already in the model). \n",
    "- identify reactions that include metabolites already in the Lst GSM (such that we do not add orphan metabolites).\n",
    "- screen reactions to make sure that there are not comparable reactions in the Lst GSM.\n",
    "- update GPR with multi-species GPRs or add reactions as necessary. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2f0e8ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "096ea2df",
   "metadata": {},
   "source": [
    "## section peruses the ylip GSM reactions and identifies ones with orthologs to Lst that are not included in the Lst model already. \n",
    "ylip - individual reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "917737be",
   "metadata": {},
   "source": [
    "the following cell blocks find reactions from ylip that have orthologs not yet added to the model. we create a dataframe with these reactions. then, we determine if the metabolites in the reactions were in the model before adding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "a74c124a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, obtain the Lst genes included in the Lst model.\n",
    "model_genes = model.genes\n",
    "model_genes = [x.id for x in model_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "547fd578",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    }
   ],
   "source": [
    "# this code finds reactions from ylip to the model that may have potential orthologs that have not been included yet. \n",
    "\n",
    "# list storage. \n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "gpr_holder = []\n",
    "\n",
    "# obtain list of reactions that had orthologs but were not added.\n",
    "# (the ylip_ortholog_available was obtained from the ortholog section looking at individual models above).\n",
    "# (the ylip_additional_rxns was obtained by through matching of the reaction names to Lst model).\n",
    "new_ylip_only_reactions = set(ylip_ortholog_available)-set(list(ylip_additional_rxns.values()))\n",
    "\n",
    "# iterate through reactions with Lst ortholog gene candidates. \n",
    "for rx in new_ylip_only_reactions:\n",
    "    \n",
    "    # obtain reaction in ylip GSM.\n",
    "    r = ylip.reactions.get_by_id(rx)\n",
    "    \n",
    " \n",
    "    \n",
    "    # counters and list tracker for new GPR. \n",
    "    count_ortho = 0\n",
    "    temp_gpr = []\n",
    "    \n",
    "    # iterate through the genes in the ylip reaction.\n",
    "    for i in range(len(r.genes)):\n",
    "        \n",
    "        # obtain a list of the ylip genes. \n",
    "        g=list(r.genes)[i]\n",
    "        g = g.id\n",
    "        \n",
    "        # check if the particular gene has an Lst ortholog. \n",
    "        try:\n",
    "            value = lst_ylip_dict[g]\n",
    "            \n",
    "            # if there is an ortholog &\n",
    "            # the gene is not in the Lst model &\n",
    "            # the reaction is in the list (assured based on iteration)\n",
    "            if (value) and (value not in model_genes) and (rx in list(new_ylip_only_reactions)):\n",
    "                \n",
    "                # count how many of the genes have orthologs.\n",
    "                count_ortho+=1\n",
    "        \n",
    "        # there is no ortholog. \n",
    "        except:\n",
    "            continue\n",
    "#             donothing=0\n",
    "    \n",
    "    # there is an ortholog for at least one gene associated with the reaction. \n",
    "    if count_ortho>0:\n",
    "        # obtain reaction metadata.\n",
    "        _names.append(r.name)\n",
    "        _id.append(r.id)  \n",
    "        _gpr.append(r.gene_reaction_rule)\n",
    "        _stoich.append(r.reaction)\n",
    "        \n",
    "#         temp = \" or \".join(sub_item.replace(\"None\", \"\").replace(\"NA\",\" \") for sub_item in temp_gpr)\n",
    "#         gpr_holder.append(temp)       \n",
    "\n",
    "\n",
    "# this next section now looks at the reactions that had an ortholog. \n",
    "new_ylip_only_reactions2 = set(_id)-set(list(ylip_additional_rxns.values()))                             \n",
    "                             \n",
    "print(len(new_ylip_only_reactions2))\n",
    "\n",
    "# re-initailize metadata storing lists. \n",
    "_names = []\n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "gpr_holder = []\n",
    "\n",
    "# iterate through the reactions that have orthologs. \n",
    "for rx in new_ylip_only_reactions2:\n",
    "    # obtain reaction. \n",
    "    r = ylip.reactions.get_by_id(rx)\n",
    "\n",
    "    # store reaction metadata. \n",
    "    _gpr.append(r.gene_reaction_rule)\n",
    "    _stoich.append(r.reaction)\n",
    "    _names.append(r.name)\n",
    "    _id.append(r.id)\n",
    "    \n",
    "    # objects for tracking and storing progress.\n",
    "    count_ortho = 0\n",
    "    temp_gpr = []\n",
    "    \n",
    "    # iterate through reaction genes. \n",
    "    for i in range(len(r.genes)):\n",
    "\n",
    "        # obtain the individual genes. \n",
    "        g=list(r.genes)[i]\n",
    "        g = g.id\n",
    "\n",
    "        # obtain the ortholog.\n",
    "        try:\n",
    "            value = lst_ylip_dict[g]\n",
    "            if (value):\n",
    "                count_ortho+=1\n",
    "        except:\n",
    "            continue\n",
    "            \n",
    "    # this implies that at least one gene has been found that was not included in the reaction. re-run section to catpure all geens in teh reaction.\n",
    "    if count_ortho>0:\n",
    "        \n",
    "        # iterate through genes again for each reaction.\n",
    "        for i in range(len(r.genes)):\n",
    "            \n",
    "            # obtain the indivdual gene. \n",
    "            g=list(r.genes)[i]\n",
    "            g = g.id\n",
    "            \n",
    "            # obtain the new GPR as a list. \n",
    "            try:\n",
    "                value = lst_ylip_dict[g]\n",
    "                if (value):\n",
    "                    temp_gpr.append(value)\n",
    "            except:\n",
    "                continue\n",
    "                \n",
    "        # create a new GPR rule. \n",
    "        temp = \" or \".join(sub_item.replace(\"None\", \"\").replace(\"NA\",\" \") for sub_item in temp_gpr)\n",
    "        gpr_holder.append(temp)\n",
    "\n",
    "# create a dataframe with the candidate reactions. \n",
    "ylip_added_orthologs_orhpans = pd.DataFrame([_id,_names,_stoich,_gpr,gpr_holder]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "98d88cf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSP_L</td>\n",
       "      <td>phosphoserine phosphatase L serine</td>\n",
       "      <td>h2o_c + pser__L_c --&gt; pi_c + ser__L_c</td>\n",
       "      <td>YALI0B20438g</td>\n",
       "      <td>Lipst1_1_1740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FAS160ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C160ACP mitochondrial</td>\n",
       "      <td>3.0 h_m + malACP_m + myrsACP_m + 2.0 nadph_m -...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATPS3g</td>\n",
       "      <td>ATP synthase Golgi Apparatus</td>\n",
       "      <td>adp_g + 3.0 h_c + pi_g --&gt; atp_g + h2o_g + 2.0...</td>\n",
       "      <td>YALI0A09900g and YALI0A11143g and YALI0B03982g...</td>\n",
       "      <td>Lipst1_1_149291 or Lipst1_1_69438 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOLK</td>\n",
       "      <td>Dolichol kinase</td>\n",
       "      <td>ctp_c + dolichol_c --&gt; cdp_c + dolp_c + h_c</td>\n",
       "      <td>YALI0E18942g</td>\n",
       "      <td>Lipst1_1_55005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PETHCT</td>\n",
       "      <td>phosphoethanolamine cytidyltransferase</td>\n",
       "      <td>ctp_c + ethamp_c + h_c --&gt; cdpea_c + ppi_c</td>\n",
       "      <td>YALI0C06303g</td>\n",
       "      <td>Lipst1_1_4188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FAS161ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C161ACP mitochondrial</td>\n",
       "      <td>4.0 h_m + malACP_m + myrsACP_m + 3.0 nadph_m +...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FAS80ACPm_L</td>\n",
       "      <td>fatty acyl ACP synthase n C80ACP mitochondrial...</td>\n",
       "      <td>acACP_m + 9.0 h_m + 3.0 malACP_m + 6.0 nadph_m...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FAS140ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C140ACP mitochondrial</td>\n",
       "      <td>ddcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m -...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACt2r</td>\n",
       "      <td>acetate reversible transport via proton symport</td>\n",
       "      <td>ac_e + h_e &lt;=&gt; ac_c + h_c</td>\n",
       "      <td>YALI0E04312g</td>\n",
       "      <td>Lipst1_1_7130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HBZOPT6</td>\n",
       "      <td>Hydroxybenzoate octaprenyltransferase</td>\n",
       "      <td>4hbz_c + octdp_5_c --&gt; 3ophb_5_c + ppi_c</td>\n",
       "      <td>(YALI0E09276g and YALI0B16126g) or (YALI0E1366...</td>\n",
       "      <td>Lipst1_1_302018 or Lipst1_1_114557 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FAS180ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C180ACP mitochondrial</td>\n",
       "      <td>3.0 h_m + malACP_m + 2.0 nadph_m + palmACP_m -...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FAS141ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C141ACP mitochondrial</td>\n",
       "      <td>ddcaACP_m + 4.0 h_m + malACP_m + 3.0 nadph_m +...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FA60tp</td>\n",
       "      <td>fatty acid peroxisomal transport YL</td>\n",
       "      <td>hca_c --&gt; hca_x</td>\n",
       "      <td>YALI0E16775g</td>\n",
       "      <td>Lipst1_1_5415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FA40tp</td>\n",
       "      <td>fatty acid peroxisomal transport YL</td>\n",
       "      <td>buca_c --&gt; buca_x</td>\n",
       "      <td>YALI0E16775g</td>\n",
       "      <td>Lipst1_1_5415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CYOOm</td>\n",
       "      <td>cytochrome c oxidase mitochondrial</td>\n",
       "      <td>4.0 focytc_m + 6.0 h_m + o2_m --&gt; 4.0 ficytc_m...</td>\n",
       "      <td>YALI0E10144g and YALI0E12628g and YALI0E16709g...</td>\n",
       "      <td>Lipst1_1_5671 or Lipst1_1_55045 or Lipst1_1_38...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FAS181ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C181ACP mitochondrial</td>\n",
       "      <td>4.0 h_m + malACP_m + 3.0 nadph_m + o2_m + palm...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TGL</td>\n",
       "      <td>TGL</td>\n",
       "      <td>h2o_c + 0.01 triglyc_SC_c --&gt; 0.01 12dgr_SC_c ...</td>\n",
       "      <td>YALI0D17534g or YALI0F10010g</td>\n",
       "      <td>Lipst1_1_161356 or Lipst1_1_253586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>exlip</td>\n",
       "      <td>extracelular lipases TAG</td>\n",
       "      <td>triglyc_SC_e --&gt; 0.02 dca_e + 0.06 ddca_e + 0....</td>\n",
       "      <td>YALI0A20350g</td>\n",
       "      <td>Lipst1_1_70518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FAS100ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C100ACP mitochondrial</td>\n",
       "      <td>3.0 h_m + malACP_m + 2.0 nadph_m + ocACP_m --&gt;...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>OHPHM6n</td>\n",
       "      <td>2 octaprenyl 6 hydroxyphenol methylase nuclear</td>\n",
       "      <td>2ohph_5_n + amet_n --&gt; 2omph_5_n + ahcys_n + h_n</td>\n",
       "      <td>YALI0E00770g</td>\n",
       "      <td>Lipst1_1_68931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FAS120ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C120ACP mitochondrial</td>\n",
       "      <td>dcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m --...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SPRMti</td>\n",
       "      <td>spermine irreversible uniport</td>\n",
       "      <td>sprm_e --&gt; sprm_c</td>\n",
       "      <td>YALI0A07447g or YALI0F03751g or YALI0C04477g o...</td>\n",
       "      <td>Lipst1_1_61594 or Lipst1_1_6033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FAS182ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C182ACP mitochondrial</td>\n",
       "      <td>5.0 h_m + malACP_m + 4.0 nadph_m + 2.0 o2_m + ...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATPCitL</td>\n",
       "      <td>ATP Citrate Lyase YL</td>\n",
       "      <td>atp_c + cit_c + coa_c --&gt; accoa_c + adp_c + oa...</td>\n",
       "      <td>YALI0E34793g and YALI0D24431g</td>\n",
       "      <td>Lipst1_1_5446 or Lipst1_1_5447</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0                                                  1  \\\n",
       "0         PSP_L                 phosphoserine phosphatase L serine   \n",
       "1    FAS160ACPm    fatty acyl ACP synthase n C160ACP mitochondrial   \n",
       "2        ATPS3g                       ATP synthase Golgi Apparatus   \n",
       "3          DOLK                                    Dolichol kinase   \n",
       "4        PETHCT             phosphoethanolamine cytidyltransferase   \n",
       "5    FAS161ACPm    fatty acyl ACP synthase n C161ACP mitochondrial   \n",
       "6   FAS80ACPm_L  fatty acyl ACP synthase n C80ACP mitochondrial...   \n",
       "7    FAS140ACPm    fatty acyl ACP synthase n C140ACP mitochondrial   \n",
       "8         ACt2r    acetate reversible transport via proton symport   \n",
       "9       HBZOPT6              Hydroxybenzoate octaprenyltransferase   \n",
       "10   FAS180ACPm    fatty acyl ACP synthase n C180ACP mitochondrial   \n",
       "11   FAS141ACPm    fatty acyl ACP synthase n C141ACP mitochondrial   \n",
       "12       FA60tp                fatty acid peroxisomal transport YL   \n",
       "13       FA40tp                fatty acid peroxisomal transport YL   \n",
       "14        CYOOm                 cytochrome c oxidase mitochondrial   \n",
       "15   FAS181ACPm    fatty acyl ACP synthase n C181ACP mitochondrial   \n",
       "16          TGL                                                TGL   \n",
       "17        exlip                           extracelular lipases TAG   \n",
       "18   FAS100ACPm    fatty acyl ACP synthase n C100ACP mitochondrial   \n",
       "19      OHPHM6n     2 octaprenyl 6 hydroxyphenol methylase nuclear   \n",
       "20   FAS120ACPm    fatty acyl ACP synthase n C120ACP mitochondrial   \n",
       "21       SPRMti                      spermine irreversible uniport   \n",
       "22   FAS182ACPm    fatty acyl ACP synthase n C182ACP mitochondrial   \n",
       "23      ATPCitL                               ATP Citrate Lyase YL   \n",
       "\n",
       "                                                    2  \\\n",
       "0               h2o_c + pser__L_c --> pi_c + ser__L_c   \n",
       "1   3.0 h_m + malACP_m + myrsACP_m + 2.0 nadph_m -...   \n",
       "2   adp_g + 3.0 h_c + pi_g --> atp_g + h2o_g + 2.0...   \n",
       "3         ctp_c + dolichol_c --> cdp_c + dolp_c + h_c   \n",
       "4          ctp_c + ethamp_c + h_c --> cdpea_c + ppi_c   \n",
       "5   4.0 h_m + malACP_m + myrsACP_m + 3.0 nadph_m +...   \n",
       "6   acACP_m + 9.0 h_m + 3.0 malACP_m + 6.0 nadph_m...   \n",
       "7   ddcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m -...   \n",
       "8                           ac_e + h_e <=> ac_c + h_c   \n",
       "9            4hbz_c + octdp_5_c --> 3ophb_5_c + ppi_c   \n",
       "10  3.0 h_m + malACP_m + 2.0 nadph_m + palmACP_m -...   \n",
       "11  ddcaACP_m + 4.0 h_m + malACP_m + 3.0 nadph_m +...   \n",
       "12                                    hca_c --> hca_x   \n",
       "13                                  buca_c --> buca_x   \n",
       "14  4.0 focytc_m + 6.0 h_m + o2_m --> 4.0 ficytc_m...   \n",
       "15  4.0 h_m + malACP_m + 3.0 nadph_m + o2_m + palm...   \n",
       "16  h2o_c + 0.01 triglyc_SC_c --> 0.01 12dgr_SC_c ...   \n",
       "17  triglyc_SC_e --> 0.02 dca_e + 0.06 ddca_e + 0....   \n",
       "18  3.0 h_m + malACP_m + 2.0 nadph_m + ocACP_m -->...   \n",
       "19   2ohph_5_n + amet_n --> 2omph_5_n + ahcys_n + h_n   \n",
       "20  dcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m --...   \n",
       "21                                  sprm_e --> sprm_c   \n",
       "22  5.0 h_m + malACP_m + 4.0 nadph_m + 2.0 o2_m + ...   \n",
       "23  atp_c + cit_c + coa_c --> accoa_c + adp_c + oa...   \n",
       "\n",
       "                                                    3  \\\n",
       "0                                        YALI0B20438g   \n",
       "1   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "2   YALI0A09900g and YALI0A11143g and YALI0B03982g...   \n",
       "3                                        YALI0E18942g   \n",
       "4                                        YALI0C06303g   \n",
       "5   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "6   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "7   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "8                                        YALI0E04312g   \n",
       "9   (YALI0E09276g and YALI0B16126g) or (YALI0E1366...   \n",
       "10  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "11  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "12                                       YALI0E16775g   \n",
       "13                                       YALI0E16775g   \n",
       "14  YALI0E10144g and YALI0E12628g and YALI0E16709g...   \n",
       "15  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "16                       YALI0D17534g or YALI0F10010g   \n",
       "17                                       YALI0A20350g   \n",
       "18  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "19                                       YALI0E00770g   \n",
       "20  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "21  YALI0A07447g or YALI0F03751g or YALI0C04477g o...   \n",
       "22  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "23                      YALI0E34793g and YALI0D24431g   \n",
       "\n",
       "                                                    4  \n",
       "0                                       Lipst1_1_1740  \n",
       "1   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "2   Lipst1_1_149291 or Lipst1_1_69438 or Lipst1_1_...  \n",
       "3                                      Lipst1_1_55005  \n",
       "4                                       Lipst1_1_4188  \n",
       "5   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "6   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "7   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "8                                       Lipst1_1_7130  \n",
       "9   Lipst1_1_302018 or Lipst1_1_114557 or Lipst1_1...  \n",
       "10  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "11  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "12                                      Lipst1_1_5415  \n",
       "13                                      Lipst1_1_5415  \n",
       "14  Lipst1_1_5671 or Lipst1_1_55045 or Lipst1_1_38...  \n",
       "15  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "16                 Lipst1_1_161356 or Lipst1_1_253586  \n",
       "17                                     Lipst1_1_70518  \n",
       "18  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "19                                     Lipst1_1_68931  \n",
       "20  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "21                    Lipst1_1_61594 or Lipst1_1_6033  \n",
       "22  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "23                     Lipst1_1_5446 or Lipst1_1_5447  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip_added_orthologs_orhpans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "9d02660f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Lipst1_1_4188'"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check one of the gene dictionaries: for PETHCT.\n",
    "lst_ylip_dict['YALI0C06303g']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f81190d0",
   "metadata": {},
   "source": [
    "#### metabolite determination. This section verifies that the reaction candidates only have metabolites that are already included in the GSM. \n",
    "metabolite section. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "2f17269e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain the metabolites in the Lst model. \n",
    "mets = model.metabolites\n",
    "mets = [x.id for x in mets]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "381b1991",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "yes -  PSP_L\n",
      "yes -  ACt2r\n",
      "yes -  CYOOm\n",
      "yes -  SPRMti\n",
      "yes -  ATPCitL\n"
     ]
    }
   ],
   "source": [
    "# this cell checks the metabolites in the reactions and ensures that they are all in the GSM. \n",
    "\n",
    "# reaction names.\n",
    "react = ylip_added_orthologs_orhpans[0].to_list()\n",
    "# new GPR.\n",
    "gpr = ylip_added_orthologs_orhpans[4].to_list()\n",
    "\n",
    "# list for tracking. \n",
    "orphan_ylip_reactions_actually_added = []\n",
    "\n",
    "# iterate through the reactions.\n",
    "for u,rx in enumerate(react):\n",
    "    \n",
    "    # obtain the ylip GSM reaction. \n",
    "    y = ylip.reactions.get_by_id(rx)\n",
    "    \n",
    "    # obtain a list of the metabolites. \n",
    "    gsm_mets = list(y.metabolites.keys())\n",
    "    gsm_mets = [x.id for x in gsm_mets]\n",
    "    \n",
    "    # counter. \n",
    "    j=0\n",
    "    \n",
    "    # iterate through the metabolites.\n",
    "    for i in range(len(gsm_mets)):\n",
    "        \n",
    "        # check if metabolite is in model.\n",
    "        if gsm_mets[i] in mets:\n",
    "            j+=1\n",
    "        else:\n",
    "#             print(gsm_mets[i])\n",
    "            continue\n",
    "    \n",
    "    # check if all the reaction metabolites were in the model. \n",
    "    if j==len(gsm_mets):\n",
    "        \n",
    "        print('yes - ', rx)\n",
    "#         r = y.copy()\n",
    "#         r.gene_reaction_rule = gpr[u]\n",
    "\n",
    "        # add to model. \n",
    "#         model.add_reactions([r])\n",
    "\n",
    "        # store the reactions. \n",
    "        orphan_ylip_reactions_actually_added.append(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "7b5a7264",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATPCitL ATP Citrate Lyase YL atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c\n",
      "ATPCitL ATP Citrate Lyase YL atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c\n",
      "ATPCitL ATP Citrate Lyase YL atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c\n",
      "ATPCitL ATP Citrate Lyase YL atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c\n",
      "ATPCitL ATP Citrate Lyase YL atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c\n"
     ]
    }
   ],
   "source": [
    "for x in orphan_ylip_reactions_actually_added:\n",
    "    print(x.id, x.name, x.reaction)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d826a1c",
   "metadata": {},
   "source": [
    "#### here we validate and potentially add reactions to the Lst model. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc311efc",
   "metadata": {},
   "source": [
    "metabolic reaction #1.\n",
    "- ACt2r acetate reversible transport via proton symport ac_e + h_e <=> ac_c + h_c\n",
    "- CYOOm cytochrome c oxidase mitochondrial 4.0 focytc_m + 6.0 h_m + o2_m --> 4.0 ficytc_m + 2.0 h2o_m + 6.0 h_c\n",
    "- PSP_L phosphoserine phosphatase L serine h2o_c + pser__L_c --> pi_c + ser__L_c\n",
    "- ATPCitL ATP Citrate Lyase YL atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c\n",
    "- SPRMti spermine irreversible uniport sprm_e --> sprm_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "f4ae2266",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ACt2r</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>acetate reversible transport via proton symport</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x17feef1d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ac_e + h_e <=> ac_c + h_c</p>\n",
       "                    <p style='text-align:right'>Acetate + H <=> Acetate + H</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0E04312g</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ACt2r at 0x17feef1d0>"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.ACt2r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "ac867a98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_ac_e: ac_e -->  Acetate exchange\n",
      "ACtr: ac_e <=> ac_c Acetate transporter\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.ac_e.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "7d0686f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ACtr</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Acetate transporter</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b813dd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ac_e <=> ac_c</p>\n",
       "                    <p style='text-align:right'>Acetate <=> Acetate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_102591</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ACtr at 0x15b813dd0>"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.ACtr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62aebc0c",
   "metadata": {},
   "source": [
    "two potential orthologs for acetate transporter. One is specified as a proton symport. Both will be included in the GSM. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "98ba7d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = ylip.reactions.ACt2r.copy()\n",
    "x = ylip_added_orthologs_orhpans.loc[ylip_added_orthologs_orhpans[0]=='ACt2r'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "5f7da26f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_7130'], dtype=object)"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "8d6385fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.ACt2r.gene_reaction_rule = 'Lipst1_1_7130'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94257c82",
   "metadata": {},
   "source": [
    "reaction #2. \n",
    "- CYOOm cytochrome c oxidase mitochondrial 4.0 focytc_m + 6.0 h_m + o2_m --> 4.0 ficytc_m + 2.0 h2o_m + 6.0 h_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "41519acc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CYOOm</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>cytochrome c oxidase mitochondrial</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a09cdc10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>4.0 focytc_m + 6.0 h_m + o2_m --> 4.0 ficytc_m + 2.0 h2o_m + 6.0 h_c</p>\n",
       "                    <p style='text-align:right'>4.0 Ferrocytochrome_c + 6.0 H + O2 --> 4.0 Ferricytochrome_c + 2.0 H2O + 6.0 H</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0E10144g and YALI0E12628g and YALI0E16709g and YALI0E19723g and YALI0F03201g and YALI0F03567g...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CYOOm at 0x2a09cdc10>"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.CYOOm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "3d0f415f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.CYOOm.check_mass_balance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "649ee953",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYOR_u9m: 2.0 ficytc_m + 2.0 h_m + q9h2_m --> 2.0 focytc_m + 4.0 h_c + q9_m Ubiquinol 6 cytochrome c reductase\n",
      "CCP2_m: 2.0 focytc_m + h2o2_m + 2.0 h_m --> 2.0 ficytc_m + 2.0 h2o_m Cytochrome c peroxidase, mitochondrial\n",
      "CYOO6m: 4.0 focytc_m + 8.0 h_m + o2_m --> 4.0 ficytc_m + 2.0 h2o_m + 4.0 h_c Cytochrome c oxidase Complex IV\n",
      "D_LACDcm: 2.0 ficytc_m + lac__D_c --> 2.0 focytc_m + 2.0 h_c + pyr_c D lactate dehydrogenase  cytosolicmitochondrial\n",
      "L_LACDcm: 2.0 ficytc_m + lac__L_c --> 2.0 focytc_m + 2.0 h_c + pyr_c L-Lactate dehydrogenase, cytosolic/mitochondrial\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.focytc_m.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2448f10",
   "metadata": {},
   "source": [
    "one potential reaction already in Lst GSM. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "800e50c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.reactions.CCP2_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "4828cf3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CYOO6m</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Cytochrome c oxidase Complex IV</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a18c2f10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>4.0 focytc_m + 8.0 h_m + o2_m --> 4.0 ficytc_m + 2.0 h2o_m + 4.0 h_c</p>\n",
       "                    <p style='text-align:right'>4.0 Ferrocytochrome c C42H53FeN8O6S2 + 8.0 H+ + Oxygen --> 4.0 Ferricytochrome c C42H52FeN8O6S2 + 2.0 H2O H2O + 4.0 H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>COX1 and COX2 and COX3 and 9165 and 13628 and 10685 and 12506 and 11352 and 13275 and 11769 and...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CYOO6m at 0x2a18c2f10>"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.CYOO6m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "c7f903a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.CYOO6m.check_mass_balance()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "551c07ba",
   "metadata": {},
   "source": [
    "these two reactions are encoding the same reaction, but with slightly different mass balances in terms of protons moved. will keep the original reaction stoichiometry but add the new GPR. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9effe153",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ead44a01",
   "metadata": {},
   "source": [
    "two potential orthologs for acetate transporter. One is specified as a proton symport. Both will be included in the GSM. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "14150911",
   "metadata": {},
   "outputs": [],
   "source": [
    "# r_copy = ylip.reactions.CYOOm.copy()\n",
    "x = ylip_added_orthologs_orhpans.loc[ylip_added_orthologs_orhpans[0]=='CYOOm'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "2ebcb617",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_5671 or Lipst1_1_55045 or Lipst1_1_3895 or Lipst1_1_54407 or Lipst1_1_65503 or Lipst1_1_67449'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "803e6546",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.CYOO6m.gene_reaction_rule = 'Lipst1_1_54407 or Lipst1_1_65503 or Lipst1_1_67449 or Lipst1_1_3895 or Lipst1_1_55045 or Lipst1_1_5671'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a797b4d9",
   "metadata": {},
   "source": [
    "reaction #3.\n",
    "- PSP_L phosphoserine phosphatase L serine h2o_c + pser__L_c --> pi_c + ser__L_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "1139504f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PSP_L</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>phosphoserine phosphatase L serine</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a7ad610</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + pser__L_c --> pi_c + ser__L_c</p>\n",
       "                    <p style='text-align:right'>H2O + O_Phospho_L_serine --> Phosphate + L_Serine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0B20438g</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PSP_L at 0x16a7ad610>"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.PSP_L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "48a3baed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PSERT: 3php_c + glu__L_c --> akg_c + pser__L_c Phosphoserine transaminase\n",
      "PSERtr: pser__L_c <=> pser__L_e Transport of O-Phospho-L-Serine, Diffusion\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.pser__L_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "184abd34",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = ylip.reactions.PSP_L.copy()\n",
    "x = ylip_added_orthologs_orhpans.loc[ylip_added_orthologs_orhpans[0]=='PSP_L'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "a941c735",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_1740'], dtype=object)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "c32ca6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.PSP_L.gene_reaction_rule = 'Lipst1_1_1740'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4466d56e",
   "metadata": {},
   "source": [
    "reaction #4. \n",
    "- ATPCitL ATP Citrate Lyase YL atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "77f62470",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ATPCitL</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ATP Citrate Lyase YL</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a4b3ed0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP + Citrate + Coenzyme_A --> Acetyl_CoA + ADP + Oxaloacetate + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0E34793g and YALI0D24431g</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ATPCitL at 0x16a4b3ed0>"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.ATPCitL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "96357cdb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACITL: atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c ATP-Citrate lyase\n",
      "CITtam: cit_c + mal__L_m <=> cit_m + mal__L_c Citrate transport  mitochondrial\n",
      "AKGCITtm: akg_c + cit_m --> akg_m + cit_c Dicarboxylate/tricarboxylate carrier (akg:cit), mitochondrial\n",
      "CITtcm: cit_c + icit_m <=> cit_m + icit_c Citrate transport  mitochondrial\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.cit_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dcd02d8",
   "metadata": {},
   "source": [
    "one reaction already included for ATP citrate lyase. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "b9989b43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ACITL</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ATP-Citrate lyase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b8b9510</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + cit_c + coa_c --> accoa_c + adp_c + oaa_c + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP C10H12N5O13P3 + Citrate + Coenzyme A --> Acetyl-CoA + ADP C10H12N5O10P2 + Oxaloacetate + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_5446</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ACITL at 0x15b8b9510>"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.ACITL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "44b8150f",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ylip_added_orthologs_orhpans.loc[ylip_added_orthologs_orhpans[0]=='ATPCitL'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "b9727fe9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_5446 or Lipst1_1_5447'], dtype=object)"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35745f94",
   "metadata": {},
   "source": [
    "update the ATPcitrate lyase model GPR. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "5eea94ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.ACITL.gene_reaction_rule = 'Lipst1_1_5446 or Lipst1_1_5447'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a0f863e",
   "metadata": {},
   "source": [
    "reaction #5.\n",
    "- SPRMti spermine irreversible uniport sprm_e --> sprm_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "d2e53940",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>SPRMti</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>spermine irreversible uniport</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15bbc4510</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>sprm_e --> sprm_c</p>\n",
       "                    <p style='text-align:right'>Spermine --> Spermine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0A07447g or YALI0F03751g or YALI0C04477g or YALI0F07062g or YALI0E10483g or YALI0B17776g or...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction SPRMti at 0x15bbc4510>"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.SPRMti"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "775c9b45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SPRMt2i: h_c + sprm_e --> h_e + sprm_c Spermine transport via proton antiport irreversible\n",
      "EX_sprm_e: sprm_e -->  Spermine C10H30N4 exchange\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.sprm_e.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "b5678247",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ylip_added_orthologs_orhpans.loc[ylip_added_orthologs_orhpans[0]=='SPRMti'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "11d58fe4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_61594 or Lipst1_1_6033'], dtype=object)"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "26a7674a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>SPRMt2i</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Spermine transport via proton antiport irreversible</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a13cbd90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h_c + sprm_e --> h_e + sprm_c</p>\n",
       "                    <p style='text-align:right'>H+ + Spermine C10H30N4 --> H+ + Spermine C10H30N4</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_2256 or Lipst1_1_61026 or Lipst1_1_6033 or Lipst1_1_52551 or Lipst1_1_97209</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction SPRMt2i at 0x2a13cbd90>"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('SPRMt2i')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2379830a",
   "metadata": {},
   "source": [
    "one additional gene not included in the transporter. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "788e0a92",
   "metadata": {},
   "source": [
    "will modify the existing genes to include the not added gene. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "db915193",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.get_by_id('SPRMt2i').gene_reaction_rule = 'Lipst1_1_2256 or Lipst1_1_61026 or Lipst1_1_6033 or Lipst1_1_97209 or Lipst1_1_52551 or Lipst1_1_61594'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b862e125",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "e3f20647",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2150"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "5baff07b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2153-2083"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06ed4a6e",
   "metadata": {},
   "source": [
    "5 additional reactions added from ylip. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "22f530ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.07800483549348808</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>6.49E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>4.095E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0004812</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.04569</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.004814</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>4.735E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.002036</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4596</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.356</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.02077</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.007819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.978E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001886</td>\n",
       "      <td>9</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0002424</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.04505</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.01144</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.008422</td>\n",
       "      <td>10</td>\n",
       "      <td>1.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.01539</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002673</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.009197</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01628</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.04003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.004771</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.01237</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.02498</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.01344</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.00445</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.007423</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.02628</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.01702</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.001819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.004865</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.02146</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-0.0001087</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001886</td>\n",
       "      <td>15</td>\n",
       "      <td>0.10%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001886</td>\n",
       "      <td>10</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-2.619</td>\n",
       "      <td>1</td>\n",
       "      <td>89.99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.0003342</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.275</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1232</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001955</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.01053</td>\n",
       "      <td>27</td>\n",
       "      <td>9.76%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a1ee6d10>"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "f06dd154",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11317 has no associated reaction\n",
      "11352 has no associated reaction\n",
      "9165 has no associated reaction\n",
      "13628 has no associated reaction\n",
      "COX3 has no associated reaction\n",
      "COX2 has no associated reaction\n",
      "12506 has no associated reaction\n",
      "COX1 has no associated reaction\n",
      "11769 has no associated reaction\n",
      "13275 has no associated reaction\n",
      "13007 has no associated reaction\n",
      "10685 has no associated reaction\n",
      "12210 has no associated reaction\n",
      "YALI0E04312g has no associated reaction\n",
      "YALI0B20438g has no associated reaction\n",
      "31\n",
      "16\n"
     ]
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))\n",
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g2)))\n",
    "print(len(set(wrong_rx2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "3db595c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACRNtp',\n",
       " 'C30CPT1',\n",
       " 'C3STKR1er',\n",
       " 'C3STKR2er',\n",
       " 'CRNtp',\n",
       " 'CYOR_u9m',\n",
       " 'DNTPPA',\n",
       " 'DRTPPD',\n",
       " 'FA160COAabcp',\n",
       " 'FA183COAabcp',\n",
       " 'FA240tp',\n",
       " 'GAPP',\n",
       " 'H2Otm',\n",
       " 'OIVD2m',\n",
       " 'PMDPHT',\n",
       " 'SERPTer'}"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2 = [x.id for x in wrong_rx2]\n",
    "set(wrong_rx2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f304756e",
   "metadata": {},
   "source": [
    "## This section repeats the process for sce. The code takes the sce reactions that have orthologs and checks that the reaction contains metabolites in the Lst GSM.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "d126967d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "457\n"
     ]
    }
   ],
   "source": [
    "# this code finds reactions from sce to the model that may have potential orthologs that have been missed. \n",
    "\n",
    "# list objects for storing information. \n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "gpr_holder = []\n",
    "\n",
    "\n",
    "# obtain reactions that have orthologs but have not already been added the GSM. \n",
    "# (sce_ortholog_available) is the complete list of sce reactions that have orthologs that arent in rto or ylip.\n",
    "# (sce_additional_rxns) is the list of sce reactions that had no ID match but were able to match rxn names. \n",
    "new_sce_only_reactions = set(sce_ortholog_available)-set(list(sce_additional_rxns.values()))\n",
    "\n",
    "# iterate through the reactions. \n",
    "for rx in new_sce_only_reactions:\n",
    "    \n",
    "    # obtain the sce reaction object.\n",
    "    r = sce.reactions.get_by_id(rx)\n",
    "    \n",
    "    \n",
    "    # objects for tracking. \n",
    "    count_ortho = 0\n",
    "    temp_gpr = []\n",
    "    \n",
    "    # iterate over the sce reaction genes. \n",
    "    for i in range(len(r.genes)):\n",
    "        \n",
    "        \n",
    "        g=list(r.genes)[i]\n",
    "        g = g.id\n",
    "        \n",
    "        # find the ortholog.\n",
    "        try:\n",
    "            \n",
    "            # find the ortholog.\n",
    "            value = lst_sce_dict[g]\n",
    "            \n",
    "            # verify that the ortholog exists & \n",
    "            # is not in the model genes &\n",
    "            # is in the ortholog reactions (guaranteed based on cell logic and for loops)\n",
    "            if (value) and (value not in model_genes) and (rx in list(new_sce_only_reactions)):\n",
    "\n",
    "                count_ortho+=1\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    # if the reaction had at least one gene ortholog in Lst. \n",
    "    if count_ortho>0:\n",
    "        # append reaction metadata. \n",
    "        _gpr.append(r.gene_reaction_rule)\n",
    "        _stoich.append(r.reaction)        \n",
    "        _names.append(r.name)\n",
    "        _id.append(r.id)  \n",
    "        \n",
    "# obtain the sce reactions that have orthologs.      \n",
    "new_sce_only_reactions2 = set(_id)-set(list(sce_additional_rxns.values()))                             \n",
    "                             \n",
    "print(len(new_sce_only_reactions2))\n",
    "\n",
    "\n",
    "# reset the information holding lists. \n",
    "_names = []\n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "gpr_holder = []\n",
    "\n",
    "# iterate over the reactions. \n",
    "for rx in new_sce_only_reactions2:\n",
    "    \n",
    "    # obtain the sce reaction object. \n",
    "    r = sce.reactions.get_by_id(rx)\n",
    "    \n",
    "    # store reaction metadata. \n",
    "    _gpr.append(r.gene_reaction_rule)\n",
    "    _stoich.append(r.reaction)\n",
    "    _names.append(r.name)\n",
    "    _id.append(r.id)\n",
    "    \n",
    "    # objects for tracking during loop. \n",
    "    count_ortho =0\n",
    "    temp_gpr = []\n",
    "    \n",
    "    # iterate over the sce reaction genes. \n",
    "    for i in range(len(r.genes)):\n",
    "\n",
    "        g=list(r.genes)[i]\n",
    "        g = g.id\n",
    "\n",
    "        try:\n",
    "            \n",
    "            # obtain the Lst ortholog. \n",
    "            value = lst_sce_dict[g]\n",
    "            \n",
    "            if (value):\n",
    "                count_ortho+=1\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "    # this implies that at least one gene has been found that was not included in the reaction.\n",
    "    if count_ortho>0:\n",
    "        \n",
    "        # re-run the gene iteration to capture all orthologs. \n",
    "        for i in range(len(r.genes)):\n",
    "            \n",
    "            g=list(r.genes)[i]\n",
    "            g = g.id\n",
    "            \n",
    "            try:\n",
    "                # obtain the Lst ortholog. \n",
    "                value = lst_sce_dict[g]\n",
    "                if (value):\n",
    "                    temp_gpr.append(value)\n",
    "            except:\n",
    "                continue\n",
    "        \n",
    "        # join multiple genes for a single reaction together using 'or' rules. \n",
    "        temp = \" or \".join(sub_item.replace(\"None\", \"\").replace(\"NA\",\" \") for sub_item in temp_gpr)\n",
    "        gpr_holder.append(temp)\n",
    "\n",
    "# pd.DataFrame() for storing the reactions.         \n",
    "sce_added_orthologs_orhpans = pd.DataFrame([_id,_names,_stoich,_gpr,gpr_holder]).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85888624",
   "metadata": {},
   "source": [
    "potentially 457 sce genes to add. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cde55a62",
   "metadata": {},
   "source": [
    "## check candidate reactions to ensure that they included metabolites already in the GSM. \n",
    "metabolite checking section. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "3b866f73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this cell checks the metabolites in the reactions and ensures that they are all in the GSM. \n",
    "\n",
    "# reaction names.\n",
    "react = sce_added_orthologs_orhpans[0].to_list()\n",
    "\n",
    "# gene reaction rules.\n",
    "gpr = sce_added_orthologs_orhpans[4].to_list()\n",
    "\n",
    "# list for storing reactions. \n",
    "orphan_sce_reactions_actually_added = []\n",
    "\n",
    "# iterate over the reactions. \n",
    "for u,rx in enumerate(react):\n",
    "    \n",
    "    # grab the sce GSM reaction object.\n",
    "    s = sce.reactions.get_by_id(rx)\n",
    "    \n",
    "    # grab the metabolites in the reaction.\n",
    "    gsm_mets = list(s.metabolites.keys())\n",
    "    gsm_mets = [x.id for x in gsm_mets]\n",
    "    \n",
    "    j=0\n",
    "    \n",
    "    # iterate over the number of metabolits in the reaction. \n",
    "    for i in range(len(gsm_mets)):\n",
    "        \n",
    "        # if metabolite is in model already.\n",
    "        if gsm_mets[i] in mets:\n",
    "            j+=1\n",
    "#         else:\n",
    "#             print(gsm_mets[i])\n",
    "\n",
    "    # store reactions that have every metabolite in the Lst model.\n",
    "    if j==len(gsm_mets):\n",
    "#         print('yes - ', gsm_mets[i])\n",
    "        r = s.copy()\n",
    "        r.gene_reaction_rule = gpr[u]\n",
    "#         print(r,r.gene_reaction_rule,'\\n',s.gene_reaction_rule)\n",
    "\n",
    "        # add to model. \n",
    "#         model.add_reactions([r])\n",
    "        orphan_sce_reactions_actually_added.append(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "5265f806",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(orphan_sce_reactions_actually_added)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "711a9da2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r2073_1 Zn(2+) transport h_e + zn2_e <=> h_c + zn2_c\n",
      "r_4162 Probable vacuolar amino acid transporter YPQ3 (PQ-loop repeat-containing protein 3) (Protein RTC2) (Restriction of telomere capping protein 2) his__L_v <=> his__L_c\n",
      "r_4331 [1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanohydrolase chtn_c + h2o_c <=> acgam_c\n",
      "GTHDH glutathione:dehydroascorbate oxidoreductase dhdascb_c + 2.0 gthrd_c <=> ascb__L_c + gthox_c\n",
      "r_1837 L-histidine transport, mitochondrial his__L_c --> his__L_m\n",
      "NTP3 nucleoside triphosphatase gtp_c + h2o_c --> gdp_c + h_c + pi_c\n",
      "MNt2 Mn(2+) transport h_e + mn2_e <=> h_c + mn2_c\n",
      "ARGt2m arginine transport arg__L_c + h_c --> arg__L_m + h_m\n",
      "r_4592 Mg(2+) transport h_e + mg2_e <=> h_c + mg2_c\n",
      "r_0438 ferrocytochrome-c:oxygen oxidoreductase focytc_m + 1.266 h_m + 0.25 o2_m --> ficytc_m + 0.5 h2o_m + 0.633 h_c\n",
      "PSP_L phosphoserine phosphatase (L-serine) h2o_c + pser__L_c --> pi_c + ser__L_c\n",
      "r_1202 L-isoleucine transport ile__L_v <=> ile__L_c\n",
      "r_4163 Probable vacuolar amino acid transporter YPQ3 (PQ-loop repeat-containing protein 3) (Protein RTC2) (Restriction of telomere capping protein 2) lys__L_v <=> lys__L_c\n",
      "GLYOX3 (R)-lactate hydro-lyase h_c + lac__D_c <=> h2o_c + mthgxl_c\n",
      "SERAT serine O-acetyltransferase accoa_c + ser__L_c --> acser_c + coa_c\n",
      "r_4275 Fe(II):NADP+ oxidoreductase 2.0 fe3_e + nadph_c <=> 2.0 fe2_c + h_c + nadp_c\n",
      "r_4589 Cu2(+) transport cu2_e + h_e <=> cu2_c + h_c\n",
      "CHOLtu Protein PNS1 (pH nine-sensitive protein 1) chol_c <=> chol_e\n",
      "r_1086 V-ATPase, vacuole atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c\n",
      "r_4587 Ca(2+) transport ca2_e + h_e <=> ca2_c + h_c\n",
      "r_1919 lysine transport lys__L_c --> lys__L_m\n",
      "r_4312 S-adenosyl-L-methionine:carnosine N-methyltransferase amet_c + carn_c <=> ahcys_c + ans_c + h_c\n",
      "r_1220 L-tyrosine transport tyr__L_v <=> tyr__L_c\n",
      "Clt chloride transport cl_e <=> cl_c\n",
      "SERD_D D-serine ammonia-lyase ser__D_c --> nh4_c + pyr_c\n",
      "GGCLUT2 Glutathione-specific gamma-glutamylcyclotransferase (Gamma-GCG) (EC 4.3.2.-) gthrd_c --> 5oxpro_c + cgly_c\n",
      "r_1085 V-ATPase, Golgi atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c\n",
      "r_1208 L-leucine transport, vacuoluar leu__L_v <=> leu__L_c\n",
      "r_4403 Triphosphate phosphohydrolase h2o_c + pppi_c --> pi_c + ppi_c\n"
     ]
    }
   ],
   "source": [
    "for x in orphan_sce_reactions_actually_added:\n",
    "    print(x.id, x.name, x.reaction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "916b3267",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "67"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)-2083"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd0665f4",
   "metadata": {},
   "source": [
    "28 addition reactions added from sce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "303054f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ee5f435c",
   "metadata": {},
   "source": [
    "## a reaction leads to more hco3 output rather than co2. lets find which added reactions does so. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "c8390ea5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.07800483549348808</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>6.49E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>4.095E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0004812</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.04569</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.004814</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>4.735E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.002036</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4596</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.356</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.02077</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.007819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.978E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001886</td>\n",
       "      <td>9</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0002424</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.04505</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.01144</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.008422</td>\n",
       "      <td>10</td>\n",
       "      <td>1.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.01539</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002673</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.009197</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01628</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.04003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.004771</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.01237</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.02498</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.01344</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.00445</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.007423</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.02628</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.01702</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.001819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.004865</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.02146</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-0.0001087</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001886</td>\n",
       "      <td>15</td>\n",
       "      <td>0.10%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001886</td>\n",
       "      <td>10</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-2.619</td>\n",
       "      <td>1</td>\n",
       "      <td>89.99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.0003342</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.275</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1232</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001955</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.01053</td>\n",
       "      <td>27</td>\n",
       "      <td>9.76%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a1e81510>"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "2c7121b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/kn/zzns_smn1q79xdnf1tgqk__r0000gn/T/tmpgmjymr5r.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1905 rows, 4300 columns, 17560 nonzeros\n"
     ]
    }
   ],
   "source": [
    "temp_model = model.copy()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "bd811001",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.010526362525668749 0.0 2.6212240608171524 0.07800483549348808\n",
      "r2073_1 Zn(2+) transport h_e + zn2_e <=> h_c + zn2_c\n",
      "0.010526362525668768 0.0 2.6212240608171564 0.07800483549348822\n",
      "r_4162 Probable vacuolar amino acid transporter YPQ3 (PQ-loop repeat-containing protein 3) (Protein RTC2) (Restriction of telomere capping protein 2) his__L_v <=> his__L_c\n",
      "0.010526362525668721 0.0 2.6212240608171467 0.0780048354934879\n",
      "r_4331 [1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanohydrolase chtn_c + h2o_c <=> acgam_c\n",
      "0.010526362525668754 0.0 2.6212240608171533 0.0780048354934881\n",
      "GTHDH glutathione:dehydroascorbate oxidoreductase dhdascb_c + 2.0 gthrd_c <=> ascb__L_c + gthox_c\n",
      "0.010526362525668775 0.0019547621750490725 2.619269298642111 0.07800483549348829\n",
      "r_1837 L-histidine transport, mitochondrial his__L_c --> his__L_m\n",
      "0.010526362525668746 0.00781904870019627 2.6134050121169596 0.07800483549348809\n",
      "NTP3 nucleoside triphosphatase gtp_c + h2o_c --> gdp_c + h_c + pi_c\n",
      "0.010526362525668796 0.0019547621750490595 2.6192692986421156 0.07800483549348844\n",
      "MNt2 Mn(2+) transport h_e + mn2_e <=> h_c + mn2_c\n",
      "0.01052636252566875 0.0019547621750490678 2.619269298642105 0.07800483549348809\n",
      "ARGt2m arginine transport arg__L_c + h_c --> arg__L_m + h_m\n",
      "0.01052636252566871 0.0 2.6212240608171444 0.07800483549348779\n",
      "r_4592 Mg(2+) transport h_e + mg2_e <=> h_c + mg2_c\n",
      "0.010526362525668699 0.0 2.6212240608171418 0.07800483549348769\n",
      "r_0438 ferrocytochrome-c:oxygen oxidoreductase focytc_m + 1.266 h_m + 0.25 o2_m --> ficytc_m + 0.5 h2o_m + 0.633 h_c\n",
      "0.010526362525668822 0.0019547621750490803 2.6192692986421204 0.0780048354934886\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'PSP_L' since it already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PSP_L phosphoserine phosphatase (L-serine) h2o_c + pser__L_c --> pi_c + ser__L_c\n",
      "0.010526362525668829 0.001954762175049065 2.6192692986421227 0.0780048354934887\n",
      "r_1202 L-isoleucine transport ile__L_v <=> ile__L_c\n",
      "0.010526362525668706 0.001954762175049053 2.6192692986420947 0.07800483549348776\n",
      "r_4163 Probable vacuolar amino acid transporter YPQ3 (PQ-loop repeat-containing protein 3) (Protein RTC2) (Restriction of telomere capping protein 2) lys__L_v <=> lys__L_c\n",
      "0.010526362525668704 0.0 2.621224060817142 0.07800483549348775\n",
      "GLYOX3 (R)-lactate hydro-lyase h_c + lac__D_c <=> h2o_c + mthgxl_c\n",
      "0.01052636252566875 0.0 2.6212240608171533 0.07800483549348809\n",
      "SERAT serine O-acetyltransferase accoa_c + ser__L_c --> acser_c + coa_c\n",
      "0.010526362525668744 0.0 2.6212240608171524 0.07800483549348805\n",
      "r_4275 Fe(II):NADP+ oxidoreductase 2.0 fe3_e + nadph_c <=> 2.0 fe2_c + h_c + nadp_c\n",
      "0.018071505434529594 0.0 0.19936681848978102 0.13391756222557039\n",
      "r_4589 Cu2(+) transport cu2_e + h_e <=> cu2_c + h_c\n",
      "0.010526362525668768 0.0 2.6212240608171564 0.07800483549348822\n",
      "CHOLtu Protein PNS1 (pH nine-sensitive protein 1) chol_c <=> chol_e\n",
      "0.010526362525668721 0.0 2.6212240608171467 0.0780048354934879\n",
      "r_1086 V-ATPase, vacuole atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c\n",
      "0.010526362525668744 0.0019547621750490496 2.619269298642104 0.07800483549348806\n",
      "r_4587 Ca(2+) transport ca2_e + h_e <=> ca2_c + h_c\n",
      "0.01052636252566872 0.0019547621750490548 2.6192692986420965 0.07800483549348786\n",
      "r_1919 lysine transport lys__L_c --> lys__L_m\n",
      "0.01052636252566876 0.003909524350098138 2.617314536467057 0.07800483549348816\n",
      "r_4312 S-adenosyl-L-methionine:carnosine N-methyltransferase amet_c + carn_c <=> ahcys_c + ans_c + h_c\n",
      "0.010526362525668752 0.003909524350098137 2.617314536467057 0.07800483549348812\n",
      "r_1220 L-tyrosine transport tyr__L_v <=> tyr__L_c\n",
      "0.010526362525668752 0.0 2.621224060817153 0.0780048354934881\n",
      "Clt chloride transport cl_e <=> cl_c\n",
      "0.010526362525668869 0.0 2.6212240608171813 0.07800483549348897\n",
      "SERD_D D-serine ammonia-lyase ser__D_c --> nh4_c + pyr_c\n",
      "0.01052636252566879 0.0 2.6212240608171626 0.07800483549348838\n",
      "GGCLUT2 Glutathione-specific gamma-glutamylcyclotransferase (Gamma-GCG) (EC 4.3.2.-) gthrd_c --> 5oxpro_c + cgly_c\n",
      "0.010526362525668808 0.0 2.621224060817166 0.07800483549348851\n",
      "r_1085 V-ATPase, Golgi atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c\n",
      "0.010526362525668742 0.001954762175049066 2.6192692986421027 0.07800483549348802\n",
      "r_1208 L-leucine transport, vacuoluar leu__L_v <=> leu__L_c\n",
      "0.010526362525668739 0.0019547621750490656 2.6192692986421022 0.078004835493488\n",
      "r_4403 Triphosphate phosphohydrolase h2o_c + pppi_c --> pi_c + ppi_c\n",
      "0.010526362525668782 0.0 2.6212240608171617 0.07800483549348833\n"
     ]
    }
   ],
   "source": [
    "sol = model.optimize()\n",
    "print(sol.fluxes['EX_zymst_e'],sol.fluxes['EX_hco3_e'],sol.fluxes['EX_co2_e'],sol.objective_value)\n",
    "for x in orphan_sce_reactions_actually_added:\n",
    "    print(x.id, x.name, x.reaction)\n",
    "\n",
    "    with temp_model:\n",
    "        temp_model.add_reactions([x])\n",
    "        sol = temp_model.optimize()\n",
    "        print(sol.fluxes['EX_zymst_e'],sol.fluxes['EX_hco3_e'],sol.fluxes['EX_co2_e'],sol.objective_value)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f80465bf",
   "metadata": {},
   "source": [
    "r_4275 Fe(II):NADP+ oxidoreductase 2.0 fe3_e + nadph_c <=> 2.0 fe2_c + h_c + nadp_c\n",
    "\n",
    "This reaction leads to no co2 outlux and to more relative outflux from zymst"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f6e2f43",
   "metadata": {},
   "source": [
    "lets see if we can figure out what is going on by adding that reaction. We will create an output flux file for visualization on escher.io."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "c971f0ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r2073_1: h_e + zn2_e <=> h_c + zn2_c\n",
      "r_4162: his__L_v <=> his__L_c\n",
      "r_4331: chtn_c + h2o_c <=> acgam_c\n",
      "GTHDH: dhdascb_c + 2.0 gthrd_c <=> ascb__L_c + gthox_c\n",
      "r_1837: his__L_c --> his__L_m\n",
      "NTP3: gtp_c + h2o_c --> gdp_c + h_c + pi_c\n",
      "MNt2: h_e + mn2_e <=> h_c + mn2_c\n",
      "ARGt2m: arg__L_c + h_c --> arg__L_m + h_m\n",
      "r_4592: h_e + mg2_e <=> h_c + mg2_c\n",
      "r_0438: focytc_m + 1.266 h_m + 0.25 o2_m --> ficytc_m + 0.5 h2o_m + 0.633 h_c\n",
      "PSP_L: h2o_c + pser__L_c --> pi_c + ser__L_c\n",
      "r_1202: ile__L_v <=> ile__L_c\n",
      "r_4163: lys__L_v <=> lys__L_c\n",
      "GLYOX3: h_c + lac__D_c <=> h2o_c + mthgxl_c\n",
      "SERAT: accoa_c + ser__L_c --> acser_c + coa_c\n",
      "r_4275: 2.0 fe3_e + nadph_c <=> 2.0 fe2_c + h_c + nadp_c\n",
      "r_4589: cu2_e + h_e <=> cu2_c + h_c\n",
      "CHOLtu: chol_c <=> chol_e\n",
      "r_1086: atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c\n",
      "r_4587: ca2_e + h_e <=> ca2_c + h_c\n",
      "r_1919: lys__L_c --> lys__L_m\n",
      "r_4312: amet_c + carn_c <=> ahcys_c + ans_c + h_c\n",
      "r_1220: tyr__L_v <=> tyr__L_c\n",
      "Clt: cl_e <=> cl_c\n",
      "SERD_D: ser__D_c --> nh4_c + pyr_c\n",
      "GGCLUT2: gthrd_c --> 5oxpro_c + cgly_c\n",
      "r_1085: atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c\n",
      "r_1208: leu__L_v <=> leu__L_c\n",
      "r_4403: h2o_c + pppi_c --> pi_c + ppi_c\n"
     ]
    }
   ],
   "source": [
    "for x in orphan_sce_reactions_actually_added:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d3770a5",
   "metadata": {},
   "source": [
    "we can separate the above reactions that need to be added into a few categories:\n",
    "\n",
    "transport (vacuole to cytoplasm) 5 rxns:\n",
    "- r_4162: his__L_v <=> his__L_c\n",
    "- r_1220: tyr__L_v <=> tyr__L_c\n",
    "- r_4163: lys__L_v <=> lys__L_c\n",
    "- r_1208: leu__L_v <=> leu__L_c\n",
    "- r_1202: ile__L_v <=> ile__L_c\n",
    "\n",
    "transport (mitochondria to cytoplasm) 3 rxns:\n",
    "- r_1919: lys__L_c --> lys__L_m\n",
    "- ARGt2m: arg__L_c + h_c --> arg__L_m + h_m\n",
    "- r_1837: his__L_c --> his__L_m\n",
    "\n",
    "transport (extracellular to cytoplasm) 8 rxns:\n",
    "- r_4587: ca2_e + h_e <=> ca2_c + h_c\n",
    "- r2073_1: h_e + zn2_e <=> h_c + zn2_c\n",
    "- MNt2: h_e + mn2_e <=> h_c + mn2_c\n",
    "- Clt: cl_e <=> cl_c\n",
    "- CHOLtu: chol_c <=> chol_e\n",
    "- r_4589: cu2_e + h_e <=> cu2_c + h_c\n",
    "- r_4592: h_e + mg2_e <=> h_c + mg2_c\n",
    "- r_4275: 2.0 fe3_e + nadph_c <=> 2.0 fe2_c + h_c + nadp_c\n",
    "\n",
    "metabolic reactions 13 rxns: \n",
    "- r_4331: chtn_c + h2o_c <=> acgam_c -- verfied, consolidated into CHTNASE. \n",
    "- r_4312: amet_c + carn_c <=> ahcys_c + ans_c + h_c\n",
    "- GGCLUT2: gthrd_c --> 5oxpro_c + cgly_c\n",
    "- NTP3: gtp_c + h2o_c --> gdp_c + h_c + pi_c\n",
    "- r_4403: h2o_c + pppi_c --> pi_c + ppi_c\n",
    "- GLYOX3: h_c + lac__D_c <=> h2o_c + mthgxl_c\n",
    "- GTHDH: dhdascb_c + 2.0 gthrd_c <=> ascb__L_c + gthox_c verfied, consolidated into DHAOX_c.\n",
    "- SERD_D: ser__D_c --> nh4_c + pyr_c\n",
    "- r_0438: focytc_m + 1.266 h_m + 0.25 o2_m --> ficytc_m + 0.5 h2o_m + 0.633 h_c\n",
    "- SERAT: accoa_c + ser__L_c --> acser_c + coa_c\n",
    "- r_1086: atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c\n",
    "- r_1085: atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c (not added to model, same genes as vacuole ATPase).\n",
    "- PSP_L: h2o_c + pser__L_c --> pi_c + ser__L_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "83a76453",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_4331</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>[1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanohydrolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16aa2e990</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>chtn_c + h2o_c <=> acgam_c</p>\n",
       "                    <p style='text-align:right'>Chitin + H2O H2O <=> N-Acetyl-D-glucosamine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YDR371W or YLR286C</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_4331 at 0x16aa2e990>"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_4331"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "244a665d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENGASE: h2o_c + s2l2n2m2m_c --> acgam_c + s2l2n2m2mn_c\n",
      "CHTNASE: chtn_c + 2.0 h2o_c --> 3.0 acgam_c\n",
      "ENGASE2: h2o_c + n2m2nm_c --> acgam_c + n2m2nmn_c\n"
     ]
    }
   ],
   "source": [
    "for m in model.metabolites.acgam_c.reactions:\n",
    "    print(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "88aa2c05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CHTNASE</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>CHTNASE</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a184bc90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>chtn_c + 2.0 h2o_c --> 3.0 acgam_c</p>\n",
       "                    <p style='text-align:right'>Chitin + 2.0 H2O H2O --> 3.0 N-Acetyl-D-glucosamine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_149679 or Lipst1_1_102785</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CHTNASE at 0x2a184bc90>"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.CHTNASE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "7b704fd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.CHTNASE.check_mass_balance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "2ef52128",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C': -16.0, 'H': -28.0, 'N': -2.0, 'O': -11.0}"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_4331.check_mass_balance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "7c3cc53d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_4331</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>[1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanohydrolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16aa2e990</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>chtn_c + h2o_c <=> acgam_c</p>\n",
       "                    <p style='text-align:right'>Chitin + H2O H2O <=> N-Acetyl-D-glucosamine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YDR371W or YLR286C</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_4331 at 0x16aa2e990>"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_4331"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6e926dc",
   "metadata": {},
   "source": [
    "sce reaction r_4331 has a different formula for chtn (c8) versus model chtn (c24). Add genes to CHTNASE in the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "9950e733",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>r_4331</td>\n",
       "      <td>[1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanoh...</td>\n",
       "      <td>chtn_c + h2o_c &lt;=&gt; acgam_c</td>\n",
       "      <td>YDR371W or YLR286C</td>\n",
       "      <td>Lipst1_1_102785 or Lipst1_1_42376</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0                                                  1  \\\n",
       "50  r_4331  [1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanoh...   \n",
       "\n",
       "                             2                   3  \\\n",
       "50  chtn_c + h2o_c <=> acgam_c  YDR371W or YLR286C   \n",
       "\n",
       "                                    4  \n",
       "50  Lipst1_1_102785 or Lipst1_1_42376  "
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='r_4331']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cd6b2ca",
   "metadata": {},
   "source": [
    "gene Lipst1_1_42376 is not yet in the model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "a650be9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.CHTNASE.gene_reaction_rule = 'Lipst1_1_149679 or Lipst1_1_102785 or Lipst1_1_42376'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a364bbb",
   "metadata": {},
   "source": [
    "second metabolic reaction.\n",
    "- r_4312 amet_c + carn_c <=> ahcys_c + ans_c + h_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "9ebd4cd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_4312</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>S-adenosyl-L-methionine:carnosine N-methyltransferase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16aa26ed0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>amet_c + carn_c <=> ahcys_c + ans_c + h_c</p>\n",
       "                    <p style='text-align:right'>S-Adenosyl-L-methionine + L-Carnosine <=> S-Adenosyl-L-homocysteine + Anserine + H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YNL092W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_4312 at 0x16aa26ed0>"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_4312"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "2de18242",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NNMT: amet_c + ncam_c --> 1mncam_c + ahcys_c Nicotinamide N methyltransferase\n",
      "SALMCOM2: adrnl_c + amet_c --> ahcys_c + h_c + mepi_c S-Adenosyl-L-methionine:catechol O-methyltransferase\n",
      "SALMCOM: amet_c + nrpphr_c --> ahcys_c + h_c + normete__L_c S-Adenosyl-L-methionine:catechol O-methyltransferase\n",
      "TM1819Z1819Z: 3.0 amet_c + dghs1819Z1819Z_c --> 3.0 ahcys_c + dgts1819Z1819Z_c + 3.0 h_c Betaine lipid synthase (trimethylase) (18:1(9Z)/18:1(9Z))\n",
      "DIPS: amet_c + caphis_c --> ahcys_c + cmaphis_c + h_c Diphthine synthase\n",
      "TM1601819Z: 3.0 amet_c + dghs1601819Z_c --> 3.0 ahcys_c + dgts1601819Z_c + 3.0 h_c Betaine lipid synthase (trimethylase) (16:0/18:1(9Z))\n",
      "AHCYStr: ahcys_c <=> ahcys_r S-Adenosyl-L-homocysteine intracellular diffusion\n",
      "MALCOAMT: amet_c + malcoa_c --> ahcys_c + malcoame_c Malonyl-CoA methyltransferase\n",
      "CCOAOMT: amet_c + caffcoa_c <=> ahcys_c + ferulcoa_c + h_c Caffeoyl-CoA O-methyltransferase\n",
      "UPP3MT: 2.0 amet_c + uppg3_c --> 2.0 ahcys_c + dscl_c + h_c Uroporphyrinogen methyltransferase\n",
      "AHCi: ahcys_c + h2o_c --> adn_c + hcys__L_c Adenosylhomocysteinase\n",
      "ACON3MT: acon_T_c + amet_c --> acon5m_c + ahcys_c Trans aconitate 3 methyltransferase\n",
      "3IPM3MT: 3c2hmp_c + amet_c --> 3ipmmest_c + ahcys_c 3 Isopropylmalate 3 methyltransferase\n",
      "ARMT: amet_c + his__L_c --> NPmehis_c + ahcys_c + h_c HnRNP arginine N methyltransferase\n",
      "34DHPHAMT: 34dhpha_c + amet_c --> ahcys_c + h_c + homoval_c 3,4-Dihydroxyphenylacetate:amet O-methyltransferase\n",
      "AMETt2m: ahcys_m + amet_c <=> ahcys_c + amet_m S-Adenosyl-L-methionine reversible transport, mitochondrial\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.ahcys_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd9b8680",
   "metadata": {},
   "source": [
    "reaction not in model. add. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "c3c8f6e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.r_4312.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "0fe694b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>r_4312</td>\n",
       "      <td>S-adenosyl-L-methionine:carnosine N-methyltran...</td>\n",
       "      <td>amet_c + carn_c &lt;=&gt; ahcys_c + ans_c + h_c</td>\n",
       "      <td>YNL092W</td>\n",
       "      <td>Lipst1_1_71580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0                                                  1  \\\n",
       "361  r_4312  S-adenosyl-L-methionine:carnosine N-methyltran...   \n",
       "\n",
       "                                             2        3               4  \n",
       "361  amet_c + carn_c <=> ahcys_c + ans_c + h_c  YNL092W  Lipst1_1_71580  "
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='r_4312']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "fa5ac2f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.r_4312.gene_reaction_rule = 'Lipst1_1_71580'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40483405",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "895d3031",
   "metadata": {},
   "source": [
    "metabolic reaction #3\n",
    "- GGCLUT2: gthrd_c --> 5oxpro_c + cgly_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "7705badd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GGCLUT2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Glutathione-specific gamma-glutamylcyclotransferase (Gamma-GCG) (EC 4.3.2.-)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a9ec410</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>gthrd_c --> 5oxpro_c + cgly_c</p>\n",
       "                    <p style='text-align:right'>Glutathione --> 5-Oxoproline + Cys-Gly</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YER163C</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GGCLUT2 at 0x16a9ec410>"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.GGCLUT2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "ea16fa74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGLYt3_2: cgly_e + 2.0 h_e --> cgly_c + 2.0 h_c Cys-Gly transport in via proton symport\n",
      "GTHRDH_syn: gthrd_c + h2o_c --> cgly_c + glu__L_c Glutathione hydralase (periplasmic)\n",
      "GTMLT: ala__L_c + gthrd_c --> cgly_c + gluala_c G glutamyltransferase\n",
      "AMPTASECG: cgly_c + h2o_c --> cys__L_c + gly_c Alanyl aminopeptidase (cys-gly)\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.cgly_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "2626d2b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OPAH: 5oxpro_c + atp_c + 2.0 h2o_c --> adp_c + glu__L_c + h_c + pi_c 5-Oxoproline amidohydrolase (ATP-hydrolysing) (ir)\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.get_by_id('5oxpro_c').reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "859e84d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>GGCLUT2</td>\n",
       "      <td>Glutathione-specific gamma-glutamylcyclotransf...</td>\n",
       "      <td>gthrd_c --&gt; 5oxpro_c + cgly_c</td>\n",
       "      <td>YER163C</td>\n",
       "      <td>Lipst1_1_47595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0                                                  1  \\\n",
       "412  GGCLUT2  Glutathione-specific gamma-glutamylcyclotransf...   \n",
       "\n",
       "                                 2        3               4  \n",
       "412  gthrd_c --> 5oxpro_c + cgly_c  YER163C  Lipst1_1_47595  "
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_copy = sce.reactions.GGCLUT2.copy()\n",
    "sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='GGCLUT2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "221706df",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.GGCLUT2.gene_reaction_rule = 'Lipst1_1_47595'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51e47497",
   "metadata": {},
   "source": [
    "metabolic reaction #4.\n",
    "- NTP3: gtp_c + h2o_c --> gdp_c + h_c + pi_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "0e603d40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>NTP3</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>nucleoside triphosphatase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a550810</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>gtp_c + h2o_c --> gdp_c + h_c + pi_c</p>\n",
       "                    <p style='text-align:right'>GTP C10H12N5O14P3 + H2O H2O --> GDP C10H12N5O11P2 + H+ + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YAL035W or YLL001W or YER005W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction NTP3 at 0x16a550810>"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.NTP3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "5b5f7d26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DOLPMTcer: dolp_c + gdpmann_c --> dolmanp_r + gdp_c Dolichyl phosphate D mannosyltransferase\n",
      "GK2: datp_c + gmp_c <=> dadp_c + gdp_c Guanylate kinase  GMPdATP \n",
      "G16MT: doldpglcnacglcnacmanman_c + gdpmann_c --> doldpglcnacglcnacman_man_man_c + gdp_c + h_c Alpha-1,6-mannosyltransferase (doldpglcnacglcnacman(man)man)\n",
      "NDPK1: atp_c + gdp_c <=> adp_c + gtp_c Nucleoside-diphosphate kinase (ATP:GDP)\n",
      "DOLPMT: dolp_c + gdpmann_c --> dolmanp_c + gdp_c Dolichyl-phosphate beta-D-mannosyltransferase\n",
      "UMPK4: gtp_c + ump_c <=> gdp_c + udp_c UMP kinase (GTP)\n",
      "GK1: atp_c + gmp_c <=> adp_c + gdp_c Guanylate kinase (GMP:ATP)\n",
      "CYTDK2: cytd_c + gtp_c --> cmp_c + gdp_c + h_c Cytidine kinase (GTP)\n",
      "G13MT_U: gdpmann_c + 0.1 mpdol_U_c --> gdp_c + h_c + 0.1 m1mpdol_U_c Glycolipid 1,3-alpha-D-mannosyltransferase (uterus)\n",
      "ATPATF3: gdp_c + gtp_c + h_c --> gp4g_c + pi_c ATP adenylyltransferase\n",
      "GTPt2m: gdp_m + gtp_c + h_c --> gdp_c + gtp_m + h_m GTPGDP translocase  mitochondrial  electroneutral \n",
      "G12MT1: doldpglcnacglcnacman_man_man_c + gdpmann_c --> doldpglcnacglcnacman_man_manman_c + gdp_c + h_c Alpha-1,2-mannosyltransferase (doldpglcnacglcnacman(man)manman)\n",
      "BDMT_U: 0.1 chito2pdol_U_c + gdpmann_c --> gdp_c + h_c + 0.1 mpdol_U_c GDPmannose:chitobiosyldiphosphodolichol beta-D-mannosyltransferase (uterus)\n",
      "RNDR2: gdp_c + trdrd_c --> dgdp_c + h2o_c + trdox_c Ribonucleoside-diphosphate reductase (GDP)\n",
      "BDMT_L: 0.1 chito2pdol__L_c + gdpmann_c --> gdp_c + h_c + 0.1 mpdol__L_c GDPmannose:chitobiosyldiphosphodolichol beta-D-mannosyltransferase (liver)\n",
      "URIK2: gtp_c + uri_c --> gdp_c + h_c + ump_c Uridine kinase (GTP:Uridine)\n",
      "FE2GTPabc: fe2_e + gtp_c + h2o_c --> fe2_c + gdp_c + h_c + pi_c Ferrous iron uptake (GTP-dependent), plasma membrane\n",
      "BDMT: doldpglcnacglcnac_c + gdpmann_c --> doldpglcnacglcnacman_c + gdp_c + h_c Chitobiosyldiphosphodolichol beta-mannosyltransferase\n",
      "ADSS: asp__L_c + gtp_c + imp_c --> dcamp_c + gdp_c + 2.0 h_c + pi_c Adenylosuccinate synthase\n",
      "G12MT2: doldpglcnacglcnacman_man_manman_c + gdpmann_c --> doldpglcnacglcnacman_man_manmanman_c + gdp_c + h_c Alpha-1,2-mannosyltransferase (doldpglcnacglcnacman(man)manmanman)\n",
      "G13MT: doldpglcnacglcnacman_c + gdpmann_c --> doldpglcnacglcnacmanman_c + gdp_c + h_c Alpha-1,3-mannosyltransferase (doldpglcnacglcnacmanman)\n",
      "G13MT_L: gdpmann_c + 0.1 mpdol__L_c --> gdp_c + h_c + 0.1 m1mpdol__L_c Glycolipid 1,3-alpha-D-mannosyltransferase (liver)\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.gdp_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "afa70a2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.NTP3.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='NTP3'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "e5ac0e0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_124163 or Lipst1_1_73932 or Lipst1_1_5321'], dtype=object)"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "9ef43a8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.NTP3.gene_reaction_rule = 'Lipst1_1_73932 or Lipst1_1_124163 or Lipst1_1_5321'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40343299",
   "metadata": {},
   "source": [
    "metabolic reaction #5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "c437a835",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_4403</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Triphosphate phosphohydrolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a8efa50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + pppi_c --> pi_c + ppi_c</p>\n",
       "                    <p style='text-align:right'>H2O H2O + Inorganic triphosphate --> Phosphate + Diphosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YHR201C or YDR452W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_4403 at 0x16a8efa50>"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_4403"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "49dcb9e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PPPIt2r: h_e + pppi_e <=> h_c + pppi_c Triphosphate reversible transport via symport\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.pppi_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "0896dae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.r_4403.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='r_4403'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "5552240d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_72748'], dtype=object)"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "5cb5a661",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.r_4403.gene_reaction_rule = 'Lipst1_1_72748'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66c78e16",
   "metadata": {},
   "source": [
    "metabolic reaction #6.\n",
    "- GLYOX3: h_c + lac__D_c <=> h2o_c + mthgxl_c\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d12c2497",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "edf44808",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYOX3</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>(R)-lactate hydro-lyase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a709e90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h_c + lac__D_c <=> h2o_c + mthgxl_c</p>\n",
       "                    <p style='text-align:right'>H+ + D-Lactate <=> H2O H2O + Methylglyoxal</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YDR533C or YMR322C or YOR391C or YPL280W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYOX3 at 0x16a709e90>"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.GLYOX3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "35cb06a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALR2: h_c + mthgxl_c + nadph_c --> acetol_c + nadp_c Aldose reductase (methylglyoxal)\n",
      "LGTHL: gthrd_c + mthgxl_c --> lgt__S_c Lactoylglutathione lyase\n",
      "LALDO3: h_c + mthgxl_c + nadph_c --> lald__L_c + nadp_c L-Lactaldehyde:NADP+ 1-oxidoreductase\n",
      "LALDO2: h_c + mthgxl_c + nadph_c --> lald__D_c + nadp_c D-Lactaldehyde:NADP+ 1-oxidoreductase\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.mthgxl_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "da9ac21e",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.GLYOX3.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='GLYOX3'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "fda26136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_3208'], dtype=object)"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "e88c5f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.GLYOX3.gene_reaction_rule = 'Lipst1_1_3208'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eabd13a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2c4ed131",
   "metadata": {},
   "source": [
    "metabolic reaction #7.\n",
    "- GTHDH: dhdascb_c + 2.0 gthrd_c <=> ascb__L_c + gthox_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69836195",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "60e52bbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GTHDH</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>glutathione:dehydroascorbate oxidoreductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a9f88d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>dhdascb_c + 2.0 gthrd_c <=> ascb__L_c + gthox_c</p>\n",
       "                    <p style='text-align:right'>Dehydroascorbate + 2.0 Glutathione <=> L-Ascorbate + Oxidized glutathione</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YKR076W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GTHDH at 0x16a9f88d0>"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.GTHDH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "80892fc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GRXR: grxox_c + 2.0 gthrd_c --> grxrd_c + gthox_c Glutaredoxin reductase\n",
      "GTHOr: gthox_c + h_c + nadph_c <=> 2.0 gthrd_c + nadp_c Glutathione oxidoreductase\n",
      "GTHPi: 2.0 gthrd_c + h2o2_c --> gthox_c + 2.0 h2o_c Glutathione peridoxase\n",
      "DHAOX_c: dhdascb_c + 2.0 gthrd_c --> ascb__L_c + gthox_c + h_c Dehydroascorbate oxidoreductase\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.gthox_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "1d738151",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DHAOX_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Dehydroascorbate oxidoreductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a191b950</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>dhdascb_c + 2.0 gthrd_c --> ascb__L_c + gthox_c + h_c</p>\n",
       "                    <p style='text-align:right'>Dehydroascorbate + 2.0 Glutathione --> L-Ascorbate + Oxidized glutathione + H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_5188</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DHAOX_c at 0x2a191b950>"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.DHAOX_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "69f4a3a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# r_copy = sce.reactions.GTHDH.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='GTHDH'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "7cff533b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_308673'], dtype=object)"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "ccc5b0b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.DHAOX_c.check_mass_balance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "c998f488",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'charge': -1.0, 'H': -1.0}"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.GTHDH.check_mass_balance()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f5ea3c",
   "metadata": {},
   "source": [
    "add reaction gene rule to model. it is the same reaction as DHAOX_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "870c9f07",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.DHAOX_c.gene_reaction_rule = 'Lipst1_1_5188 or Lipst1_1_308673'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21342998",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d818b02b",
   "metadata": {},
   "source": [
    "metabolic reaction #8.\n",
    "- SERD_D: ser__D_c --> nh4_c + pyr_c\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c141e177",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "04deccea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>SERD_D</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-serine ammonia-lyase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16aa0c290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ser__D_c --> nh4_c + pyr_c</p>\n",
       "                    <p style='text-align:right'>D-Serine --> Ammonium + Pyruvate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YGL196W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction SERD_D at 0x16aa0c290>"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.SERD_D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "9870eec2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DSERDHr: nadp_c + ser__D_c <=> 2amsa_c + h_c + nadph_c D-serine dehydrogenase\n",
      "SER_Dt: ser__D_e --> ser__D_c D-serine transport\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.ser__D_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "de700de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.SERD_D.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='SERD_D'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "80659589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_1096'], dtype=object)"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "7f8b8855",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.SERD_D.gene_reaction_rule = 'Lipst1_1_1096'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdb7cd45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fcc282e9",
   "metadata": {},
   "source": [
    "metabolic reaction #9.\n",
    "- r_0438: focytc_m + 1.266 h_m + 0.25 o2_m --> ficytc_m + 0.5 h2o_m + 0.633 h_c\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "109f09bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "1b5ccacb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_0438</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ferrocytochrome-c:oxygen oxidoreductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b04a550</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>focytc_m + 1.266 h_m + 0.25 o2_m --> ficytc_m + 0.5 h2o_m + 0.633 h_c</p>\n",
       "                    <p style='text-align:right'>Ferrocytochrome c C42H53FeN8O6S2 + 1.266 H+ + 0.25 Oxygen --> Ferricytochrome c C42H52FeN8O6S2 + 0.5 H2O H2O + 0.633 H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>(Q0045 and Q0250 and Q0275 and YDL067C and YEL039C and YGL187C and YGL191W and YHR051W and...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_0438 at 0x15b04a550>"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_0438"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "07a699f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYOR_u9m: 2.0 ficytc_m + 2.0 h_m + q9h2_m --> 2.0 focytc_m + 4.0 h_c + q9_m Ubiquinol 6 cytochrome c reductase\n",
      "CCP2_m: 2.0 focytc_m + h2o2_m + 2.0 h_m --> 2.0 ficytc_m + 2.0 h2o_m Cytochrome c peroxidase, mitochondrial\n",
      "CYOO6m: 4.0 focytc_m + 8.0 h_m + o2_m --> 4.0 ficytc_m + 2.0 h2o_m + 4.0 h_c Cytochrome c oxidase Complex IV\n",
      "D_LACDcm: 2.0 ficytc_m + lac__D_c --> 2.0 focytc_m + 2.0 h_c + pyr_c D lactate dehydrogenase  cytosolicmitochondrial\n",
      "L_LACDcm: 2.0 ficytc_m + lac__L_c --> 2.0 focytc_m + 2.0 h_c + pyr_c L-Lactate dehydrogenase, cytosolic/mitochondrial\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.focytc_m.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "c4e40c8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.CYOO6m.check_mass_balance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "a57cf1cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'charge': 0.367, 'H': 0.367}"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_0438.check_mass_balance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "555b3b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# r_copy = sce.reactions.SERD_D.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='r_0438'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "3fccd5e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_3895 or Lipst1_1_65274 or Lipst1_1_65503 or Lipst1_1_54407 or Lipst1_1_25868 or Lipst1_1_4546 or Lipst1_1_67449 or Lipst1_1_55045 or Lipst1_1_3870 or Lipst1_1_41423 or Lipst1_1_106947'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "789596da",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = x.values\n",
    "temp = [t.split(' or ') for t in temp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "1501b7ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2 = []\n",
    "for g in model.reactions.CYOO6m.genes:\n",
    "    temp2.append(g.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "4bbeb9db",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp=temp[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "2268526c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Lipst1_1_106947',\n",
       " 'Lipst1_1_25868',\n",
       " 'Lipst1_1_3870',\n",
       " 'Lipst1_1_41423',\n",
       " 'Lipst1_1_4546',\n",
       " 'Lipst1_1_65274'}"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(temp)-set(temp2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "05f3a9d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Lipst1_1_5671'}"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(temp2)-set(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "bafa2be7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Lipst1_1_54407 or Lipst1_1_65503 or Lipst1_1_67449 or Lipst1_1_3895 or Lipst1_1_55045 or Lipst1_1_5671'"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.CYOO6m.gene_reaction_rule"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "656ffdaf",
   "metadata": {},
   "source": [
    "These reactions above need consolidation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "6e1f085b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.add_reactions([r_copy])\n",
    "model.reactions.CYOO6m.gene_reaction_rule = 'Lipst1_1_3870 or Lipst1_1_54407 or Lipst1_1_65503 or Lipst1_1_25868 or Lipst1_1_4546 or Lipst1_1_55045 or Lipst1_1_106947 or Lipst1_1_3895 or Lipst1_1_41423 or Lipst1_1_65274 or Lipst1_1_67449 or Lipst1_1_5671'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bc4c5a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7302f0fb",
   "metadata": {},
   "source": [
    "metabolic reaction #9.\n",
    "- SERAT: accoa_c + ser__L_c --> acser_c + coa_c\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab67439c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "c126b637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>SERAT</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>serine O-acetyltransferase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a57a6d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>accoa_c + ser__L_c --> acser_c + coa_c</p>\n",
       "                    <p style='text-align:right'>Acetyl-CoA + L-Serine --> O-Acetyl-L-serine + Coenzyme A</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YHR013C and YDL040C and YOR253W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction SERAT at 0x16a57a6d0>"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.SERAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "9c2a7cc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYSS: acser_c + h2s_c --> ac_c + cys__L_c + h_c Cysteine synthase\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.acser_c.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "54165ada",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.SERAT.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='SERAT'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "edb7da98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_4043 or Lipst1_1_164418'], dtype=object)"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "294ffed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([r_copy])\n",
    "model.reactions.SERAT.gene_reaction_rule = 'Lipst1_1_164418 or Lipst1_1_4043'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513633d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "706c6358",
   "metadata": {},
   "source": [
    "metabolic reaction #10.\n",
    "- r_1086: atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c\n",
    "- r_1085: atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7710ed3f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "7b17cbe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_1086</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>V-ATPase, vacuole</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a593c50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP C10H12N5O13P3 + H2O H2O + H+ --> ADP C10H12N5O10P2 + 2.0 H+ + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>(YBR127C and YDL185W and YEL027W and YEL051W and YGR020C and YHR026W and YHR039C-A and YKL080W...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_1086 at 0x16a593c50>"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_1086"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "69b77e7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GLUt7: glu__L_v + h_v --> glu__L_c + h_c L glutamate transport out via proton symport  vacuolar\n",
      "PSERDv_RT: h_v + 0.01 ps_RT_v --> co2_v + 0.01 pe_RT_v Phosphatidylserine decarboxylase  yeast specific  vacuolar\n",
      "TYRt6: h_v + tyr__L_c --> h_c + tyr__L_v L tyrosine transport in via proton antiport  vacuolar \n",
      "TYRt7: h_v + tyr__L_v --> h_c + tyr__L_c L tyrosine transport out via proton symport  vacuolar\n",
      "ATPS2v: atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c Vacuolar H+-ATPase\n",
      "ILEt7: h_v + ile__L_v --> h_c + ile__L_c L isoleucine transport out via proton symport  vacuolar\n",
      "LEUt7: h_v + leu__L_v --> h_c + leu__L_c L leucine transport out via proton symport  vacuolar\n",
      "LYSt6: h_v + lys__L_c --> h_c + lys__L_v L lysine transport in via proton antiport  vacuolar \n",
      "GLNt6: gln__L_c + h_v --> gln__L_v + h_c L glutamine transport in via proton antiport  vacuolar \n",
      "ASPt7: asp__L_v + h_v --> asp__L_c + h_c L aspartate transport out via proton symport  vacuolar\n",
      "GTHRDabcv: atp_v + gthrd_c + h2o_v --> adp_v + gthrd_v + h_v + pi_v Reduced glutathione via ABC system  vacuolar \n",
      "GLNt7: gln__L_v + h_v --> gln__L_c + h_c L glutamine transport out via proton symport  vacuolar\n",
      "ASNt7: asn__L_v + h_v --> asn__L_c + h_c L asparagine transport out via proton symport  vacuolar\n",
      "TCHOLAabcv: atp_v + h2o_v + tchola_c --> adp_v + h_v + pi_v + tchola_v Taurcholate via ABC system  vacuolar \n",
      "HISt6: h_v + his__L_c --> h_c + his__L_v L histidine transport in via proton antiport  vacuolar \n",
      "ASNt6: asn__L_c + h_v --> asn__L_v + h_c L asparagine transport in via proton antiport  vacuolar \n",
      "LEUt6: h_v + leu__L_c --> h_c + leu__L_v L leucine transport in via proton antiport  vacuolar \n",
      "ILEt6: h_v + ile__L_c --> h_c + ile__L_v L isoleucine transport in via proton antiport  vacuolar \n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.h_v.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "49845045",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Lipst1_1_27636 and Lipst1_1_4281 and Lipst1_1_103286 and Lipst1_1_4502 and Lipst1_1_72599 and Lipst1_1_71403 and Lipst1_1_149291 and Lipst1_1_67780 and Lipst1_1_4471 and Lipst1_1_2781 and Lipst1_1_60868 and Lipst1_1_69438 and Lipst1_1_102936 and Lipst1_1_64406'"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.ATPS2v.gene_reaction_rule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "7e651432",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.r_1086.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='r_1086'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "2c0c095c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_2781 or Lipst1_1_102936 or Lipst1_1_60868 or Lipst1_1_71403 or Lipst1_1_69438 or Lipst1_1_4502 or Lipst1_1_149291 or Lipst1_1_72599 or Lipst1_1_103286 or Lipst1_1_4471 or Lipst1_1_67780 or Lipst1_1_4281 or Lipst1_1_27636 or Lipst1_1_64406 or Lipst1_1_51333'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "40cdbf32",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = x.values\n",
    "temp = [t.split(' or ') for t in temp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "0f184019",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Lipst1_1_2781',\n",
       "  'Lipst1_1_102936',\n",
       "  'Lipst1_1_60868',\n",
       "  'Lipst1_1_71403',\n",
       "  'Lipst1_1_69438',\n",
       "  'Lipst1_1_4502',\n",
       "  'Lipst1_1_149291',\n",
       "  'Lipst1_1_72599',\n",
       "  'Lipst1_1_103286',\n",
       "  'Lipst1_1_4471',\n",
       "  'Lipst1_1_67780',\n",
       "  'Lipst1_1_4281',\n",
       "  'Lipst1_1_27636',\n",
       "  'Lipst1_1_64406',\n",
       "  'Lipst1_1_51333']]"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "b247ae43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Lipst1_1_102936',\n",
       " 'Lipst1_1_103286',\n",
       " 'Lipst1_1_149291',\n",
       " 'Lipst1_1_27636',\n",
       " 'Lipst1_1_2781',\n",
       " 'Lipst1_1_4281',\n",
       " 'Lipst1_1_4471',\n",
       " 'Lipst1_1_4502',\n",
       " 'Lipst1_1_51333',\n",
       " 'Lipst1_1_60868',\n",
       " 'Lipst1_1_64406',\n",
       " 'Lipst1_1_67780',\n",
       " 'Lipst1_1_69438',\n",
       " 'Lipst1_1_71403',\n",
       " 'Lipst1_1_72599'}"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = set(temp[0])\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "8d10d9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2 = []\n",
    "for g in model.reactions.ATPS2v.genes:\n",
    "    temp2.append(g.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "4f27bd88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Lipst1_1_51333'}"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp-set(temp2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "69e32250",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(temp2)-temp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "700b8334",
   "metadata": {},
   "source": [
    "one extra gene was not included in the model. will add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "69879465",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.add_reactions([r_copy])\n",
    "model.reactions.ATPS2v.gene_reaction_rule = 'Lipst1_1_71403 or Lipst1_1_51333 or Lipst1_1_103286 or Lipst1_1_27636 or Lipst1_1_149291 or Lipst1_1_72599 or Lipst1_1_2781 or Lipst1_1_4471 or Lipst1_1_67780 or Lipst1_1_4502 or Lipst1_1_102936 or Lipst1_1_69438 or Lipst1_1_4281 or Lipst1_1_60868 or Lipst1_1_64406'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53887424",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3cbe4200",
   "metadata": {},
   "source": [
    "metabolic reaction #10.\n",
    "- r_1086: atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c\n",
    "- r_1085: atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6cae54e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "a3e09acf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_1085</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>V-ATPase, Golgi</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a57b850</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP C10H12N5O13P3 + H2O H2O + H+ --> ADP C10H12N5O10P2 + 2.0 H+ + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>(YPR170W-B and YBR127C and YDL185W and YOR332W and YHR026W and YEL027W and YKL080W and YLR447C...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_1085 at 0x16a57b850>"
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_1085"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "c77767ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CERGT18g: cer6_18_g + udpg_g --> glccer_18_g + h_g + udp_g Ceramide glucosyltransferase  18C\n",
      "MIPCS326g_RT: gdpmann_g + 0.01 ipc326_RT_g --> gdp_g + h_g + 0.01 mipc326_RT_g Mannose inositol phosphorylceramide synthase  ceramide 3  26C   yeast specific\n",
      "NAt3_1g: h_g + na1_c <=> h_c + na1_g Sodium proton antiporter (H:NA is 1:1)\n",
      "PSERDg_RT: h_g + 0.01 ps_RT_g --> co2_g + 0.01 pe_RT_g Phosphatidylserine decarboxylase  yeast specific  Golgi\n",
      "MIPCS324g_RT: gdpmann_g + 0.01 ipc324_RT_g --> gdp_g + h_g + 0.01 mipc324_RT_g Mannose inositol phosphorylceramide synthase  ceramide 3  24C   yeast specific\n",
      "IPC324PLCg_RT: h2o_g + 0.01 ipc324_RT_g --> cer3_24_g + h_g + mi1p__D_g Inositol phosphorylceramide  ceramide 3  24C  phospholipase C\n",
      "MIPCS126g_RT: gdpmann_g + 0.01 ipc126_RT_g --> gdp_g + h_g + 0.01 mipc126_RT_g Mannose inositol phosphorylceramide synthase  ceramide 1  26C   yeast specific\n",
      "UDPGLDCg: h_g + udpglcur_g --> co2_g + udpxyl_g UDP-D-glucuronate decarboxylase, Golgi apparatus\n",
      "MIPC124PLCg_RT: h2o_g + 0.01 mipc124_RT_g --> cer1_24_g + h_g + manmi1p__D_g Mannose inositol phosphorylceramide  ceramide 1  24C  phospholipase C\n",
      "MIP2C126PLCg_RT: h2o_g + 0.01 mip2c126_RT_g --> cer1_26_g + h_g + man2mi1p__D_g Mannose  inositol P 2 ceramide  ceramide 1  26C  phospholipase C\n",
      "MIP2C324PLCg_RT: h2o_g + 0.01 mip2c324_RT_g --> cer3_24_g + h_g + man2mi1p__D_g Mannose  inositol P 2 ceramide  ceramide 3  24C  phospholipase C\n",
      "MIP2C226PLCg_RT: h2o_g + 0.01 mip2c226_RT_g --> cer2_26_g + h_g + man2mi1p__D_g Mannose  inositol P 2 ceramide  ceramide 2  26C  phospholipase C\n",
      "MIPC226PLCg_RT: h2o_g + 0.01 mipc226_RT_g --> cer2_26_g + h_g + manmi1p__D_g Mannose inositol phosphorylceramide  ceramide 2  26C  phospholipase C\n",
      "IPC224PLCg_RT: h2o_g + 0.01 ipc224_RT_g --> cer2_24_g + h_g + mi1p__D_g Inositol phosphorylceramide  ceramide 2  24C  phospholipase C\n",
      "NDP3g: gdp_g + h2o_g --> gmp_g + h_g + pi_g Nucleoside diphosphatase  GDP   Golgi\n",
      "MIP2C224PLCg_RT: h2o_g + 0.01 mip2c224_RT_g --> cer2_24_g + h_g + man2mi1p__D_g Mannose  inositol P 2 ceramide  ceramide 2  24C  phospholipase C\n",
      "IPC326PLCg_RT: h2o_g + 0.01 ipc326_RT_g --> cer3_26_g + h_g + mi1p__D_g Inositol phosphorylceramide  ceramide 3  26C  phospholipase C\n",
      "MIPC224PLCg_RT: h2o_g + 0.01 mipc224_RT_g --> cer2_24_g + h_g + manmi1p__D_g Mannose inositol phosphorylceramide  ceramide 2  24C  phospholipase C\n",
      "IPC124PLCg_RT: h2o_g + 0.01 ipc124_RT_g --> cer1_24_g + h_g + mi1p__D_g Inositol phosphorylceramide  ceramide 1  24C  phospholipase C\n",
      "NDP7g: h2o_g + udp_g --> h_g + pi_g + ump_g Nucleoside-diphosphatase (UDP), Golgi apparatus\n",
      "IPC226PLCg_RT: h2o_g + 0.01 ipc226_RT_g --> cer2_26_g + h_g + mi1p__D_g Inositol phosphorylceramide  ceramide 2  26C  phospholipase C\n",
      "MIPCS226g_RT: gdpmann_g + 0.01 ipc226_RT_g --> gdp_g + h_g + 0.01 mipc226_RT_g Mannose inositol phosphorylceramide synthase  ceramide 2  26C   yeast specific\n",
      "MIPC326PLCg_RT: h2o_g + 0.01 mipc326_RT_g --> cer3_26_g + h_g + manmi1p__D_g Mannose inositol phosphorylceramide  ceramide 3  26C  phospholipase C\n",
      "MIPCS124g_RT: gdpmann_g + 0.01 ipc124_RT_g --> gdp_g + h_g + 0.01 mipc124_RT_g Mannose inositiol phophorylceramide synthase  ceramide 1  24C   yeast specific\n",
      "MIPC324PLCg_RT: h2o_g + 0.01 mipc324_RT_g --> cer3_24_g + h_g + manmi1p__D_g Mannose inositol phosphorylceramide  ceramide 3  24C  phospholipase C\n",
      "MIP2C326PLCg_RT: h2o_g + 0.01 mip2c326_RT_g --> cer3_26_g + h_g + man2mi1p__D_g Mannose  inositol P 2 ceramide  ceramide 3  26C  phospholipase C\n",
      "G12MT2g: gdpmann_g + m3macchitppdol_g --> gdp_g + h_g + m4macchitppdol_g Glycolipid 1 2 alpha D mannosyltransferase  Golgi apparatus\n",
      "G12MT1g: gdpmann_g + m2macchitppdol_g --> gdp_g + h_g + m3macchitppdol_g Glycolipid 1 2 alpha D mannosyltransferase  Golgi apparatus\n",
      "MIPCS224g_RT: gdpmann_g + 0.01 ipc224_RT_g --> gdp_g + h_g + 0.01 mipc224_RT_g Mannose inositol phosphorylceramide synthase  ceramide 2  24C   yeast specific\n",
      "MIPC126PLCg_RT: h2o_g + 0.01 mipc126_RT_g --> cer1_26_g + h_g + manmi1p__D_g Mannose inositol phosphorylceramide  ceramide 1  26C  phospholipase C\n",
      "MIP2C124PLCg_RT: h2o_g + 0.01 mip2c124_RT_g --> cer1_24_g + h_g + man2mi1p__D_g Mannose  inositol P 2 ceramide  ceramide 1  24C  phospholipase C\n",
      "IPC126PLCg_RT: h2o_g + 0.01 ipc126_RT_g --> cer1_26_g + h_g + mi1p__D_g Inositol phosphorylceramide  ceramide 1  26C  phospholipase C\n",
      "Kt3g: h_g + k_c <=> h_c + k_g Potassium transport  via proton antiport\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.h_g.reactions:\n",
    "    print(r, r.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f49f5158",
   "metadata": {},
   "source": [
    "no atpase reaction in golgi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "2746ca38",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.r_1085.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='r_1085'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "be13bee5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_2781 or Lipst1_1_102936 or Lipst1_1_60868 or Lipst1_1_71403 or Lipst1_1_69438 or Lipst1_1_4502 or Lipst1_1_149291 or Lipst1_1_72599 or Lipst1_1_103286 or Lipst1_1_4471 or Lipst1_1_4281 or Lipst1_1_27636 or Lipst1_1_64406 or Lipst1_1_51333'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 261,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "de5ca3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = x.values\n",
    "temp = [t.split(' or ') for t in temp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "3ecc3276",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Lipst1_1_2781',\n",
       "  'Lipst1_1_102936',\n",
       "  'Lipst1_1_60868',\n",
       "  'Lipst1_1_71403',\n",
       "  'Lipst1_1_69438',\n",
       "  'Lipst1_1_4502',\n",
       "  'Lipst1_1_149291',\n",
       "  'Lipst1_1_72599',\n",
       "  'Lipst1_1_103286',\n",
       "  'Lipst1_1_4471',\n",
       "  'Lipst1_1_4281',\n",
       "  'Lipst1_1_27636',\n",
       "  'Lipst1_1_64406',\n",
       "  'Lipst1_1_51333']]"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "aced87cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Lipst1_1_102936',\n",
       " 'Lipst1_1_103286',\n",
       " 'Lipst1_1_149291',\n",
       " 'Lipst1_1_27636',\n",
       " 'Lipst1_1_2781',\n",
       " 'Lipst1_1_4281',\n",
       " 'Lipst1_1_4471',\n",
       " 'Lipst1_1_4502',\n",
       " 'Lipst1_1_51333',\n",
       " 'Lipst1_1_60868',\n",
       " 'Lipst1_1_64406',\n",
       " 'Lipst1_1_69438',\n",
       " 'Lipst1_1_71403',\n",
       " 'Lipst1_1_72599'}"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = set(temp[0])\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "1237856d",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2 = []\n",
    "for g in model.reactions.ATPS2v.genes:\n",
    "    temp2.append(g.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "61752a59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp-set(temp2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a08d6971",
   "metadata": {},
   "source": [
    "no differences from the vacuoloe ATPASE. not added to model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "bc2122f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ATPS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ATPase cytosolic</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b7fdb10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + h2o_c --> adp_c + h_e + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP C10H12N5O13P3 + H2O H2O --> ADP C10H12N5O10P2 + H+ + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_3900 or Lipst1_1_73932</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ATPS at 0x15b7fdb10>"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.ATPS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "e8e1800d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ATPS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ATPase cytosolic</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15a0ef290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + h2o_c --> adp_c + h_e + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP C10H12N5O13P3 + H2O H2O --> ADP C10H12N5O10P2 + H+ + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>13617 or 15584</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ATPS at 0x15a0ef290>"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rto.reactions.ATPS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ec707ad",
   "metadata": {},
   "source": [
    "cytosolic ATPase may needed looking at for the GPR. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "6237ae2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.add_reactions([r_copy])\n",
    "# model.reactions.ATPS2v.gene_reaction_rule = 'Lipst1_1_71403 or Lipst1_1_51333 or Lipst1_1_103286 or Lipst1_1_27636 or Lipst1_1_149291 or Lipst1_1_72599 or Lipst1_1_2781 or Lipst1_1_4471 or Lipst1_1_67780 or Lipst1_1_4502 or Lipst1_1_102936 or Lipst1_1_69438 or Lipst1_1_4281 or Lipst1_1_60868 or Lipst1_1_64406'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30a808b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "48fe6e57",
   "metadata": {},
   "source": [
    "reaction 13.\n",
    "- PSP_L: h2o_c + pser__L_c --> pi_c + ser__L_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8289a9e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb8c6e99",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "f0e4aa97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PSP_L</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>phosphoserine phosphatase (L-serine)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a4fe890</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + pser__L_c --> pi_c + ser__L_c</p>\n",
       "                    <p style='text-align:right'>H2O + 3-phospho-serine --> phosphate + L-serine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YGR208W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PSP_L at 0x16a4fe890>"
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.PSP_L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "b7be96e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PSP_L</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>phosphoserine phosphatase L serine</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a6985910</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + pser__L_c --> pi_c + ser__L_c</p>\n",
       "                    <p style='text-align:right'>H2O H2O + O-Phospho-L-serine --> Phosphate + L-Serine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_1740</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PSP_L at 0x2a6985910>"
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.PSP_L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "8cf6e21b",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_copy = sce.reactions.PSP_L.copy()\n",
    "x = sce_added_orthologs_orhpans.loc[sce_added_orthologs_orhpans[0]=='PSP_L'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "0d9d81f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_1740'], dtype=object)"
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "bb9c9659",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.add_reactions([r_copy])\n",
    "# model.reactions.PSP_L.gene_reaction_rule = 'Lipst1_1_1740'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "id": "3561c846",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PSP_L</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>phosphoserine phosphatase L serine</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a6985910</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + pser__L_c --> pi_c + ser__L_c</p>\n",
       "                    <p style='text-align:right'>H2O H2O + O-Phospho-L-serine --> Phosphate + L-Serine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_1740</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PSP_L at 0x2a6985910>"
      ]
     },
     "execution_count": 275,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.PSP_L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6221afd2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "a4fd1ee9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YNL092W has no associated reaction\n",
      "YER163C has no associated reaction\n",
      "YAL035W has no associated reaction\n",
      "YER005W has no associated reaction\n",
      "YLL001W has no associated reaction\n",
      "YHR201C has no associated reaction\n",
      "YDR452W has no associated reaction\n",
      "YOR391C has no associated reaction\n",
      "YDR533C has no associated reaction\n",
      "YMR322C has no associated reaction\n",
      "YPL280W has no associated reaction\n",
      "YGL196W has no associated reaction\n",
      "YHR013C has no associated reaction\n",
      "YDL040C has no associated reaction\n",
      "YOR253W has no associated reaction\n",
      "31\n",
      "16\n"
     ]
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))\n",
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g2)))\n",
    "print(len(set(wrong_rx2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "e4daa5a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(wrong_rx2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "fff5d165",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACRNtp',\n",
       " 'C30CPT1',\n",
       " 'C3STKR1er',\n",
       " 'C3STKR2er',\n",
       " 'CRNtp',\n",
       " 'CYOR_u9m',\n",
       " 'DNTPPA',\n",
       " 'DRTPPD',\n",
       " 'FA160COAabcp',\n",
       " 'FA183COAabcp',\n",
       " 'FA240tp',\n",
       " 'GAPP',\n",
       " 'H2Otm',\n",
       " 'OIVD2m',\n",
       " 'PMDPHT',\n",
       " 'SERPTer'}"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2 = [x.id for x in wrong_rx2]\n",
    "set(wrong_rx2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "819f1d43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['H2Otm',\n",
       " 'DNTPPA',\n",
       " 'PMDPHT',\n",
       " 'ACRNtp',\n",
       " 'CRNtp',\n",
       " 'FA240tp',\n",
       " 'FA160COAabcp',\n",
       " 'DRTPPD']"
      ]
     },
     "execution_count": 279,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maintained_v01_gpr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc1e83b",
   "metadata": {},
   "source": [
    "## here we will try to adjust the 16 reactions to look for Lst GPRs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "f91f05d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# obtain reactions to iterate over.\n",
    "iterate_rxns = set(wrong_rx2)\n",
    "\n",
    "# reset the information holding lists. \n",
    "_names = []\n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "gpr_holder = []\n",
    "\n",
    "# iterate through the reactions. \n",
    "for rx in iterate_rxns:\n",
    "    \n",
    "    # obtain the rto reaction object.\n",
    "    r = rto.reactions.get_by_id(rx)\n",
    "    \n",
    "    \n",
    "    # objects for tracking. \n",
    "    count_ortho = 0\n",
    "    temp_gpr = []\n",
    "    \n",
    "    # iterate over the rto reaction genes. \n",
    "    for i in range(len(r.genes)):\n",
    "        \n",
    "        \n",
    "        g=list(r.genes)[i]\n",
    "        g = g.id\n",
    "        \n",
    "        # find the ortholog.\n",
    "        try:\n",
    "            \n",
    "            # find the ortholog.\n",
    "            value = lst_rto_dict[g]\n",
    "            \n",
    "            # verify that the ortholog exists & \n",
    "            # is not in the model genes &\n",
    "            if (value):\n",
    "\n",
    "                count_ortho+=1\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    # if the reaction had at least one gene ortholog in Lst. \n",
    "    if count_ortho>0:\n",
    "        # append reaction metadata. \n",
    "        _gpr.append(r.gene_reaction_rule)\n",
    "        _stoich.append(r.reaction)        \n",
    "        _names.append(r.name)\n",
    "        _id.append(r.id)  \n",
    "        \n",
    "# obtain the sce reactions that have orthologs.      \n",
    "new_iterate_rxns = set(_id)                            \n",
    "                             \n",
    "print(len(new_iterate_rxns))\n",
    "\n",
    "\n",
    "# reset the information holding lists. \n",
    "_names = []\n",
    "_gpr = []\n",
    "_id = []\n",
    "_stoich = []\n",
    "_subsystem = []\n",
    "gpr_holder = []\n",
    "\n",
    "# iterate over the reactions. \n",
    "for rx in new_iterate_rxns:\n",
    "    \n",
    "    # obtain the rto reaction object. \n",
    "    r = rto.reactions.get_by_id(rx)\n",
    "    \n",
    "    # store reaction metadata. \n",
    "    _gpr.append(r.gene_reaction_rule)\n",
    "    _stoich.append(r.reaction)\n",
    "    _names.append(r.name)\n",
    "    _id.append(r.id)\n",
    "    \n",
    "    # objects for tracking during loop. \n",
    "    count_ortho =0\n",
    "    temp_gpr = []\n",
    "    \n",
    "    # iterate over the sce reaction genes. \n",
    "    for i in range(len(r.genes)):\n",
    "\n",
    "        g=list(r.genes)[i]\n",
    "        g = g.id\n",
    "\n",
    "        try:\n",
    "            \n",
    "            # obtain the Lst ortholog. \n",
    "            value = lst_rto_dict[g]\n",
    "            \n",
    "            if (value):\n",
    "                count_ortho+=1\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "    # this implies that at least one gene has been found that was not included in the reaction.\n",
    "    if count_ortho>0:\n",
    "        \n",
    "        # re-run the gene iteration to capture all orthologs. \n",
    "        for i in range(len(r.genes)):\n",
    "            \n",
    "            g=list(r.genes)[i]\n",
    "            g = g.id\n",
    "            \n",
    "            try:\n",
    "                # obtain the Lst ortholog. \n",
    "                value = lst_rto_dict[g]\n",
    "                if (value):\n",
    "                    temp_gpr.append(value)\n",
    "            except:\n",
    "                continue\n",
    "        \n",
    "        # join multiple genes for a single reaction together using 'or' rules. \n",
    "        temp = \" or \".join(sub_item.replace(\"None\", \"\").replace(\"NA\",\" \") for sub_item in temp_gpr)\n",
    "        gpr_holder.append(temp)\n",
    "\n",
    "# pd.DataFrame() for storing the reactions.         \n",
    "rto_added_orthologs_orhpans = pd.DataFrame([_id,_names,_stoich,_gpr,gpr_holder]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "1db73f07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SERPTer</td>\n",
       "      <td>Palmitoyl-CoA:L-serine C-palmitoyltransferase ...</td>\n",
       "      <td>h_r + pmtcoa_r + ser__L_r --&gt; 3dsphgn_r + co2_...</td>\n",
       "      <td>10303 and 9394 and 9425</td>\n",
       "      <td>Lipst1_1_72156 or Lipst1_1_71191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYOR_u9m</td>\n",
       "      <td>Ubiquinol 6 cytochrome c reductase</td>\n",
       "      <td>2.0 ficytc_m + 2.0 h_m + q9h2_m --&gt; 2.0 focytc...</td>\n",
       "      <td>COB and 9705 and 9198 and 11618 and 15758 and ...</td>\n",
       "      <td>Lipst1_1_70538 or Lipst1_1_112667 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OIVD2m</td>\n",
       "      <td>2-oxoisovalerate dehydrogenase (acylating; 3-m...</td>\n",
       "      <td>3mob_m + coa_m + nad_m --&gt; co2_m + ibcoa_m + n...</td>\n",
       "      <td>(10040 and 11183 and 12566 and 15436) or (1004...</td>\n",
       "      <td>Lipst1_1_67480</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0                                                  1  \\\n",
       "0   SERPTer  Palmitoyl-CoA:L-serine C-palmitoyltransferase ...   \n",
       "1  CYOR_u9m                 Ubiquinol 6 cytochrome c reductase   \n",
       "2    OIVD2m  2-oxoisovalerate dehydrogenase (acylating; 3-m...   \n",
       "\n",
       "                                                   2  \\\n",
       "0  h_r + pmtcoa_r + ser__L_r --> 3dsphgn_r + co2_...   \n",
       "1  2.0 ficytc_m + 2.0 h_m + q9h2_m --> 2.0 focytc...   \n",
       "2  3mob_m + coa_m + nad_m --> co2_m + ibcoa_m + n...   \n",
       "\n",
       "                                                   3  \\\n",
       "0                            10303 and 9394 and 9425   \n",
       "1  COB and 9705 and 9198 and 11618 and 15758 and ...   \n",
       "2  (10040 and 11183 and 12566 and 15436) or (1004...   \n",
       "\n",
       "                                                   4  \n",
       "0                   Lipst1_1_72156 or Lipst1_1_71191  \n",
       "1  Lipst1_1_70538 or Lipst1_1_112667 or Lipst1_1_...  \n",
       "2                                     Lipst1_1_67480  "
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rto_added_orthologs_orhpans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "b7955d90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNTPPA 15385\n",
      "ACRNtp 10149\n",
      "CRNtp 10149\n",
      "FA240tp 9912\n",
      "H2Otm 9014 or Lipst1_1_156565\n",
      "SERPTer 10303 and 9394 and 9425\n",
      "CYOR_u9m 15681 or 15758 or 13608 or 13614 or 15231 or COB or 9830 or 9705 or 8802 or 11618 or 12966 or 9198 or Lipst1_1_6808 or Lipst1_1_38038 or Lipst1_1_3376 or Lipst1_1_70538 or Lipst1_1_42388 or Lipst1_1_112667\n",
      "PMDPHT 11513\n",
      "C3STKR2er 14213 or Lipst1_1_50180\n",
      "C30CPT1 9315\n",
      "GAPP 8460\n",
      "FA160COAabcp 13167 and 9637\n",
      "C3STKR1er 14213 or Lipst1_1_50180\n",
      "DRTPPD 15506\n",
      "OIVD2m (10040 and 11183 and 12566 and 15436) or (10040 and 11188 and 12566 and 15436)\n",
      "FA183COAabcp 13167 and 9637\n"
     ]
    }
   ],
   "source": [
    "for r in iterate_rxns:\n",
    "    print(r, model.reactions.get_by_id(r).gene_reaction_rule)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "6a7c4af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.H2Otm.gene_reaction_rule = 'Lipst1_1_156565'\n",
    "model.reactions.C3STKR2er.gene_reaction_rule = 'Lipst1_1_50180'\n",
    "model.reactions.C3STKR1er.gene_reaction_rule = 'Lipst1_1_50180'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "2814b190",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.SERPTer.gene_reaction_rule = 'Lipst1_1_71191 or Lipst1_1_72156'\n",
    "model.reactions.OIVD2m.gene_reaction_rule = 'Lipst1_1_67480'\n",
    "# model.reactions.C3STKR1er.gene_reaction_rule = 'Lipst1_1_50180'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "dd4d176f",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = rto_added_orthologs_orhpans[rto_added_orthologs_orhpans[0]=='CYOR_u9m'][4].values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c526f57",
   "metadata": {},
   "source": [
    "15681 or 15758 or 13608 or 12966 or 15231 or 9198 or 9830 or 9705 or 11618 or COB or 8802 or 13614 or Lipst1_1_3376 or Lipst1_1_38038 or Lipst1_1_42388 or Lipst1_1_70538 or Lipst1_1_6808 or Lipst1_1_112667"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "8b5c8a18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Lipst1_1_70538 or Lipst1_1_112667 or Lipst1_1_49747 or Lipst1_1_72848 or Lipst1_1_3376 or Lipst1_1_65274 or Lipst1_1_42388 or Lipst1_1_6808 or Lipst1_1_38038 or Lipst1_1_47735'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "b07058aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = x\n",
    "temp = [t.split(' or ') for t in temp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63c28381",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "0e0c9cd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp2 = []\n",
    "for g in model.reactions.CYOR_u9m.genes:\n",
    "    temp2.append(g.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "99d84c3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp=temp[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "89c9a495",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Lipst1_1_47735', 'Lipst1_1_49747', 'Lipst1_1_65274', 'Lipst1_1_72848'}"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(temp)-set(temp2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "0219b515",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'11618',\n",
       " '12966',\n",
       " '13608',\n",
       " '13614',\n",
       " '15231',\n",
       " '15681',\n",
       " '15758',\n",
       " '8802',\n",
       " '9198',\n",
       " '9705',\n",
       " '9830',\n",
       " 'COB'}"
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(temp2)-set(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "985de475",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.CYOR_u9m.gene_reaction_rule = 'Lipst1_1_49747 or Lipst1_1_112667 or Lipst1_1_70538 or Lipst1_1_42388 or Lipst1_1_65274 or Lipst1_1_6808 or Lipst1_1_38038 or Lipst1_1_72848 or Lipst1_1_3376 or Lipst1_1_47735'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c504d62e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "id": "11ffd5ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15231 has no associated reaction\n",
      "15681 has no associated reaction\n",
      "15758 has no associated reaction\n",
      "8802 has no associated reaction\n",
      "9198 has no associated reaction\n",
      "9705 has no associated reaction\n",
      "9830 has no associated reaction\n",
      "13608 has no associated reaction\n",
      "11618 has no associated reaction\n",
      "COB has no associated reaction\n",
      "12966 has no associated reaction\n",
      "13614 has no associated reaction\n",
      "9014 has no associated reaction\n",
      "9394 has no associated reaction\n",
      "10303 has no associated reaction\n",
      "9425 has no associated reaction\n",
      "11183 has no associated reaction\n",
      "11188 has no associated reaction\n",
      "15436 has no associated reaction\n",
      "12566 has no associated reaction\n",
      "10040 has no associated reaction\n",
      "14213 has no associated reaction\n",
      "9\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))\n",
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g2)))\n",
    "print(len(set(wrong_rx2)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "47e0e009",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACRNtp',\n",
       " 'C30CPT1',\n",
       " 'CRNtp',\n",
       " 'DNTPPA',\n",
       " 'DRTPPD',\n",
       " 'FA160COAabcp',\n",
       " 'FA183COAabcp',\n",
       " 'FA240tp',\n",
       " 'GAPP',\n",
       " 'PMDPHT'}"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2 = [x.id for x in wrong_rx2]\n",
    "set(wrong_rx2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc9b2af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "id": "8d5a11ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lst_rto_dict['9705']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "3a05949d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# set(wrong_rx2)-set(maintained_v01_gpr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "ba137db1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNTPPA\n",
      "Dihydroneopterin triphosphate pyrophosphatase\n",
      "15385\n",
      "\n",
      "\n",
      "ACRNtp\n",
      "Acetylcarnitine transport out of peroxisome\n",
      "10149\n",
      "\n",
      "\n",
      "FA240tp\n",
      "Fatty acid peroxisomal transport\n",
      "9912\n",
      "\n",
      "\n",
      "PMDPHT\n",
      "Pyrimidine phosphatase\n",
      "11513\n",
      "\n",
      "\n",
      "C30CPT1\n",
      "Production of propionylcarnitine\n",
      "9315\n",
      "\n",
      "\n",
      "GAPP\n",
      "Glyceraldehyde-3-phosphate phosphatase\n",
      "8460\n",
      "\n",
      "\n",
      "FA160COAabcp\n",
      "Fatty acyl CoA peroxisomal transport via ABC system\n",
      "13167 and 9637\n",
      "\n",
      "\n",
      "DRTPPD\n",
      "2 5 diamino 6 ribitylamino 4 3H  pyrimidinone 5  phosphate deaminase\n",
      "15506\n",
      "\n",
      "\n",
      "CRNtp\n",
      "Carnitine transport into peroxsiome\n",
      "10149\n",
      "\n",
      "\n",
      "FA183COAabcp\n",
      "Fatty acid peroxisomal transport via ABC system\n",
      "13167 and 9637\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "temp = set(wrong_rx2)\n",
    "for x in temp:\n",
    "    r = model.reactions.get_by_id(x)\n",
    "    print(r.id)\n",
    "    print(r.name)\n",
    "    print(r.gene_reaction_rule)\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "a56a2016",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSP_L</td>\n",
       "      <td>phosphoserine phosphatase L serine</td>\n",
       "      <td>h2o_c + pser__L_c --&gt; pi_c + ser__L_c</td>\n",
       "      <td>YALI0B20438g</td>\n",
       "      <td>Lipst1_1_1740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FAS160ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C160ACP mitochondrial</td>\n",
       "      <td>3.0 h_m + malACP_m + myrsACP_m + 2.0 nadph_m -...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATPS3g</td>\n",
       "      <td>ATP synthase Golgi Apparatus</td>\n",
       "      <td>adp_g + 3.0 h_c + pi_g --&gt; atp_g + h2o_g + 2.0...</td>\n",
       "      <td>YALI0A09900g and YALI0A11143g and YALI0B03982g...</td>\n",
       "      <td>Lipst1_1_149291 or Lipst1_1_69438 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOLK</td>\n",
       "      <td>Dolichol kinase</td>\n",
       "      <td>ctp_c + dolichol_c --&gt; cdp_c + dolp_c + h_c</td>\n",
       "      <td>YALI0E18942g</td>\n",
       "      <td>Lipst1_1_55005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PETHCT</td>\n",
       "      <td>phosphoethanolamine cytidyltransferase</td>\n",
       "      <td>ctp_c + ethamp_c + h_c --&gt; cdpea_c + ppi_c</td>\n",
       "      <td>YALI0C06303g</td>\n",
       "      <td>Lipst1_1_4188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FAS161ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C161ACP mitochondrial</td>\n",
       "      <td>4.0 h_m + malACP_m + myrsACP_m + 3.0 nadph_m +...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FAS80ACPm_L</td>\n",
       "      <td>fatty acyl ACP synthase n C80ACP mitochondrial...</td>\n",
       "      <td>acACP_m + 9.0 h_m + 3.0 malACP_m + 6.0 nadph_m...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FAS140ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C140ACP mitochondrial</td>\n",
       "      <td>ddcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m -...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACt2r</td>\n",
       "      <td>acetate reversible transport via proton symport</td>\n",
       "      <td>ac_e + h_e &lt;=&gt; ac_c + h_c</td>\n",
       "      <td>YALI0E04312g</td>\n",
       "      <td>Lipst1_1_7130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HBZOPT6</td>\n",
       "      <td>Hydroxybenzoate octaprenyltransferase</td>\n",
       "      <td>4hbz_c + octdp_5_c --&gt; 3ophb_5_c + ppi_c</td>\n",
       "      <td>(YALI0E09276g and YALI0B16126g) or (YALI0E1366...</td>\n",
       "      <td>Lipst1_1_302018 or Lipst1_1_114557 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FAS180ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C180ACP mitochondrial</td>\n",
       "      <td>3.0 h_m + malACP_m + 2.0 nadph_m + palmACP_m -...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FAS141ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C141ACP mitochondrial</td>\n",
       "      <td>ddcaACP_m + 4.0 h_m + malACP_m + 3.0 nadph_m +...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FA60tp</td>\n",
       "      <td>fatty acid peroxisomal transport YL</td>\n",
       "      <td>hca_c --&gt; hca_x</td>\n",
       "      <td>YALI0E16775g</td>\n",
       "      <td>Lipst1_1_5415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FA40tp</td>\n",
       "      <td>fatty acid peroxisomal transport YL</td>\n",
       "      <td>buca_c --&gt; buca_x</td>\n",
       "      <td>YALI0E16775g</td>\n",
       "      <td>Lipst1_1_5415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CYOOm</td>\n",
       "      <td>cytochrome c oxidase mitochondrial</td>\n",
       "      <td>4.0 focytc_m + 6.0 h_m + o2_m --&gt; 4.0 ficytc_m...</td>\n",
       "      <td>YALI0E10144g and YALI0E12628g and YALI0E16709g...</td>\n",
       "      <td>Lipst1_1_5671 or Lipst1_1_55045 or Lipst1_1_38...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FAS181ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C181ACP mitochondrial</td>\n",
       "      <td>4.0 h_m + malACP_m + 3.0 nadph_m + o2_m + palm...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TGL</td>\n",
       "      <td>TGL</td>\n",
       "      <td>h2o_c + 0.01 triglyc_SC_c --&gt; 0.01 12dgr_SC_c ...</td>\n",
       "      <td>YALI0D17534g or YALI0F10010g</td>\n",
       "      <td>Lipst1_1_161356 or Lipst1_1_253586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>exlip</td>\n",
       "      <td>extracelular lipases TAG</td>\n",
       "      <td>triglyc_SC_e --&gt; 0.02 dca_e + 0.06 ddca_e + 0....</td>\n",
       "      <td>YALI0A20350g</td>\n",
       "      <td>Lipst1_1_70518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FAS100ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C100ACP mitochondrial</td>\n",
       "      <td>3.0 h_m + malACP_m + 2.0 nadph_m + ocACP_m --&gt;...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>OHPHM6n</td>\n",
       "      <td>2 octaprenyl 6 hydroxyphenol methylase nuclear</td>\n",
       "      <td>2ohph_5_n + amet_n --&gt; 2omph_5_n + ahcys_n + h_n</td>\n",
       "      <td>YALI0E00770g</td>\n",
       "      <td>Lipst1_1_68931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FAS120ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C120ACP mitochondrial</td>\n",
       "      <td>dcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m --...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SPRMti</td>\n",
       "      <td>spermine irreversible uniport</td>\n",
       "      <td>sprm_e --&gt; sprm_c</td>\n",
       "      <td>YALI0A07447g or YALI0F03751g or YALI0C04477g o...</td>\n",
       "      <td>Lipst1_1_61594 or Lipst1_1_6033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FAS182ACPm</td>\n",
       "      <td>fatty acyl ACP synthase n C182ACP mitochondrial</td>\n",
       "      <td>5.0 h_m + malACP_m + 4.0 nadph_m + 2.0 o2_m + ...</td>\n",
       "      <td>YALI0D14850g and YALI0D24629g and YALI0F30679g...</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATPCitL</td>\n",
       "      <td>ATP Citrate Lyase YL</td>\n",
       "      <td>atp_c + cit_c + coa_c --&gt; accoa_c + adp_c + oa...</td>\n",
       "      <td>YALI0E34793g and YALI0D24431g</td>\n",
       "      <td>Lipst1_1_5446 or Lipst1_1_5447</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0                                                  1  \\\n",
       "0         PSP_L                 phosphoserine phosphatase L serine   \n",
       "1    FAS160ACPm    fatty acyl ACP synthase n C160ACP mitochondrial   \n",
       "2        ATPS3g                       ATP synthase Golgi Apparatus   \n",
       "3          DOLK                                    Dolichol kinase   \n",
       "4        PETHCT             phosphoethanolamine cytidyltransferase   \n",
       "5    FAS161ACPm    fatty acyl ACP synthase n C161ACP mitochondrial   \n",
       "6   FAS80ACPm_L  fatty acyl ACP synthase n C80ACP mitochondrial...   \n",
       "7    FAS140ACPm    fatty acyl ACP synthase n C140ACP mitochondrial   \n",
       "8         ACt2r    acetate reversible transport via proton symport   \n",
       "9       HBZOPT6              Hydroxybenzoate octaprenyltransferase   \n",
       "10   FAS180ACPm    fatty acyl ACP synthase n C180ACP mitochondrial   \n",
       "11   FAS141ACPm    fatty acyl ACP synthase n C141ACP mitochondrial   \n",
       "12       FA60tp                fatty acid peroxisomal transport YL   \n",
       "13       FA40tp                fatty acid peroxisomal transport YL   \n",
       "14        CYOOm                 cytochrome c oxidase mitochondrial   \n",
       "15   FAS181ACPm    fatty acyl ACP synthase n C181ACP mitochondrial   \n",
       "16          TGL                                                TGL   \n",
       "17        exlip                           extracelular lipases TAG   \n",
       "18   FAS100ACPm    fatty acyl ACP synthase n C100ACP mitochondrial   \n",
       "19      OHPHM6n     2 octaprenyl 6 hydroxyphenol methylase nuclear   \n",
       "20   FAS120ACPm    fatty acyl ACP synthase n C120ACP mitochondrial   \n",
       "21       SPRMti                      spermine irreversible uniport   \n",
       "22   FAS182ACPm    fatty acyl ACP synthase n C182ACP mitochondrial   \n",
       "23      ATPCitL                               ATP Citrate Lyase YL   \n",
       "\n",
       "                                                    2  \\\n",
       "0               h2o_c + pser__L_c --> pi_c + ser__L_c   \n",
       "1   3.0 h_m + malACP_m + myrsACP_m + 2.0 nadph_m -...   \n",
       "2   adp_g + 3.0 h_c + pi_g --> atp_g + h2o_g + 2.0...   \n",
       "3         ctp_c + dolichol_c --> cdp_c + dolp_c + h_c   \n",
       "4          ctp_c + ethamp_c + h_c --> cdpea_c + ppi_c   \n",
       "5   4.0 h_m + malACP_m + myrsACP_m + 3.0 nadph_m +...   \n",
       "6   acACP_m + 9.0 h_m + 3.0 malACP_m + 6.0 nadph_m...   \n",
       "7   ddcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m -...   \n",
       "8                           ac_e + h_e <=> ac_c + h_c   \n",
       "9            4hbz_c + octdp_5_c --> 3ophb_5_c + ppi_c   \n",
       "10  3.0 h_m + malACP_m + 2.0 nadph_m + palmACP_m -...   \n",
       "11  ddcaACP_m + 4.0 h_m + malACP_m + 3.0 nadph_m +...   \n",
       "12                                    hca_c --> hca_x   \n",
       "13                                  buca_c --> buca_x   \n",
       "14  4.0 focytc_m + 6.0 h_m + o2_m --> 4.0 ficytc_m...   \n",
       "15  4.0 h_m + malACP_m + 3.0 nadph_m + o2_m + palm...   \n",
       "16  h2o_c + 0.01 triglyc_SC_c --> 0.01 12dgr_SC_c ...   \n",
       "17  triglyc_SC_e --> 0.02 dca_e + 0.06 ddca_e + 0....   \n",
       "18  3.0 h_m + malACP_m + 2.0 nadph_m + ocACP_m -->...   \n",
       "19   2ohph_5_n + amet_n --> 2omph_5_n + ahcys_n + h_n   \n",
       "20  dcaACP_m + 3.0 h_m + malACP_m + 2.0 nadph_m --...   \n",
       "21                                  sprm_e --> sprm_c   \n",
       "22  5.0 h_m + malACP_m + 4.0 nadph_m + 2.0 o2_m + ...   \n",
       "23  atp_c + cit_c + coa_c --> accoa_c + adp_c + oa...   \n",
       "\n",
       "                                                    3  \\\n",
       "0                                        YALI0B20438g   \n",
       "1   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "2   YALI0A09900g and YALI0A11143g and YALI0B03982g...   \n",
       "3                                        YALI0E18942g   \n",
       "4                                        YALI0C06303g   \n",
       "5   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "6   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "7   YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "8                                        YALI0E04312g   \n",
       "9   (YALI0E09276g and YALI0B16126g) or (YALI0E1366...   \n",
       "10  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "11  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "12                                       YALI0E16775g   \n",
       "13                                       YALI0E16775g   \n",
       "14  YALI0E10144g and YALI0E12628g and YALI0E16709g...   \n",
       "15  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "16                       YALI0D17534g or YALI0F10010g   \n",
       "17                                       YALI0A20350g   \n",
       "18  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "19                                       YALI0E00770g   \n",
       "20  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "21  YALI0A07447g or YALI0F03751g or YALI0C04477g o...   \n",
       "22  YALI0D14850g and YALI0D24629g and YALI0F30679g...   \n",
       "23                      YALI0E34793g and YALI0D24431g   \n",
       "\n",
       "                                                    4  \n",
       "0                                       Lipst1_1_1740  \n",
       "1   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "2   Lipst1_1_149291 or Lipst1_1_69438 or Lipst1_1_...  \n",
       "3                                      Lipst1_1_55005  \n",
       "4                                       Lipst1_1_4188  \n",
       "5   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "6   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "7   Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "8                                       Lipst1_1_7130  \n",
       "9   Lipst1_1_302018 or Lipst1_1_114557 or Lipst1_1...  \n",
       "10  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "11  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "12                                      Lipst1_1_5415  \n",
       "13                                      Lipst1_1_5415  \n",
       "14  Lipst1_1_5671 or Lipst1_1_55045 or Lipst1_1_38...  \n",
       "15  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "16                 Lipst1_1_161356 or Lipst1_1_253586  \n",
       "17                                     Lipst1_1_70518  \n",
       "18  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "19                                     Lipst1_1_68931  \n",
       "20  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "21                    Lipst1_1_61594 or Lipst1_1_6033  \n",
       "22  Lipst1_1_3273 or Lipst1_1_2955 or Lipst1_1_106903  \n",
       "23                     Lipst1_1_5446 or Lipst1_1_5447  "
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip_added_orthologs_orhpans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "a4b3d775",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>r_2594</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3250 --&gt; h_mm + ocdca_mm + s_3275</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>r_2852</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-16:0, 2-16:1)...</td>\n",
       "      <td>atp_vm + h_vm + s_3340 --&gt; adp_vm + s_3392</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>r_2586</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3236 --&gt; h_mm + hdca_mm + s_3273</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>r_2719</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3284 + s_3302 &lt;=&gt; 1agpc181_mm + s_3235</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>r_2731</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3285 + s_3307 &lt;=&gt; 1agpc181_mm + s_3314</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>r_2632</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3272 + s_3304 &lt;=&gt; 1agpc160_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>r_3266</td>\n",
       "      <td>TAG lipase (1-16:0, 2-16:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3067 --&gt; h_lp + ocdca_lp + s_3043</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>r_2764</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3290 + s_3296 &lt;=&gt; 1agpc160_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>r_3293</td>\n",
       "      <td>TAG lipase (1-18:1, 2-18:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3066 --&gt; 12dgr181_lp + h_lp + hdcea_lp</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>r_2328</td>\n",
       "      <td>acyl dhap reductase (16:0), lipid particle</td>\n",
       "      <td>h_lp + nadph_lp + s_2948 --&gt; 1hdecg3p_lp + nad...</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>r_4194</td>\n",
       "      <td>Zinc transporter YKE4</td>\n",
       "      <td>zn2_c &lt;=&gt; zn2_er</td>\n",
       "      <td>YIL023C</td>\n",
       "      <td>Lipst1_1_69058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>r_4200</td>\n",
       "      <td>Glutathione S-transferase 1 (EC 2.5.1.18) (GST-I)</td>\n",
       "      <td>gthrd_c + s_3834 &lt;=&gt; s_3835 + s_3836</td>\n",
       "      <td>YMR251W or YKR076W</td>\n",
       "      <td>Lipst1_1_308673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>r_2327</td>\n",
       "      <td>acyl dhap reductase (18:1), ER membrane</td>\n",
       "      <td>h_erm + nadph_erm + s_2943 --&gt; nadp_erm + s_2938</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>r_3279</td>\n",
       "      <td>TAG lipase (1-18:1, 2-16:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + tag1819Z1619Z1819Z_lp --&gt; h_lp + ocdc...</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>r2073_1</td>\n",
       "      <td>Zn(2+) transport</td>\n",
       "      <td>h_e + zn2_e &lt;=&gt; h_c + zn2_c</td>\n",
       "      <td>YGL255W or YLR130C</td>\n",
       "      <td>Lipst1_1_71180 or Lipst1_1_2861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>r_3281</td>\n",
       "      <td>TAG lipase (1-16:0, 2-18:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3060 --&gt; h_lp + hdcea_lp + s_3045</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>r_2703</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3282 + s_3302 &lt;=&gt; 1agpc181_mm + s_3268</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>r_2675</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3278 + s_3307 &lt;=&gt; 1agpc181_mm + s_3308</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>r_3286</td>\n",
       "      <td>TAG lipase (1-16:1, 2-18:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3070 --&gt; h_lp + ocdca_lp + s_3049</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>r_2745</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3287 + s_3305 &lt;=&gt; 1agpc161_mm + s_3315</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>r_2712</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3283 + s_3304 &lt;=&gt; 1agpc160_mm + s_3313</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>r_2678</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3279 + s_3300 &lt;=&gt; 1agpc180_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>r_4162</td>\n",
       "      <td>Probable vacuolar amino acid transporter YPQ3 ...</td>\n",
       "      <td>his__L_v &lt;=&gt; his__L_c</td>\n",
       "      <td>YBR147W or YDR352W</td>\n",
       "      <td>Lipst1_1_211414 or Lipst1_1_62919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>r_4174</td>\n",
       "      <td>Probable ATP-dependent permease</td>\n",
       "      <td>chsterol_er &lt;=&gt; chsterol_erm</td>\n",
       "      <td>YCR011C</td>\n",
       "      <td>Lipst1_1_235077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>r_3252</td>\n",
       "      <td>ergosteryl ester hydrolase (16:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3656 --&gt; ergst_lp + h_lp + hdcea_lp</td>\n",
       "      <td>YKL140W or YLL012W</td>\n",
       "      <td>Lipst1_1_1895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>r_2859</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-18:1, 2-18:1)...</td>\n",
       "      <td>atp_vm + h_vm + s_3356 --&gt; adp_vm + s_3399</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>r_2647</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3274 + s_3302 &lt;=&gt; 1agpc181_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BZALDH</td>\n",
       "      <td>Benzyl alcohol:NAD+ oxidoreductase</td>\n",
       "      <td>bzalc_c + nad_c &lt;=&gt; bzal_c + h_c + nadh_c</td>\n",
       "      <td>YCR107W or YDL243C or YJR155W or YNL331C or YO...</td>\n",
       "      <td>Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>r_2324</td>\n",
       "      <td>acyl dhap reductase (16:0), ER membrane</td>\n",
       "      <td>h_erm + nadph_erm + s_2940 --&gt; 1hdecg3p_erm + ...</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>r_3264</td>\n",
       "      <td>TAG lipase (1-16:0, 2-16:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3044 --&gt; h_lp + hdca_lp + s_3043</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C or YC...</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>r_2600</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3262 --&gt; h_mm + hdca_mm + s_3282</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>r_2750</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3288 + s_3300 &lt;=&gt; 1agpc180_mm + s_3247</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>r_2715</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3283 + s_3307 &lt;=&gt; 1agpc181_mm + s_3313</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>r_4219</td>\n",
       "      <td>D-aminoacyl-tRNA deacylase (DTD) (EC 3.1.1.96)...</td>\n",
       "      <td>h2o_c + s_3857 &lt;=&gt; h_c + s_3858 + trnatyr_c</td>\n",
       "      <td>YDL219W</td>\n",
       "      <td>Lipst1_1_50667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>r_2736</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3286 + s_3304 &lt;=&gt; 1agpc160_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>r_2608</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3246 --&gt; h_mm + hdca_mm + s_3288</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>r_3300</td>\n",
       "      <td>DAG lipase (1-16:0, 2-18:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3045 --&gt; h_lp + mag160_lp + ocdcea_lp</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>r_2699</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3281 + s_3307 &lt;=&gt; 1agpc181_mm + s_3311</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>r_3291</td>\n",
       "      <td>TAG lipase (1-18:0, 2-18:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3080 --&gt; h_lp + ocdcea_lp + s_3053</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>r_2645</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3274 &lt;=&gt; 1agpc161_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>r_0044</td>\n",
       "      <td>3-hexaprenyl-4-hydroxy-5-methoxybenzoate decar...</td>\n",
       "      <td>3hph5mb_m + h_m --&gt; 2hp6mp_m + co2_m</td>\n",
       "      <td>YDR538W and YDR539W</td>\n",
       "      <td>Lipst1_1_198908 or Lipst1_1_76260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>r_2796</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3294 + s_3296 &lt;=&gt; 1agpc160_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>r_2656</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3275 + s_3304 &lt;=&gt; 1agpc160_mm + s_3256</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>r_0963</td>\n",
       "      <td>quinone oxidoreductase</td>\n",
       "      <td>2hpmmbq_m + h_m + nadh_m + o2_m --&gt; 2hpmhmbq_m...</td>\n",
       "      <td>YDR204W and YGR255C and YLR201C and YML110C an...</td>\n",
       "      <td>Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>r_2590</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3246 --&gt; h_mm + ocdca_mm + s_3270</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>r_2633</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3272 + s_3305 &lt;=&gt; 1agpc161_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>r_4333</td>\n",
       "      <td>Polyphosphate phosphohydrolase</td>\n",
       "      <td>h2o_c + s_4018 &lt;=&gt; h_c + 2.0 pi_c</td>\n",
       "      <td>YHR201C or YDR452W</td>\n",
       "      <td>Lipst1_1_72748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>r_3215</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-18:1, 2-18:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3615 --&gt; 2.0 h_erm + pi_erm + s_3591</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>r_2698</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3281 + s_3306 &lt;=&gt; 1agpc180_mm + s_3311</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>r_3817</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:0, 2-18:1) transpo...</td>\n",
       "      <td>s_3113 &lt;=&gt; s_3155</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>r_4331</td>\n",
       "      <td>[1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanoh...</td>\n",
       "      <td>chtn_c + h2o_c &lt;=&gt; acgam_c</td>\n",
       "      <td>YDR371W or YLR286C</td>\n",
       "      <td>Lipst1_1_102785 or Lipst1_1_42376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>FMNRx</td>\n",
       "      <td>FMN reductase</td>\n",
       "      <td>fmn_c + 2.0 h_c + nadh_c --&gt; fmnh2_c + nad_c</td>\n",
       "      <td>YLR011W</td>\n",
       "      <td>Lipst1_1_52431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>r_2694</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3281 + s_3300 &lt;=&gt; 1agpc180_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>G6PI_1</td>\n",
       "      <td>alpha-D-Glucose 6-phosphate ketol-isomerase</td>\n",
       "      <td>g6p_A_c &lt;=&gt; g6p_B_c</td>\n",
       "      <td>YMR099C</td>\n",
       "      <td>Lipst1_1_68948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>r_2622</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3270 + s_3300 &lt;=&gt; 1agpc180_mm + s_3240</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>r_4242</td>\n",
       "      <td>Initiation-specific alpha-1,6-mannosyltransfer...</td>\n",
       "      <td>gdpmann_er + s_3891 &lt;=&gt; gdp_er + h_er + s_3893</td>\n",
       "      <td>YGL038C</td>\n",
       "      <td>Lipst1_1_37797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>r_2720</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3284 + s_3304 &lt;=&gt; 1agpc160_mm + s_3237</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>r_2757</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3289 &lt;=&gt; 1agpc161_mm + s_3251</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>r_2643</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3273 + s_3307 &lt;=&gt; 1agpc181_mm + s_3254</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>r_4199</td>\n",
       "      <td>Glutathione S-transferase 1 (EC 2.5.1.18) (GST-I)</td>\n",
       "      <td>gthrd_er + s_3830 &lt;=&gt; s_3832 + s_3833</td>\n",
       "      <td>YIR038C</td>\n",
       "      <td>Lipst1_1_74576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>FMNRx2</td>\n",
       "      <td>FMN reductase</td>\n",
       "      <td>fmn_c + 2.0 h_c + nadph_c --&gt; fmnh2_c + nadp_c</td>\n",
       "      <td>YLR011W</td>\n",
       "      <td>Lipst1_1_52431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>GTHDH</td>\n",
       "      <td>glutathione:dehydroascorbate oxidoreductase</td>\n",
       "      <td>dhdascb_c + 2.0 gthrd_c &lt;=&gt; ascb__L_c + gthox_c</td>\n",
       "      <td>YKR076W</td>\n",
       "      <td>Lipst1_1_308673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>r_4159</td>\n",
       "      <td>UDP-N-acetylglucosamine transferase subunit AL...</td>\n",
       "      <td>naglc2p_er + uacgam_er &lt;=&gt; h_er + s_3767 + udp_er</td>\n",
       "      <td>YBR070C and YGL047W</td>\n",
       "      <td>Lipst1_1_5477 or Lipst1_1_74719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>r_0993</td>\n",
       "      <td>serine palmitotransferase</td>\n",
       "      <td>h_er + pmtcoa_er + ser__L_er --&gt; 3dsphgn_er + ...</td>\n",
       "      <td>YMR296C and YKL212W and YDR062W and YGR038W an...</td>\n",
       "      <td>Lipst1_1_71191 or Lipst1_1_72156 or Lipst1_1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>CYANSTm</td>\n",
       "      <td>thiosulfate:cyanide sulfurtranserase</td>\n",
       "      <td>cyan_m + tsul_m &lt;=&gt; 2.0 h_m + so3_m + tcynt_m</td>\n",
       "      <td>YOR251C or YOR285W or YOR286W</td>\n",
       "      <td>Lipst1_1_965 or Lipst1_1_60674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>r_4176</td>\n",
       "      <td>Putative aryl-alcohol dehydrogenase AAD3 (EC 1...</td>\n",
       "      <td>nad_c + s_3791 &lt;=&gt; h_c + nadh_c + s_3792</td>\n",
       "      <td>YCR107W or YDL243C or YJR155W or YNL331C or YO...</td>\n",
       "      <td>Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>r_2772</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3291 + s_3296 &lt;=&gt; 1agpc160_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>r_2748</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3288 + s_3296 &lt;=&gt; 1agpc160_mm + s_3247</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>r_2646</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3274 + s_3300 &lt;=&gt; 1agpc180_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>r_1837</td>\n",
       "      <td>L-histidine transport, mitochondrial</td>\n",
       "      <td>his__L_c --&gt; his__L_m</td>\n",
       "      <td>YBR147W</td>\n",
       "      <td>Lipst1_1_211414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NTP3</td>\n",
       "      <td>nucleoside triphosphatase</td>\n",
       "      <td>gtp_c + h2o_c --&gt; gdp_c + h_c + pi_c</td>\n",
       "      <td>YAL035W or YLL001W or YER005W</td>\n",
       "      <td>Lipst1_1_124163 or Lipst1_1_73932 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>r_2692</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3281 + s_3296 &lt;=&gt; 1agpc160_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>r_2727</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3285 + s_3302 &lt;=&gt; 1agpc181_mm + s_3239</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>r_2840</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-16:0, 2-18:1)...</td>\n",
       "      <td>atp_gm + h_gm + s_3368 --&gt; adp_gm + s_3369</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>r_2673</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3278 + s_3305 &lt;=&gt; 1agpc161_mm + s_3308</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>r_2855</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-18:1, 2-16:1)...</td>\n",
       "      <td>atp_vm + h_vm + s_3348 --&gt; adp_vm + s_3395</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>r_2741</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3287 &lt;=&gt; 1agpc161_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>r_0192</td>\n",
       "      <td>alpha 1,2-mannosyltransferase</td>\n",
       "      <td>gdpmann_g + h_g + mpdol_g --&gt; gdp_g + s_0414</td>\n",
       "      <td>YBR199W or YBR205W or YDR483W or YJL139C or YK...</td>\n",
       "      <td>Lipst1_1_384 or Lipst1_1_4409 or Lipst1_1_67994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>r_2701</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3282 &lt;=&gt; 1agpc161_mm + s_3268</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>r_2771</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3290 + s_3307 &lt;=&gt; 1agpc181_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>r_2662</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3276 + s_3300 &lt;=&gt; 1agpc180_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>r_2682</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3279 + s_3306 &lt;=&gt; 1agpc180_mm + s_3309</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>r_2691</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3280 + s_3307 &lt;=&gt; 1agpc181_mm + s_3310</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>r_2615</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3260 --&gt; h_mm + ocdca_mm + s_3292</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>r_2751</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3288 + s_3302 &lt;=&gt; 1agpc181_mm + s_3247</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>r_2659</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3275 + s_3307 &lt;=&gt; 1agpc181_mm + s_3256</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>r_2790</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3293 + s_3300 &lt;=&gt; 1agpc180_mm + s_3263</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>r_3815</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:0, 2-16:1) transpo...</td>\n",
       "      <td>s_3111 &lt;=&gt; s_3151</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>r_2810</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3295 + s_3306 &lt;=&gt; 1agpc180_mm + s_3319</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>r_3220</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-16:0, 2-18:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3420 --&gt; 2.0 h_vm + pi_vm + s_3396</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>r_4250</td>\n",
       "      <td>Dolichyl-diphosphate phosphohydrolase</td>\n",
       "      <td>h2o_er + h_er + s_3888 &lt;=&gt; dolp_er + pi_er</td>\n",
       "      <td>YGR036C</td>\n",
       "      <td>Lipst1_1_52818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>MNt2</td>\n",
       "      <td>Mn(2+) transport</td>\n",
       "      <td>h_e + mn2_e &lt;=&gt; h_c + mn2_c</td>\n",
       "      <td>YOL130W</td>\n",
       "      <td>Lipst1_1_141125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>r_2326</td>\n",
       "      <td>acyl dhap reductase (18:0), ER membrane</td>\n",
       "      <td>h_erm + nadph_erm + s_2942 --&gt; 1odecg3p_erm + ...</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>ARGt2m</td>\n",
       "      <td>arginine transport</td>\n",
       "      <td>arg__L_c + h_c --&gt; arg__L_m + h_m</td>\n",
       "      <td>YBR147W</td>\n",
       "      <td>Lipst1_1_211414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>r_3313</td>\n",
       "      <td>lysoPE (1-16:1) oleoyl-CoA acyltransferase, li...</td>\n",
       "      <td>1agpe161_lp + ocdce9coa_lp --&gt; coa_lp + s_3688</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>r_2854</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-18:0, 2-16:1)...</td>\n",
       "      <td>atp_vm + h_vm + s_3346 --&gt; adp_vm + s_3394</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>r_2584</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3234 --&gt; h_mm + hdca_mm + s_3270</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>r_3217</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-16:1, 2-16:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3417 --&gt; 2.0 h_vm + pail3p161_vm + ...</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>r_3222</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-18:0, 2-18:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3422 --&gt; 2.0 h_vm + pi_vm + s_3398</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>r_2809</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3295 + s_3305 &lt;=&gt; 1agpc161_mm + s_3319</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>r_2733</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3286 &lt;=&gt; 1agpc161_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>r_2711</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3283 + s_3302 &lt;=&gt; 1agpc181_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>r_4161</td>\n",
       "      <td>Probable vacuolar amino acid transporter YPQ3 ...</td>\n",
       "      <td>arg__L_v &lt;=&gt; arg__L_c</td>\n",
       "      <td>YBR147W or YDR352W</td>\n",
       "      <td>Lipst1_1_211414 or Lipst1_1_62919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>r_0089</td>\n",
       "      <td>5PP-IP5 pyrophosphorylation to 5,6-PP2-IP4</td>\n",
       "      <td>2.0 h_c + pi_c + ppmi12346p_c &lt;=&gt; h2o_c + s_0309</td>\n",
       "      <td>YLR410W</td>\n",
       "      <td>Lipst1_1_57858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>r_2668</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3278 + s_3296 &lt;=&gt; 1agpc160_mm + s_3264</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>r_3296</td>\n",
       "      <td>DAG lipase (1-16:0, 2-16:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3043 --&gt; h_lp + hdcea_lp + mag160_lp</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>r_2621</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3270 &lt;=&gt; 1agpc161_mm + s_3240</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>r_2794</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3293 + s_3306 &lt;=&gt; 1agpc180_mm + s_3318</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>r_2680</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3279 + s_3304 &lt;=&gt; 1agpc160_mm + s_3309</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>r_2747</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3287 + s_3307 &lt;=&gt; 1agpc181_mm + s_3315</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>r_2679</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3279 + s_3302 &lt;=&gt; 1agpc181_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>r_2763</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3289 + s_3307 &lt;=&gt; 1agpc181_mm + s_3316</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>r_2585</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-16:1, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3235 --&gt; h_mm + hdca_mm + s_3272</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>r_0520</td>\n",
       "      <td>GPI-anchor assembly, step 6</td>\n",
       "      <td>pe_hs_er + s_0334 &lt;=&gt; s_0337</td>\n",
       "      <td>YKL165C</td>\n",
       "      <td>Lipst1_1_70311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>r_2592</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3248 --&gt; h_mm + ocdca_mm + s_3273</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>r_2755</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3288 + s_3307 &lt;=&gt; 1agpc181_mm + s_3249</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>r_3298</td>\n",
       "      <td>DAG lipase (1-18:0, 2-16:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3051 --&gt; h_lp + hdcea_lp + mag180_lp</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>r_2726</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3285 + s_3300 &lt;=&gt; 1agpc180_mm + s_3239</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>r_0249</td>\n",
       "      <td>CAAX farnesyltransferase</td>\n",
       "      <td>4hbz_c + s_1231 --&gt; ppi_c + s_1230</td>\n",
       "      <td>YDL090C and YKL019W</td>\n",
       "      <td>Lipst1_1_302018 or Lipst1_1_4158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>r_2754</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3288 + s_3306 &lt;=&gt; 1agpc180_mm + s_3249</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>r_2787</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3292 + s_3307 &lt;=&gt; 1agpc181_mm + s_3261</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>r_2702</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3282 + s_3300 &lt;=&gt; 1agpc180_mm + s_3268</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>r_2774</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3291 + s_3300 &lt;=&gt; 1agpc180_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>r_3292</td>\n",
       "      <td>TAG lipase (1-18:1, 2-18:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3058 --&gt; 12dgr181_lp + h_lp + hdca_lp</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>r_3894</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:1, 2-16:1) transpo...</td>\n",
       "      <td>s_3110 &lt;=&gt; s_3480</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>r_2671</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3278 + s_3302 &lt;=&gt; 1agpc181_mm + s_3264</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>r_2710</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3283 + s_3300 &lt;=&gt; 1agpc180_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>r_3287</td>\n",
       "      <td>TAG lipase (1-16:1, 2-18:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3078 --&gt; h_lp + ocdcea_lp + s_3049</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>r_3900</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:1, 2-18:1) transpo...</td>\n",
       "      <td>s_3116 &lt;=&gt; s_3489</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>r_2672</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3278 + s_3304 &lt;=&gt; 1agpc160_mm + s_3308</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>r_2778</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3291 + s_3306 &lt;=&gt; 1agpc180_mm + s_3317</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>r_2800</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3294 + s_3304 &lt;=&gt; 1agpc160_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>r_4204</td>\n",
       "      <td>Vacuolar aminopeptidase 1 (EC 3.4.11.22) (Amin...</td>\n",
       "      <td>ala_L_ile__L_v + h2o_v &lt;=&gt; ala__L_v + leu__L_v</td>\n",
       "      <td>YKL103C</td>\n",
       "      <td>Lipst1_1_106140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>r_3214</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-18:0, 2-18:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3614 --&gt; 2.0 h_erm + pi_erm + s_3590</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>r_2612</td>\n",
       "      <td>CL (1-18:1, 2-16:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3254 --&gt; h_mm + ocdca_mm + s_3290</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>r_3295</td>\n",
       "      <td>TAG lipase (1-18:1, 2-18:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3082 --&gt; 12dgr181_lp + h_lp + ocdce...</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>r_4704</td>\n",
       "      <td>alkanesulfonate transport</td>\n",
       "      <td>s_4266 --&gt; s_4265</td>\n",
       "      <td>YIL166C</td>\n",
       "      <td>Lipst1_1_6611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>PETHCT</td>\n",
       "      <td>phosphoethanolamine cytidyltransferase</td>\n",
       "      <td>ctp_c + ethamp_c + h_c --&gt; cdpea_c + ppi_c</td>\n",
       "      <td>YGR007W</td>\n",
       "      <td>Lipst1_1_4188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>r_0518</td>\n",
       "      <td>GPI-anchor assembly, step 2</td>\n",
       "      <td>acgpail_hs_er + h2o_er &lt;=&gt; ac_er + gpail_hs_er...</td>\n",
       "      <td>YMR281W</td>\n",
       "      <td>Lipst1_1_294677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>r_2690</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3280 + s_3306 &lt;=&gt; 1agpc180_mm + s_3310</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>r_2661</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3276 &lt;=&gt; 1agpc161_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>3OAS40_m</td>\n",
       "      <td>B-ketoacyl-ACP synthase (acetoacetyl-ACP)</td>\n",
       "      <td>acACP_m + h_m + malACP_m --&gt; ACP_m + actACP_m ...</td>\n",
       "      <td>YER061C and YKL192C</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>r_2144</td>\n",
       "      <td>B-ketoacyl-ACP synthase (3-oxo-octanoyl-ACP)</td>\n",
       "      <td>h_m + malACP_m + s_2773 --&gt; 3ooctACP_m + ACP_m...</td>\n",
       "      <td>YER061C and YKL192C</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>r_2331</td>\n",
       "      <td>acyl dhap reductase (18:1), lipid particle</td>\n",
       "      <td>h_lp + nadph_lp + s_2951 --&gt; nadp_lp + s_2947</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>r_2602</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3234 --&gt; h_mm + hdca_mm + s_3284</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>r_2792</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3293 + s_3304 &lt;=&gt; 1agpc160_mm + s_3318</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>r_2729</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3285 + s_3305 &lt;=&gt; 1agpc161_mm + s_3314</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>r_2760</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3289 + s_3304 &lt;=&gt; 1agpc160_mm + s_3316</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>r_2713</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3283 + s_3305 &lt;=&gt; 1agpc161_mm + s_3313</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>r_3209</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-16:1, 2-16:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3609 --&gt; 2.0 h_erm + pail4p161_erm...</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>r_2836</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-16:0, 2-16:1)...</td>\n",
       "      <td>atp_gm + h_gm + s_3358 --&gt; adp_gm + s_3361</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>r_2738</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3286 + s_3306 &lt;=&gt; 1agpc180_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>r_2588</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3238 --&gt; h_mm + hdca_mm + s_3275</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>r_2740</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3287 + s_3296 &lt;=&gt; 1agpc160_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>r_0767</td>\n",
       "      <td>NAD nucleosidase nuclear</td>\n",
       "      <td>h2o_n + nad_n --&gt; adprib_n + h_n + ncam_n</td>\n",
       "      <td>YDL042C or YDR191W or YOL068C or YOR025W or YP...</td>\n",
       "      <td>Lipst1_1_336196 or Lipst1_1_2970 or Lipst1_1_7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>r_2700</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3282 + s_3296 &lt;=&gt; 1agpc160_mm + s_3268</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>r_3267</td>\n",
       "      <td>TAG lipase (1-16:0, 2-16:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3075 --&gt; h_lp + ocdcea_lp + s_3043</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>r_4592</td>\n",
       "      <td>Mg(2+) transport</td>\n",
       "      <td>h_e + mg2_e &lt;=&gt; h_c + mg2_c</td>\n",
       "      <td>YGR191W or YOL130W</td>\n",
       "      <td>Lipst1_1_141125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>r_4218</td>\n",
       "      <td>D-aminoacyl-tRNA deacylase (DTD) (EC 3.1.1.96)...</td>\n",
       "      <td>h2o_c + s_3856 &lt;=&gt; gly_c + h_c + trnaala_c</td>\n",
       "      <td>YDL219W</td>\n",
       "      <td>Lipst1_1_50667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>r_3897</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:0, 2-18:1) transpo...</td>\n",
       "      <td>s_3113 &lt;=&gt; s_3486</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>r_2786</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3292 + s_3306 &lt;=&gt; 1agpc180_mm + s_3261</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>r_2665</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3276 + s_3305 &lt;=&gt; 1agpc161_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>r_4308</td>\n",
       "      <td>Probable phospholipid-transporting ATPase DNF3...</td>\n",
       "      <td>atp_erm + h2o_erm + s_3985 --&gt; adp_erm + h_erm...</td>\n",
       "      <td>YMR162C</td>\n",
       "      <td>Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>r_2784</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3292 + s_3304 &lt;=&gt; 1agpc160_mm + s_3261</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>r_4751</td>\n",
       "      <td>D-phenylalanine N-acetyltransferase</td>\n",
       "      <td>accoa_c + s_4296 --&gt; coa_c + s_4297</td>\n",
       "      <td>YEL066W or YPR193C</td>\n",
       "      <td>Lipst1_1_48516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>r_2717</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3284 &lt;=&gt; 1agpc161_mm + s_3235</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>r_2782</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3292 + s_3300 &lt;=&gt; 1agpc180_mm + s_3259</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>r_2143</td>\n",
       "      <td>B-ketoacyl-ACP synthase (3-oxo-hexanoyl-ACP)</td>\n",
       "      <td>h_m + malACP_m + s_2771 --&gt; 3ohexACP_m + ACP_m...</td>\n",
       "      <td>YER061C and YKL192C</td>\n",
       "      <td>Lipst1_1_3273 or Lipst1_1_2955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>23CAPPD</td>\n",
       "      <td>ADP-ribose 2,3-cyclic phosphodiesterase</td>\n",
       "      <td>23camp_c + h2o_c --&gt; amp2p_c + h_c</td>\n",
       "      <td>YGR247W</td>\n",
       "      <td>Lipst1_1_72954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>r_2752</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3288 + s_3304 &lt;=&gt; 1agpc160_mm + s_3249</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>r_2627</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3270 + s_3307 &lt;=&gt; 1agpc181_mm + s_3252</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>r_2707</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3282 + s_3307 &lt;=&gt; 1agpc181_mm + s_3312</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>r_3211</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-18:1, 2-16:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3611 --&gt; 2.0 h_erm + pi_erm + s_3587</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>r_2808</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3295 + s_3304 &lt;=&gt; 1agpc160_mm + s_3319</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>r_4155</td>\n",
       "      <td>Glutamyl-tRNA(Gln) amidotransferase subunit B,...</td>\n",
       "      <td>atp_m + gln__L_m + glutrna_gln_m + h2o_m --&gt; a...</td>\n",
       "      <td>YBL080C</td>\n",
       "      <td>Lipst1_1_110148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>r_2596</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3258 --&gt; h_mm + hdca_mm + s_3278</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>r_2604</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3238 --&gt; h_mm + hdca_mm + s_3285</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>r_3818</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:1, 2-18:1) transpo...</td>\n",
       "      <td>s_3114 &lt;=&gt; s_3157</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>r_2618</td>\n",
       "      <td>CL (1-16:1, 2-18:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3266 --&gt; h_mm + ocdca_mm + s_3294</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>r_0438</td>\n",
       "      <td>ferrocytochrome-c:oxygen oxidoreductase</td>\n",
       "      <td>focytc_m + 1.266 h_m + 0.25 o2_m --&gt; ficytc_m ...</td>\n",
       "      <td>(Q0045 and Q0250 and Q0275 and YDL067C and YEL...</td>\n",
       "      <td>Lipst1_1_3895 or Lipst1_1_65274 or Lipst1_1_65...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>r_2785</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3292 + s_3305 &lt;=&gt; 1agpc161_mm + s_3261</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>r_4217</td>\n",
       "      <td>Fe(II):oxygen oxidoreductase; Fe2+:oxygen oxid...</td>\n",
       "      <td>4.0 fe2_m + 4.0 h_m + o2_m &lt;=&gt; 4.0 fe3_m + 2.0...</td>\n",
       "      <td>YDL120W</td>\n",
       "      <td>Lipst1_1_69822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>r_2811</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3295 + s_3307 &lt;=&gt; 1agpc181_mm + s_3319</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>r_2610</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3250 --&gt; h_mm + hdca_mm + s_3289</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>r_2620</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3270 + s_3296 &lt;=&gt; 1agpc160_mm + s_3240</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>r_4181</td>\n",
       "      <td>Manganese-transporting ATPase 1 (EC 3.6.3.-)</td>\n",
       "      <td>atp_c + h2o_c + mn2_c &lt;=&gt; adp_c + h_c + mn2_er...</td>\n",
       "      <td>YEL031W</td>\n",
       "      <td>Lipst1_1_75754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>r_2634</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3272 + s_3306 &lt;=&gt; 1agpc180_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>r_3284</td>\n",
       "      <td>TAG lipase (1-16:1, 2-18:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3050 --&gt; h_lp + hdca_lp + s_3049</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>r_2773</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3291 &lt;=&gt; 1agpc161_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>PSP_L</td>\n",
       "      <td>phosphoserine phosphatase (L-serine)</td>\n",
       "      <td>h2o_c + pser__L_c --&gt; pi_c + ser__L_c</td>\n",
       "      <td>YGR208W</td>\n",
       "      <td>Lipst1_1_1740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>r_2797</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3294 &lt;=&gt; 1agpc161_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>4MBZALDH</td>\n",
       "      <td>Putative aryl-alcohol dehydrogenase AAD3 (EC 1...</td>\n",
       "      <td>4mbzalc_c + nad_c &lt;=&gt; 4mbzald_c + h_c + nadh_c</td>\n",
       "      <td>YCR107W or YDL243C or YJR155W or YNL331C or YO...</td>\n",
       "      <td>Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>r_4313</td>\n",
       "      <td>Alpha-1,2-mannosyltransferase ALG9 (EC 2.4.1.2...</td>\n",
       "      <td>dolmanp_er + s_3993 &lt;=&gt; dolp_er + s_3994</td>\n",
       "      <td>YNL219C</td>\n",
       "      <td>Lipst1_1_91236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>r_2841</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-16:1, 2-18:1)...</td>\n",
       "      <td>atp_gm + h_gm + s_3370 --&gt; adp_gm + s_3371</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>r_2637</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3273 &lt;=&gt; 1agpc161_mm + s_3242</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>r_2663</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3276 + s_3302 &lt;=&gt; 1agpc181_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>r_2597</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-16:1, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3259 --&gt; h_mm + hdca_mm + s_3279</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>r_2599</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-18:1, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3261 --&gt; h_mm + hdca_mm + s_3281</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>r_2709</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3283 &lt;=&gt; 1agpc161_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>r_2657</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3275 + s_3305 &lt;=&gt; 1agpc161_mm + s_3256</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>r_2329</td>\n",
       "      <td>acyl dhap reductase (16:1), lipid particle</td>\n",
       "      <td>h_lp + nadph_lp + s_2949 --&gt; nadp_lp + s_2945</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>r_3312</td>\n",
       "      <td>lysoPE (1-16:0) oleoyl-CoA acyltransferase, li...</td>\n",
       "      <td>1agpe160_lp + ocdce9coa_lp --&gt; coa_lp + s_3686</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>r_3269</td>\n",
       "      <td>TAG lipase (1-16:1, 2-16:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + tag1619Z1619Z1619Z_lp --&gt; 12dgr161_lp...</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>r_2807</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3295 + s_3302 &lt;=&gt; 1agpc181_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>r_3276</td>\n",
       "      <td>TAG lipase (1-18:1, 2-16:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3056 --&gt; h_lp + hdca_lp + s_3055</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>r_2706</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3282 + s_3306 &lt;=&gt; 1agpc180_mm + s_3312</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>r_2687</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3280 + s_3302 &lt;=&gt; 1agpc181_mm + s_3266</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>r_1202</td>\n",
       "      <td>L-isoleucine transport</td>\n",
       "      <td>ile__L_v &lt;=&gt; ile__L_c</td>\n",
       "      <td>YCL038C</td>\n",
       "      <td>Lipst1_1_106989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>r_3218</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-18:0, 2-16:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3418 --&gt; 2.0 h_vm + pi_vm + s_3394</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>r_0073</td>\n",
       "      <td>4PP-IP5 depyrophosphorylation to IP6</td>\n",
       "      <td>h2o_c + s_0279 --&gt; h_c + minohp_c + pi_c</td>\n",
       "      <td>YLR410W or YOR163W</td>\n",
       "      <td>Lipst1_1_57858 or Lipst1_1_105766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>r_3896</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:1, 2-16:1) transpo...</td>\n",
       "      <td>s_3112 &lt;=&gt; s_3484</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>r_0523</td>\n",
       "      <td>GPI-anchor assembly, step 9</td>\n",
       "      <td>pe_hs_er + s_0338 &lt;=&gt; dag_hs_er + s_0332</td>\n",
       "      <td>YDR302W or YLL031C</td>\n",
       "      <td>Lipst1_1_141087 or Lipst1_1_38480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>r_2652</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3275 + s_3296 &lt;=&gt; 1agpc160_mm + s_3244</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>r_3208</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-16:0, 2-16:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3608 --&gt; 2.0 h_erm + pi_erm + s_3584</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>r_4723</td>\n",
       "      <td>polyphosphate phosphohydrolase, v</td>\n",
       "      <td>h2o_v + s_4290 &lt;=&gt; h_v + 2.0 pi_v</td>\n",
       "      <td>YHR201C or YDR452W</td>\n",
       "      <td>Lipst1_1_72748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>r_2648</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3274 + s_3304 &lt;=&gt; 1agpc160_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>r_2639</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3273 + s_3302 &lt;=&gt; 1agpc181_mm + s_3242</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>r_4167</td>\n",
       "      <td>Vacuolar cation-chloride cotransporter 1 (Vacu...</td>\n",
       "      <td>cl_v + k_v + na1_v &lt;=&gt; cl_c + k_c + na1_c</td>\n",
       "      <td>YBR235W</td>\n",
       "      <td>Lipst1_1_118138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>r_2744</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3287 + s_3304 &lt;=&gt; 1agpc160_mm + s_3315</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>r_2654</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3275 + s_3300 &lt;=&gt; 1agpc180_mm + s_3244</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>r_4156</td>\n",
       "      <td>Methionine aminopeptidase 2 (MAP 2) (MetAP 2) ...</td>\n",
       "      <td>h2o_v + met_L_ala__L_v &lt;=&gt; ala__L_v + met__L_v</td>\n",
       "      <td>YBL091C or YLR244C</td>\n",
       "      <td>Lipst1_1_338327 or Lipst1_1_904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>r_2641</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3273 + s_3305 &lt;=&gt; 1agpc161_mm + s_3254</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>r_2616</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3262 --&gt; h_mm + hdca_mm + s_3293</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>r_2843</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-18:1, 2-18:1)...</td>\n",
       "      <td>atp_gm + h_gm + s_3374 --&gt; adp_gm + s_3375</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>r_3271</td>\n",
       "      <td>TAG lipase (1-16:1, 2-16:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + tag1619Z1619Z1819Z_lp --&gt; 12dgr161_lp...</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>r_3210</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-18:0, 2-16:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3610 --&gt; 2.0 h_erm + pi_erm + s_3586</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>r_2688</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3280 + s_3304 &lt;=&gt; 1agpc160_mm + s_3310</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>r_3213</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-16:1, 2-18:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3613 --&gt; 2.0 h_erm + pi_erm + s_3589</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>r_2685</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3280 &lt;=&gt; 1agpc161_mm + s_3266</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>r_2677</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3279 &lt;=&gt; 1agpc161_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>r_2684</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3280 + s_3296 &lt;=&gt; 1agpc160_mm + s_3266</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>r_2759</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3289 + s_3302 &lt;=&gt; 1agpc181_mm + s_3251</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>r_2617</td>\n",
       "      <td>CL (1-16:1, 2-18:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3264 --&gt; h_mm + hdca_mm + s_3294</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>r_4163</td>\n",
       "      <td>Probable vacuolar amino acid transporter YPQ3 ...</td>\n",
       "      <td>lys__L_v &lt;=&gt; lys__L_c</td>\n",
       "      <td>YBR147W or YDR352W</td>\n",
       "      <td>Lipst1_1_211414 or Lipst1_1_62919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>r_2593</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-18:1, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3249 --&gt; h_mm + ocdca_mm + s_3274</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>r_2670</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3278 + s_3300 &lt;=&gt; 1agpc180_mm + s_3264</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>r_2636</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3273 + s_3296 &lt;=&gt; 1agpc160_mm + s_3242</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>GLYOX3</td>\n",
       "      <td>(R)-lactate hydro-lyase</td>\n",
       "      <td>h_c + lac__D_c &lt;=&gt; h2o_c + mthgxl_c</td>\n",
       "      <td>YDR533C or YMR322C or YOR391C or YPL280W</td>\n",
       "      <td>Lipst1_1_3208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>r_2669</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3278 &lt;=&gt; 1agpc161_mm + s_3264</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>r_3303</td>\n",
       "      <td>DAG lipase (1-18:1, 2-18:1), lipid particle</td>\n",
       "      <td>12dgr181_lp + h2o_lp --&gt; h_lp + magole_hs_lp +...</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>r_2716</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3284 + s_3296 &lt;=&gt; 1agpc160_mm + s_3235</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>r_2624</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3270 + s_3304 &lt;=&gt; 1agpc160_mm + s_3252</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>r_2765</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3290 &lt;=&gt; 1agpc161_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>r_2742</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3287 + s_3300 &lt;=&gt; 1agpc180_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>r_2611</td>\n",
       "      <td>CL (1-18:1, 2-16:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3252 --&gt; h_mm + hdca_mm + s_3290</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>r_2607</td>\n",
       "      <td>CL (1-16:1, 2-16:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3244 --&gt; h_mm + hdca_mm + s_3287</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>r_2676</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3279 + s_3296 &lt;=&gt; 1agpc160_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>r_4327</td>\n",
       "      <td>Magnesium transporter MRS2, mitochondrial (RNA...</td>\n",
       "      <td>mg2_c &lt;=&gt; mg2_m</td>\n",
       "      <td>YOR334W or YPL060W</td>\n",
       "      <td>Lipst1_1_1637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>r_2783</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3292 + s_3302 &lt;=&gt; 1agpc181_mm + s_3259</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>r_2649</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3274 + s_3305 &lt;=&gt; 1agpc161_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>r_2725</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3285 &lt;=&gt; 1agpc161_mm + s_3239</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>r_0357</td>\n",
       "      <td>diphosphoinositol-1,3,4,6-tetrakisphosphate di...</td>\n",
       "      <td>3.0 h2o_c + s_0309 --&gt; 3.0 h_c + inospp1_c + 3...</td>\n",
       "      <td>YOR163W</td>\n",
       "      <td>Lipst1_1_105766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>r_2793</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3293 + s_3305 &lt;=&gt; 1agpc161_mm + s_3318</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>r_2704</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3282 + s_3304 &lt;=&gt; 1agpc160_mm + s_3312</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>r_2718</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3284 + s_3300 &lt;=&gt; 1agpc180_mm + s_3235</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>r_2858</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-18:0, 2-18:1)...</td>\n",
       "      <td>atp_vm + h_vm + s_3354 --&gt; adp_vm + s_3398</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>r_2777</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3291 + s_3305 &lt;=&gt; 1agpc161_mm + s_3317</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>r_2462</td>\n",
       "      <td>lysoPI acyltransferase (1-18:0, 2-16:1), ER me...</td>\n",
       "      <td>s_3126 + stcoa_erm &lt;=&gt; coa_erm + s_3120</td>\n",
       "      <td>YBR042C</td>\n",
       "      <td>Lipst1_1_68496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>r_4178</td>\n",
       "      <td>Putative aryl-alcohol dehydrogenase AAD3 (EC 1...</td>\n",
       "      <td>nad_c + s_3795 &lt;=&gt; h_c + nadh_c + s_3796</td>\n",
       "      <td>YCR107W or YDL243C or YJR155W or YNL331C or YO...</td>\n",
       "      <td>Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>r_4239</td>\n",
       "      <td>L-arginyl-tRNA(Arg):protein arginyltransferase</td>\n",
       "      <td>argtrna_c + s_3884 &lt;=&gt; h_c + s_3885 + trnaarg_c</td>\n",
       "      <td>YGL017W</td>\n",
       "      <td>Lipst1_1_67104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>r_2801</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3294 + s_3305 &lt;=&gt; 1agpc161_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>r_2325</td>\n",
       "      <td>acyl dhap reductase (16:1), ER membrane</td>\n",
       "      <td>h_erm + nadph_erm + s_2941 --&gt; nadp_erm + s_2936</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>r_3302</td>\n",
       "      <td>DAG lipase (1-18:0, 2-18:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3053 --&gt; h_lp + mag180_lp + ocdcea_lp</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>r_4388</td>\n",
       "      <td>cyclic phosphodiesterase</td>\n",
       "      <td>23ccmp_c + h2o_c --&gt; h_c + s_4100</td>\n",
       "      <td>YGR247W</td>\n",
       "      <td>Lipst1_1_72954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>THMPPt2m</td>\n",
       "      <td>thiamine diphosphate transport</td>\n",
       "      <td>thmpp_c --&gt; thmpp_m</td>\n",
       "      <td>YGR096W</td>\n",
       "      <td>Lipst1_1_5945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>r_4158</td>\n",
       "      <td>NADPH2:quinone oxidoreductase</td>\n",
       "      <td>h_c + nadph_c + 2.0 s_3762 &lt;=&gt; nadp_c + 2.0 s_...</td>\n",
       "      <td>YBR046C</td>\n",
       "      <td>Lipst1_1_6135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>r_0532</td>\n",
       "      <td>hexaprenyldihydroxybenzoate methyltransferase</td>\n",
       "      <td>2hpmhmbq_m + amet_m + h_m --&gt; ahcys_m + s_1535</td>\n",
       "      <td>YDR204W and YGR255C and YLR201C and YML110C an...</td>\n",
       "      <td>Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>r_2625</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3270 + s_3305 &lt;=&gt; 1agpc161_mm + s_3252</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>r_2664</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3276 + s_3304 &lt;=&gt; 1agpc160_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>r_2589</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-16:1, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3239 --&gt; h_mm + hdca_mm + s_3276</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>r_2776</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3291 + s_3304 &lt;=&gt; 1agpc160_mm + s_3317</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>r_2799</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3294 + s_3302 &lt;=&gt; 1agpc181_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>SERAT</td>\n",
       "      <td>serine O-acetyltransferase</td>\n",
       "      <td>accoa_c + ser__L_c --&gt; acser_c + coa_c</td>\n",
       "      <td>YHR013C and YDL040C and YOR253W</td>\n",
       "      <td>Lipst1_1_4043 or Lipst1_1_164418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>r_2804</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3295 + s_3296 &lt;=&gt; 1agpc160_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>r_2756</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3289 + s_3296 &lt;=&gt; 1agpc160_mm + s_3251</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>r_2761</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3289 + s_3305 &lt;=&gt; 1agpc161_mm + s_3316</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>r_4722</td>\n",
       "      <td>polyphosphate polymerase, c</td>\n",
       "      <td>2.0 atp_c + h2o_c --&gt; h_c + s_4290</td>\n",
       "      <td>YJL012C</td>\n",
       "      <td>Lipst1_1_106087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>r_2601</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-16:1, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3263 --&gt; h_mm + hdca_mm + s_3283</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>r_2838</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-18:0, 2-16:1)...</td>\n",
       "      <td>atp_gm + h_gm + s_3364 --&gt; adp_gm + s_3365</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>r_0082</td>\n",
       "      <td>5-diphosphoinositol-1,2,3,4,6-pentakisphosphat...</td>\n",
       "      <td>h2o_c + ppmi12346p_c --&gt; h_c + minohp_c + pi_c</td>\n",
       "      <td>YOR163W</td>\n",
       "      <td>Lipst1_1_105766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>r_2693</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3281 &lt;=&gt; 1agpc161_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>r_3819</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:0, 2-18:1) transpo...</td>\n",
       "      <td>s_3115 &lt;=&gt; s_3159</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>r_2330</td>\n",
       "      <td>acyl dhap reductase (18:0), lipid particle</td>\n",
       "      <td>h_lp + nadph_lp + s_2950 --&gt; 1odecg3p_lp + nad...</td>\n",
       "      <td>YIL124W</td>\n",
       "      <td>Lipst1_1_71724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>r_3820</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:1, 2-18:1) transpo...</td>\n",
       "      <td>s_3116 &lt;=&gt; s_3161</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>r_2638</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3273 + s_3300 &lt;=&gt; 1agpc180_mm + s_3242</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>r_0521</td>\n",
       "      <td>GPI-anchor assembly, step 7</td>\n",
       "      <td>dolmanp_er + s_0337 &lt;=&gt; dolp_er + s_0339</td>\n",
       "      <td>YBR004C</td>\n",
       "      <td>Lipst1_1_74168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>r_2708</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3283 + s_3296 &lt;=&gt; 1agpc160_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>r_2714</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3283 + s_3306 &lt;=&gt; 1agpc180_mm + s_3313</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>2HP6MPMOm</td>\n",
       "      <td>2-hexaprenyl-6-methoxyphenol monooxygenase</td>\n",
       "      <td>2hp6mp_m + o2_m --&gt; 2hp6mbq_m + h2o_m</td>\n",
       "      <td>YDR204W and YGR255C and YLR201C and YML110C an...</td>\n",
       "      <td>Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>r_3277</td>\n",
       "      <td>TAG lipase (1-18:1, 2-16:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + tag1819Z1619Z1619Z_lp --&gt; h_lp + hdce...</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>r_2642</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3273 + s_3306 &lt;=&gt; 1agpc180_mm + s_3254</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>r_2721</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3284 + s_3305 &lt;=&gt; 1agpc161_mm + s_3237</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>r_3219</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-18:1, 2-16:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3419 --&gt; 2.0 h_vm + pi_vm + s_3395</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>r_2681</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3279 + s_3305 &lt;=&gt; 1agpc161_mm + s_3309</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>r_3275</td>\n",
       "      <td>TAG lipase (1-18:0, 2-16:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3079 --&gt; h_lp + ocdcea_lp + s_3051</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>r_2722</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3284 + s_3306 &lt;=&gt; 1agpc180_mm + s_3237</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>PAPtm</td>\n",
       "      <td>adenosine 3',5'-bisphosphate transport</td>\n",
       "      <td>pap_c &lt;=&gt; pap_m</td>\n",
       "      <td>YPR011C</td>\n",
       "      <td>Lipst1_1_180684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>r_2609</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3248 --&gt; h_mm + ocdca_mm + s_3288</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>r_4275</td>\n",
       "      <td>Fe(II):NADP+ oxidoreductase</td>\n",
       "      <td>2.0 fe3_e + nadph_c &lt;=&gt; 2.0 fe2_c + h_c + nadp_c</td>\n",
       "      <td>YLR047C or YKL220C or YLR214W or YNR060W or YO...</td>\n",
       "      <td>Lipst1_1_68023 or Lipst1_1_64448 or Lipst1_1_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>r_4589</td>\n",
       "      <td>Cu2(+) transport</td>\n",
       "      <td>cu2_e + h_e &lt;=&gt; cu2_c + h_c</td>\n",
       "      <td>YOL122C</td>\n",
       "      <td>Lipst1_1_67749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>r_3301</td>\n",
       "      <td>DAG lipase (1-16:1, 2-18:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3049 --&gt; h_lp + ocdcea_lp + s_3672</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>r_2644</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3274 + s_3296 &lt;=&gt; 1agpc160_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>r_2629</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3272 &lt;=&gt; 1agpc161_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>3MBZALDH</td>\n",
       "      <td>Putative aryl-alcohol dehydrogenase AAD3 (EC 1...</td>\n",
       "      <td>3mbzalc_c + nad_c &lt;=&gt; 3mbzald_c + h_c + nadh_c</td>\n",
       "      <td>YCR107W or YDL243C or YJR155W or YNL331C or YO...</td>\n",
       "      <td>Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>r_3289</td>\n",
       "      <td>TAG lipase (1-18:0, 2-18:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3064 --&gt; h_lp + hdcea_lp + s_3053</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>r_2767</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3290 + s_3302 &lt;=&gt; 1agpc181_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>r_3273</td>\n",
       "      <td>TAG lipase (1-18:0, 2-16:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3063 --&gt; h_lp + hdcea_lp + s_3051</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>r_3813</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:0, 2-16:1) transpo...</td>\n",
       "      <td>s_3109 &lt;=&gt; s_3145</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>r_2803</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3294 + s_3307 &lt;=&gt; 1agpc181_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>r_2779</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3291 + s_3307 &lt;=&gt; 1agpc181_mm + s_3317</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>r_2628</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3272 + s_3296 &lt;=&gt; 1agpc160_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>r_0088</td>\n",
       "      <td>5PP-IP5 pyrophosphorylation to 4,5-PP2-IP4</td>\n",
       "      <td>2.0 h_c + pi_c + ppmi12346p_c &lt;=&gt; h2o_c + s_0263</td>\n",
       "      <td>YLR410W</td>\n",
       "      <td>Lipst1_1_57858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>r_3280</td>\n",
       "      <td>TAG lipase (1-16:0, 2-18:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3046 --&gt; h_lp + hdca_lp + s_3045</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>r_0340</td>\n",
       "      <td>dihydroceramidase</td>\n",
       "      <td>lgnc_er + sphgn_er &lt;=&gt; cer1_24_er + h2o_er</td>\n",
       "      <td>YPL087W</td>\n",
       "      <td>Lipst1_1_296629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>r_2739</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3286 + s_3307 &lt;=&gt; 1agpc181_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>r_2689</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3280 + s_3305 &lt;=&gt; 1agpc161_mm + s_3310</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>r_3315</td>\n",
       "      <td>lysoPE (1-18:1) oleoyl-CoA acyltransferase, li...</td>\n",
       "      <td>1agpe181_lp + ocdce9coa_lp --&gt; coa_lp + s_3692</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>r_0013</td>\n",
       "      <td>2,3-diketo-5-methylthio-1-phosphopentane degra...</td>\n",
       "      <td>dkmpp_c + 3.0 h2o_c --&gt; 4met2obut_c + for_c + ...</td>\n",
       "      <td>YEL038W and YMR009W</td>\n",
       "      <td>Lipst1_1_67605 or Lipst1_1_103456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>r_4263</td>\n",
       "      <td>Anion/proton exchange transporter GEF1 (CLC pr...</td>\n",
       "      <td>cl_g &lt;=&gt; cl_c</td>\n",
       "      <td>YJR040W</td>\n",
       "      <td>Lipst1_1_263735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>r_2735</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3286 + s_3302 &lt;=&gt; 1agpc181_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>r_3893</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:0, 2-16:1) transpo...</td>\n",
       "      <td>s_3109 &lt;=&gt; s_3478</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>3DH5HPBMTm</td>\n",
       "      <td>S-adenosyl-L-methionine:3-hexaprenyl-4,5-dihyd...</td>\n",
       "      <td>3dh5hpb_m + amet_m --&gt; 3hph5mb_m + ahcys_m + h_m</td>\n",
       "      <td>YDR204W and YGR255C and YLR201C and YML110C an...</td>\n",
       "      <td>Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>r_0137</td>\n",
       "      <td>acylation of GPI inositol at 2 position, GPI-a...</td>\n",
       "      <td>gpail_hs_er + pmtcoa_er &lt;=&gt; coa_er + s_0330</td>\n",
       "      <td>YJL091C</td>\n",
       "      <td>Lipst1_1_70451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>r_2746</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3287 + s_3306 &lt;=&gt; 1agpc180_mm + s_3315</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>r_3274</td>\n",
       "      <td>TAG lipase (1-18:0, 2-16:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3071 --&gt; h_lp + ocdca_lp + s_3051</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>r_2839</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-18:1, 2-16:1)...</td>\n",
       "      <td>atp_gm + h_gm + s_3366 --&gt; adp_gm + s_3367</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>r_2768</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3290 + s_3304 &lt;=&gt; 1agpc160_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>r_3272</td>\n",
       "      <td>TAG lipase (1-18:0, 2-16:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3052 --&gt; h_lp + hdca_lp + s_3051</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>CHOLtu</td>\n",
       "      <td>Protein PNS1 (pH nine-sensitive protein 1)</td>\n",
       "      <td>chol_c &lt;=&gt; chol_e</td>\n",
       "      <td>YOR161C</td>\n",
       "      <td>Lipst1_1_159073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>r_2770</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3290 + s_3306 &lt;=&gt; 1agpc180_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>r_2613</td>\n",
       "      <td>CL (1-18:1, 2-16:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3256 --&gt; h_mm + hdca_mm + s_3291</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>r_2603</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3236 --&gt; h_mm + ocdca_mm + s_3284</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>r_2651</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3274 + s_3307 &lt;=&gt; 1agpc181_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>r_1086</td>\n",
       "      <td>V-ATPase, vacuole</td>\n",
       "      <td>atp_c + h2o_c + h_c --&gt; adp_c + 2.0 h_v + pi_c</td>\n",
       "      <td>(YBR127C and YDL185W and YEL027W and YEL051W a...</td>\n",
       "      <td>Lipst1_1_2781 or Lipst1_1_102936 or Lipst1_1_6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>r_2697</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3281 + s_3305 &lt;=&gt; 1agpc161_mm + s_3311</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>r_2791</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3293 + s_3302 &lt;=&gt; 1agpc181_mm + s_3263</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>r_4700</td>\n",
       "      <td>hydrogen sulfide transport</td>\n",
       "      <td>h2s_c --&gt; h2s_e</td>\n",
       "      <td>YIL166C</td>\n",
       "      <td>Lipst1_1_6611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>r_3212</td>\n",
       "      <td>PI 4,5-P2 phosphatase (1-16:0, 2-18:1), ER mem...</td>\n",
       "      <td>h2o_erm + s_3612 --&gt; 2.0 h_erm + pi_erm + s_3588</td>\n",
       "      <td>YOL065C</td>\n",
       "      <td>Lipst1_1_110448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>r_2650</td>\n",
       "      <td>MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3274 + s_3306 &lt;=&gt; 1agpc180_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>r_2769</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3290 + s_3305 &lt;=&gt; 1agpc161_mm + s_3255</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>r_3288</td>\n",
       "      <td>TAG lipase (1-18:0, 2-18:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3054 --&gt; h_lp + hdca_lp + s_3053</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>r_3899</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:0, 2-18:1) transpo...</td>\n",
       "      <td>s_3115 &lt;=&gt; s_3488</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>r_2853</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-16:1, 2-16:1)...</td>\n",
       "      <td>atp_vm + h_vm + pail161_vm --&gt; adp_vm + pail3p...</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>r_2802</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3294 + s_3306 &lt;=&gt; 1agpc180_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>r_3895</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:0, 2-16:1) transpo...</td>\n",
       "      <td>s_3111 &lt;=&gt; s_3482</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>r_2788</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3293 + s_3296 &lt;=&gt; 1agpc160_mm + s_3263</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>r_4754</td>\n",
       "      <td>flavin prenyltransferase</td>\n",
       "      <td>fmnh2_m + s_4295 &lt;=&gt; pi_m + s_4312</td>\n",
       "      <td>YDR538W</td>\n",
       "      <td>Lipst1_1_76260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>r_2626</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3270 + s_3306 &lt;=&gt; 1agpc180_mm + s_3252</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>r_4587</td>\n",
       "      <td>Ca(2+) transport</td>\n",
       "      <td>ca2_e + h_e &lt;=&gt; ca2_c + h_c</td>\n",
       "      <td>YOL122C</td>\n",
       "      <td>Lipst1_1_67749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>r_3816</td>\n",
       "      <td>phosphatidyl-L-serine (1-18:1, 2-16:1) transpo...</td>\n",
       "      <td>s_3112 &lt;=&gt; s_3153</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>r_4314</td>\n",
       "      <td>Alpha-1,2-mannosyltransferase ALG9 (EC 2.4.1.2...</td>\n",
       "      <td>dolmanp_er + s_3995 &lt;=&gt; dolp_er + s_3996</td>\n",
       "      <td>YNL219C</td>\n",
       "      <td>Lipst1_1_91236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>r_2631</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3272 + s_3302 &lt;=&gt; 1agpc181_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>r_2619</td>\n",
       "      <td>CL (1-16:1, 2-18:1, 3-16:0, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3268 --&gt; h_mm + hdca_mm + s_3295</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>r_0341</td>\n",
       "      <td>dihydroceramidase</td>\n",
       "      <td>hexc_er + sphgn_er &lt;=&gt; cer1_26_er + h2o_er</td>\n",
       "      <td>YPL087W</td>\n",
       "      <td>Lipst1_1_296629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>r_2695</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3281 + s_3302 &lt;=&gt; 1agpc181_mm + s_3267</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>r_2640</td>\n",
       "      <td>MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3273 + s_3304 &lt;=&gt; 1agpc160_mm + s_3254</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>r_4705</td>\n",
       "      <td>taurocholate transport</td>\n",
       "      <td>tchola_e --&gt; tchola_c_</td>\n",
       "      <td>YIL166C</td>\n",
       "      <td>Lipst1_1_6611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>r_1919</td>\n",
       "      <td>lysine transport</td>\n",
       "      <td>lys__L_c --&gt; lys__L_m</td>\n",
       "      <td>YBR147W</td>\n",
       "      <td>Lipst1_1_211414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>r_2595</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-16:1, 4-18:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3251 --&gt; h_mm + ocdca_mm + s_3276</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>r_3898</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:1, 2-18:1) transpo...</td>\n",
       "      <td>s_3114 &lt;=&gt; s_3487</td>\n",
       "      <td>YIL048W</td>\n",
       "      <td>Lipst1_1_49943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>r_2762</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3289 + s_3306 &lt;=&gt; 1agpc180_mm + s_3316</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>r_4312</td>\n",
       "      <td>S-adenosyl-L-methionine:carnosine N-methyltran...</td>\n",
       "      <td>amet_c + carn_c &lt;=&gt; ahcys_c + ans_c + h_c</td>\n",
       "      <td>YNL092W</td>\n",
       "      <td>Lipst1_1_71580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>r_2728</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3285 + s_3304 &lt;=&gt; 1agpc160_mm + s_3314</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>r_2630</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3272 + s_3300 &lt;=&gt; 1agpc180_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>r_1220</td>\n",
       "      <td>L-tyrosine transport</td>\n",
       "      <td>tyr__L_v &lt;=&gt; tyr__L_c</td>\n",
       "      <td>YCL038C</td>\n",
       "      <td>Lipst1_1_106989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>r_2795</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3293 + s_3307 &lt;=&gt; 1agpc181_mm + s_3318</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>r_4372</td>\n",
       "      <td>cyclic phosphodiesterase</td>\n",
       "      <td>23cump_c + h2o_c --&gt; h_c + s_4081</td>\n",
       "      <td>YGR247W</td>\n",
       "      <td>Lipst1_1_72954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>r_2635</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3272 + s_3307 &lt;=&gt; 1agpc181_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>r_2730</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3285 + s_3306 &lt;=&gt; 1agpc180_mm + s_3314</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>r_2734</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3286 + s_3300 &lt;=&gt; 1agpc180_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>r_2686</td>\n",
       "      <td>MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3280 + s_3300 &lt;=&gt; 1agpc180_mm + s_3266</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>r_4234</td>\n",
       "      <td>Phosphatidylinositol N-acetylglucosaminyltrans...</td>\n",
       "      <td>pail_cho_er + uacgam_er &lt;=&gt; h_er + s_3879 + ud...</td>\n",
       "      <td>YDR437W and YGR216C and YNL038W and YPL076W an...</td>\n",
       "      <td>Lipst1_1_56084 or Lipst1_1_6286 or Lipst1_1_13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>r_0343</td>\n",
       "      <td>dihydroceramidase</td>\n",
       "      <td>cer2_26_er + h2o_er --&gt; hexc_er + psphings_er</td>\n",
       "      <td>YPL087W</td>\n",
       "      <td>Lipst1_1_296629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>r_2749</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3288 &lt;=&gt; 1agpc161_mm + s_3247</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>r_2737</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3286 + s_3305 &lt;=&gt; 1agpc161_mm + s_3243</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>Clt</td>\n",
       "      <td>chloride transport</td>\n",
       "      <td>cl_e &lt;=&gt; cl_c</td>\n",
       "      <td>YNL275W</td>\n",
       "      <td>Lipst1_1_170576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>r_4193</td>\n",
       "      <td>S-Adenosyl-L-methionine:histone-L-lysine N6-me...</td>\n",
       "      <td>amet_n + s_3819 &lt;=&gt; ahcys_n + h_n + s_3821</td>\n",
       "      <td>YHR119W or YDR440W or YJL168C</td>\n",
       "      <td>Lipst1_1_103455 or Lipst1_1_106915 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>r_2705</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3282 + s_3305 &lt;=&gt; 1agpc161_mm + s_3312</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>SERD_D</td>\n",
       "      <td>D-serine ammonia-lyase</td>\n",
       "      <td>ser__D_c --&gt; nh4_c + pyr_c</td>\n",
       "      <td>YGL196W</td>\n",
       "      <td>Lipst1_1_1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>r_2623</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3270 + s_3302 &lt;=&gt; 1agpc181_mm + s_3240</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>r_3270</td>\n",
       "      <td>TAG lipase (1-16:1, 2-16:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3069 --&gt; 12dgr161_lp + h_lp + ocdca_lp</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>r_3216</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-16:0, 2-16:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3416 --&gt; 2.0 h_vm + pi_vm + s_3392</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>r_4322</td>\n",
       "      <td>GPI mannosyltransferase 4 (EC 2.4.1.-) (GPI ma...</td>\n",
       "      <td>dolmanp_er + s_4002 &lt;=&gt; dolp_er + h_er + s_4003</td>\n",
       "      <td>YOR149C</td>\n",
       "      <td>Lipst1_1_62314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>r_2606</td>\n",
       "      <td>CL (1-16:1, 2-16:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3242 --&gt; h_mm + ocdca_mm + s_3286</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>r_2723</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3284 + s_3307 &lt;=&gt; 1agpc181_mm + s_3237</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>r_3299</td>\n",
       "      <td>DAG lipase (1-18:1, 2-16:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3055 --&gt; h_lp + hdcea_lp + magole_h...</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>r_2766</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3290 + s_3300 &lt;=&gt; 1agpc180_mm + s_3253</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>r_4175</td>\n",
       "      <td>Putative aryl-alcohol dehydrogenase AAD3 (EC 1...</td>\n",
       "      <td>nad_c + s_3789 &lt;=&gt; h_c + nadh_c + s_3790</td>\n",
       "      <td>YCR107W or YDL243C or YJR155W or YNL331C or YO...</td>\n",
       "      <td>Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>r_4210</td>\n",
       "      <td>urea hydro-lyase (cyanamide-forming)</td>\n",
       "      <td>urea_c &lt;=&gt; h2o_c + s_3851</td>\n",
       "      <td>YFL061W or YNL335W</td>\n",
       "      <td>Lipst1_1_50907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>r_4225</td>\n",
       "      <td>Broad-range acid phosphatase DET1 (EC 3.1.3.-)...</td>\n",
       "      <td>atp_c + h2o_c + s_3865 &lt;=&gt; adp_c + h_c + pi_c ...</td>\n",
       "      <td>YDR051C or YOL075C</td>\n",
       "      <td>Lipst1_1_92 or Lipst1_1_71475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>r_3223</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-18:1, 2-18:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3423 --&gt; 2.0 h_vm + pi_vm + s_3399</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>r_2857</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-16:1, 2-18:1)...</td>\n",
       "      <td>atp_vm + h_vm + s_3352 --&gt; adp_vm + s_3397</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>r_4337</td>\n",
       "      <td>Iron-sulfur clusters transporter ATM1, mitocho...</td>\n",
       "      <td>atp_m + h2o_m + s_4022 &lt;=&gt; adp_m + h_m + pi_m ...</td>\n",
       "      <td>YMR301C</td>\n",
       "      <td>Lipst1_1_318678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>r_3265</td>\n",
       "      <td>TAG lipase (1-16:0, 2-16:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3059 --&gt; h_lp + hdcea_lp + s_3043</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>CD2abc1</td>\n",
       "      <td>P-type cation-transporting ATPase (EC 3.6.3.3)...</td>\n",
       "      <td>atp_c + cd2_c + h2o_c --&gt; adp_c + cd2_e + h_c ...</td>\n",
       "      <td>YBR295W</td>\n",
       "      <td>Lipst1_1_1388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>r_2724</td>\n",
       "      <td>MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3285 + s_3296 &lt;=&gt; 1agpc160_mm + s_3239</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>r_2856</td>\n",
       "      <td>phosphatidylinositol 3-kinase (1-16:0, 2-18:1)...</td>\n",
       "      <td>atp_vm + h_vm + s_3350 --&gt; adp_vm + s_3396</td>\n",
       "      <td>(YBR097W and YLR240W) or (YBR097W and YLR240W ...</td>\n",
       "      <td>Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>r_2667</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3276 + s_3307 &lt;=&gt; 1agpc181_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>r_3282</td>\n",
       "      <td>TAG lipase (1-16:0, 2-18:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3068 --&gt; h_lp + ocdca_lp + s_3045</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>r_2658</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3275 + s_3306 &lt;=&gt; 1agpc180_mm + s_3256</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>r_3814</td>\n",
       "      <td>phosphatidyl-L-serine (1-16:1, 2-16:1) transpo...</td>\n",
       "      <td>s_3110 &lt;=&gt; s_3149</td>\n",
       "      <td>YIL048W or YMR162C</td>\n",
       "      <td>Lipst1_1_49943 or Lipst1_1_984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>r_2789</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3293 &lt;=&gt; 1agpc161_mm + s_3263</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>r_0331</td>\n",
       "      <td>deoxyhypusine synthase, cytosolic/mitochondrial</td>\n",
       "      <td>h2o_c + q6_m + spmd_c --&gt; 4abutn_c + s_1526 + ...</td>\n",
       "      <td>YHR068W</td>\n",
       "      <td>Lipst1_1_222261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>r_4724</td>\n",
       "      <td>polyphosphate phosphohydrolase, m</td>\n",
       "      <td>h2o_m + s_4291 &lt;=&gt; h_m + 2.0 pi_m</td>\n",
       "      <td>YHR201C or YDR452W</td>\n",
       "      <td>Lipst1_1_72748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>r_4412</td>\n",
       "      <td>cyclic phosphodiesterase</td>\n",
       "      <td>23cgmp_c + h2o_c --&gt; h_c + s_4066</td>\n",
       "      <td>YGR247W</td>\n",
       "      <td>Lipst1_1_72954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>r_2674</td>\n",
       "      <td>MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3278 + s_3306 &lt;=&gt; 1agpc180_mm + s_3308</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>r_2696</td>\n",
       "      <td>MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:0, 2-18...</td>\n",
       "      <td>s_3281 + s_3304 &lt;=&gt; 1agpc160_mm + s_3311</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>r_2842</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-18:0, 2-18:1)...</td>\n",
       "      <td>atp_gm + h_gm + s_3372 --&gt; adp_gm + s_3373</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>r_2758</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3289 + s_3300 &lt;=&gt; 1agpc180_mm + s_3251</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>r_2666</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:0, 2-18...</td>\n",
       "      <td>s_3276 + s_3306 &lt;=&gt; 1agpc180_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>r_3278</td>\n",
       "      <td>TAG lipase (1-18:1, 2-16:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3073 --&gt; h_lp + ocdca_lp + s_3055</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>FCYTOB5OXR_m</td>\n",
       "      <td>NADH-cytochrome b5 reductase 1 (EC 1.6.2.2) (M...</td>\n",
       "      <td>2.0 ficytb5_m + nadh_m &lt;=&gt; 2.0 focytb5_m + h_m...</td>\n",
       "      <td>YIL043C or YKL150W or YML125C</td>\n",
       "      <td>Lipst1_1_6803 or Lipst1_1_2352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>GGCLUT2</td>\n",
       "      <td>Glutathione-specific gamma-glutamylcyclotransf...</td>\n",
       "      <td>gthrd_c --&gt; 5oxpro_c + cgly_c</td>\n",
       "      <td>YER163C</td>\n",
       "      <td>Lipst1_1_47595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>r_2732</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3286 + s_3296 &lt;=&gt; 1agpc160_mm + clpn161_mm</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>r_2655</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3275 + s_3302 &lt;=&gt; 1agpc181_mm + s_3244</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>r_2660</td>\n",
       "      <td>MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3276 + s_3296 &lt;=&gt; 1agpc160_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>r_4203</td>\n",
       "      <td>Vacuolar aminopeptidase 1 (EC 3.4.11.22) (Amin...</td>\n",
       "      <td>L_alagly_v + h2o_v &lt;=&gt; ala__L_v + gly_v</td>\n",
       "      <td>YKL103C</td>\n",
       "      <td>Lipst1_1_106140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>r_2837</td>\n",
       "      <td>phosphatidylinositol 4-kinase (1-16:1, 2-16:1)...</td>\n",
       "      <td>atp_gm + h_gm + pail161_gm --&gt; adp_gm + pail4p...</td>\n",
       "      <td>YDR373W and YNL267W</td>\n",
       "      <td>Lipst1_1_2990 or Lipst1_1_218588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>r_2753</td>\n",
       "      <td>MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:1, 2-18...</td>\n",
       "      <td>s_3288 + s_3305 &lt;=&gt; 1agpc161_mm + s_3249</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>r_1085</td>\n",
       "      <td>V-ATPase, Golgi</td>\n",
       "      <td>atp_c + h2o_c + h_c --&gt; adp_c + 2.0 h_g + pi_c</td>\n",
       "      <td>(YPR170W-B and YBR127C and YDL185W and YOR332W...</td>\n",
       "      <td>Lipst1_1_2781 or Lipst1_1_102936 or Lipst1_1_6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>r_2463</td>\n",
       "      <td>lysoPI acyltransferase (1-18:0, 2-18:1), ER me...</td>\n",
       "      <td>s_3127 + stcoa_erm &lt;=&gt; coa_erm + s_3124</td>\n",
       "      <td>YBR042C</td>\n",
       "      <td>Lipst1_1_68496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>r_2798</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3294 + s_3300 &lt;=&gt; 1agpc180_mm + s_3265</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>r_2781</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3292 &lt;=&gt; 1agpc161_mm + s_3259</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>r_2587</td>\n",
       "      <td>CL (1-16:0, 2-16:1, 3-18:1, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3237 --&gt; h_mm + hdca_mm + s_3274</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>r_0522</td>\n",
       "      <td>GPI-anchor assembly, step 8</td>\n",
       "      <td>dolmanp_er + s_0339 &lt;=&gt; s_0338</td>\n",
       "      <td>YGL142C</td>\n",
       "      <td>Lipst1_1_103747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>r_3297</td>\n",
       "      <td>DAG lipase (1-16:1, 2-16:1), lipid particle</td>\n",
       "      <td>12dgr161_lp + h2o_lp --&gt; h_lp + hdcea_lp + s_3672</td>\n",
       "      <td>YCR068W or YMR313C</td>\n",
       "      <td>Lipst1_1_47671 or Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>r_2653</td>\n",
       "      <td>MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3275 &lt;=&gt; 1agpc161_mm + s_3244</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>r_2683</td>\n",
       "      <td>MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:1, 2-18...</td>\n",
       "      <td>s_3279 + s_3307 &lt;=&gt; 1agpc181_mm + s_3309</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>r_1208</td>\n",
       "      <td>L-leucine transport, vacuoluar</td>\n",
       "      <td>leu__L_v &lt;=&gt; leu__L_c</td>\n",
       "      <td>YCL038C</td>\n",
       "      <td>Lipst1_1_106989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>r_0092</td>\n",
       "      <td>6PP-IP5 depyrophosphorylation to IP6</td>\n",
       "      <td>h2o_c + s_0333 --&gt; h_c + minohp_c + pi_c</td>\n",
       "      <td>YLR410W or YOR163W</td>\n",
       "      <td>Lipst1_1_57858 or Lipst1_1_105766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>r_1684</td>\n",
       "      <td>choline transport</td>\n",
       "      <td>chol_ce &lt;=&gt; chol_c</td>\n",
       "      <td>YGL077C or YOR161C</td>\n",
       "      <td>Lipst1_1_159073 or Lipst1_1_71543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>r_3314</td>\n",
       "      <td>lysoPE (1-18:0) oleoyl-CoA acyltransferase, li...</td>\n",
       "      <td>1agpe180_lp + ocdce9coa_lp --&gt; coa_lp + s_3690</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>Lipst1_1_161356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>r_4725</td>\n",
       "      <td>polyphosphate phosphohydrolase, n</td>\n",
       "      <td>h2o_n + s_4292 &lt;=&gt; h_n + 2.0 pi_n</td>\n",
       "      <td>YHR201C or YDR452W</td>\n",
       "      <td>Lipst1_1_72748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>r_2780</td>\n",
       "      <td>MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:0, 2-16...</td>\n",
       "      <td>s_3292 + s_3296 &lt;=&gt; 1agpc160_mm + s_3259</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>r_3294</td>\n",
       "      <td>TAG lipase (1-18:1, 2-18:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3074 --&gt; 12dgr181_lp + h_lp + ocdca_lp</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>r_2598</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-18:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3260 --&gt; h_mm + hdca_mm + s_3280</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>r_2805</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:1, 2-16...</td>\n",
       "      <td>pc161_mm + s_3295 &lt;=&gt; 1agpc161_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>r_4403</td>\n",
       "      <td>Triphosphate phosphohydrolase</td>\n",
       "      <td>h2o_c + pppi_c --&gt; pi_c + ppi_c</td>\n",
       "      <td>YHR201C or YDR452W</td>\n",
       "      <td>Lipst1_1_72748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>r_3283</td>\n",
       "      <td>TAG lipase (1-16:0, 2-18:1, 3-18:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3076 --&gt; h_lp + ocdcea_lp + s_3045</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>r_2614</td>\n",
       "      <td>CL (1-16:0, 2-18:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3258 --&gt; h_mm + hdca_mm + s_3292</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>r_4247</td>\n",
       "      <td>ATP:L-threonyl,bicarbonate adenylyltransferase</td>\n",
       "      <td>atp_c + hco3_c + thr__L_c --&gt; h2o_c + ppi_c + ...</td>\n",
       "      <td>YGL169W</td>\n",
       "      <td>Lipst1_1_42443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>r_4339</td>\n",
       "      <td>UDP-N-acetylglucosamine transporter YEA4</td>\n",
       "      <td>uacgam_er &lt;=&gt; uacgam_g</td>\n",
       "      <td>YEL004W</td>\n",
       "      <td>Lipst1_1_99002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>r_4233</td>\n",
       "      <td>S-Adenosyl-L-methionine:protein-C-terminal-S-f...</td>\n",
       "      <td>amet_c + s_3876 &lt;=&gt; ahcys_c + s_3877</td>\n",
       "      <td>YDR410C</td>\n",
       "      <td>Lipst1_1_76042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>2HPMBQMTm</td>\n",
       "      <td>2-hexaprenyl-6-methoxy-1,4-benzoquinone methyl...</td>\n",
       "      <td>2hp6mbq_m + amet_m --&gt; 2hpmmbq_m + ahcys_m + h_m</td>\n",
       "      <td>YDR204W and YGR255C and YLR201C and YML110C an...</td>\n",
       "      <td>Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>r_3285</td>\n",
       "      <td>TAG lipase (1-16:1, 2-18:1, 3-16:1), lipid par...</td>\n",
       "      <td>h2o_lp + s_3062 --&gt; h_lp + hdcea_lp + s_3049</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>r_2806</td>\n",
       "      <td>MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:0, 2-16...</td>\n",
       "      <td>s_3295 + s_3300 &lt;=&gt; 1agpc180_mm + s_3269</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>r_4278</td>\n",
       "      <td>diphthine:ammonia ligase (AMP-forming)</td>\n",
       "      <td>atp_c + nh4_c + s_3940 --&gt; amp_c + h_c + ppi_c...</td>\n",
       "      <td>YLR143W</td>\n",
       "      <td>Lipst1_1_281312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>r_2775</td>\n",
       "      <td>MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3291 + s_3302 &lt;=&gt; 1agpc181_mm + s_3257</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>r_2743</td>\n",
       "      <td>MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:1, 2-16...</td>\n",
       "      <td>s_3287 + s_3302 &lt;=&gt; 1agpc181_mm + s_3245</td>\n",
       "      <td>YPR140W</td>\n",
       "      <td>Lipst1_1_68925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>r_0342</td>\n",
       "      <td>dihydroceramidase</td>\n",
       "      <td>cer2_24_er + h2o_er --&gt; lgnc_er + psphings_er</td>\n",
       "      <td>YPL087W</td>\n",
       "      <td>Lipst1_1_296629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>r_3268</td>\n",
       "      <td>TAG lipase (1-16:1, 2-16:1, 3-16:0), lipid par...</td>\n",
       "      <td>h2o_lp + tag1619Z1619Z160_lp --&gt; 12dgr161_lp +...</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>r_4195</td>\n",
       "      <td>Zinc transporter YKE4</td>\n",
       "      <td>zn2_m &lt;=&gt; zn2_v</td>\n",
       "      <td>YMR243C or YOR316C</td>\n",
       "      <td>Lipst1_1_3875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>r_2591</td>\n",
       "      <td>CL (1-18:0, 2-16:1, 3-16:1, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3247 --&gt; h_mm + ocdca_mm + s_3272</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>r_3221</td>\n",
       "      <td>PI 3,5-P2 phosphatase (1-16:1, 2-18:1), vacuol...</td>\n",
       "      <td>h2o_vm + s_3421 --&gt; 2.0 h_vm + pi_vm + s_3397</td>\n",
       "      <td>YFR019W and YNL325C and YFR021W and YNL054W an...</td>\n",
       "      <td>Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>r_3253</td>\n",
       "      <td>ergosteryl ester hydrolase (18:1), lipid particle</td>\n",
       "      <td>h2o_lp + s_3658 --&gt; ergst_lp + h_lp + ocdcea_lp</td>\n",
       "      <td>YKL140W or YLL012W</td>\n",
       "      <td>Lipst1_1_1895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>r_2605</td>\n",
       "      <td>CL (1-16:1, 2-16:1, 3-16:0, 4-16:1) phospholip...</td>\n",
       "      <td>h2o_mm + s_3240 --&gt; h_mm + hdca_mm + s_3286</td>\n",
       "      <td>YGR110W</td>\n",
       "      <td>Lipst1_1_102653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>r_3290</td>\n",
       "      <td>TAG lipase (1-18:0, 2-18:1, 3-18:0), lipid par...</td>\n",
       "      <td>h2o_lp + s_3072 --&gt; h_lp + ocdca_lp + s_3053</td>\n",
       "      <td>YBR204C or YKR089C or YMR313C or YOR081C</td>\n",
       "      <td>Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                0                                                  1  \\\n",
       "0          r_2594  CL (1-18:0, 2-16:1, 3-16:0, 4-18:1) phospholip...   \n",
       "1          r_2852  phosphatidylinositol 3-kinase (1-16:0, 2-16:1)...   \n",
       "2          r_2586  CL (1-16:0, 2-16:1, 3-18:0, 4-16:1) phospholip...   \n",
       "3          r_2719  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "4          r_2731  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "5          r_2632  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "6          r_3266  TAG lipase (1-16:0, 2-16:1, 3-18:0), lipid par...   \n",
       "7          r_2764  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "8          r_3293  TAG lipase (1-18:1, 2-18:1, 3-16:1), lipid par...   \n",
       "9          r_2328         acyl dhap reductase (16:0), lipid particle   \n",
       "10         r_4194                              Zinc transporter YKE4   \n",
       "11         r_4200  Glutathione S-transferase 1 (EC 2.5.1.18) (GST-I)   \n",
       "12         r_2327            acyl dhap reductase (18:1), ER membrane   \n",
       "13         r_3279  TAG lipase (1-18:1, 2-16:1, 3-18:1), lipid par...   \n",
       "14        r2073_1                                   Zn(2+) transport   \n",
       "15         r_3281  TAG lipase (1-16:0, 2-18:1, 3-16:1), lipid par...   \n",
       "16         r_2703  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:1, 2-16...   \n",
       "17         r_2675  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:1, 2-18...   \n",
       "18         r_3286  TAG lipase (1-16:1, 2-18:1, 3-18:0), lipid par...   \n",
       "19         r_2745  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "20         r_2712  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "21         r_2678  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "22         r_4162  Probable vacuolar amino acid transporter YPQ3 ...   \n",
       "23         r_4174                    Probable ATP-dependent permease   \n",
       "24         r_3252  ergosteryl ester hydrolase (16:1), lipid particle   \n",
       "25         r_2859  phosphatidylinositol 3-kinase (1-18:1, 2-18:1)...   \n",
       "26         r_2647  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "27         BZALDH                 Benzyl alcohol:NAD+ oxidoreductase   \n",
       "28         r_2324            acyl dhap reductase (16:0), ER membrane   \n",
       "29         r_3264  TAG lipase (1-16:0, 2-16:1, 3-16:0), lipid par...   \n",
       "30         r_2600  CL (1-16:0, 2-18:1, 3-16:0, 4-18:1) phospholip...   \n",
       "31         r_2750  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "32         r_2715  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "33         r_4219  D-aminoacyl-tRNA deacylase (DTD) (EC 3.1.1.96)...   \n",
       "34         r_2736  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "35         r_2608  CL (1-18:0, 2-16:1, 3-16:0, 4-16:1) phospholip...   \n",
       "36         r_3300        DAG lipase (1-16:0, 2-18:1), lipid particle   \n",
       "37         r_2699  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "38         r_3291  TAG lipase (1-18:0, 2-18:1, 3-18:1), lipid par...   \n",
       "39         r_2645  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "40         r_0044  3-hexaprenyl-4-hydroxy-5-methoxybenzoate decar...   \n",
       "41         r_2796  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "42         r_2656  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:0, 2-18...   \n",
       "43         r_0963                             quinone oxidoreductase   \n",
       "44         r_2590  CL (1-18:0, 2-16:1, 3-16:0, 4-16:1) phospholip...   \n",
       "45         r_2633  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "46         r_4333                     Polyphosphate phosphohydrolase   \n",
       "47         r_3215  PI 4,5-P2 phosphatase (1-18:1, 2-18:1), ER mem...   \n",
       "48         r_2698  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "49         r_3817  phosphatidyl-L-serine (1-16:0, 2-18:1) transpo...   \n",
       "50         r_4331  [1,4-(N-Acetyl-beta-D-glucosaminyl)]n glycanoh...   \n",
       "51          FMNRx                                      FMN reductase   \n",
       "52         r_2694  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "53         G6PI_1        alpha-D-Glucose 6-phosphate ketol-isomerase   \n",
       "54         r_2622  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:0, 2-16...   \n",
       "55         r_4242  Initiation-specific alpha-1,6-mannosyltransfer...   \n",
       "56         r_2720  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "57         r_2757  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "58         r_2643  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:1, 2-18...   \n",
       "59         r_4199  Glutathione S-transferase 1 (EC 2.5.1.18) (GST-I)   \n",
       "60         FMNRx2                                      FMN reductase   \n",
       "61          GTHDH        glutathione:dehydroascorbate oxidoreductase   \n",
       "62         r_4159  UDP-N-acetylglucosamine transferase subunit AL...   \n",
       "63         r_0993                          serine palmitotransferase   \n",
       "64        CYANSTm               thiosulfate:cyanide sulfurtranserase   \n",
       "65         r_4176  Putative aryl-alcohol dehydrogenase AAD3 (EC 1...   \n",
       "66         r_2772  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "67         r_2748  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "68         r_2646  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "69         r_1837               L-histidine transport, mitochondrial   \n",
       "70           NTP3                          nucleoside triphosphatase   \n",
       "71         r_2692  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "72         r_2727  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "73         r_2840  phosphatidylinositol 4-kinase (1-16:0, 2-18:1)...   \n",
       "74         r_2673  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:1, 2-18...   \n",
       "75         r_2855  phosphatidylinositol 3-kinase (1-18:1, 2-16:1)...   \n",
       "76         r_2741  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "77         r_0192                      alpha 1,2-mannosyltransferase   \n",
       "78         r_2701  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:1, 2-16...   \n",
       "79         r_2771  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "80         r_2662  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "81         r_2682  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "82         r_2691  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:1, 2-18...   \n",
       "83         r_2615  CL (1-16:0, 2-18:1, 3-18:0, 4-16:1) phospholip...   \n",
       "84         r_2751  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "85         r_2659  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:1, 2-18...   \n",
       "86         r_2790  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "87         r_3815  phosphatidyl-L-serine (1-18:0, 2-16:1) transpo...   \n",
       "88         r_2810  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "89         r_3220  PI 3,5-P2 phosphatase (1-16:0, 2-18:1), vacuol...   \n",
       "90         r_4250              Dolichyl-diphosphate phosphohydrolase   \n",
       "91           MNt2                                   Mn(2+) transport   \n",
       "92         r_2326            acyl dhap reductase (18:0), ER membrane   \n",
       "93         ARGt2m                                 arginine transport   \n",
       "94         r_3313  lysoPE (1-16:1) oleoyl-CoA acyltransferase, li...   \n",
       "95         r_2854  phosphatidylinositol 3-kinase (1-18:0, 2-16:1)...   \n",
       "96         r_2584  CL (1-16:0, 2-16:1, 3-16:0, 4-16:1) phospholip...   \n",
       "97         r_3217  PI 3,5-P2 phosphatase (1-16:1, 2-16:1), vacuol...   \n",
       "98         r_3222  PI 3,5-P2 phosphatase (1-18:0, 2-18:1), vacuol...   \n",
       "99         r_2809  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "100        r_2733  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "101        r_2711  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "102        r_4161  Probable vacuolar amino acid transporter YPQ3 ...   \n",
       "103        r_0089         5PP-IP5 pyrophosphorylation to 5,6-PP2-IP4   \n",
       "104        r_2668  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:0, 2-16...   \n",
       "105        r_3296        DAG lipase (1-16:0, 2-16:1), lipid particle   \n",
       "106        r_2621  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:1, 2-16...   \n",
       "107        r_2794  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "108        r_2680  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "109        r_2747  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "110        r_2679  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "111        r_2763  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "112        r_2585  CL (1-16:0, 2-16:1, 3-16:1, 4-16:1) phospholip...   \n",
       "113        r_0520                        GPI-anchor assembly, step 6   \n",
       "114        r_2592  CL (1-18:0, 2-16:1, 3-18:0, 4-16:1) phospholip...   \n",
       "115        r_2755  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "116        r_3298        DAG lipase (1-18:0, 2-16:1), lipid particle   \n",
       "117        r_2726  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "118        r_0249                           CAAX farnesyltransferase   \n",
       "119        r_2754  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "120        r_2787  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "121        r_2702  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:0, 2-16...   \n",
       "122        r_2774  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "123        r_3292  TAG lipase (1-18:1, 2-18:1, 3-16:0), lipid par...   \n",
       "124        r_3894  phosphatidyl-L-serine (1-16:1, 2-16:1) transpo...   \n",
       "125        r_2671  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:1, 2-16...   \n",
       "126        r_2710  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "127        r_3287  TAG lipase (1-16:1, 2-18:1, 3-18:1), lipid par...   \n",
       "128        r_3900  phosphatidyl-L-serine (1-18:1, 2-18:1) transpo...   \n",
       "129        r_2672  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:0, 2-18...   \n",
       "130        r_2778  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "131        r_2800  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "132        r_4204  Vacuolar aminopeptidase 1 (EC 3.4.11.22) (Amin...   \n",
       "133        r_3214  PI 4,5-P2 phosphatase (1-18:0, 2-18:1), ER mem...   \n",
       "134        r_2612  CL (1-18:1, 2-16:1, 3-18:0, 4-16:1) phospholip...   \n",
       "135        r_3295  TAG lipase (1-18:1, 2-18:1, 3-18:1), lipid par...   \n",
       "136        r_4704                          alkanesulfonate transport   \n",
       "137        PETHCT             phosphoethanolamine cytidyltransferase   \n",
       "138        r_0518                        GPI-anchor assembly, step 2   \n",
       "139        r_2690  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:0, 2-18...   \n",
       "140        r_2661  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "141      3OAS40_m          B-ketoacyl-ACP synthase (acetoacetyl-ACP)   \n",
       "142        r_2144       B-ketoacyl-ACP synthase (3-oxo-octanoyl-ACP)   \n",
       "143        r_2331         acyl dhap reductase (18:1), lipid particle   \n",
       "144        r_2602  CL (1-16:0, 2-16:1, 3-16:0, 4-16:1) phospholip...   \n",
       "145        r_2792  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "146        r_2729  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "147        r_2760  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "148        r_2713  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "149        r_3209  PI 4,5-P2 phosphatase (1-16:1, 2-16:1), ER mem...   \n",
       "150        r_2836  phosphatidylinositol 4-kinase (1-16:0, 2-16:1)...   \n",
       "151        r_2738  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "152        r_2588  CL (1-16:0, 2-16:1, 3-16:0, 4-18:1) phospholip...   \n",
       "153        r_2740  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "154        r_0767                           NAD nucleosidase nuclear   \n",
       "155        r_2700  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:0, 2-16...   \n",
       "156        r_3267  TAG lipase (1-16:0, 2-16:1, 3-18:1), lipid par...   \n",
       "157        r_4592                                   Mg(2+) transport   \n",
       "158        r_4218  D-aminoacyl-tRNA deacylase (DTD) (EC 3.1.1.96)...   \n",
       "159        r_3897  phosphatidyl-L-serine (1-16:0, 2-18:1) transpo...   \n",
       "160        r_2786  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "161        r_2665  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "162        r_4308  Probable phospholipid-transporting ATPase DNF3...   \n",
       "163        r_2784  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "164        r_4751                D-phenylalanine N-acetyltransferase   \n",
       "165        r_2717  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "166        r_2782  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "167        r_2143       B-ketoacyl-ACP synthase (3-oxo-hexanoyl-ACP)   \n",
       "168       23CAPPD            ADP-ribose 2,3-cyclic phosphodiesterase   \n",
       "169        r_2752  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "170        r_2627  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:1, 2-18...   \n",
       "171        r_2707  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:1, 2-18...   \n",
       "172        r_3211  PI 4,5-P2 phosphatase (1-18:1, 2-16:1), ER mem...   \n",
       "173        r_2808  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "174        r_4155  Glutamyl-tRNA(Gln) amidotransferase subunit B,...   \n",
       "175        r_2596  CL (1-16:0, 2-18:1, 3-16:0, 4-16:1) phospholip...   \n",
       "176        r_2604  CL (1-16:0, 2-16:1, 3-16:0, 4-18:1) phospholip...   \n",
       "177        r_3818  phosphatidyl-L-serine (1-16:1, 2-18:1) transpo...   \n",
       "178        r_2618  CL (1-16:1, 2-18:1, 3-18:0, 4-16:1) phospholip...   \n",
       "179        r_0438            ferrocytochrome-c:oxygen oxidoreductase   \n",
       "180        r_2785  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "181        r_4217  Fe(II):oxygen oxidoreductase; Fe2+:oxygen oxid...   \n",
       "182        r_2811  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "183        r_2610  CL (1-18:0, 2-16:1, 3-16:0, 4-18:1) phospholip...   \n",
       "184        r_2620  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:0, 2-16...   \n",
       "185        r_4181       Manganese-transporting ATPase 1 (EC 3.6.3.-)   \n",
       "186        r_2634  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "187        r_3284  TAG lipase (1-16:1, 2-18:1, 3-16:0), lipid par...   \n",
       "188        r_2773  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "189         PSP_L               phosphoserine phosphatase (L-serine)   \n",
       "190        r_2797  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "191      4MBZALDH  Putative aryl-alcohol dehydrogenase AAD3 (EC 1...   \n",
       "192        r_4313  Alpha-1,2-mannosyltransferase ALG9 (EC 2.4.1.2...   \n",
       "193        r_2841  phosphatidylinositol 4-kinase (1-16:1, 2-18:1)...   \n",
       "194        r_2637  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:1, 2-16...   \n",
       "195        r_2663  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "196        r_2597  CL (1-16:0, 2-18:1, 3-16:1, 4-16:1) phospholip...   \n",
       "197        r_2599  CL (1-16:0, 2-18:1, 3-18:1, 4-16:1) phospholip...   \n",
       "198        r_2709  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "199        r_2657  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:1, 2-18...   \n",
       "200        r_2329         acyl dhap reductase (16:1), lipid particle   \n",
       "201        r_3312  lysoPE (1-16:0) oleoyl-CoA acyltransferase, li...   \n",
       "202        r_3269  TAG lipase (1-16:1, 2-16:1, 3-16:1), lipid par...   \n",
       "203        r_2807  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "204        r_3276  TAG lipase (1-18:1, 2-16:1, 3-16:0), lipid par...   \n",
       "205        r_2706  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-18:0, 2-18...   \n",
       "206        r_2687  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:1, 2-16...   \n",
       "207        r_1202                             L-isoleucine transport   \n",
       "208        r_3218  PI 3,5-P2 phosphatase (1-18:0, 2-16:1), vacuol...   \n",
       "209        r_0073               4PP-IP5 depyrophosphorylation to IP6   \n",
       "210        r_3896  phosphatidyl-L-serine (1-18:1, 2-16:1) transpo...   \n",
       "211        r_0523                        GPI-anchor assembly, step 9   \n",
       "212        r_2652  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:0, 2-16...   \n",
       "213        r_3208  PI 4,5-P2 phosphatase (1-16:0, 2-16:1), ER mem...   \n",
       "214        r_4723                  polyphosphate phosphohydrolase, v   \n",
       "215        r_2648  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "216        r_2639  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:1, 2-16...   \n",
       "217        r_4167  Vacuolar cation-chloride cotransporter 1 (Vacu...   \n",
       "218        r_2744  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "219        r_2654  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:0, 2-16...   \n",
       "220        r_4156  Methionine aminopeptidase 2 (MAP 2) (MetAP 2) ...   \n",
       "221        r_2641  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:1, 2-18...   \n",
       "222        r_2616  CL (1-16:0, 2-18:1, 3-16:0, 4-18:1) phospholip...   \n",
       "223        r_2843  phosphatidylinositol 4-kinase (1-18:1, 2-18:1)...   \n",
       "224        r_3271  TAG lipase (1-16:1, 2-16:1, 3-18:1), lipid par...   \n",
       "225        r_3210  PI 4,5-P2 phosphatase (1-18:0, 2-16:1), ER mem...   \n",
       "226        r_2688  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:0, 2-18...   \n",
       "227        r_3213  PI 4,5-P2 phosphatase (1-16:1, 2-18:1), ER mem...   \n",
       "228        r_2685  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:1, 2-16...   \n",
       "229        r_2677  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "230        r_2684  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:0, 2-16...   \n",
       "231        r_2759  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "232        r_2617  CL (1-16:1, 2-18:1, 3-16:0, 4-16:1) phospholip...   \n",
       "233        r_4163  Probable vacuolar amino acid transporter YPQ3 ...   \n",
       "234        r_2593  CL (1-18:0, 2-16:1, 3-18:1, 4-16:1) phospholip...   \n",
       "235        r_2670  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:0, 2-16...   \n",
       "236        r_2636  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:0, 2-16...   \n",
       "237        GLYOX3                            (R)-lactate hydro-lyase   \n",
       "238        r_2669  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-16:1, 2-16...   \n",
       "239        r_3303        DAG lipase (1-18:1, 2-18:1), lipid particle   \n",
       "240        r_2716  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "241        r_2624  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:0, 2-18...   \n",
       "242        r_2765  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "243        r_2742  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "244        r_2611  CL (1-18:1, 2-16:1, 3-16:0, 4-16:1) phospholip...   \n",
       "245        r_2607  CL (1-16:1, 2-16:1, 3-16:0, 4-18:1) phospholip...   \n",
       "246        r_2676  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "247        r_4327  Magnesium transporter MRS2, mitochondrial (RNA...   \n",
       "248        r_2783  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "249        r_2649  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "250        r_2725  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "251        r_0357  diphosphoinositol-1,3,4,6-tetrakisphosphate di...   \n",
       "252        r_2793  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "253        r_2704  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:0, 2-18...   \n",
       "254        r_2718  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "255        r_2858  phosphatidylinositol 3-kinase (1-18:0, 2-18:1)...   \n",
       "256        r_2777  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "257        r_2462  lysoPI acyltransferase (1-18:0, 2-16:1), ER me...   \n",
       "258        r_4178  Putative aryl-alcohol dehydrogenase AAD3 (EC 1...   \n",
       "259        r_4239     L-arginyl-tRNA(Arg):protein arginyltransferase   \n",
       "260        r_2801  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "261        r_2325            acyl dhap reductase (16:1), ER membrane   \n",
       "262        r_3302        DAG lipase (1-18:0, 2-18:1), lipid particle   \n",
       "263        r_4388                           cyclic phosphodiesterase   \n",
       "264      THMPPt2m                     thiamine diphosphate transport   \n",
       "265        r_4158                      NADPH2:quinone oxidoreductase   \n",
       "266        r_0532      hexaprenyldihydroxybenzoate methyltransferase   \n",
       "267        r_2625  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-16:1, 2-18...   \n",
       "268        r_2664  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "269        r_2589  CL (1-16:0, 2-16:1, 3-16:1, 4-18:1) phospholip...   \n",
       "270        r_2776  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "271        r_2799  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "272         SERAT                         serine O-acetyltransferase   \n",
       "273        r_2804  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "274        r_2756  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "275        r_2761  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-16:1, 2-18...   \n",
       "276        r_4722                        polyphosphate polymerase, c   \n",
       "277        r_2601  CL (1-16:0, 2-18:1, 3-16:1, 4-18:1) phospholip...   \n",
       "278        r_2838  phosphatidylinositol 4-kinase (1-18:0, 2-16:1)...   \n",
       "279        r_0082  5-diphosphoinositol-1,2,3,4,6-pentakisphosphat...   \n",
       "280        r_2693  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "281        r_3819  phosphatidyl-L-serine (1-18:0, 2-18:1) transpo...   \n",
       "282        r_2330         acyl dhap reductase (18:0), lipid particle   \n",
       "283        r_3820  phosphatidyl-L-serine (1-18:1, 2-18:1) transpo...   \n",
       "284        r_2638  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:0, 2-16...   \n",
       "285        r_0521                        GPI-anchor assembly, step 7   \n",
       "286        r_2708  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "287        r_2714  MLCL (2-18:1, 3-16:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "288     2HP6MPMOm         2-hexaprenyl-6-methoxyphenol monooxygenase   \n",
       "289        r_3277  TAG lipase (1-18:1, 2-16:1, 3-16:1), lipid par...   \n",
       "290        r_2642  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-18:0, 2-18...   \n",
       "291        r_2721  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "292        r_3219  PI 3,5-P2 phosphatase (1-18:1, 2-16:1), vacuol...   \n",
       "293        r_2681  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "294        r_3275  TAG lipase (1-18:0, 2-16:1, 3-18:1), lipid par...   \n",
       "295        r_2722  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "296         PAPtm             adenosine 3',5'-bisphosphate transport   \n",
       "297        r_2609  CL (1-18:0, 2-16:1, 3-18:0, 4-16:1) phospholip...   \n",
       "298        r_4275                        Fe(II):NADP+ oxidoreductase   \n",
       "299        r_4589                                   Cu2(+) transport   \n",
       "300        r_3301        DAG lipase (1-16:1, 2-18:1), lipid particle   \n",
       "301        r_2644  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "302        r_2629  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "303      3MBZALDH  Putative aryl-alcohol dehydrogenase AAD3 (EC 1...   \n",
       "304        r_3289  TAG lipase (1-18:0, 2-18:1, 3-16:1), lipid par...   \n",
       "305        r_2767  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "306        r_3273  TAG lipase (1-18:0, 2-16:1, 3-16:1), lipid par...   \n",
       "307        r_3813  phosphatidyl-L-serine (1-16:0, 2-16:1) transpo...   \n",
       "308        r_2803  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "309        r_2779  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "310        r_2628  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "311        r_0088         5PP-IP5 pyrophosphorylation to 4,5-PP2-IP4   \n",
       "312        r_3280  TAG lipase (1-16:0, 2-18:1, 3-16:0), lipid par...   \n",
       "313        r_0340                                  dihydroceramidase   \n",
       "314        r_2739  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "315        r_2689  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-16:1, 2-18...   \n",
       "316        r_3315  lysoPE (1-18:1) oleoyl-CoA acyltransferase, li...   \n",
       "317        r_0013  2,3-diketo-5-methylthio-1-phosphopentane degra...   \n",
       "318        r_4263  Anion/proton exchange transporter GEF1 (CLC pr...   \n",
       "319        r_2735  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "320        r_3893  phosphatidyl-L-serine (1-16:0, 2-16:1) transpo...   \n",
       "321    3DH5HPBMTm  S-adenosyl-L-methionine:3-hexaprenyl-4,5-dihyd...   \n",
       "322        r_0137  acylation of GPI inositol at 2 position, GPI-a...   \n",
       "323        r_2746  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "324        r_3274  TAG lipase (1-18:0, 2-16:1, 3-18:0), lipid par...   \n",
       "325        r_2839  phosphatidylinositol 4-kinase (1-18:1, 2-16:1)...   \n",
       "326        r_2768  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "327        r_3272  TAG lipase (1-18:0, 2-16:1, 3-16:0), lipid par...   \n",
       "328        CHOLtu         Protein PNS1 (pH nine-sensitive protein 1)   \n",
       "329        r_2770  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "330        r_2613  CL (1-18:1, 2-16:1, 3-16:0, 4-18:1) phospholip...   \n",
       "331        r_2603  CL (1-16:0, 2-16:1, 3-18:0, 4-16:1) phospholip...   \n",
       "332        r_2651  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "333        r_1086                                  V-ATPase, vacuole   \n",
       "334        r_2697  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "335        r_2791  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "336        r_4700                         hydrogen sulfide transport   \n",
       "337        r_3212  PI 4,5-P2 phosphatase (1-16:0, 2-18:1), ER mem...   \n",
       "338        r_2650  MLCL (2-16:1, 3-18:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "339        r_2769  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "340        r_3288  TAG lipase (1-18:0, 2-18:1, 3-16:0), lipid par...   \n",
       "341        r_3899  phosphatidyl-L-serine (1-18:0, 2-18:1) transpo...   \n",
       "342        r_2853  phosphatidylinositol 3-kinase (1-16:1, 2-16:1)...   \n",
       "343        r_2802  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:0, 2-18...   \n",
       "344        r_3895  phosphatidyl-L-serine (1-18:0, 2-16:1) transpo...   \n",
       "345        r_2788  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "346        r_4754                           flavin prenyltransferase   \n",
       "347        r_2626  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:0, 2-18...   \n",
       "348        r_4587                                   Ca(2+) transport   \n",
       "349        r_3816  phosphatidyl-L-serine (1-18:1, 2-16:1) transpo...   \n",
       "350        r_4314  Alpha-1,2-mannosyltransferase ALG9 (EC 2.4.1.2...   \n",
       "351        r_2631  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "352        r_2619  CL (1-16:1, 2-18:1, 3-16:0, 4-18:1) phospholip...   \n",
       "353        r_0341                                  dihydroceramidase   \n",
       "354        r_2695  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-18:1, 2-16...   \n",
       "355        r_2640  MLCL (2-16:1, 3-18:0, 4-16:1):PC (1-16:0, 2-18...   \n",
       "356        r_4705                             taurocholate transport   \n",
       "357        r_1919                                   lysine transport   \n",
       "358        r_2595  CL (1-18:0, 2-16:1, 3-16:1, 4-18:1) phospholip...   \n",
       "359        r_3898  phosphatidyl-L-serine (1-16:1, 2-18:1) transpo...   \n",
       "360        r_2762  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "361        r_4312  S-adenosyl-L-methionine:carnosine N-methyltran...   \n",
       "362        r_2728  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:0, 2-18...   \n",
       "363        r_2630  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "364        r_1220                               L-tyrosine transport   \n",
       "365        r_2795  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "366        r_4372                           cyclic phosphodiesterase   \n",
       "367        r_2635  MLCL (2-16:1, 3-16:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "368        r_2730  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "369        r_2734  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "370        r_2686  MLCL (2-18:1, 3-18:0, 4-16:1):PC (1-18:0, 2-16...   \n",
       "371        r_4234  Phosphatidylinositol N-acetylglucosaminyltrans...   \n",
       "372        r_0343                                  dihydroceramidase   \n",
       "373        r_2749  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "374        r_2737  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "375           Clt                                 chloride transport   \n",
       "376        r_4193  S-Adenosyl-L-methionine:histone-L-lysine N6-me...   \n",
       "377        r_2705  MLCL (2-18:1, 3-16:0, 4-18:1):PC (1-16:1, 2-18...   \n",
       "378        SERD_D                             D-serine ammonia-lyase   \n",
       "379        r_2623  MLCL (2-16:1, 3-16:0, 4-16:1):PC (1-18:1, 2-16...   \n",
       "380        r_3270  TAG lipase (1-16:1, 2-16:1, 3-18:0), lipid par...   \n",
       "381        r_3216  PI 3,5-P2 phosphatase (1-16:0, 2-16:1), vacuol...   \n",
       "382        r_4322  GPI mannosyltransferase 4 (EC 2.4.1.-) (GPI ma...   \n",
       "383        r_2606  CL (1-16:1, 2-16:1, 3-18:0, 4-16:1) phospholip...   \n",
       "384        r_2723  MLCL (1-16:0, 2-16:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "385        r_3299        DAG lipase (1-18:1, 2-16:1), lipid particle   \n",
       "386        r_2766  MLCL (1-18:1, 2-16:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "387        r_4175  Putative aryl-alcohol dehydrogenase AAD3 (EC 1...   \n",
       "388        r_4210               urea hydro-lyase (cyanamide-forming)   \n",
       "389        r_4225  Broad-range acid phosphatase DET1 (EC 3.1.3.-)...   \n",
       "390        r_3223  PI 3,5-P2 phosphatase (1-18:1, 2-18:1), vacuol...   \n",
       "391        r_2857  phosphatidylinositol 3-kinase (1-16:1, 2-18:1)...   \n",
       "392        r_4337  Iron-sulfur clusters transporter ATM1, mitocho...   \n",
       "393        r_3265  TAG lipase (1-16:0, 2-16:1, 3-16:1), lipid par...   \n",
       "394       CD2abc1  P-type cation-transporting ATPase (EC 3.6.3.3)...   \n",
       "395        r_2724  MLCL (1-16:0, 2-16:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "396        r_2856  phosphatidylinositol 3-kinase (1-16:0, 2-18:1)...   \n",
       "397        r_2667  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:1, 2-18...   \n",
       "398        r_3282  TAG lipase (1-16:0, 2-18:1, 3-18:0), lipid par...   \n",
       "399        r_2658  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:0, 2-18...   \n",
       "400        r_3814  phosphatidyl-L-serine (1-16:1, 2-16:1) transpo...   \n",
       "401        r_2789  MLCL (1-16:0, 2-18:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "402        r_0331    deoxyhypusine synthase, cytosolic/mitochondrial   \n",
       "403        r_4724                  polyphosphate phosphohydrolase, m   \n",
       "404        r_4412                           cyclic phosphodiesterase   \n",
       "405        r_2674  MLCL (2-18:1, 3-16:0, 4-16:1):PC (1-18:0, 2-18...   \n",
       "406        r_2696  MLCL (2-18:1, 3-18:1, 4-16:1):PC (1-16:0, 2-18...   \n",
       "407        r_2842  phosphatidylinositol 4-kinase (1-18:0, 2-18:1)...   \n",
       "408        r_2758  MLCL (1-18:0, 2-16:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "409        r_2666  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:0, 2-18...   \n",
       "410        r_3278  TAG lipase (1-18:1, 2-16:1, 3-18:0), lipid par...   \n",
       "411  FCYTOB5OXR_m  NADH-cytochrome b5 reductase 1 (EC 1.6.2.2) (M...   \n",
       "412       GGCLUT2  Glutathione-specific gamma-glutamylcyclotransf...   \n",
       "413        r_2732  MLCL (1-16:1, 2-16:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "414        r_2655  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-18:1, 2-16...   \n",
       "415        r_2660  MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:0, 2-16...   \n",
       "416        r_4203  Vacuolar aminopeptidase 1 (EC 3.4.11.22) (Amin...   \n",
       "417        r_2837  phosphatidylinositol 4-kinase (1-16:1, 2-16:1)...   \n",
       "418        r_2753  MLCL (1-18:0, 2-16:1, 4-16:1):PC (1-16:1, 2-18...   \n",
       "419        r_1085                                    V-ATPase, Golgi   \n",
       "420        r_2463  lysoPI acyltransferase (1-18:0, 2-18:1), ER me...   \n",
       "421        r_2798  MLCL (1-16:1, 2-18:1, 4-16:1):PC (1-18:0, 2-16...   \n",
       "422        r_2781  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:1, 2-16...   \n",
       "423        r_2587  CL (1-16:0, 2-16:1, 3-18:1, 4-16:1) phospholip...   \n",
       "424        r_0522                        GPI-anchor assembly, step 8   \n",
       "425        r_3297        DAG lipase (1-16:1, 2-16:1), lipid particle   \n",
       "426        r_2653  MLCL (2-16:1, 3-16:0, 4-18:1):PC (1-16:1, 2-16...   \n",
       "427        r_2683  MLCL (2-18:1, 3-16:1, 4-16:1):PC (1-18:1, 2-18...   \n",
       "428        r_1208                     L-leucine transport, vacuoluar   \n",
       "429        r_0092               6PP-IP5 depyrophosphorylation to IP6   \n",
       "430        r_1684                                  choline transport   \n",
       "431        r_3314  lysoPE (1-18:0) oleoyl-CoA acyltransferase, li...   \n",
       "432        r_4725                  polyphosphate phosphohydrolase, n   \n",
       "433        r_2780  MLCL (1-16:0, 2-18:1, 4-16:1):PC (1-16:0, 2-16...   \n",
       "434        r_3294  TAG lipase (1-18:1, 2-18:1, 3-18:0), lipid par...   \n",
       "435        r_2598  CL (1-16:0, 2-18:1, 3-18:0, 4-16:1) phospholip...   \n",
       "436        r_2805  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-16:1, 2-16...   \n",
       "437        r_4403                      Triphosphate phosphohydrolase   \n",
       "438        r_3283  TAG lipase (1-16:0, 2-18:1, 3-18:1), lipid par...   \n",
       "439        r_2614  CL (1-16:0, 2-18:1, 3-16:0, 4-16:1) phospholip...   \n",
       "440        r_4247     ATP:L-threonyl,bicarbonate adenylyltransferase   \n",
       "441        r_4339           UDP-N-acetylglucosamine transporter YEA4   \n",
       "442        r_4233  S-Adenosyl-L-methionine:protein-C-terminal-S-f...   \n",
       "443     2HPMBQMTm  2-hexaprenyl-6-methoxy-1,4-benzoquinone methyl...   \n",
       "444        r_3285  TAG lipase (1-16:1, 2-18:1, 3-16:1), lipid par...   \n",
       "445        r_2806  MLCL (1-16:1, 2-18:1, 4-18:1):PC (1-18:0, 2-16...   \n",
       "446        r_4278             diphthine:ammonia ligase (AMP-forming)   \n",
       "447        r_2775  MLCL (1-18:1, 2-16:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "448        r_2743  MLCL (1-16:1, 2-16:1, 4-18:1):PC (1-18:1, 2-16...   \n",
       "449        r_0342                                  dihydroceramidase   \n",
       "450        r_3268  TAG lipase (1-16:1, 2-16:1, 3-16:0), lipid par...   \n",
       "451        r_4195                              Zinc transporter YKE4   \n",
       "452        r_2591  CL (1-18:0, 2-16:1, 3-16:1, 4-16:1) phospholip...   \n",
       "453        r_3221  PI 3,5-P2 phosphatase (1-16:1, 2-18:1), vacuol...   \n",
       "454        r_3253  ergosteryl ester hydrolase (18:1), lipid particle   \n",
       "455        r_2605  CL (1-16:1, 2-16:1, 3-16:0, 4-16:1) phospholip...   \n",
       "456        r_3290  TAG lipase (1-18:0, 2-18:1, 3-18:0), lipid par...   \n",
       "\n",
       "                                                     2  \\\n",
       "0         h2o_mm + s_3250 --> h_mm + ocdca_mm + s_3275   \n",
       "1           atp_vm + h_vm + s_3340 --> adp_vm + s_3392   \n",
       "2          h2o_mm + s_3236 --> h_mm + hdca_mm + s_3273   \n",
       "3             s_3284 + s_3302 <=> 1agpc181_mm + s_3235   \n",
       "4             s_3285 + s_3307 <=> 1agpc181_mm + s_3314   \n",
       "5             s_3272 + s_3304 <=> 1agpc160_mm + s_3253   \n",
       "6         h2o_lp + s_3067 --> h_lp + ocdca_lp + s_3043   \n",
       "7             s_3290 + s_3296 <=> 1agpc160_mm + s_3253   \n",
       "8    h2o_lp + s_3066 --> 12dgr181_lp + h_lp + hdcea_lp   \n",
       "9    h_lp + nadph_lp + s_2948 --> 1hdecg3p_lp + nad...   \n",
       "10                                    zn2_c <=> zn2_er   \n",
       "11                gthrd_c + s_3834 <=> s_3835 + s_3836   \n",
       "12    h_erm + nadph_erm + s_2943 --> nadp_erm + s_2938   \n",
       "13   h2o_lp + tag1819Z1619Z1819Z_lp --> h_lp + ocdc...   \n",
       "14                         h_e + zn2_e <=> h_c + zn2_c   \n",
       "15        h2o_lp + s_3060 --> h_lp + hdcea_lp + s_3045   \n",
       "16            s_3282 + s_3302 <=> 1agpc181_mm + s_3268   \n",
       "17            s_3278 + s_3307 <=> 1agpc181_mm + s_3308   \n",
       "18        h2o_lp + s_3070 --> h_lp + ocdca_lp + s_3049   \n",
       "19            s_3287 + s_3305 <=> 1agpc161_mm + s_3315   \n",
       "20            s_3283 + s_3304 <=> 1agpc160_mm + s_3313   \n",
       "21            s_3279 + s_3300 <=> 1agpc180_mm + s_3265   \n",
       "22                               his__L_v <=> his__L_c   \n",
       "23                        chsterol_er <=> chsterol_erm   \n",
       "24      h2o_lp + s_3656 --> ergst_lp + h_lp + hdcea_lp   \n",
       "25          atp_vm + h_vm + s_3356 --> adp_vm + s_3399   \n",
       "26            s_3274 + s_3302 <=> 1agpc181_mm + s_3243   \n",
       "27           bzalc_c + nad_c <=> bzal_c + h_c + nadh_c   \n",
       "28   h_erm + nadph_erm + s_2940 --> 1hdecg3p_erm + ...   \n",
       "29         h2o_lp + s_3044 --> h_lp + hdca_lp + s_3043   \n",
       "30         h2o_mm + s_3262 --> h_mm + hdca_mm + s_3282   \n",
       "31            s_3288 + s_3300 <=> 1agpc180_mm + s_3247   \n",
       "32            s_3283 + s_3307 <=> 1agpc181_mm + s_3313   \n",
       "33         h2o_c + s_3857 <=> h_c + s_3858 + trnatyr_c   \n",
       "34            s_3286 + s_3304 <=> 1agpc160_mm + s_3243   \n",
       "35         h2o_mm + s_3246 --> h_mm + hdca_mm + s_3288   \n",
       "36    h2o_lp + s_3045 --> h_lp + mag160_lp + ocdcea_lp   \n",
       "37            s_3281 + s_3307 <=> 1agpc181_mm + s_3311   \n",
       "38       h2o_lp + s_3080 --> h_lp + ocdcea_lp + s_3053   \n",
       "39          pc161_mm + s_3274 <=> 1agpc161_mm + s_3243   \n",
       "40                3hph5mb_m + h_m --> 2hp6mp_m + co2_m   \n",
       "41            s_3294 + s_3296 <=> 1agpc160_mm + s_3265   \n",
       "42            s_3275 + s_3304 <=> 1agpc160_mm + s_3256   \n",
       "43   2hpmmbq_m + h_m + nadh_m + o2_m --> 2hpmhmbq_m...   \n",
       "44        h2o_mm + s_3246 --> h_mm + ocdca_mm + s_3270   \n",
       "45            s_3272 + s_3305 <=> 1agpc161_mm + s_3253   \n",
       "46                   h2o_c + s_4018 <=> h_c + 2.0 pi_c   \n",
       "47    h2o_erm + s_3615 --> 2.0 h_erm + pi_erm + s_3591   \n",
       "48            s_3281 + s_3306 <=> 1agpc180_mm + s_3311   \n",
       "49                                   s_3113 <=> s_3155   \n",
       "50                          chtn_c + h2o_c <=> acgam_c   \n",
       "51        fmn_c + 2.0 h_c + nadh_c --> fmnh2_c + nad_c   \n",
       "52            s_3281 + s_3300 <=> 1agpc180_mm + s_3267   \n",
       "53                                 g6p_A_c <=> g6p_B_c   \n",
       "54            s_3270 + s_3300 <=> 1agpc180_mm + s_3240   \n",
       "55      gdpmann_er + s_3891 <=> gdp_er + h_er + s_3893   \n",
       "56            s_3284 + s_3304 <=> 1agpc160_mm + s_3237   \n",
       "57          pc161_mm + s_3289 <=> 1agpc161_mm + s_3251   \n",
       "58            s_3273 + s_3307 <=> 1agpc181_mm + s_3254   \n",
       "59               gthrd_er + s_3830 <=> s_3832 + s_3833   \n",
       "60      fmn_c + 2.0 h_c + nadph_c --> fmnh2_c + nadp_c   \n",
       "61     dhdascb_c + 2.0 gthrd_c <=> ascb__L_c + gthox_c   \n",
       "62   naglc2p_er + uacgam_er <=> h_er + s_3767 + udp_er   \n",
       "63   h_er + pmtcoa_er + ser__L_er --> 3dsphgn_er + ...   \n",
       "64       cyan_m + tsul_m <=> 2.0 h_m + so3_m + tcynt_m   \n",
       "65            nad_c + s_3791 <=> h_c + nadh_c + s_3792   \n",
       "66            s_3291 + s_3296 <=> 1agpc160_mm + s_3257   \n",
       "67            s_3288 + s_3296 <=> 1agpc160_mm + s_3247   \n",
       "68            s_3274 + s_3300 <=> 1agpc180_mm + s_3243   \n",
       "69                               his__L_c --> his__L_m   \n",
       "70                gtp_c + h2o_c --> gdp_c + h_c + pi_c   \n",
       "71            s_3281 + s_3296 <=> 1agpc160_mm + s_3267   \n",
       "72            s_3285 + s_3302 <=> 1agpc181_mm + s_3239   \n",
       "73          atp_gm + h_gm + s_3368 --> adp_gm + s_3369   \n",
       "74            s_3278 + s_3305 <=> 1agpc161_mm + s_3308   \n",
       "75          atp_vm + h_vm + s_3348 --> adp_vm + s_3395   \n",
       "76          pc161_mm + s_3287 <=> 1agpc161_mm + s_3245   \n",
       "77        gdpmann_g + h_g + mpdol_g --> gdp_g + s_0414   \n",
       "78          pc161_mm + s_3282 <=> 1agpc161_mm + s_3268   \n",
       "79            s_3290 + s_3307 <=> 1agpc181_mm + s_3255   \n",
       "80            s_3276 + s_3300 <=> 1agpc180_mm + s_3245   \n",
       "81            s_3279 + s_3306 <=> 1agpc180_mm + s_3309   \n",
       "82            s_3280 + s_3307 <=> 1agpc181_mm + s_3310   \n",
       "83        h2o_mm + s_3260 --> h_mm + ocdca_mm + s_3292   \n",
       "84            s_3288 + s_3302 <=> 1agpc181_mm + s_3247   \n",
       "85            s_3275 + s_3307 <=> 1agpc181_mm + s_3256   \n",
       "86            s_3293 + s_3300 <=> 1agpc180_mm + s_3263   \n",
       "87                                   s_3111 <=> s_3151   \n",
       "88            s_3295 + s_3306 <=> 1agpc180_mm + s_3319   \n",
       "89       h2o_vm + s_3420 --> 2.0 h_vm + pi_vm + s_3396   \n",
       "90          h2o_er + h_er + s_3888 <=> dolp_er + pi_er   \n",
       "91                         h_e + mn2_e <=> h_c + mn2_c   \n",
       "92   h_erm + nadph_erm + s_2942 --> 1odecg3p_erm + ...   \n",
       "93                   arg__L_c + h_c --> arg__L_m + h_m   \n",
       "94      1agpe161_lp + ocdce9coa_lp --> coa_lp + s_3688   \n",
       "95          atp_vm + h_vm + s_3346 --> adp_vm + s_3394   \n",
       "96         h2o_mm + s_3234 --> h_mm + hdca_mm + s_3270   \n",
       "97   h2o_vm + s_3417 --> 2.0 h_vm + pail3p161_vm + ...   \n",
       "98       h2o_vm + s_3422 --> 2.0 h_vm + pi_vm + s_3398   \n",
       "99            s_3295 + s_3305 <=> 1agpc161_mm + s_3319   \n",
       "100     pc161_mm + s_3286 <=> 1agpc161_mm + clpn161_mm   \n",
       "101           s_3283 + s_3302 <=> 1agpc181_mm + s_3269   \n",
       "102                              arg__L_v <=> arg__L_c   \n",
       "103   2.0 h_c + pi_c + ppmi12346p_c <=> h2o_c + s_0309   \n",
       "104           s_3278 + s_3296 <=> 1agpc160_mm + s_3264   \n",
       "105    h2o_lp + s_3043 --> h_lp + hdcea_lp + mag160_lp   \n",
       "106         pc161_mm + s_3270 <=> 1agpc161_mm + s_3240   \n",
       "107           s_3293 + s_3306 <=> 1agpc180_mm + s_3318   \n",
       "108           s_3279 + s_3304 <=> 1agpc160_mm + s_3309   \n",
       "109           s_3287 + s_3307 <=> 1agpc181_mm + s_3315   \n",
       "110           s_3279 + s_3302 <=> 1agpc181_mm + s_3265   \n",
       "111           s_3289 + s_3307 <=> 1agpc181_mm + s_3316   \n",
       "112        h2o_mm + s_3235 --> h_mm + hdca_mm + s_3272   \n",
       "113                       pe_hs_er + s_0334 <=> s_0337   \n",
       "114       h2o_mm + s_3248 --> h_mm + ocdca_mm + s_3273   \n",
       "115           s_3288 + s_3307 <=> 1agpc181_mm + s_3249   \n",
       "116    h2o_lp + s_3051 --> h_lp + hdcea_lp + mag180_lp   \n",
       "117           s_3285 + s_3300 <=> 1agpc180_mm + s_3239   \n",
       "118                 4hbz_c + s_1231 --> ppi_c + s_1230   \n",
       "119           s_3288 + s_3306 <=> 1agpc180_mm + s_3249   \n",
       "120           s_3292 + s_3307 <=> 1agpc181_mm + s_3261   \n",
       "121           s_3282 + s_3300 <=> 1agpc180_mm + s_3268   \n",
       "122           s_3291 + s_3300 <=> 1agpc180_mm + s_3257   \n",
       "123   h2o_lp + s_3058 --> 12dgr181_lp + h_lp + hdca_lp   \n",
       "124                                  s_3110 <=> s_3480   \n",
       "125           s_3278 + s_3302 <=> 1agpc181_mm + s_3264   \n",
       "126           s_3283 + s_3300 <=> 1agpc180_mm + s_3269   \n",
       "127      h2o_lp + s_3078 --> h_lp + ocdcea_lp + s_3049   \n",
       "128                                  s_3116 <=> s_3489   \n",
       "129           s_3278 + s_3304 <=> 1agpc160_mm + s_3308   \n",
       "130           s_3291 + s_3306 <=> 1agpc180_mm + s_3317   \n",
       "131           s_3294 + s_3304 <=> 1agpc160_mm + s_3267   \n",
       "132     ala_L_ile__L_v + h2o_v <=> ala__L_v + leu__L_v   \n",
       "133   h2o_erm + s_3614 --> 2.0 h_erm + pi_erm + s_3590   \n",
       "134       h2o_mm + s_3254 --> h_mm + ocdca_mm + s_3290   \n",
       "135  h2o_lp + s_3082 --> 12dgr181_lp + h_lp + ocdce...   \n",
       "136                                  s_4266 --> s_4265   \n",
       "137         ctp_c + ethamp_c + h_c --> cdpea_c + ppi_c   \n",
       "138  acgpail_hs_er + h2o_er <=> ac_er + gpail_hs_er...   \n",
       "139           s_3280 + s_3306 <=> 1agpc180_mm + s_3310   \n",
       "140         pc161_mm + s_3276 <=> 1agpc161_mm + s_3245   \n",
       "141  acACP_m + h_m + malACP_m --> ACP_m + actACP_m ...   \n",
       "142  h_m + malACP_m + s_2773 --> 3ooctACP_m + ACP_m...   \n",
       "143      h_lp + nadph_lp + s_2951 --> nadp_lp + s_2947   \n",
       "144        h2o_mm + s_3234 --> h_mm + hdca_mm + s_3284   \n",
       "145           s_3293 + s_3304 <=> 1agpc160_mm + s_3318   \n",
       "146           s_3285 + s_3305 <=> 1agpc161_mm + s_3314   \n",
       "147           s_3289 + s_3304 <=> 1agpc160_mm + s_3316   \n",
       "148           s_3283 + s_3305 <=> 1agpc161_mm + s_3313   \n",
       "149  h2o_erm + s_3609 --> 2.0 h_erm + pail4p161_erm...   \n",
       "150         atp_gm + h_gm + s_3358 --> adp_gm + s_3361   \n",
       "151           s_3286 + s_3306 <=> 1agpc180_mm + s_3243   \n",
       "152        h2o_mm + s_3238 --> h_mm + hdca_mm + s_3275   \n",
       "153           s_3287 + s_3296 <=> 1agpc160_mm + s_3245   \n",
       "154          h2o_n + nad_n --> adprib_n + h_n + ncam_n   \n",
       "155           s_3282 + s_3296 <=> 1agpc160_mm + s_3268   \n",
       "156      h2o_lp + s_3075 --> h_lp + ocdcea_lp + s_3043   \n",
       "157                        h_e + mg2_e <=> h_c + mg2_c   \n",
       "158         h2o_c + s_3856 <=> gly_c + h_c + trnaala_c   \n",
       "159                                  s_3113 <=> s_3486   \n",
       "160           s_3292 + s_3306 <=> 1agpc180_mm + s_3261   \n",
       "161           s_3276 + s_3305 <=> 1agpc161_mm + s_3257   \n",
       "162  atp_erm + h2o_erm + s_3985 --> adp_erm + h_erm...   \n",
       "163           s_3292 + s_3304 <=> 1agpc160_mm + s_3261   \n",
       "164                accoa_c + s_4296 --> coa_c + s_4297   \n",
       "165         pc161_mm + s_3284 <=> 1agpc161_mm + s_3235   \n",
       "166           s_3292 + s_3300 <=> 1agpc180_mm + s_3259   \n",
       "167  h_m + malACP_m + s_2771 --> 3ohexACP_m + ACP_m...   \n",
       "168                 23camp_c + h2o_c --> amp2p_c + h_c   \n",
       "169           s_3288 + s_3304 <=> 1agpc160_mm + s_3249   \n",
       "170           s_3270 + s_3307 <=> 1agpc181_mm + s_3252   \n",
       "171           s_3282 + s_3307 <=> 1agpc181_mm + s_3312   \n",
       "172   h2o_erm + s_3611 --> 2.0 h_erm + pi_erm + s_3587   \n",
       "173           s_3295 + s_3304 <=> 1agpc160_mm + s_3319   \n",
       "174  atp_m + gln__L_m + glutrna_gln_m + h2o_m --> a...   \n",
       "175        h2o_mm + s_3258 --> h_mm + hdca_mm + s_3278   \n",
       "176        h2o_mm + s_3238 --> h_mm + hdca_mm + s_3285   \n",
       "177                                  s_3114 <=> s_3157   \n",
       "178       h2o_mm + s_3266 --> h_mm + ocdca_mm + s_3294   \n",
       "179  focytc_m + 1.266 h_m + 0.25 o2_m --> ficytc_m ...   \n",
       "180           s_3292 + s_3305 <=> 1agpc161_mm + s_3261   \n",
       "181  4.0 fe2_m + 4.0 h_m + o2_m <=> 4.0 fe3_m + 2.0...   \n",
       "182           s_3295 + s_3307 <=> 1agpc181_mm + s_3319   \n",
       "183        h2o_mm + s_3250 --> h_mm + hdca_mm + s_3289   \n",
       "184           s_3270 + s_3296 <=> 1agpc160_mm + s_3240   \n",
       "185  atp_c + h2o_c + mn2_c <=> adp_c + h_c + mn2_er...   \n",
       "186           s_3272 + s_3306 <=> 1agpc180_mm + s_3253   \n",
       "187        h2o_lp + s_3050 --> h_lp + hdca_lp + s_3049   \n",
       "188         pc161_mm + s_3291 <=> 1agpc161_mm + s_3257   \n",
       "189              h2o_c + pser__L_c --> pi_c + ser__L_c   \n",
       "190         pc161_mm + s_3294 <=> 1agpc161_mm + s_3265   \n",
       "191     4mbzalc_c + nad_c <=> 4mbzald_c + h_c + nadh_c   \n",
       "192           dolmanp_er + s_3993 <=> dolp_er + s_3994   \n",
       "193         atp_gm + h_gm + s_3370 --> adp_gm + s_3371   \n",
       "194         pc161_mm + s_3273 <=> 1agpc161_mm + s_3242   \n",
       "195           s_3276 + s_3302 <=> 1agpc181_mm + s_3245   \n",
       "196        h2o_mm + s_3259 --> h_mm + hdca_mm + s_3279   \n",
       "197        h2o_mm + s_3261 --> h_mm + hdca_mm + s_3281   \n",
       "198         pc161_mm + s_3283 <=> 1agpc161_mm + s_3269   \n",
       "199           s_3275 + s_3305 <=> 1agpc161_mm + s_3256   \n",
       "200      h_lp + nadph_lp + s_2949 --> nadp_lp + s_2945   \n",
       "201     1agpe160_lp + ocdce9coa_lp --> coa_lp + s_3686   \n",
       "202  h2o_lp + tag1619Z1619Z1619Z_lp --> 12dgr161_lp...   \n",
       "203           s_3295 + s_3302 <=> 1agpc181_mm + s_3269   \n",
       "204        h2o_lp + s_3056 --> h_lp + hdca_lp + s_3055   \n",
       "205           s_3282 + s_3306 <=> 1agpc180_mm + s_3312   \n",
       "206           s_3280 + s_3302 <=> 1agpc181_mm + s_3266   \n",
       "207                              ile__L_v <=> ile__L_c   \n",
       "208      h2o_vm + s_3418 --> 2.0 h_vm + pi_vm + s_3394   \n",
       "209           h2o_c + s_0279 --> h_c + minohp_c + pi_c   \n",
       "210                                  s_3112 <=> s_3484   \n",
       "211           pe_hs_er + s_0338 <=> dag_hs_er + s_0332   \n",
       "212           s_3275 + s_3296 <=> 1agpc160_mm + s_3244   \n",
       "213   h2o_erm + s_3608 --> 2.0 h_erm + pi_erm + s_3584   \n",
       "214                  h2o_v + s_4290 <=> h_v + 2.0 pi_v   \n",
       "215           s_3274 + s_3304 <=> 1agpc160_mm + s_3255   \n",
       "216           s_3273 + s_3302 <=> 1agpc181_mm + s_3242   \n",
       "217          cl_v + k_v + na1_v <=> cl_c + k_c + na1_c   \n",
       "218           s_3287 + s_3304 <=> 1agpc160_mm + s_3315   \n",
       "219           s_3275 + s_3300 <=> 1agpc180_mm + s_3244   \n",
       "220     h2o_v + met_L_ala__L_v <=> ala__L_v + met__L_v   \n",
       "221           s_3273 + s_3305 <=> 1agpc161_mm + s_3254   \n",
       "222        h2o_mm + s_3262 --> h_mm + hdca_mm + s_3293   \n",
       "223         atp_gm + h_gm + s_3374 --> adp_gm + s_3375   \n",
       "224  h2o_lp + tag1619Z1619Z1819Z_lp --> 12dgr161_lp...   \n",
       "225   h2o_erm + s_3610 --> 2.0 h_erm + pi_erm + s_3586   \n",
       "226           s_3280 + s_3304 <=> 1agpc160_mm + s_3310   \n",
       "227   h2o_erm + s_3613 --> 2.0 h_erm + pi_erm + s_3589   \n",
       "228         pc161_mm + s_3280 <=> 1agpc161_mm + s_3266   \n",
       "229         pc161_mm + s_3279 <=> 1agpc161_mm + s_3265   \n",
       "230           s_3280 + s_3296 <=> 1agpc160_mm + s_3266   \n",
       "231           s_3289 + s_3302 <=> 1agpc181_mm + s_3251   \n",
       "232        h2o_mm + s_3264 --> h_mm + hdca_mm + s_3294   \n",
       "233                              lys__L_v <=> lys__L_c   \n",
       "234       h2o_mm + s_3249 --> h_mm + ocdca_mm + s_3274   \n",
       "235           s_3278 + s_3300 <=> 1agpc180_mm + s_3264   \n",
       "236           s_3273 + s_3296 <=> 1agpc160_mm + s_3242   \n",
       "237                h_c + lac__D_c <=> h2o_c + mthgxl_c   \n",
       "238         pc161_mm + s_3278 <=> 1agpc161_mm + s_3264   \n",
       "239  12dgr181_lp + h2o_lp --> h_lp + magole_hs_lp +...   \n",
       "240           s_3284 + s_3296 <=> 1agpc160_mm + s_3235   \n",
       "241           s_3270 + s_3304 <=> 1agpc160_mm + s_3252   \n",
       "242         pc161_mm + s_3290 <=> 1agpc161_mm + s_3253   \n",
       "243           s_3287 + s_3300 <=> 1agpc180_mm + s_3245   \n",
       "244        h2o_mm + s_3252 --> h_mm + hdca_mm + s_3290   \n",
       "245        h2o_mm + s_3244 --> h_mm + hdca_mm + s_3287   \n",
       "246           s_3279 + s_3296 <=> 1agpc160_mm + s_3265   \n",
       "247                                    mg2_c <=> mg2_m   \n",
       "248           s_3292 + s_3302 <=> 1agpc181_mm + s_3259   \n",
       "249           s_3274 + s_3305 <=> 1agpc161_mm + s_3255   \n",
       "250         pc161_mm + s_3285 <=> 1agpc161_mm + s_3239   \n",
       "251  3.0 h2o_c + s_0309 --> 3.0 h_c + inospp1_c + 3...   \n",
       "252           s_3293 + s_3305 <=> 1agpc161_mm + s_3318   \n",
       "253           s_3282 + s_3304 <=> 1agpc160_mm + s_3312   \n",
       "254           s_3284 + s_3300 <=> 1agpc180_mm + s_3235   \n",
       "255         atp_vm + h_vm + s_3354 --> adp_vm + s_3398   \n",
       "256           s_3291 + s_3305 <=> 1agpc161_mm + s_3317   \n",
       "257            s_3126 + stcoa_erm <=> coa_erm + s_3120   \n",
       "258           nad_c + s_3795 <=> h_c + nadh_c + s_3796   \n",
       "259    argtrna_c + s_3884 <=> h_c + s_3885 + trnaarg_c   \n",
       "260           s_3294 + s_3305 <=> 1agpc161_mm + s_3267   \n",
       "261   h_erm + nadph_erm + s_2941 --> nadp_erm + s_2936   \n",
       "262   h2o_lp + s_3053 --> h_lp + mag180_lp + ocdcea_lp   \n",
       "263                  23ccmp_c + h2o_c --> h_c + s_4100   \n",
       "264                                thmpp_c --> thmpp_m   \n",
       "265  h_c + nadph_c + 2.0 s_3762 <=> nadp_c + 2.0 s_...   \n",
       "266     2hpmhmbq_m + amet_m + h_m --> ahcys_m + s_1535   \n",
       "267           s_3270 + s_3305 <=> 1agpc161_mm + s_3252   \n",
       "268           s_3276 + s_3304 <=> 1agpc160_mm + s_3257   \n",
       "269        h2o_mm + s_3239 --> h_mm + hdca_mm + s_3276   \n",
       "270           s_3291 + s_3304 <=> 1agpc160_mm + s_3317   \n",
       "271           s_3294 + s_3302 <=> 1agpc181_mm + s_3265   \n",
       "272             accoa_c + ser__L_c --> acser_c + coa_c   \n",
       "273           s_3295 + s_3296 <=> 1agpc160_mm + s_3269   \n",
       "274           s_3289 + s_3296 <=> 1agpc160_mm + s_3251   \n",
       "275           s_3289 + s_3305 <=> 1agpc161_mm + s_3316   \n",
       "276                 2.0 atp_c + h2o_c --> h_c + s_4290   \n",
       "277        h2o_mm + s_3263 --> h_mm + hdca_mm + s_3283   \n",
       "278         atp_gm + h_gm + s_3364 --> adp_gm + s_3365   \n",
       "279     h2o_c + ppmi12346p_c --> h_c + minohp_c + pi_c   \n",
       "280         pc161_mm + s_3281 <=> 1agpc161_mm + s_3267   \n",
       "281                                  s_3115 <=> s_3159   \n",
       "282  h_lp + nadph_lp + s_2950 --> 1odecg3p_lp + nad...   \n",
       "283                                  s_3116 <=> s_3161   \n",
       "284           s_3273 + s_3300 <=> 1agpc180_mm + s_3242   \n",
       "285           dolmanp_er + s_0337 <=> dolp_er + s_0339   \n",
       "286           s_3283 + s_3296 <=> 1agpc160_mm + s_3269   \n",
       "287           s_3283 + s_3306 <=> 1agpc180_mm + s_3313   \n",
       "288              2hp6mp_m + o2_m --> 2hp6mbq_m + h2o_m   \n",
       "289  h2o_lp + tag1819Z1619Z1619Z_lp --> h_lp + hdce...   \n",
       "290           s_3273 + s_3306 <=> 1agpc180_mm + s_3254   \n",
       "291           s_3284 + s_3305 <=> 1agpc161_mm + s_3237   \n",
       "292      h2o_vm + s_3419 --> 2.0 h_vm + pi_vm + s_3395   \n",
       "293           s_3279 + s_3305 <=> 1agpc161_mm + s_3309   \n",
       "294      h2o_lp + s_3079 --> h_lp + ocdcea_lp + s_3051   \n",
       "295           s_3284 + s_3306 <=> 1agpc180_mm + s_3237   \n",
       "296                                    pap_c <=> pap_m   \n",
       "297       h2o_mm + s_3248 --> h_mm + ocdca_mm + s_3288   \n",
       "298   2.0 fe3_e + nadph_c <=> 2.0 fe2_c + h_c + nadp_c   \n",
       "299                        cu2_e + h_e <=> cu2_c + h_c   \n",
       "300      h2o_lp + s_3049 --> h_lp + ocdcea_lp + s_3672   \n",
       "301           s_3274 + s_3296 <=> 1agpc160_mm + s_3243   \n",
       "302     pc161_mm + s_3272 <=> 1agpc161_mm + clpn161_mm   \n",
       "303     3mbzalc_c + nad_c <=> 3mbzald_c + h_c + nadh_c   \n",
       "304       h2o_lp + s_3064 --> h_lp + hdcea_lp + s_3053   \n",
       "305           s_3290 + s_3302 <=> 1agpc181_mm + s_3253   \n",
       "306       h2o_lp + s_3063 --> h_lp + hdcea_lp + s_3051   \n",
       "307                                  s_3109 <=> s_3145   \n",
       "308           s_3294 + s_3307 <=> 1agpc181_mm + s_3267   \n",
       "309           s_3291 + s_3307 <=> 1agpc181_mm + s_3317   \n",
       "310       s_3272 + s_3296 <=> 1agpc160_mm + clpn161_mm   \n",
       "311   2.0 h_c + pi_c + ppmi12346p_c <=> h2o_c + s_0263   \n",
       "312        h2o_lp + s_3046 --> h_lp + hdca_lp + s_3045   \n",
       "313         lgnc_er + sphgn_er <=> cer1_24_er + h2o_er   \n",
       "314           s_3286 + s_3307 <=> 1agpc181_mm + s_3243   \n",
       "315           s_3280 + s_3305 <=> 1agpc161_mm + s_3310   \n",
       "316     1agpe181_lp + ocdce9coa_lp --> coa_lp + s_3692   \n",
       "317  dkmpp_c + 3.0 h2o_c --> 4met2obut_c + for_c + ...   \n",
       "318                                      cl_g <=> cl_c   \n",
       "319       s_3286 + s_3302 <=> 1agpc181_mm + clpn161_mm   \n",
       "320                                  s_3109 <=> s_3478   \n",
       "321   3dh5hpb_m + amet_m --> 3hph5mb_m + ahcys_m + h_m   \n",
       "322        gpail_hs_er + pmtcoa_er <=> coa_er + s_0330   \n",
       "323           s_3287 + s_3306 <=> 1agpc180_mm + s_3315   \n",
       "324       h2o_lp + s_3071 --> h_lp + ocdca_lp + s_3051   \n",
       "325         atp_gm + h_gm + s_3366 --> adp_gm + s_3367   \n",
       "326           s_3290 + s_3304 <=> 1agpc160_mm + s_3255   \n",
       "327        h2o_lp + s_3052 --> h_lp + hdca_lp + s_3051   \n",
       "328                                  chol_c <=> chol_e   \n",
       "329           s_3290 + s_3306 <=> 1agpc180_mm + s_3255   \n",
       "330        h2o_mm + s_3256 --> h_mm + hdca_mm + s_3291   \n",
       "331       h2o_mm + s_3236 --> h_mm + ocdca_mm + s_3284   \n",
       "332           s_3274 + s_3307 <=> 1agpc181_mm + s_3255   \n",
       "333     atp_c + h2o_c + h_c --> adp_c + 2.0 h_v + pi_c   \n",
       "334           s_3281 + s_3305 <=> 1agpc161_mm + s_3311   \n",
       "335           s_3293 + s_3302 <=> 1agpc181_mm + s_3263   \n",
       "336                                    h2s_c --> h2s_e   \n",
       "337   h2o_erm + s_3612 --> 2.0 h_erm + pi_erm + s_3588   \n",
       "338           s_3274 + s_3306 <=> 1agpc180_mm + s_3255   \n",
       "339           s_3290 + s_3305 <=> 1agpc161_mm + s_3255   \n",
       "340        h2o_lp + s_3054 --> h_lp + hdca_lp + s_3053   \n",
       "341                                  s_3115 <=> s_3488   \n",
       "342  atp_vm + h_vm + pail161_vm --> adp_vm + pail3p...   \n",
       "343           s_3294 + s_3306 <=> 1agpc180_mm + s_3267   \n",
       "344                                  s_3111 <=> s_3482   \n",
       "345           s_3293 + s_3296 <=> 1agpc160_mm + s_3263   \n",
       "346                 fmnh2_m + s_4295 <=> pi_m + s_4312   \n",
       "347           s_3270 + s_3306 <=> 1agpc180_mm + s_3252   \n",
       "348                        ca2_e + h_e <=> ca2_c + h_c   \n",
       "349                                  s_3112 <=> s_3153   \n",
       "350           dolmanp_er + s_3995 <=> dolp_er + s_3996   \n",
       "351       s_3272 + s_3302 <=> 1agpc181_mm + clpn161_mm   \n",
       "352        h2o_mm + s_3268 --> h_mm + hdca_mm + s_3295   \n",
       "353         hexc_er + sphgn_er <=> cer1_26_er + h2o_er   \n",
       "354           s_3281 + s_3302 <=> 1agpc181_mm + s_3267   \n",
       "355           s_3273 + s_3304 <=> 1agpc160_mm + s_3254   \n",
       "356                             tchola_e --> tchola_c_   \n",
       "357                              lys__L_c --> lys__L_m   \n",
       "358       h2o_mm + s_3251 --> h_mm + ocdca_mm + s_3276   \n",
       "359                                  s_3114 <=> s_3487   \n",
       "360           s_3289 + s_3306 <=> 1agpc180_mm + s_3316   \n",
       "361          amet_c + carn_c <=> ahcys_c + ans_c + h_c   \n",
       "362           s_3285 + s_3304 <=> 1agpc160_mm + s_3314   \n",
       "363       s_3272 + s_3300 <=> 1agpc180_mm + clpn161_mm   \n",
       "364                              tyr__L_v <=> tyr__L_c   \n",
       "365           s_3293 + s_3307 <=> 1agpc181_mm + s_3318   \n",
       "366                  23cump_c + h2o_c --> h_c + s_4081   \n",
       "367           s_3272 + s_3307 <=> 1agpc181_mm + s_3253   \n",
       "368           s_3285 + s_3306 <=> 1agpc180_mm + s_3314   \n",
       "369       s_3286 + s_3300 <=> 1agpc180_mm + clpn161_mm   \n",
       "370           s_3280 + s_3300 <=> 1agpc180_mm + s_3266   \n",
       "371  pail_cho_er + uacgam_er <=> h_er + s_3879 + ud...   \n",
       "372      cer2_26_er + h2o_er --> hexc_er + psphings_er   \n",
       "373         pc161_mm + s_3288 <=> 1agpc161_mm + s_3247   \n",
       "374           s_3286 + s_3305 <=> 1agpc161_mm + s_3243   \n",
       "375                                      cl_e <=> cl_c   \n",
       "376         amet_n + s_3819 <=> ahcys_n + h_n + s_3821   \n",
       "377           s_3282 + s_3305 <=> 1agpc161_mm + s_3312   \n",
       "378                         ser__D_c --> nh4_c + pyr_c   \n",
       "379           s_3270 + s_3302 <=> 1agpc181_mm + s_3240   \n",
       "380  h2o_lp + s_3069 --> 12dgr161_lp + h_lp + ocdca_lp   \n",
       "381      h2o_vm + s_3416 --> 2.0 h_vm + pi_vm + s_3392   \n",
       "382    dolmanp_er + s_4002 <=> dolp_er + h_er + s_4003   \n",
       "383       h2o_mm + s_3242 --> h_mm + ocdca_mm + s_3286   \n",
       "384           s_3284 + s_3307 <=> 1agpc181_mm + s_3237   \n",
       "385  h2o_lp + s_3055 --> h_lp + hdcea_lp + magole_h...   \n",
       "386           s_3290 + s_3300 <=> 1agpc180_mm + s_3253   \n",
       "387           nad_c + s_3789 <=> h_c + nadh_c + s_3790   \n",
       "388                          urea_c <=> h2o_c + s_3851   \n",
       "389  atp_c + h2o_c + s_3865 <=> adp_c + h_c + pi_c ...   \n",
       "390      h2o_vm + s_3423 --> 2.0 h_vm + pi_vm + s_3399   \n",
       "391         atp_vm + h_vm + s_3352 --> adp_vm + s_3397   \n",
       "392  atp_m + h2o_m + s_4022 <=> adp_m + h_m + pi_m ...   \n",
       "393       h2o_lp + s_3059 --> h_lp + hdcea_lp + s_3043   \n",
       "394  atp_c + cd2_c + h2o_c --> adp_c + cd2_e + h_c ...   \n",
       "395           s_3285 + s_3296 <=> 1agpc160_mm + s_3239   \n",
       "396         atp_vm + h_vm + s_3350 --> adp_vm + s_3396   \n",
       "397           s_3276 + s_3307 <=> 1agpc181_mm + s_3257   \n",
       "398       h2o_lp + s_3068 --> h_lp + ocdca_lp + s_3045   \n",
       "399           s_3275 + s_3306 <=> 1agpc180_mm + s_3256   \n",
       "400                                  s_3110 <=> s_3149   \n",
       "401         pc161_mm + s_3293 <=> 1agpc161_mm + s_3263   \n",
       "402  h2o_c + q6_m + spmd_c --> 4abutn_c + s_1526 + ...   \n",
       "403                  h2o_m + s_4291 <=> h_m + 2.0 pi_m   \n",
       "404                  23cgmp_c + h2o_c --> h_c + s_4066   \n",
       "405           s_3278 + s_3306 <=> 1agpc180_mm + s_3308   \n",
       "406           s_3281 + s_3304 <=> 1agpc160_mm + s_3311   \n",
       "407         atp_gm + h_gm + s_3372 --> adp_gm + s_3373   \n",
       "408           s_3289 + s_3300 <=> 1agpc180_mm + s_3251   \n",
       "409           s_3276 + s_3306 <=> 1agpc180_mm + s_3257   \n",
       "410       h2o_lp + s_3073 --> h_lp + ocdca_lp + s_3055   \n",
       "411  2.0 ficytb5_m + nadh_m <=> 2.0 focytb5_m + h_m...   \n",
       "412                      gthrd_c --> 5oxpro_c + cgly_c   \n",
       "413       s_3286 + s_3296 <=> 1agpc160_mm + clpn161_mm   \n",
       "414           s_3275 + s_3302 <=> 1agpc181_mm + s_3244   \n",
       "415           s_3276 + s_3296 <=> 1agpc160_mm + s_3245   \n",
       "416            L_alagly_v + h2o_v <=> ala__L_v + gly_v   \n",
       "417  atp_gm + h_gm + pail161_gm --> adp_gm + pail4p...   \n",
       "418           s_3288 + s_3305 <=> 1agpc161_mm + s_3249   \n",
       "419     atp_c + h2o_c + h_c --> adp_c + 2.0 h_g + pi_c   \n",
       "420            s_3127 + stcoa_erm <=> coa_erm + s_3124   \n",
       "421           s_3294 + s_3300 <=> 1agpc180_mm + s_3265   \n",
       "422         pc161_mm + s_3292 <=> 1agpc161_mm + s_3259   \n",
       "423        h2o_mm + s_3237 --> h_mm + hdca_mm + s_3274   \n",
       "424                     dolmanp_er + s_0339 <=> s_0338   \n",
       "425  12dgr161_lp + h2o_lp --> h_lp + hdcea_lp + s_3672   \n",
       "426         pc161_mm + s_3275 <=> 1agpc161_mm + s_3244   \n",
       "427           s_3279 + s_3307 <=> 1agpc181_mm + s_3309   \n",
       "428                              leu__L_v <=> leu__L_c   \n",
       "429           h2o_c + s_0333 --> h_c + minohp_c + pi_c   \n",
       "430                                 chol_ce <=> chol_c   \n",
       "431     1agpe180_lp + ocdce9coa_lp --> coa_lp + s_3690   \n",
       "432                  h2o_n + s_4292 <=> h_n + 2.0 pi_n   \n",
       "433           s_3292 + s_3296 <=> 1agpc160_mm + s_3259   \n",
       "434  h2o_lp + s_3074 --> 12dgr181_lp + h_lp + ocdca_lp   \n",
       "435        h2o_mm + s_3260 --> h_mm + hdca_mm + s_3280   \n",
       "436         pc161_mm + s_3295 <=> 1agpc161_mm + s_3269   \n",
       "437                    h2o_c + pppi_c --> pi_c + ppi_c   \n",
       "438      h2o_lp + s_3076 --> h_lp + ocdcea_lp + s_3045   \n",
       "439        h2o_mm + s_3258 --> h_mm + hdca_mm + s_3292   \n",
       "440  atp_c + hco3_c + thr__L_c --> h2o_c + ppi_c + ...   \n",
       "441                             uacgam_er <=> uacgam_g   \n",
       "442               amet_c + s_3876 <=> ahcys_c + s_3877   \n",
       "443   2hp6mbq_m + amet_m --> 2hpmmbq_m + ahcys_m + h_m   \n",
       "444       h2o_lp + s_3062 --> h_lp + hdcea_lp + s_3049   \n",
       "445           s_3295 + s_3300 <=> 1agpc180_mm + s_3269   \n",
       "446  atp_c + nh4_c + s_3940 --> amp_c + h_c + ppi_c...   \n",
       "447           s_3291 + s_3302 <=> 1agpc181_mm + s_3257   \n",
       "448           s_3287 + s_3302 <=> 1agpc181_mm + s_3245   \n",
       "449      cer2_24_er + h2o_er --> lgnc_er + psphings_er   \n",
       "450  h2o_lp + tag1619Z1619Z160_lp --> 12dgr161_lp +...   \n",
       "451                                    zn2_m <=> zn2_v   \n",
       "452       h2o_mm + s_3247 --> h_mm + ocdca_mm + s_3272   \n",
       "453      h2o_vm + s_3421 --> 2.0 h_vm + pi_vm + s_3397   \n",
       "454    h2o_lp + s_3658 --> ergst_lp + h_lp + ocdcea_lp   \n",
       "455        h2o_mm + s_3240 --> h_mm + hdca_mm + s_3286   \n",
       "456       h2o_lp + s_3072 --> h_lp + ocdca_lp + s_3053   \n",
       "\n",
       "                                                     3  \\\n",
       "0                                              YGR110W   \n",
       "1    (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "2                                              YGR110W   \n",
       "3                                              YPR140W   \n",
       "4                                              YPR140W   \n",
       "5                                              YPR140W   \n",
       "6             YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "7                                              YPR140W   \n",
       "8             YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "9                                              YIL124W   \n",
       "10                                             YIL023C   \n",
       "11                                  YMR251W or YKR076W   \n",
       "12                                             YIL124W   \n",
       "13            YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "14                                  YGL255W or YLR130C   \n",
       "15            YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "16                                             YPR140W   \n",
       "17                                             YPR140W   \n",
       "18            YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "19                                             YPR140W   \n",
       "20                                             YPR140W   \n",
       "21                                             YPR140W   \n",
       "22                                  YBR147W or YDR352W   \n",
       "23                                             YCR011C   \n",
       "24                                  YKL140W or YLL012W   \n",
       "25   (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "26                                             YPR140W   \n",
       "27   YCR107W or YDL243C or YJR155W or YNL331C or YO...   \n",
       "28                                             YIL124W   \n",
       "29   YBR204C or YKR089C or YMR313C or YOR081C or YC...   \n",
       "30                                             YGR110W   \n",
       "31                                             YPR140W   \n",
       "32                                             YPR140W   \n",
       "33                                             YDL219W   \n",
       "34                                             YPR140W   \n",
       "35                                             YGR110W   \n",
       "36                                             YMR313C   \n",
       "37                                             YPR140W   \n",
       "38            YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "39                                             YPR140W   \n",
       "40                                 YDR538W and YDR539W   \n",
       "41                                             YPR140W   \n",
       "42                                             YPR140W   \n",
       "43   YDR204W and YGR255C and YLR201C and YML110C an...   \n",
       "44                                             YGR110W   \n",
       "45                                             YPR140W   \n",
       "46                                  YHR201C or YDR452W   \n",
       "47                                             YOL065C   \n",
       "48                                             YPR140W   \n",
       "49                                  YIL048W or YMR162C   \n",
       "50                                  YDR371W or YLR286C   \n",
       "51                                             YLR011W   \n",
       "52                                             YPR140W   \n",
       "53                                             YMR099C   \n",
       "54                                             YPR140W   \n",
       "55                                             YGL038C   \n",
       "56                                             YPR140W   \n",
       "57                                             YPR140W   \n",
       "58                                             YPR140W   \n",
       "59                                             YIR038C   \n",
       "60                                             YLR011W   \n",
       "61                                             YKR076W   \n",
       "62                                 YBR070C and YGL047W   \n",
       "63   YMR296C and YKL212W and YDR062W and YGR038W an...   \n",
       "64                       YOR251C or YOR285W or YOR286W   \n",
       "65   YCR107W or YDL243C or YJR155W or YNL331C or YO...   \n",
       "66                                             YPR140W   \n",
       "67                                             YPR140W   \n",
       "68                                             YPR140W   \n",
       "69                                             YBR147W   \n",
       "70                       YAL035W or YLL001W or YER005W   \n",
       "71                                             YPR140W   \n",
       "72                                             YPR140W   \n",
       "73                                 YDR373W and YNL267W   \n",
       "74                                             YPR140W   \n",
       "75   (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "76                                             YPR140W   \n",
       "77   YBR199W or YBR205W or YDR483W or YJL139C or YK...   \n",
       "78                                             YPR140W   \n",
       "79                                             YPR140W   \n",
       "80                                             YPR140W   \n",
       "81                                             YPR140W   \n",
       "82                                             YPR140W   \n",
       "83                                             YGR110W   \n",
       "84                                             YPR140W   \n",
       "85                                             YPR140W   \n",
       "86                                             YPR140W   \n",
       "87                                  YIL048W or YMR162C   \n",
       "88                                             YPR140W   \n",
       "89   YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "90                                             YGR036C   \n",
       "91                                             YOL130W   \n",
       "92                                             YIL124W   \n",
       "93                                             YBR147W   \n",
       "94                                             YMR313C   \n",
       "95   (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "96                                             YGR110W   \n",
       "97   YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "98   YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "99                                             YPR140W   \n",
       "100                                            YPR140W   \n",
       "101                                            YPR140W   \n",
       "102                                 YBR147W or YDR352W   \n",
       "103                                            YLR410W   \n",
       "104                                            YPR140W   \n",
       "105                                            YMR313C   \n",
       "106                                            YPR140W   \n",
       "107                                            YPR140W   \n",
       "108                                            YPR140W   \n",
       "109                                            YPR140W   \n",
       "110                                            YPR140W   \n",
       "111                                            YPR140W   \n",
       "112                                            YGR110W   \n",
       "113                                            YKL165C   \n",
       "114                                            YGR110W   \n",
       "115                                            YPR140W   \n",
       "116                                            YMR313C   \n",
       "117                                            YPR140W   \n",
       "118                                YDL090C and YKL019W   \n",
       "119                                            YPR140W   \n",
       "120                                            YPR140W   \n",
       "121                                            YPR140W   \n",
       "122                                            YPR140W   \n",
       "123           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "124                                            YIL048W   \n",
       "125                                            YPR140W   \n",
       "126                                            YPR140W   \n",
       "127           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "128                                            YIL048W   \n",
       "129                                            YPR140W   \n",
       "130                                            YPR140W   \n",
       "131                                            YPR140W   \n",
       "132                                            YKL103C   \n",
       "133                                            YOL065C   \n",
       "134                                            YGR110W   \n",
       "135           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "136                                            YIL166C   \n",
       "137                                            YGR007W   \n",
       "138                                            YMR281W   \n",
       "139                                            YPR140W   \n",
       "140                                            YPR140W   \n",
       "141                                YER061C and YKL192C   \n",
       "142                                YER061C and YKL192C   \n",
       "143                                            YIL124W   \n",
       "144                                            YGR110W   \n",
       "145                                            YPR140W   \n",
       "146                                            YPR140W   \n",
       "147                                            YPR140W   \n",
       "148                                            YPR140W   \n",
       "149                                            YOL065C   \n",
       "150                                YDR373W and YNL267W   \n",
       "151                                            YPR140W   \n",
       "152                                            YGR110W   \n",
       "153                                            YPR140W   \n",
       "154  YDL042C or YDR191W or YOL068C or YOR025W or YP...   \n",
       "155                                            YPR140W   \n",
       "156           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "157                                 YGR191W or YOL130W   \n",
       "158                                            YDL219W   \n",
       "159                                            YIL048W   \n",
       "160                                            YPR140W   \n",
       "161                                            YPR140W   \n",
       "162                                            YMR162C   \n",
       "163                                            YPR140W   \n",
       "164                                 YEL066W or YPR193C   \n",
       "165                                            YPR140W   \n",
       "166                                            YPR140W   \n",
       "167                                YER061C and YKL192C   \n",
       "168                                            YGR247W   \n",
       "169                                            YPR140W   \n",
       "170                                            YPR140W   \n",
       "171                                            YPR140W   \n",
       "172                                            YOL065C   \n",
       "173                                            YPR140W   \n",
       "174                                            YBL080C   \n",
       "175                                            YGR110W   \n",
       "176                                            YGR110W   \n",
       "177                                 YIL048W or YMR162C   \n",
       "178                                            YGR110W   \n",
       "179  (Q0045 and Q0250 and Q0275 and YDL067C and YEL...   \n",
       "180                                            YPR140W   \n",
       "181                                            YDL120W   \n",
       "182                                            YPR140W   \n",
       "183                                            YGR110W   \n",
       "184                                            YPR140W   \n",
       "185                                            YEL031W   \n",
       "186                                            YPR140W   \n",
       "187           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "188                                            YPR140W   \n",
       "189                                            YGR208W   \n",
       "190                                            YPR140W   \n",
       "191  YCR107W or YDL243C or YJR155W or YNL331C or YO...   \n",
       "192                                            YNL219C   \n",
       "193                                YDR373W and YNL267W   \n",
       "194                                            YPR140W   \n",
       "195                                            YPR140W   \n",
       "196                                            YGR110W   \n",
       "197                                            YGR110W   \n",
       "198                                            YPR140W   \n",
       "199                                            YPR140W   \n",
       "200                                            YIL124W   \n",
       "201                                            YMR313C   \n",
       "202           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "203                                            YPR140W   \n",
       "204           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "205                                            YPR140W   \n",
       "206                                            YPR140W   \n",
       "207                                            YCL038C   \n",
       "208  YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "209                                 YLR410W or YOR163W   \n",
       "210                                            YIL048W   \n",
       "211                                 YDR302W or YLL031C   \n",
       "212                                            YPR140W   \n",
       "213                                            YOL065C   \n",
       "214                                 YHR201C or YDR452W   \n",
       "215                                            YPR140W   \n",
       "216                                            YPR140W   \n",
       "217                                            YBR235W   \n",
       "218                                            YPR140W   \n",
       "219                                            YPR140W   \n",
       "220                                 YBL091C or YLR244C   \n",
       "221                                            YPR140W   \n",
       "222                                            YGR110W   \n",
       "223                                YDR373W and YNL267W   \n",
       "224           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "225                                            YOL065C   \n",
       "226                                            YPR140W   \n",
       "227                                            YOL065C   \n",
       "228                                            YPR140W   \n",
       "229                                            YPR140W   \n",
       "230                                            YPR140W   \n",
       "231                                            YPR140W   \n",
       "232                                            YGR110W   \n",
       "233                                 YBR147W or YDR352W   \n",
       "234                                            YGR110W   \n",
       "235                                            YPR140W   \n",
       "236                                            YPR140W   \n",
       "237           YDR533C or YMR322C or YOR391C or YPL280W   \n",
       "238                                            YPR140W   \n",
       "239                                            YMR313C   \n",
       "240                                            YPR140W   \n",
       "241                                            YPR140W   \n",
       "242                                            YPR140W   \n",
       "243                                            YPR140W   \n",
       "244                                            YGR110W   \n",
       "245                                            YGR110W   \n",
       "246                                            YPR140W   \n",
       "247                                 YOR334W or YPL060W   \n",
       "248                                            YPR140W   \n",
       "249                                            YPR140W   \n",
       "250                                            YPR140W   \n",
       "251                                            YOR163W   \n",
       "252                                            YPR140W   \n",
       "253                                            YPR140W   \n",
       "254                                            YPR140W   \n",
       "255  (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "256                                            YPR140W   \n",
       "257                                            YBR042C   \n",
       "258  YCR107W or YDL243C or YJR155W or YNL331C or YO...   \n",
       "259                                            YGL017W   \n",
       "260                                            YPR140W   \n",
       "261                                            YIL124W   \n",
       "262                                            YMR313C   \n",
       "263                                            YGR247W   \n",
       "264                                            YGR096W   \n",
       "265                                            YBR046C   \n",
       "266  YDR204W and YGR255C and YLR201C and YML110C an...   \n",
       "267                                            YPR140W   \n",
       "268                                            YPR140W   \n",
       "269                                            YGR110W   \n",
       "270                                            YPR140W   \n",
       "271                                            YPR140W   \n",
       "272                    YHR013C and YDL040C and YOR253W   \n",
       "273                                            YPR140W   \n",
       "274                                            YPR140W   \n",
       "275                                            YPR140W   \n",
       "276                                            YJL012C   \n",
       "277                                            YGR110W   \n",
       "278                                YDR373W and YNL267W   \n",
       "279                                            YOR163W   \n",
       "280                                            YPR140W   \n",
       "281                                 YIL048W or YMR162C   \n",
       "282                                            YIL124W   \n",
       "283                                 YIL048W or YMR162C   \n",
       "284                                            YPR140W   \n",
       "285                                            YBR004C   \n",
       "286                                            YPR140W   \n",
       "287                                            YPR140W   \n",
       "288  YDR204W and YGR255C and YLR201C and YML110C an...   \n",
       "289           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "290                                            YPR140W   \n",
       "291                                            YPR140W   \n",
       "292  YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "293                                            YPR140W   \n",
       "294           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "295                                            YPR140W   \n",
       "296                                            YPR011C   \n",
       "297                                            YGR110W   \n",
       "298  YLR047C or YKL220C or YLR214W or YNR060W or YO...   \n",
       "299                                            YOL122C   \n",
       "300                                            YMR313C   \n",
       "301                                            YPR140W   \n",
       "302                                            YPR140W   \n",
       "303  YCR107W or YDL243C or YJR155W or YNL331C or YO...   \n",
       "304           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "305                                            YPR140W   \n",
       "306           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "307                                 YIL048W or YMR162C   \n",
       "308                                            YPR140W   \n",
       "309                                            YPR140W   \n",
       "310                                            YPR140W   \n",
       "311                                            YLR410W   \n",
       "312           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "313                                            YPL087W   \n",
       "314                                            YPR140W   \n",
       "315                                            YPR140W   \n",
       "316                                            YMR313C   \n",
       "317                                YEL038W and YMR009W   \n",
       "318                                            YJR040W   \n",
       "319                                            YPR140W   \n",
       "320                                            YIL048W   \n",
       "321  YDR204W and YGR255C and YLR201C and YML110C an...   \n",
       "322                                            YJL091C   \n",
       "323                                            YPR140W   \n",
       "324           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "325                                YDR373W and YNL267W   \n",
       "326                                            YPR140W   \n",
       "327           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "328                                            YOR161C   \n",
       "329                                            YPR140W   \n",
       "330                                            YGR110W   \n",
       "331                                            YGR110W   \n",
       "332                                            YPR140W   \n",
       "333  (YBR127C and YDL185W and YEL027W and YEL051W a...   \n",
       "334                                            YPR140W   \n",
       "335                                            YPR140W   \n",
       "336                                            YIL166C   \n",
       "337                                            YOL065C   \n",
       "338                                            YPR140W   \n",
       "339                                            YPR140W   \n",
       "340           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "341                                            YIL048W   \n",
       "342  (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "343                                            YPR140W   \n",
       "344                                            YIL048W   \n",
       "345                                            YPR140W   \n",
       "346                                            YDR538W   \n",
       "347                                            YPR140W   \n",
       "348                                            YOL122C   \n",
       "349                                 YIL048W or YMR162C   \n",
       "350                                            YNL219C   \n",
       "351                                            YPR140W   \n",
       "352                                            YGR110W   \n",
       "353                                            YPL087W   \n",
       "354                                            YPR140W   \n",
       "355                                            YPR140W   \n",
       "356                                            YIL166C   \n",
       "357                                            YBR147W   \n",
       "358                                            YGR110W   \n",
       "359                                            YIL048W   \n",
       "360                                            YPR140W   \n",
       "361                                            YNL092W   \n",
       "362                                            YPR140W   \n",
       "363                                            YPR140W   \n",
       "364                                            YCL038C   \n",
       "365                                            YPR140W   \n",
       "366                                            YGR247W   \n",
       "367                                            YPR140W   \n",
       "368                                            YPR140W   \n",
       "369                                            YPR140W   \n",
       "370                                            YPR140W   \n",
       "371  YDR437W and YGR216C and YNL038W and YPL076W an...   \n",
       "372                                            YPL087W   \n",
       "373                                            YPR140W   \n",
       "374                                            YPR140W   \n",
       "375                                            YNL275W   \n",
       "376                      YHR119W or YDR440W or YJL168C   \n",
       "377                                            YPR140W   \n",
       "378                                            YGL196W   \n",
       "379                                            YPR140W   \n",
       "380           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "381  YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "382                                            YOR149C   \n",
       "383                                            YGR110W   \n",
       "384                                            YPR140W   \n",
       "385                                            YMR313C   \n",
       "386                                            YPR140W   \n",
       "387  YCR107W or YDL243C or YJR155W or YNL331C or YO...   \n",
       "388                                 YFL061W or YNL335W   \n",
       "389                                 YDR051C or YOL075C   \n",
       "390  YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "391  (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "392                                            YMR301C   \n",
       "393           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "394                                            YBR295W   \n",
       "395                                            YPR140W   \n",
       "396  (YBR097W and YLR240W) or (YBR097W and YLR240W ...   \n",
       "397                                            YPR140W   \n",
       "398           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "399                                            YPR140W   \n",
       "400                                 YIL048W or YMR162C   \n",
       "401                                            YPR140W   \n",
       "402                                            YHR068W   \n",
       "403                                 YHR201C or YDR452W   \n",
       "404                                            YGR247W   \n",
       "405                                            YPR140W   \n",
       "406                                            YPR140W   \n",
       "407                                YDR373W and YNL267W   \n",
       "408                                            YPR140W   \n",
       "409                                            YPR140W   \n",
       "410           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "411                      YIL043C or YKL150W or YML125C   \n",
       "412                                            YER163C   \n",
       "413                                            YPR140W   \n",
       "414                                            YPR140W   \n",
       "415                                            YPR140W   \n",
       "416                                            YKL103C   \n",
       "417                                YDR373W and YNL267W   \n",
       "418                                            YPR140W   \n",
       "419  (YPR170W-B and YBR127C and YDL185W and YOR332W...   \n",
       "420                                            YBR042C   \n",
       "421                                            YPR140W   \n",
       "422                                            YPR140W   \n",
       "423                                            YGR110W   \n",
       "424                                            YGL142C   \n",
       "425                                 YCR068W or YMR313C   \n",
       "426                                            YPR140W   \n",
       "427                                            YPR140W   \n",
       "428                                            YCL038C   \n",
       "429                                 YLR410W or YOR163W   \n",
       "430                                 YGL077C or YOR161C   \n",
       "431                                            YMR313C   \n",
       "432                                 YHR201C or YDR452W   \n",
       "433                                            YPR140W   \n",
       "434           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "435                                            YGR110W   \n",
       "436                                            YPR140W   \n",
       "437                                 YHR201C or YDR452W   \n",
       "438           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "439                                            YGR110W   \n",
       "440                                            YGL169W   \n",
       "441                                            YEL004W   \n",
       "442                                            YDR410C   \n",
       "443  YDR204W and YGR255C and YLR201C and YML110C an...   \n",
       "444           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "445                                            YPR140W   \n",
       "446                                            YLR143W   \n",
       "447                                            YPR140W   \n",
       "448                                            YPR140W   \n",
       "449                                            YPL087W   \n",
       "450           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "451                                 YMR243C or YOR316C   \n",
       "452                                            YGR110W   \n",
       "453  YFR019W and YNL325C and YFR021W and YNL054W an...   \n",
       "454                                 YKL140W or YLL012W   \n",
       "455                                            YGR110W   \n",
       "456           YBR204C or YKR089C or YMR313C or YOR081C   \n",
       "\n",
       "                                                     4  \n",
       "0                                      Lipst1_1_102653  \n",
       "1     Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "2                                      Lipst1_1_102653  \n",
       "3                                       Lipst1_1_68925  \n",
       "4                                       Lipst1_1_68925  \n",
       "5                                       Lipst1_1_68925  \n",
       "6    Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "7                                       Lipst1_1_68925  \n",
       "8    Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "9                                       Lipst1_1_71724  \n",
       "10                                      Lipst1_1_69058  \n",
       "11                                     Lipst1_1_308673  \n",
       "12                                      Lipst1_1_71724  \n",
       "13   Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "14                     Lipst1_1_71180 or Lipst1_1_2861  \n",
       "15   Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "16                                      Lipst1_1_68925  \n",
       "17                                      Lipst1_1_68925  \n",
       "18   Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "19                                      Lipst1_1_68925  \n",
       "20                                      Lipst1_1_68925  \n",
       "21                                      Lipst1_1_68925  \n",
       "22                   Lipst1_1_211414 or Lipst1_1_62919  \n",
       "23                                     Lipst1_1_235077  \n",
       "24                                       Lipst1_1_1895  \n",
       "25    Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "26                                      Lipst1_1_68925  \n",
       "27   Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...  \n",
       "28                                      Lipst1_1_71724  \n",
       "29   Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "30                                     Lipst1_1_102653  \n",
       "31                                      Lipst1_1_68925  \n",
       "32                                      Lipst1_1_68925  \n",
       "33                                      Lipst1_1_50667  \n",
       "34                                      Lipst1_1_68925  \n",
       "35                                     Lipst1_1_102653  \n",
       "36                                     Lipst1_1_161356  \n",
       "37                                      Lipst1_1_68925  \n",
       "38   Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "39                                      Lipst1_1_68925  \n",
       "40                   Lipst1_1_198908 or Lipst1_1_76260  \n",
       "41                                      Lipst1_1_68925  \n",
       "42                                      Lipst1_1_68925  \n",
       "43   Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...  \n",
       "44                                     Lipst1_1_102653  \n",
       "45                                      Lipst1_1_68925  \n",
       "46                                      Lipst1_1_72748  \n",
       "47                                     Lipst1_1_110448  \n",
       "48                                      Lipst1_1_68925  \n",
       "49                      Lipst1_1_49943 or Lipst1_1_984  \n",
       "50                   Lipst1_1_102785 or Lipst1_1_42376  \n",
       "51                                      Lipst1_1_52431  \n",
       "52                                      Lipst1_1_68925  \n",
       "53                                      Lipst1_1_68948  \n",
       "54                                      Lipst1_1_68925  \n",
       "55                                      Lipst1_1_37797  \n",
       "56                                      Lipst1_1_68925  \n",
       "57                                      Lipst1_1_68925  \n",
       "58                                      Lipst1_1_68925  \n",
       "59                                      Lipst1_1_74576  \n",
       "60                                      Lipst1_1_52431  \n",
       "61                                     Lipst1_1_308673  \n",
       "62                     Lipst1_1_5477 or Lipst1_1_74719  \n",
       "63   Lipst1_1_71191 or Lipst1_1_72156 or Lipst1_1_1...  \n",
       "64                      Lipst1_1_965 or Lipst1_1_60674  \n",
       "65   Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...  \n",
       "66                                      Lipst1_1_68925  \n",
       "67                                      Lipst1_1_68925  \n",
       "68                                      Lipst1_1_68925  \n",
       "69                                     Lipst1_1_211414  \n",
       "70   Lipst1_1_124163 or Lipst1_1_73932 or Lipst1_1_...  \n",
       "71                                      Lipst1_1_68925  \n",
       "72                                      Lipst1_1_68925  \n",
       "73                    Lipst1_1_2990 or Lipst1_1_218588  \n",
       "74                                      Lipst1_1_68925  \n",
       "75    Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "76                                      Lipst1_1_68925  \n",
       "77     Lipst1_1_384 or Lipst1_1_4409 or Lipst1_1_67994  \n",
       "78                                      Lipst1_1_68925  \n",
       "79                                      Lipst1_1_68925  \n",
       "80                                      Lipst1_1_68925  \n",
       "81                                      Lipst1_1_68925  \n",
       "82                                      Lipst1_1_68925  \n",
       "83                                     Lipst1_1_102653  \n",
       "84                                      Lipst1_1_68925  \n",
       "85                                      Lipst1_1_68925  \n",
       "86                                      Lipst1_1_68925  \n",
       "87                      Lipst1_1_49943 or Lipst1_1_984  \n",
       "88                                      Lipst1_1_68925  \n",
       "89   Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "90                                      Lipst1_1_52818  \n",
       "91                                     Lipst1_1_141125  \n",
       "92                                      Lipst1_1_71724  \n",
       "93                                     Lipst1_1_211414  \n",
       "94                                     Lipst1_1_161356  \n",
       "95    Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "96                                     Lipst1_1_102653  \n",
       "97   Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "98   Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "99                                      Lipst1_1_68925  \n",
       "100                                     Lipst1_1_68925  \n",
       "101                                     Lipst1_1_68925  \n",
       "102                  Lipst1_1_211414 or Lipst1_1_62919  \n",
       "103                                     Lipst1_1_57858  \n",
       "104                                     Lipst1_1_68925  \n",
       "105                                    Lipst1_1_161356  \n",
       "106                                     Lipst1_1_68925  \n",
       "107                                     Lipst1_1_68925  \n",
       "108                                     Lipst1_1_68925  \n",
       "109                                     Lipst1_1_68925  \n",
       "110                                     Lipst1_1_68925  \n",
       "111                                     Lipst1_1_68925  \n",
       "112                                    Lipst1_1_102653  \n",
       "113                                     Lipst1_1_70311  \n",
       "114                                    Lipst1_1_102653  \n",
       "115                                     Lipst1_1_68925  \n",
       "116                                    Lipst1_1_161356  \n",
       "117                                     Lipst1_1_68925  \n",
       "118                   Lipst1_1_302018 or Lipst1_1_4158  \n",
       "119                                     Lipst1_1_68925  \n",
       "120                                     Lipst1_1_68925  \n",
       "121                                     Lipst1_1_68925  \n",
       "122                                     Lipst1_1_68925  \n",
       "123  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "124                                     Lipst1_1_49943  \n",
       "125                                     Lipst1_1_68925  \n",
       "126                                     Lipst1_1_68925  \n",
       "127  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "128                                     Lipst1_1_49943  \n",
       "129                                     Lipst1_1_68925  \n",
       "130                                     Lipst1_1_68925  \n",
       "131                                     Lipst1_1_68925  \n",
       "132                                    Lipst1_1_106140  \n",
       "133                                    Lipst1_1_110448  \n",
       "134                                    Lipst1_1_102653  \n",
       "135  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "136                                      Lipst1_1_6611  \n",
       "137                                      Lipst1_1_4188  \n",
       "138                                    Lipst1_1_294677  \n",
       "139                                     Lipst1_1_68925  \n",
       "140                                     Lipst1_1_68925  \n",
       "141                     Lipst1_1_3273 or Lipst1_1_2955  \n",
       "142                     Lipst1_1_3273 or Lipst1_1_2955  \n",
       "143                                     Lipst1_1_71724  \n",
       "144                                    Lipst1_1_102653  \n",
       "145                                     Lipst1_1_68925  \n",
       "146                                     Lipst1_1_68925  \n",
       "147                                     Lipst1_1_68925  \n",
       "148                                     Lipst1_1_68925  \n",
       "149                                    Lipst1_1_110448  \n",
       "150                   Lipst1_1_2990 or Lipst1_1_218588  \n",
       "151                                     Lipst1_1_68925  \n",
       "152                                    Lipst1_1_102653  \n",
       "153                                     Lipst1_1_68925  \n",
       "154  Lipst1_1_336196 or Lipst1_1_2970 or Lipst1_1_7...  \n",
       "155                                     Lipst1_1_68925  \n",
       "156  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "157                                    Lipst1_1_141125  \n",
       "158                                     Lipst1_1_50667  \n",
       "159                                     Lipst1_1_49943  \n",
       "160                                     Lipst1_1_68925  \n",
       "161                                     Lipst1_1_68925  \n",
       "162                                       Lipst1_1_984  \n",
       "163                                     Lipst1_1_68925  \n",
       "164                                     Lipst1_1_48516  \n",
       "165                                     Lipst1_1_68925  \n",
       "166                                     Lipst1_1_68925  \n",
       "167                     Lipst1_1_3273 or Lipst1_1_2955  \n",
       "168                                     Lipst1_1_72954  \n",
       "169                                     Lipst1_1_68925  \n",
       "170                                     Lipst1_1_68925  \n",
       "171                                     Lipst1_1_68925  \n",
       "172                                    Lipst1_1_110448  \n",
       "173                                     Lipst1_1_68925  \n",
       "174                                    Lipst1_1_110148  \n",
       "175                                    Lipst1_1_102653  \n",
       "176                                    Lipst1_1_102653  \n",
       "177                     Lipst1_1_49943 or Lipst1_1_984  \n",
       "178                                    Lipst1_1_102653  \n",
       "179  Lipst1_1_3895 or Lipst1_1_65274 or Lipst1_1_65...  \n",
       "180                                     Lipst1_1_68925  \n",
       "181                                     Lipst1_1_69822  \n",
       "182                                     Lipst1_1_68925  \n",
       "183                                    Lipst1_1_102653  \n",
       "184                                     Lipst1_1_68925  \n",
       "185                                     Lipst1_1_75754  \n",
       "186                                     Lipst1_1_68925  \n",
       "187  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "188                                     Lipst1_1_68925  \n",
       "189                                      Lipst1_1_1740  \n",
       "190                                     Lipst1_1_68925  \n",
       "191  Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...  \n",
       "192                                     Lipst1_1_91236  \n",
       "193                   Lipst1_1_2990 or Lipst1_1_218588  \n",
       "194                                     Lipst1_1_68925  \n",
       "195                                     Lipst1_1_68925  \n",
       "196                                    Lipst1_1_102653  \n",
       "197                                    Lipst1_1_102653  \n",
       "198                                     Lipst1_1_68925  \n",
       "199                                     Lipst1_1_68925  \n",
       "200                                     Lipst1_1_71724  \n",
       "201                                    Lipst1_1_161356  \n",
       "202  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "203                                     Lipst1_1_68925  \n",
       "204  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "205                                     Lipst1_1_68925  \n",
       "206                                     Lipst1_1_68925  \n",
       "207                                    Lipst1_1_106989  \n",
       "208  Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "209                  Lipst1_1_57858 or Lipst1_1_105766  \n",
       "210                                     Lipst1_1_49943  \n",
       "211                  Lipst1_1_141087 or Lipst1_1_38480  \n",
       "212                                     Lipst1_1_68925  \n",
       "213                                    Lipst1_1_110448  \n",
       "214                                     Lipst1_1_72748  \n",
       "215                                     Lipst1_1_68925  \n",
       "216                                     Lipst1_1_68925  \n",
       "217                                    Lipst1_1_118138  \n",
       "218                                     Lipst1_1_68925  \n",
       "219                                     Lipst1_1_68925  \n",
       "220                    Lipst1_1_338327 or Lipst1_1_904  \n",
       "221                                     Lipst1_1_68925  \n",
       "222                                    Lipst1_1_102653  \n",
       "223                   Lipst1_1_2990 or Lipst1_1_218588  \n",
       "224  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "225                                    Lipst1_1_110448  \n",
       "226                                     Lipst1_1_68925  \n",
       "227                                    Lipst1_1_110448  \n",
       "228                                     Lipst1_1_68925  \n",
       "229                                     Lipst1_1_68925  \n",
       "230                                     Lipst1_1_68925  \n",
       "231                                     Lipst1_1_68925  \n",
       "232                                    Lipst1_1_102653  \n",
       "233                  Lipst1_1_211414 or Lipst1_1_62919  \n",
       "234                                    Lipst1_1_102653  \n",
       "235                                     Lipst1_1_68925  \n",
       "236                                     Lipst1_1_68925  \n",
       "237                                      Lipst1_1_3208  \n",
       "238                                     Lipst1_1_68925  \n",
       "239                                    Lipst1_1_161356  \n",
       "240                                     Lipst1_1_68925  \n",
       "241                                     Lipst1_1_68925  \n",
       "242                                     Lipst1_1_68925  \n",
       "243                                     Lipst1_1_68925  \n",
       "244                                    Lipst1_1_102653  \n",
       "245                                    Lipst1_1_102653  \n",
       "246                                     Lipst1_1_68925  \n",
       "247                                      Lipst1_1_1637  \n",
       "248                                     Lipst1_1_68925  \n",
       "249                                     Lipst1_1_68925  \n",
       "250                                     Lipst1_1_68925  \n",
       "251                                    Lipst1_1_105766  \n",
       "252                                     Lipst1_1_68925  \n",
       "253                                     Lipst1_1_68925  \n",
       "254                                     Lipst1_1_68925  \n",
       "255   Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "256                                     Lipst1_1_68925  \n",
       "257                                     Lipst1_1_68496  \n",
       "258  Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...  \n",
       "259                                     Lipst1_1_67104  \n",
       "260                                     Lipst1_1_68925  \n",
       "261                                     Lipst1_1_71724  \n",
       "262                                    Lipst1_1_161356  \n",
       "263                                     Lipst1_1_72954  \n",
       "264                                      Lipst1_1_5945  \n",
       "265                                      Lipst1_1_6135  \n",
       "266  Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...  \n",
       "267                                     Lipst1_1_68925  \n",
       "268                                     Lipst1_1_68925  \n",
       "269                                    Lipst1_1_102653  \n",
       "270                                     Lipst1_1_68925  \n",
       "271                                     Lipst1_1_68925  \n",
       "272                   Lipst1_1_4043 or Lipst1_1_164418  \n",
       "273                                     Lipst1_1_68925  \n",
       "274                                     Lipst1_1_68925  \n",
       "275                                     Lipst1_1_68925  \n",
       "276                                    Lipst1_1_106087  \n",
       "277                                    Lipst1_1_102653  \n",
       "278                   Lipst1_1_2990 or Lipst1_1_218588  \n",
       "279                                    Lipst1_1_105766  \n",
       "280                                     Lipst1_1_68925  \n",
       "281                     Lipst1_1_49943 or Lipst1_1_984  \n",
       "282                                     Lipst1_1_71724  \n",
       "283                     Lipst1_1_49943 or Lipst1_1_984  \n",
       "284                                     Lipst1_1_68925  \n",
       "285                                     Lipst1_1_74168  \n",
       "286                                     Lipst1_1_68925  \n",
       "287                                     Lipst1_1_68925  \n",
       "288  Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...  \n",
       "289  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "290                                     Lipst1_1_68925  \n",
       "291                                     Lipst1_1_68925  \n",
       "292  Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "293                                     Lipst1_1_68925  \n",
       "294  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "295                                     Lipst1_1_68925  \n",
       "296                                    Lipst1_1_180684  \n",
       "297                                    Lipst1_1_102653  \n",
       "298  Lipst1_1_68023 or Lipst1_1_64448 or Lipst1_1_5...  \n",
       "299                                     Lipst1_1_67749  \n",
       "300                                    Lipst1_1_161356  \n",
       "301                                     Lipst1_1_68925  \n",
       "302                                     Lipst1_1_68925  \n",
       "303  Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...  \n",
       "304  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "305                                     Lipst1_1_68925  \n",
       "306  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "307                     Lipst1_1_49943 or Lipst1_1_984  \n",
       "308                                     Lipst1_1_68925  \n",
       "309                                     Lipst1_1_68925  \n",
       "310                                     Lipst1_1_68925  \n",
       "311                                     Lipst1_1_57858  \n",
       "312  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "313                                    Lipst1_1_296629  \n",
       "314                                     Lipst1_1_68925  \n",
       "315                                     Lipst1_1_68925  \n",
       "316                                    Lipst1_1_161356  \n",
       "317                  Lipst1_1_67605 or Lipst1_1_103456  \n",
       "318                                    Lipst1_1_263735  \n",
       "319                                     Lipst1_1_68925  \n",
       "320                                     Lipst1_1_49943  \n",
       "321  Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...  \n",
       "322                                     Lipst1_1_70451  \n",
       "323                                     Lipst1_1_68925  \n",
       "324  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "325                   Lipst1_1_2990 or Lipst1_1_218588  \n",
       "326                                     Lipst1_1_68925  \n",
       "327  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "328                                    Lipst1_1_159073  \n",
       "329                                     Lipst1_1_68925  \n",
       "330                                    Lipst1_1_102653  \n",
       "331                                    Lipst1_1_102653  \n",
       "332                                     Lipst1_1_68925  \n",
       "333  Lipst1_1_2781 or Lipst1_1_102936 or Lipst1_1_6...  \n",
       "334                                     Lipst1_1_68925  \n",
       "335                                     Lipst1_1_68925  \n",
       "336                                      Lipst1_1_6611  \n",
       "337                                    Lipst1_1_110448  \n",
       "338                                     Lipst1_1_68925  \n",
       "339                                     Lipst1_1_68925  \n",
       "340  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "341                                     Lipst1_1_49943  \n",
       "342   Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "343                                     Lipst1_1_68925  \n",
       "344                                     Lipst1_1_49943  \n",
       "345                                     Lipst1_1_68925  \n",
       "346                                     Lipst1_1_76260  \n",
       "347                                     Lipst1_1_68925  \n",
       "348                                     Lipst1_1_67749  \n",
       "349                     Lipst1_1_49943 or Lipst1_1_984  \n",
       "350                                     Lipst1_1_91236  \n",
       "351                                     Lipst1_1_68925  \n",
       "352                                    Lipst1_1_102653  \n",
       "353                                    Lipst1_1_296629  \n",
       "354                                     Lipst1_1_68925  \n",
       "355                                     Lipst1_1_68925  \n",
       "356                                      Lipst1_1_6611  \n",
       "357                                    Lipst1_1_211414  \n",
       "358                                    Lipst1_1_102653  \n",
       "359                                     Lipst1_1_49943  \n",
       "360                                     Lipst1_1_68925  \n",
       "361                                     Lipst1_1_71580  \n",
       "362                                     Lipst1_1_68925  \n",
       "363                                     Lipst1_1_68925  \n",
       "364                                    Lipst1_1_106989  \n",
       "365                                     Lipst1_1_68925  \n",
       "366                                     Lipst1_1_72954  \n",
       "367                                     Lipst1_1_68925  \n",
       "368                                     Lipst1_1_68925  \n",
       "369                                     Lipst1_1_68925  \n",
       "370                                     Lipst1_1_68925  \n",
       "371  Lipst1_1_56084 or Lipst1_1_6286 or Lipst1_1_13...  \n",
       "372                                    Lipst1_1_296629  \n",
       "373                                     Lipst1_1_68925  \n",
       "374                                     Lipst1_1_68925  \n",
       "375                                    Lipst1_1_170576  \n",
       "376  Lipst1_1_103455 or Lipst1_1_106915 or Lipst1_1...  \n",
       "377                                     Lipst1_1_68925  \n",
       "378                                      Lipst1_1_1096  \n",
       "379                                     Lipst1_1_68925  \n",
       "380  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "381  Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "382                                     Lipst1_1_62314  \n",
       "383                                    Lipst1_1_102653  \n",
       "384                                     Lipst1_1_68925  \n",
       "385                                    Lipst1_1_161356  \n",
       "386                                     Lipst1_1_68925  \n",
       "387  Lipst1_1_61079 or Lipst1_1_155238 or Lipst1_1_...  \n",
       "388                                     Lipst1_1_50907  \n",
       "389                      Lipst1_1_92 or Lipst1_1_71475  \n",
       "390  Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "391   Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "392                                    Lipst1_1_318678  \n",
       "393  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "394                                      Lipst1_1_1388  \n",
       "395                                     Lipst1_1_68925  \n",
       "396   Lipst1_1_68143 or Lipst1_1_5996 or Lipst1_1_6287  \n",
       "397                                     Lipst1_1_68925  \n",
       "398  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "399                                     Lipst1_1_68925  \n",
       "400                     Lipst1_1_49943 or Lipst1_1_984  \n",
       "401                                     Lipst1_1_68925  \n",
       "402                                    Lipst1_1_222261  \n",
       "403                                     Lipst1_1_72748  \n",
       "404                                     Lipst1_1_72954  \n",
       "405                                     Lipst1_1_68925  \n",
       "406                                     Lipst1_1_68925  \n",
       "407                   Lipst1_1_2990 or Lipst1_1_218588  \n",
       "408                                     Lipst1_1_68925  \n",
       "409                                     Lipst1_1_68925  \n",
       "410  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "411                     Lipst1_1_6803 or Lipst1_1_2352  \n",
       "412                                     Lipst1_1_47595  \n",
       "413                                     Lipst1_1_68925  \n",
       "414                                     Lipst1_1_68925  \n",
       "415                                     Lipst1_1_68925  \n",
       "416                                    Lipst1_1_106140  \n",
       "417                   Lipst1_1_2990 or Lipst1_1_218588  \n",
       "418                                     Lipst1_1_68925  \n",
       "419  Lipst1_1_2781 or Lipst1_1_102936 or Lipst1_1_6...  \n",
       "420                                     Lipst1_1_68496  \n",
       "421                                     Lipst1_1_68925  \n",
       "422                                     Lipst1_1_68925  \n",
       "423                                    Lipst1_1_102653  \n",
       "424                                    Lipst1_1_103747  \n",
       "425                  Lipst1_1_47671 or Lipst1_1_161356  \n",
       "426                                     Lipst1_1_68925  \n",
       "427                                     Lipst1_1_68925  \n",
       "428                                    Lipst1_1_106989  \n",
       "429                  Lipst1_1_57858 or Lipst1_1_105766  \n",
       "430                  Lipst1_1_159073 or Lipst1_1_71543  \n",
       "431                                    Lipst1_1_161356  \n",
       "432                                     Lipst1_1_72748  \n",
       "433                                     Lipst1_1_68925  \n",
       "434  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "435                                    Lipst1_1_102653  \n",
       "436                                     Lipst1_1_68925  \n",
       "437                                     Lipst1_1_72748  \n",
       "438  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "439                                    Lipst1_1_102653  \n",
       "440                                     Lipst1_1_42443  \n",
       "441                                     Lipst1_1_99002  \n",
       "442                                     Lipst1_1_76042  \n",
       "443  Lipst1_1_63052 or Lipst1_1_108075 or Lipst1_1_...  \n",
       "444  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "445                                     Lipst1_1_68925  \n",
       "446                                    Lipst1_1_281312  \n",
       "447                                     Lipst1_1_68925  \n",
       "448                                     Lipst1_1_68925  \n",
       "449                                    Lipst1_1_296629  \n",
       "450  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  \n",
       "451                                      Lipst1_1_3875  \n",
       "452                                    Lipst1_1_102653  \n",
       "453  Lipst1_1_71408 or Lipst1_1_78240 or Lipst1_1_3...  \n",
       "454                                      Lipst1_1_1895  \n",
       "455                                    Lipst1_1_102653  \n",
       "456  Lipst1_1_167596 or Lipst1_1_253586 or Lipst1_1...  "
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sce_added_orthologs_orhpans\n",
    "pd.set_option('display.max_rows', None)\n",
    "sce_added_orthologs_orhpans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "49664840",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cobra.io.save_json_model(model,'Lst_v0.1_model_withOrthoMCLallAdded_noSCEtransporters.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa3f73a3",
   "metadata": {},
   "source": [
    "## try removing the reactions with only rto genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "8b9f7777",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACRNtp',\n",
       " 'C30CPT1',\n",
       " 'CRNtp',\n",
       " 'DNTPPA',\n",
       " 'DRTPPD',\n",
       " 'FA160COAabcp',\n",
       " 'FA183COAabcp',\n",
       " 'FA240tp',\n",
       " 'GAPP',\n",
       " 'PMDPHT'}"
      ]
     },
     "execution_count": 301,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "d6371571",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.07800483549348808 default\n",
      "\n",
      "DNTPPA\n",
      "Dihydroneopterin triphosphate pyrophosphatase\n",
      "15385\n",
      "0.0\n",
      "\n",
      "\n",
      "ACRNtp\n",
      "Acetylcarnitine transport out of peroxisome\n",
      "10149\n",
      "0.0\n",
      "\n",
      "\n",
      "FA240tp\n",
      "Fatty acid peroxisomal transport\n",
      "9912\n",
      "0.07800483549348884\n",
      "\n",
      "\n",
      "PMDPHT\n",
      "Pyrimidine phosphatase\n",
      "11513\n",
      "0.0\n",
      "\n",
      "\n",
      "C30CPT1\n",
      "Production of propionylcarnitine\n",
      "9315\n",
      "0.0\n",
      "\n",
      "\n",
      "GAPP\n",
      "Glyceraldehyde-3-phosphate phosphatase\n",
      "8460\n",
      "0.07800318735919652\n",
      "\n",
      "\n",
      "FA160COAabcp\n",
      "Fatty acyl CoA peroxisomal transport via ABC system\n",
      "13167 and 9637\n",
      "0.07679940446092794\n",
      "\n",
      "\n",
      "DRTPPD\n",
      "2 5 diamino 6 ribitylamino 4 3H  pyrimidinone 5  phosphate deaminase\n",
      "15506\n",
      "0.0\n",
      "\n",
      "\n",
      "CRNtp\n",
      "Carnitine transport into peroxsiome\n",
      "10149\n",
      "0.0\n",
      "\n",
      "\n",
      "FA183COAabcp\n",
      "Fatty acid peroxisomal transport via ABC system\n",
      "13167 and 9637\n",
      "0.07800483549348679\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "temp = set(wrong_rx2)\n",
    "\n",
    "print(model.optimize().objective_value,'default\\n')\n",
    "\n",
    "for x in temp:\n",
    "    with model:\n",
    "        r = model.reactions.get_by_id(x)\n",
    "        print(r.id)\n",
    "        print(r.name)\n",
    "        print(r.gene_reaction_rule)\n",
    "        model.remove_reactions([r])\n",
    "        print(model.optimize().objective_value)\n",
    "        print('\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd1ad7fd",
   "metadata": {},
   "source": [
    "Fatty acid transporters can be removed. however will set the gene_reaction_rule to none."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "4bd39260",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.FA183COAabcp.gene_reaction_rule = ''\n",
    "model.reactions.FA240tp.gene_reaction_rule = ''\n",
    "model.reactions.FA160COAabcp.gene_reaction_rule = ''\n",
    "\n",
    "\n",
    "model.reactions.ACRNtp.gene_reaction_rule = ''\n",
    "model.reactions.CRNtp.gene_reaction_rule = ''\n",
    "# model.reactions.FA160COAabcp.gene_reaction_rule = ''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "288a1f85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9912 has no associated reaction\n",
      "13167 has no associated reaction\n",
      "9637 has no associated reaction\n",
      "10149 has no associated reaction\n",
      "5\n",
      "5\n"
     ]
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))\n",
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g2)))\n",
    "print(len(set(wrong_rx2)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "f89a3d70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C30CPT1', 'DNTPPA', 'DRTPPD', 'GAPP', 'PMDPHT'}"
      ]
     },
     "execution_count": 305,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2 = [x.id for x in wrong_rx2]\n",
    "set(wrong_rx2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "id": "d0a5e1e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DRTPPD: 25dthpp_c + h2o_c + h_c --> 5aprbu_c + nh4_c DRTPPD 2 5 diamino 6 ribitylamino 4 3H  pyrimidinone 5  phosphate deaminase\n",
      "25dthpp_c not in ylip\n",
      "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c DNTPPA Dihydroneopterin triphosphate pyrophosphatase\n",
      "PMDPHT: 5aprbu_c + h2o_c --> 4r5au_c + pi_c PMDPHT Pyrimidine phosphatase\n",
      "C30CPT1: crn_c + ppcoa_c <=> coa_c + pcrn_c C30CPT1 Production of propionylcarnitine\n",
      "pcrn_c not in ylip\n",
      "ppcoa_c not in ylip\n",
      "GAPP: g3p_c + h2o_c --> glyald_c + pi_c GAPP Glyceraldehyde-3-phosphate phosphatase\n"
     ]
    }
   ],
   "source": [
    "for rx in wrong_rx2:\n",
    "    r = model.reactions.get_by_id(rx)\n",
    "    print(r,r.id,r.name)\n",
    "    for metabolites in r.metabolites:\n",
    "        try:\n",
    "            ylip.metabolites.get_by_id(metabolites.id)\n",
    "        except:\n",
    "            print(f'{metabolites} not in ylip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "f4f87404",
   "metadata": {},
   "outputs": [],
   "source": [
    "with model:\n",
    "    model.remove_reactions(['GAPP'])\n",
    "    gapfiller = cobra.flux_analysis.gapfilling.GapFiller(model, sce, demand_reactions=False, integer_threshold=1e-10)\n",
    "    gapfiller.model.solver.configuration.tolerances.feasibility = 1e-9\n",
    "    gapfiller.model.solver.configuration.tolerances.integrality = 1e-10\n",
    "#     gapfiller.model.solver.configuration.tolerances.optimality = 1e-9\n",
    "    solution = gapfiller.fill()\n",
    "\n",
    "for reaction in solution[0]:\n",
    "    print(reaction.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "b118ca5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "with model:\n",
    "    model.remove_reactions(['GAPP'])\n",
    "    gapfiller = cobra.flux_analysis.gapfilling.GapFiller(model, ylip, demand_reactions=False, integer_threshold=1e-10)\n",
    "    gapfiller.model.solver.configuration.tolerances.feasibility = 1e-9\n",
    "    gapfiller.model.solver.configuration.tolerances.integrality = 1e-10\n",
    "#     gapfiller.model.solver.configuration.tolerances.optimality = 1e-9\n",
    "    solution = gapfiller.fill()\n",
    "\n",
    "for reaction in solution[0]:\n",
    "    print(reaction.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "a080760d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>G3PT</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>glycerol-3-phosphatase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a4f7dd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>glyc3p_c + h2o_c --> glyc_c + pi_c</p>\n",
       "                    <p style='text-align:right'>glycerol 3-phosphate + H2O --> glycerol + phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YER062C or YIL053W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction G3PT at 0x16a4f7dd0>"
      ]
     },
     "execution_count": 307,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.G3PT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "bf34ad05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# neither sce gene is available. \n",
    "# lst_sce_dict['YER062C']\n",
    "# lst_sce_dict['YIL053W']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "2410e897",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYC3Ptm</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>glycerol 3 phosphate shuttle</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15bb3d1d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>glyc3p_c --> glyc3p_m</p>\n",
       "                    <p style='text-align:right'>Glycerol_3_phosphate --> Glycerol_3_phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYC3Ptm at 0x15bb3d1d0>"
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.GLYC3Ptm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "674c4777",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>Lipst1_1_67195</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b7a3490</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 9 reaction(s)</strong></td><td>\n",
       "                    F1PP, GNP, G6PP, R5PP, G1PP, E4PP, F6PP, 2DOXG6PP, RU5PP\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene Lipst1_1_67195 at 0x15b7a3490>"
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# JGI shows the following gene as a glycerol-1-phosphatase.\n",
    "model.genes.Lipst1_1_67195"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "991949fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>G6PP</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Glucose-6-phosphate phosphatase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a150b4d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>g6p_c + h2o_c --> glc__D_c + pi_c</p>\n",
       "                    <p style='text-align:right'>D-Glucose 6-phosphate + H2O H2O --> alpha-D-Glucose + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_67195</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction G6PP at 0x2a150b4d0>"
      ]
     },
     "execution_count": 311,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.G6PP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0e350c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b32ffd82",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b45494",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84f0d3e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7292f80",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd036034",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "51cfacf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DRTPPD sce\n",
      "DRTPPD\n",
      "DRTPPD ylip\n",
      "APRAUR\n",
      "DHPPDA\n",
      "DNTPPA sce\n",
      "DNTPPA\n",
      "DNTPPA ylip\n",
      "AKP1\n",
      "PMDPHT sce\n",
      "PMDPHT\n",
      "PMDPHT ylip\n",
      "PMDPHT\n",
      "C30CPT1 sce\n",
      "r_4570\n",
      "r_4571\n",
      "C30CPT1 ylip\n",
      "GAPP sce\n",
      "GLYALDDr\n",
      "GAPP ylip\n",
      "GLYC3Ptm\n"
     ]
    }
   ],
   "source": [
    "for rx in wrong_rx2:\n",
    "    print(rx,'sce')\n",
    "    with model:\n",
    "        model.remove_reactions([rx])\n",
    "        gapfiller = cobra.flux_analysis.gapfilling.GapFiller(model, sce, demand_reactions=False, integer_threshold=1e-10)\n",
    "        gapfiller.model.solver.configuration.tolerances.feasibility = 1e-9\n",
    "        gapfiller.model.solver.configuration.tolerances.integrality = 1e-10\n",
    "    #     gapfiller.model.solver.configuration.tolerances.optimality = 1e-9\n",
    "        solution = gapfiller.fill()\n",
    "\n",
    "    for reaction in solution[0]:\n",
    "        print(reaction.id)\n",
    "\n",
    "    print(rx,'ylip')\n",
    "    \n",
    "    if rx!='C30CPT1': # reaction fails in ylip. \n",
    "        with model:\n",
    "            model.remove_reactions([rx])\n",
    "            gapfiller = cobra.flux_analysis.gapfilling.GapFiller(model, ylip, demand_reactions=False, integer_threshold=1e-10)\n",
    "            gapfiller.model.solver.configuration.tolerances.feasibility = 1e-9\n",
    "            gapfiller.model.solver.configuration.tolerances.integrality = 1e-10\n",
    "        #     gapfiller.model.solver.configuration.tolerances.optimality = 1e-9\n",
    "            solution = gapfiller.fill()\n",
    "\n",
    "        for reaction in solution[0]:\n",
    "            print(reaction.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "19c8a90e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/czaj603/miniconda3/envs/lipGSM/lib/python3.11/site-packages/cobra/core/group.py:147: UserWarning: need to pass in a list\n",
      "  warn(\"need to pass in a list\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ylip growth: 1.1398166174414033\n",
      "Model growth: 0.0\n"
     ]
    }
   ],
   "source": [
    "with ylip, model:\n",
    "    ylip.remove_reactions(['PMDPHT'])\n",
    "    model.remove_reactions(['PMDPHT'])\n",
    "    print(\"Ylip growth:\", ylip.slim_optimize())\n",
    "    print(\"Model growth:\", model.slim_optimize())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "4bfa2243",
   "metadata": {},
   "outputs": [
    {
     "ename": "Infeasible",
     "evalue": "gap filling optimization failed (infeasible).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInfeasible\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[336], line 9\u001b[0m\n\u001b[1;32m      7\u001b[0m     gapfiller\u001b[38;5;241m.\u001b[39mmodel\u001b[38;5;241m.\u001b[39msolver\u001b[38;5;241m.\u001b[39mconfiguration\u001b[38;5;241m.\u001b[39mtolerances\u001b[38;5;241m.\u001b[39mintegrality \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1e-10\u001b[39m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m#     gapfiller.model.solver.configuration.tolerances.optimality = 1e-9\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m     solution \u001b[38;5;241m=\u001b[39m gapfiller\u001b[38;5;241m.\u001b[39mfill()\n",
      "File \u001b[0;32m~/miniconda3/envs/lipGSM/lib/python3.11/site-packages/cobra/flux_analysis/gapfilling.py:296\u001b[0m, in \u001b[0;36mGapFiller.fill\u001b[0;34m(self, iterations)\u001b[0m\n\u001b[1;32m    294\u001b[0m used_reactions \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    295\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m _ \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(iterations):\n\u001b[0;32m--> 296\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel\u001b[38;5;241m.\u001b[39mslim_optimize(\n\u001b[1;32m    297\u001b[0m         error_value\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, message\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgap filling optimization failed\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    298\u001b[0m     )\n\u001b[1;32m    299\u001b[0m     solution \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m    300\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel\u001b[38;5;241m.\u001b[39mreactions\u001b[38;5;241m.\u001b[39mget_by_id(ind\u001b[38;5;241m.\u001b[39mrxn_id)\n\u001b[1;32m    301\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m ind \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindicators\n\u001b[1;32m    302\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m ind\u001b[38;5;241m.\u001b[39m_get_primal() \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minteger_threshold\n\u001b[1;32m    303\u001b[0m     ]\n\u001b[1;32m    304\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalidate(solution):\n",
      "File \u001b[0;32m~/miniconda3/envs/lipGSM/lib/python3.11/site-packages/cobra/core/model.py:1200\u001b[0m, in \u001b[0;36mModel.slim_optimize\u001b[0;34m(self, error_value, message)\u001b[0m\n\u001b[1;32m   1198\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m error_value\n\u001b[1;32m   1199\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1200\u001b[0m     assert_optimal(\u001b[38;5;28mself\u001b[39m, message)\n",
      "File \u001b[0;32m~/miniconda3/envs/lipGSM/lib/python3.11/site-packages/cobra/util/solver.py:590\u001b[0m, in \u001b[0;36massert_optimal\u001b[0;34m(model, message)\u001b[0m\n\u001b[1;32m    588\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m OPTIMAL:\n\u001b[1;32m    589\u001b[0m     exception_cls \u001b[38;5;241m=\u001b[39m OPTLANG_TO_EXCEPTIONS_DICT\u001b[38;5;241m.\u001b[39mget(status, OptimizationError)\n\u001b[0;32m--> 590\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exception_cls(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmessage\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m (\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstatus\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m).\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mInfeasible\u001b[0m: gap filling optimization failed (infeasible)."
     ]
    }
   ],
   "source": [
    "# # becomes infeasible. \n",
    "\n",
    "#     with ylip,model:\n",
    "#         ylip.remove_reactions(['PMDPHT'])\n",
    "#         model.remove_reactions(['PMDPHT'])\n",
    "        \n",
    "#         gapfiller = cobra.flux_analysis.gapfilling.GapFiller(model, ylip, demand_reactions=False, integer_threshold=1e-10)\n",
    "#         gapfiller.model.solver.configuration.tolerances.feasibility = 1e-9\n",
    "#         gapfiller.model.solver.configuration.tolerances.integrality = 1e-10\n",
    "#     #     gapfiller.model.solver.configuration.tolerances.optimality = 1e-9\n",
    "#         solution = gapfiller.fill()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fafb2ac5",
   "metadata": {},
   "source": [
    "the sce replacements seem simplier to implement overall. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "6fe25f2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DRTPPD</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>2 5 diamino 6 ribitylamino 4 3H pyrimidinone 5 phosphate deaminase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a1b92f90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>25dthpp_c + h2o_c + h_c --> 5aprbu_c + nh4_c</p>\n",
       "                    <p style='text-align:right'>2 5 diamino 6 ribitylamino 4 3H pyrimidinone 5 phosphate C9H16N5O8P + H2O H2O + H+ --> 5-Amino-6-(5'-phosphoribitylamino)uracil + Ammonium</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>15506</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DRTPPD at 0x2a1b92f90>"
      ]
     },
     "execution_count": 312,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.DRTPPD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "cab44975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>APRAUR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>5 amino 6 5 phosphoribosylamino uracil reductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x17fefe090</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>5apru_c + h_c + nadph_c --> 5aprbu_c + nadp_c</p>\n",
       "                    <p style='text-align:right'>5_Amino_6__5__phosphoribosylamino_uracil + H + Nicotinamide_adenine_dinucleotide_phosphate___reduced --> 5_Amino_6__5__phosphoribitylamino_uracil + Nicotinamide_adenine_dinucleotide_phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0D06501g</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction APRAUR at 0x17fefe090>"
      ]
     },
     "execution_count": 313,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.APRAUR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "db47f069",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Lipst1_1_74905'"
      ]
     },
     "execution_count": 314,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lst_ylip_dict['YALI0D06501g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "7d1d9f83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>Lipst1_1_74905</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b78ed10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 1 reaction(s)</strong></td><td>\n",
       "                    DROPPRx\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene Lipst1_1_74905 at 0x15b78ed10>"
      ]
     },
     "execution_count": 315,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.genes.Lipst1_1_74905"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "4e266b07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DROPPRx</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>2 5 diamino 6 ribosylamino 4 3H pyrimidinone 5 phosphate reductase nadh</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b9efe90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>25dhpp_c + h_c + nadh_c --> 25dthpp_c + nad_c</p>\n",
       "                    <p style='text-align:right'>2 5 Diamino 6 hydroxy 4 5 phosphoribosylamino pyrimidine C9H14N5O8P + H+ + Nicotinamide adenine dinucleotide - reduced --> 2 5 diamino 6 ribitylamino 4 3H pyrimidinone 5 phosphate C9H16N5O8P +...</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_74905</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DROPPRx at 0x15b9efe90>"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.DROPPRx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "e302f137",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DHPPDA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>diaminohydroxyphosphoribosylaminopyrimidine deaminase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a7ad410</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>25dhpp_c + h2o_c + h_c --> 5apru_c + nh4_c</p>\n",
       "                    <p style='text-align:right'>2_5_Diamino_6_hydroxy_4__5__phosphoribosylamino__pyrimidine + H2O + H --> 5_Amino_6__5__phosphoribosylamino_uracil + Ammonium</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DHPPDA at 0x16a7ad410>"
      ]
     },
     "execution_count": 317,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.DHPPDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "02d6a4b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DRTPPD', 'DNTPPA', 'PMDPHT', 'C30CPT1', 'GAPP']"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "id": "498ac70a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>AKP1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>alkaline phosphatase (dihydroneopterin)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b095a90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ahdt_c + 3.0 h2o_c --> dhnpt_c + 2.0 h_c + 3.0 pi_c</p>\n",
       "                    <p style='text-align:right'>7,8-dihydroneopterin 3'-triphosphate + 3.0 H2O --> 7,8-dihydroneopterin + 2.0 H+ + 3.0 phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YDR481C</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction AKP1 at 0x15b095a90>"
      ]
     },
     "execution_count": 319,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.AKP1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "id": "4972d778",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Lipst1_1_2594'"
      ]
     },
     "execution_count": 320,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lst_sce_dict['YDR481C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "eee9c0f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>Lipst1_1_2594</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b787910</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 2 reaction(s)</strong></td><td>\n",
       "                    NMNHYD, ALKP\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene Lipst1_1_2594 at 0x15b787910>"
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.genes.Lipst1_1_2594"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "id": "3a76edea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ALKP</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Alkaline phosphatase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x17ff19a50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>dhap_c + h2o_c --> dha_c + pi_c</p>\n",
       "                    <p style='text-align:right'>Dihydroxyacetone phosphate + H2O H2O --> Dihydroxyacetone + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_2594</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ALKP at 0x17ff19a50>"
      ]
     },
     "execution_count": 322,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.ALKP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "id": "61b60067",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Lipst1_1_2594'"
      ]
     },
     "execution_count": 323,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.remove_reactions(['DRTPPD'])\n",
    "r = sce.reactions.AKP1.copy()\n",
    "r.gene_reaction_rule = 'Lipst1_1_2594'\n",
    "model.add_reactions([r])\n",
    "model.reactions.get_by_id(r.id).gene_reaction_rule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "id": "b9548913",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.remove_reactions(['DNTPPA'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "15c59a8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DRTPPD', 'DNTPPA', 'PMDPHT', 'C30CPT1', 'GAPP']"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c072502",
   "metadata": {},
   "source": [
    "PMDPHT seems to be an essential reaction with no associated gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "a3b5f39a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PMDPHT</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>pyrimidine phosphatase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a641310</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>5aprbu_c + h2o_c --> 4r5au_c + pi_c</p>\n",
       "                    <p style='text-align:right'>5-amino-6-(5-phosphoribitylamino)uracil + H2O --> 5-amino-6-(D-ribitylamino)uracil + phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PMDPHT at 0x16a641310>"
      ]
     },
     "execution_count": 326,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.get_by_id('PMDPHT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "15eb472c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PMDPHT</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>pyrimidine phosphatase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15bba2410</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>5aprbu_c + h2o_c --> 4r5au_c + pi_c</p>\n",
       "                    <p style='text-align:right'>5_Amino_6__5__phosphoribitylamino_uracil + H2O --> 4__1_D_Ribitylamino__5_aminouracil + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PMDPHT at 0x15bba2410>"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.get_by_id('PMDPHT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "1916f34d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PMDPHT</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Pyrimidine phosphatase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a1bc2350</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>5aprbu_c + h2o_c --> 4r5au_c + pi_c</p>\n",
       "                    <p style='text-align:right'>5-Amino-6-(5'-phosphoribitylamino)uracil + H2O H2O --> 4-(1-D-Ribitylamino)-5-aminouracil + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>11513</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PMDPHT at 0x2a1bc2350>"
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('PMDPHT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "48a8deba",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.get_by_id('PMDPHT').gene_reaction_rule = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aad59e76",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "c902b8f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DRTPPD', 'DNTPPA', 'PMDPHT', 'C30CPT1', 'GAPP']"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wrong_rx2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3f74ee0",
   "metadata": {},
   "source": [
    "C30CPT1 - r_4570\n",
    "r_4571"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "43493a37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>C30CPT1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Production of propionylcarnitine</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x2a1bc32d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>crn_c + ppcoa_c <=> coa_c + pcrn_c</p>\n",
       "                    <p style='text-align:right'>L-Carnitine + Propanoyl-CoA <=> Coenzyme A + Propionyl-carnitine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>9315</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction C30CPT1 at 0x2a1bc32d0>"
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('C30CPT1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "c1eab298",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_4570</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>phosphoglycerate dehydrogenase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16aa73650</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>2hog_c + nad_c <=> akg_c + h_c + nadh_c</p>\n",
       "                    <p style='text-align:right'>2-hydroxyglutarate + NAD <=> 2-oxoglutarate + H+ + NADH</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YIL074C or YER081W</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_4570 at 0x16aa73650>"
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_4570"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "8ecd9b1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_4571</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>seed:rxn00681</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a71bc50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>2hog_c + coa_c + h_c <=> glx_c + h2o_c + ppcoa_c</p>\n",
       "                    <p style='text-align:right'>2-hydroxyglutarate + coenzyme A + H+ <=> glyoxylate + H2O + propionyl-CoA</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_4571 at 0x16a71bc50>"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.r_4571"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa9f3bad",
   "metadata": {},
   "source": [
    "gap filled sce solution has no gpr for second reaction. remove gpr for first reaction. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "6f3a15eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.get_by_id('C30CPT1').gene_reaction_rule = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "95686ec5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYALDDr</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-Glyceraldehyde:NAD+ oxidoreductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16aa1b990</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>glyald_c + h2o_c + nad_c <=> glyc__R_c + 2.0 h_c + nadh_c</p>\n",
       "                    <p style='text-align:right'>D-glyceraldehyde + H2O + NAD <=> D-Glycerate + 2.0 H+ + NADH</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YMR110C</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYALDDr at 0x16aa1b990>"
      ]
     },
     "execution_count": 335,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce.reactions.get_by_id('GLYALDDr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "id": "78c93b98",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'YMR110C'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[358], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m lst_sce_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mYMR110C\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "\u001b[0;31mKeyError\u001b[0m: 'YMR110C'"
     ]
    }
   ],
   "source": [
    "# not present in the ortholog dict. \n",
    "# lst_sce_dict['YMR110C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "id": "da9f3d2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "APRAUR\n",
      "DHPPDA\n"
     ]
    }
   ],
   "source": [
    "        rx = model.reactions.get_by_id('GAPP')\n",
    "        with model:\n",
    "            model.remove_reactions([rx])\n",
    "            gapfiller = cobra.flux_analysis.gapfilling.GapFiller(model, ylip, demand_reactions=False, integer_threshold=1e-10)\n",
    "            gapfiller.model.solver.configuration.tolerances.feasibility = 1e-9\n",
    "            gapfiller.model.solver.configuration.tolerances.integrality = 1e-10\n",
    "        #     gapfiller.model.solver.configuration.tolerances.optimality = 1e-9\n",
    "            solution = gapfiller.fill()\n",
    "\n",
    "        for reaction in solution[0]:\n",
    "            print(reaction.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26a02d52",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "67b80c24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYC3Ptm</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>glycerol 3 phosphate shuttle</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15bb3d1d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>glyc3p_c --> glyc3p_m</p>\n",
       "                    <p style='text-align:right'>Glycerol_3_phosphate --> Glycerol_3_phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYC3Ptm at 0x15bb3d1d0>"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.get_by_id('GLYC3Ptm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "857e9f6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DHPPDA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>diaminohydroxyphosphoribosylaminopyrimidine deaminase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x16a7ad410</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>25dhpp_c + h2o_c + h_c --> 5apru_c + nh4_c</p>\n",
       "                    <p style='text-align:right'>2_5_Diamino_6_hydroxy_4__5__phosphoribosylamino__pyrimidine + H2O + H --> 5_Amino_6__5__phosphoribosylamino_uracil + Ammonium</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DHPPDA at 0x16a7ad410>"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.get_by_id('DHPPDA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "f50c6192",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>5aprbu_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>5-Amino-6-(5'-phosphoribitylamino)uracil</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b7227d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C9H15N4O9P</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 1 reaction(s)</strong></td><td>\n",
       "                    PMDPHT\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite 5aprbu_c at 0x15b7227d0>"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('5aprbu_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94effbf8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "d3b0c96c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>APRAUR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>5 amino 6 5 phosphoribosylamino uracil reductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x17fefe090</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>5apru_c + h_c + nadph_c --> 5aprbu_c + nadp_c</p>\n",
       "                    <p style='text-align:right'>5_Amino_6__5__phosphoribosylamino_uracil + H + Nicotinamide_adenine_dinucleotide_phosphate___reduced --> 5_Amino_6__5__phosphoribitylamino_uracil + Nicotinamide_adenine_dinucleotide_phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0D06501g</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction APRAUR at 0x17fefe090>"
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ylip.reactions.get_by_id('APRAUR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "982f3e2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Lipst1_1_74905'"
      ]
     },
     "execution_count": 340,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lst_ylip_dict['YALI0D06501g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "44693f0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>Lipst1_1_74905</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b78ed10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 1 reaction(s)</strong></td><td>\n",
       "                    DROPPRx\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene Lipst1_1_74905 at 0x15b78ed10>"
      ]
     },
     "execution_count": 341,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.genes.Lipst1_1_74905"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "id": "b2fd086f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DROPPRx</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>2 5 diamino 6 ribosylamino 4 3H pyrimidinone 5 phosphate reductase nadh</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15b9efe90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>25dhpp_c + h_c + nadh_c --> 25dthpp_c + nad_c</p>\n",
       "                    <p style='text-align:right'>2 5 Diamino 6 hydroxy 4 5 phosphoribosylamino pyrimidine C9H14N5O8P + H+ + Nicotinamide adenine dinucleotide - reduced --> 2 5 diamino 6 ribitylamino 4 3H pyrimidinone 5 phosphate C9H16N5O8P +...</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Lipst1_1_74905</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DROPPRx at 0x15b9efe90>"
      ]
     },
     "execution_count": 342,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.DROPPRx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "id": "a6538b68",
   "metadata": {},
   "outputs": [],
   "source": [
    "r = model.reactions.get_by_id('GAPP')\n",
    "model.remove_reactions([r])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "id": "4bfac764",
   "metadata": {},
   "outputs": [],
   "source": [
    "y = ylip.reactions.get_by_id('GLYC3Ptm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "id": "0e020750",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions([y])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "78ac0789",
   "metadata": {},
   "outputs": [],
   "source": [
    "y = ylip.reactions.get_by_id('DHPPDA')\n",
    "model.add_reactions([y])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "b80034ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "y = ylip.reactions.get_by_id('APRAUR').copy()\n",
    "y.gene_reaction_rule = 'Lipst1_1_74905'\n",
    "model.add_reactions([y])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "a1b6ee64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.078</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALCD25yi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTHFCm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMPN</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.022961e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DAGCPTer_RT</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PYRt2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FA183COAabcp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKP1</th>\n",
       "      <td>0.000304</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLYC3Ptm</th>\n",
       "      <td>0.000140</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHPPDA</th>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APRAUR</th>\n",
       "      <td>0.000136</td>\n",
       "      <td>-2.775558e-17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2158 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.078 at 0x2a903f410>"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "2a6815ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15506 has no associated reaction\n",
      "15385 has no associated reaction\n",
      "11513 has no associated reaction\n",
      "9315 has no associated reaction\n",
      "8460 has no associated reaction\n",
      "0\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 349,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = model.genes\n",
    "genes = [g.id for g in genes]\n",
    "wrong = [g for g in genes if 'Lipst1' not in g]\n",
    "len(set(wrong))\n",
    "wrong_g2 = []\n",
    "wrong_rx2 = []\n",
    "for g in wrong:\n",
    "    reac = model.genes.get_by_id(g).reactions\n",
    "    if not reac:\n",
    "        print(f'{g} has no associated reaction')\n",
    "        \n",
    "        cobra.manipulation.delete.remove_genes(model,[g])\n",
    "    for rx in reac:\n",
    "        if rx:\n",
    "#             print(rx,g)\n",
    "            wrong_g2.append(g)\n",
    "            wrong_rx2.append(rx)\n",
    "print(len(set(wrong_g2)))\n",
    "print(len(set(wrong_rx2)))\n",
    "wrong_rx2 = [x.id for x in wrong_rx2]\n",
    "set(wrong_rx2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cba5a44e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "bbc80eb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ls = 0.07800521206597387</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>6.49E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>4.095E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.0004812</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>98.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.04569</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.004814</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>4.735E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>na1_e</td>\n",
       "      <td>EX_na1_e</td>\n",
       "      <td>0.002036</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>0.4596</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>2.356</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.02077</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.007819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>3.978E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8aonn_m</td>\n",
       "      <td>SRC_8aonn_m</td>\n",
       "      <td>0.0001886</td>\n",
       "      <td>9</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lipoate_m</td>\n",
       "      <td>SRC_lipoate_m</td>\n",
       "      <td>0.0002424</td>\n",
       "      <td>8</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaala_c</td>\n",
       "      <td>SRC_trnaala_c</td>\n",
       "      <td>0.04505</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaarg_c</td>\n",
       "      <td>SRC_trnaarg_c</td>\n",
       "      <td>0.01144</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasn_c</td>\n",
       "      <td>SRC_trnaasn_c</td>\n",
       "      <td>0.008422</td>\n",
       "      <td>10</td>\n",
       "      <td>1.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaasp_c</td>\n",
       "      <td>SRC_trnaasp_c</td>\n",
       "      <td>0.01539</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnacys_c</td>\n",
       "      <td>SRC_trnacys_c</td>\n",
       "      <td>0.002673</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagln_c</td>\n",
       "      <td>SRC_trnagln_c</td>\n",
       "      <td>0.009197</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaglu_c</td>\n",
       "      <td>SRC_trnaglu_c</td>\n",
       "      <td>0.01628</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnagly_c</td>\n",
       "      <td>SRC_trnagly_c</td>\n",
       "      <td>0.04003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnahis_c</td>\n",
       "      <td>SRC_trnahis_c</td>\n",
       "      <td>0.004771</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaile_c</td>\n",
       "      <td>SRC_trnaile_c</td>\n",
       "      <td>0.01237</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaleu_c</td>\n",
       "      <td>SRC_trnaleu_c</td>\n",
       "      <td>0.02498</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnalys_c</td>\n",
       "      <td>SRC_trnalys_c</td>\n",
       "      <td>0.01344</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnamet_c</td>\n",
       "      <td>SRC_trnamet_c</td>\n",
       "      <td>0.00445</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaphe_c</td>\n",
       "      <td>SRC_trnaphe_c</td>\n",
       "      <td>0.007423</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnapro_c</td>\n",
       "      <td>SRC_trnapro_c</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaser_c</td>\n",
       "      <td>SRC_trnaser_c</td>\n",
       "      <td>0.02628</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnathr_c</td>\n",
       "      <td>SRC_trnathr_c</td>\n",
       "      <td>0.01702</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatrp_c</td>\n",
       "      <td>SRC_trnatrp_c</td>\n",
       "      <td>0.001819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnatyr_c</td>\n",
       "      <td>SRC_trnatyr_c</td>\n",
       "      <td>0.004865</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trnaval_c</td>\n",
       "      <td>SRC_trnaval_c</td>\n",
       "      <td>0.02146</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4oglu_c</td>\n",
       "      <td>DM_4oglu_c</td>\n",
       "      <td>-0.0001087</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_m</td>\n",
       "      <td>DM_amob_m</td>\n",
       "      <td>-0.0001886</td>\n",
       "      <td>15</td>\n",
       "      <td>0.10%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>dad_5_m</td>\n",
       "      <td>DM_dad_5_m</td>\n",
       "      <td>-0.0001886</td>\n",
       "      <td>10</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-2.619</td>\n",
       "      <td>1</td>\n",
       "      <td>89.99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>-0.0003342</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-4.275</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-0.1232</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hco3_e</td>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>-0.001955</td>\n",
       "      <td>1</td>\n",
       "      <td>0.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zymst_e</td>\n",
       "      <td>EX_zymst_e</td>\n",
       "      <td>-0.01053</td>\n",
       "      <td>27</td>\n",
       "      <td>9.76%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x2a903ab50>"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "96552a81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.078</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALCD25yi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTHFCm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMPN</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.022961e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DAGCPTer_RT</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PYRt2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FA183COAabcp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKP1</th>\n",
       "      <td>0.000304</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLYC3Ptm</th>\n",
       "      <td>0.000140</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHPPDA</th>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APRAUR</th>\n",
       "      <td>0.000136</td>\n",
       "      <td>-2.775558e-17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2158 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.078 at 0x2aa45fb50>"
      ]
     },
     "execution_count": 351,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6862386",
   "metadata": {},
   "source": [
    "## adjust model subsystems."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "93d9a6af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THRtm: thr__L_c <=> thr__L_m\n",
      "GLYCLTtm: glyclt_c <=> glyclt_m\n",
      "GLXtm: glx_c <=> glx_m\n",
      "METtm: met__L_c <=> met__L_m\n",
      "CYStm: cys__L_c <=> cys__L_m\n",
      "ACACt2m: acac_c + h_c <=> acac_m + h_m\n",
      "BHBtm: bhb_c + h_c <=> bhb_m + h_m\n",
      "GHBtm: ghb_c + h_c <=> ghb_m + h_m\n",
      "4ABUTtm: 4abut_c <=> 4abut_m\n",
      "AGMt_m: agm_c <=> agm_m\n",
      "NADtm: nad_c --> nad_m\n"
     ]
    }
   ],
   "source": [
    "x = []\n",
    "for r in model.reactions:\n",
    "    if (r.subsystem == 'Cysteine and methionine Metabolism') | (r.subsystem == 'Cysteine and methionine metabolism'):\n",
    "        r.subsystem = 'Cysteine and Methionine Metabolism'\n",
    "    if (r.subsystem == 'Glycine, serine and threonine metabolism'):\n",
    "        r.subsystem = 'Glycine, Serine, and Threonine Metabolism'\n",
    "    if (r.subsystem == 'Histidine metabolism'):\n",
    "        r.subsystem = 'Histidine Metabolism'\n",
    "    if (r.subsystem == 'Purine metabolism'):\n",
    "        r.subsystem = 'Purine Metabolism'\n",
    "    if (r.subsystem == 'Pyruvate metabolism'):\n",
    "        r.subsystem = 'Pyruvate Metabolism'\n",
    "        \n",
    "    if (r.subsystem == 'Transport, nucleus'):\n",
    "        r.subsystem = 'Transport, Nuclear'\n",
    "    if (r.subsystem == 'Transport, nuclear'):\n",
    "        r.subsystem = 'Transport, Nuclear'\n",
    "    if (r.subsystem == 'Transport, Glyoxysome'):\n",
    "        r.subsystem = 'Transport, Peroxisomal'\n",
    "        print(r)\n",
    "    if (r.subsystem == 'Transport, Mitochondriall'):\n",
    "        r.subsystem = 'Transport, Mitochondrial'\n",
    "        print(r)\n",
    "    if (r.subsystem == 'Glycine, Serine, Alanine, and threonine Metabolism'):\n",
    "        r.subsystem = 'Glycine, Serine, Alanine, and Threonine Metabolism'\n",
    "#     if (r.subsystem == 'Glyoxylate Metabolism'):\n",
    "#         print(r)\n",
    "#     if (r.subsystem == 'Glyoxylate and Dicarboxylate Metabolism'):\n",
    "#         r.subsystem == 'Glyoxylate Metabolism'\n",
    "        \n",
    "#         r.subsystem = 'Histidine Metabolism'\n",
    "    x.append(r.subsystem)\n",
    "    \n",
    "# 'Methionine Metabolism',\n",
    "#  'Methionine and cysteine Metabolism',"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f04d442",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "d8000de1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " 'Alanine and Aspartate Metabolism',\n",
       " 'Alternate Carbon Metabolism',\n",
       " 'Amino sugar and nucleotide sugar Metabolism',\n",
       " 'Aminosugar Metabolism',\n",
       " 'Anaplerotic Reactions',\n",
       " 'Arabinose Metabolism',\n",
       " 'Arginine and Proline Metabolism',\n",
       " 'Ascorbate Metabolism',\n",
       " 'B-Ketoadipate Pathway',\n",
       " 'Bile Acid Biosynthesis',\n",
       " 'Biomass and maintenance functions',\n",
       " 'Biosynthesis of steroids: Terpenoid backbone synthesis',\n",
       " 'Biosynthesis of steroids: Ubiquinone/Terpenoid-quinone Biosynthesis',\n",
       " 'Biosynthesis of unsaturated fatty acids',\n",
       " 'Biotin Metabolism',\n",
       " 'Butanoate Metabolism',\n",
       " 'Calvin cycle/Pentose Phosphate Pathway',\n",
       " 'Carbohydrates and related molecules',\n",
       " 'Carbon fixation;Glutamate Metabolism;Alanine and Aspartate Metabolism',\n",
       " 'Carnitine shuttle',\n",
       " 'Cell Envelope Biosynthesis',\n",
       " 'Cholesterol Metabolism',\n",
       " 'Citric Acid Cycle',\n",
       " 'Cofactor Biosynthesis: Carnitine Biosynthesis',\n",
       " 'Cofactor and Prosthetic Group Biosynthesis',\n",
       " 'Cofactor recycling',\n",
       " 'Complex Alcohol Metabolism',\n",
       " 'Cysteine Metabolism',\n",
       " 'Cysteine and Methionine Metabolism',\n",
       " 'Drug Metabolism - other enzymes',\n",
       " 'Extracellular exchange',\n",
       " 'Fatty Acid Degradation',\n",
       " 'Fatty Acid Metabolism',\n",
       " 'Fatty Acid, Biosynthesis',\n",
       " 'Fatty acid Metabolism',\n",
       " 'Fatty acid activation',\n",
       " 'Fatty acid elongation',\n",
       " 'Fatty acid elongation in mitochondria',\n",
       " 'Fatty acid oxidation',\n",
       " 'Fatty acid oxidation, peroxisome',\n",
       " 'Fatty acid synthesis',\n",
       " 'Folate Biosynthesis',\n",
       " 'Folate Metabolism',\n",
       " 'Folate biosynthesis',\n",
       " 'Fructose and Mannose Metabolism',\n",
       " 'Galactose Metabolism',\n",
       " 'Glutamate Metabolism',\n",
       " 'Glutamine Metabolism',\n",
       " 'Glutathione metabolism',\n",
       " 'Glycerolipid Metabolism',\n",
       " 'Glycerophospholipid Metabolism',\n",
       " 'Glycine and Serine Metabolism',\n",
       " 'Glycine, Serine, Alanine, and Threonine Metabolism',\n",
       " 'Glycine, Serine, and Threonine Metabolism',\n",
       " 'Glycolysis',\n",
       " 'Glycolysis/Gluconeogenesis',\n",
       " 'Glycoprotein Metabolism',\n",
       " 'Glyoxylate Metabolism',\n",
       " 'Glyoxylate and Dicarboxylate Metabolism',\n",
       " 'High-mannose type N-glycan Biosynthesis',\n",
       " 'Histidine Metabolism',\n",
       " 'Homogentisate Pathway',\n",
       " 'Inorganic Ion Transport and Metabolism',\n",
       " 'Inositol Phosphate Metabolism',\n",
       " 'Intracellular demand',\n",
       " 'Lipopolysaccharide Biosynthesis / Recycling',\n",
       " 'Lysine Biosynthesis',\n",
       " 'Lysine Metabolism',\n",
       " 'Membrane Lipid Metabolism',\n",
       " 'Metabolite Repair',\n",
       " 'Methane Metabolism',\n",
       " 'Methionine Metabolism',\n",
       " 'Methionine and cysteine Metabolism',\n",
       " 'Methylglyoxal Metabolism',\n",
       " 'N-Glycan Biosynthesis',\n",
       " 'N-Glycan Degradation',\n",
       " 'NAD Biosynthesis',\n",
       " 'NAD Metabolism',\n",
       " 'Nicotinate and nicotinamide Metabolism',\n",
       " 'Nitrogen Metabolism',\n",
       " 'Nucleotide Salvage Pathway',\n",
       " 'Nucleotide Sugar Metabolism',\n",
       " 'Nucleotides',\n",
       " 'One carbon pool by folate',\n",
       " 'Other',\n",
       " 'Other Amino Acid Metabolism',\n",
       " 'Oxidative Phosphorylation',\n",
       " 'Pantothenate and CoA Biosynthesis',\n",
       " 'Pentose Phosphate Pathway',\n",
       " 'Pentose and Glucuronate Interconversions',\n",
       " 'Phenylalanine Metabolism',\n",
       " 'Phospholipid Biosynthesis',\n",
       " 'Phospholipid Metabolism',\n",
       " 'Porphyrin and Chlorophyll Metabolism',\n",
       " 'Propanoate Metabolism',\n",
       " 'Purine Metabolism',\n",
       " 'Purine and Pyrimidine Biosynthesis',\n",
       " 'Pyridoxine Metabolism',\n",
       " 'Pyrimidine Catabolism',\n",
       " 'Pyrimidine Metabolism',\n",
       " 'Pyruvate Metabolism',\n",
       " 'Quinone Biosynthesis',\n",
       " 'ROS Detoxification',\n",
       " 'Riboflavin Metabolism',\n",
       " 'Selenoamino acid Metabolism',\n",
       " 'Sphingolipid Metabolism',\n",
       " 'Starch and Sucrose Metabolism',\n",
       " 'Sterol Metabolism',\n",
       " 'Sulfur Metabolism',\n",
       " 'TCA cycle;Carbon fixation;CO2 fixation',\n",
       " 'Taurine and hypotaurine Metabolism',\n",
       " 'Terpenoid backbone Biosynthesis',\n",
       " 'Tetrahydrobiopterin',\n",
       " 'Thiamine Metabolism',\n",
       " 'Threonine and Lysine Metabolism',\n",
       " 'Transport',\n",
       " 'Transport Extracellular',\n",
       " 'Transport Mitochondrial',\n",
       " 'Transport, Endoplasmic Reticular',\n",
       " 'Transport, Extracellular',\n",
       " 'Transport, Golgi Apparatus',\n",
       " 'Transport, Inner Membrane',\n",
       " 'Transport, Mitochondrial',\n",
       " 'Transport, Nuclear',\n",
       " 'Transport, Outer Membrane Porin',\n",
       " 'Transport, Peroxisomal',\n",
       " 'Transport, Vacuolar',\n",
       " 'Transport, glyoxysome',\n",
       " 'Tryptophan Metabolism',\n",
       " 'Tyrosine Metabolism',\n",
       " 'Tyrosine, Tryptophan, and Phenylalanine Metabolism',\n",
       " 'Ubiquinone synthesis',\n",
       " 'Unassigned',\n",
       " 'Urea cycle/amino group Metabolism',\n",
       " 'Valine, Leucine, and Isoleucine Degradation',\n",
       " 'Valine, Leucine, and Isoleucine Metabolism',\n",
       " 'Vitamin B2 Metabolism',\n",
       " 'Vitamin B6 Metabolism',\n",
       " 'Xylose Metabolism',\n",
       " 'beta-Alanine Metabolism',\n",
       " 'tRNA Charging'}"
      ]
     },
     "execution_count": 353,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08d6bcd7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39114dbe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8544485e",
   "metadata": {},
   "source": [
    "## lets clean the model to remove orphan metabolites. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "130fe7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find orphan metabolites\n",
    "orphan_metabolites = [m for m in model.metabolites if len(m.reactions) == 0]\n",
    "\n",
    "# Remove orphan metabolites from the model\n",
    "model.remove_metabolites(orphan_metabolites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "7e631e72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orphan_metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eb13378",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "id": "9e7e8dd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.save_json_model(model,'../models/Lst_v0.1_model_withOrthoMCLallAdded_forPUB.json')\n",
    "# cobra.io.write_sbml_model(model, \"Lst_v0.1_model_verify.xml\")\n",
    "cobra.io.save_json_model(model,'../modelsLst_v0.2_model_forPUB_copy.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f60d946",
   "metadata": {},
   "source": [
    "## finally, we will not add transporters here. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24c4f801",
   "metadata": {},
   "source": [
    "we can separate the above reactions that need to be added into a few categories:\n",
    "\n",
    "transport (vacuole to cytoplasm) 5 rxns:\n",
    "- r_4162: his__L_v <=> his__L_c\n",
    "- r_1220: tyr__L_v <=> tyr__L_c\n",
    "- r_4163: lys__L_v <=> lys__L_c\n",
    "- r_1208: leu__L_v <=> leu__L_c\n",
    "- r_1202: ile__L_v <=> ile__L_c\n",
    "\n",
    "transport (mitochondria to cytoplasm) 3 rxns:\n",
    "- r_1919: lys__L_c --> lys__L_m\n",
    "- ARGt2m: arg__L_c + h_c --> arg__L_m + h_m\n",
    "- r_1837: his__L_c --> his__L_m\n",
    "\n",
    "transport (extracellular to cytoplasm) 8 rxns:\n",
    "- r_4587: ca2_e + h_e <=> ca2_c + h_c\n",
    "- r2073_1: h_e + zn2_e <=> h_c + zn2_c\n",
    "- MNt2: h_e + mn2_e <=> h_c + mn2_c\n",
    "- Clt: cl_e <=> cl_c\n",
    "- CHOLtu: chol_c <=> chol_e\n",
    "- r_4589: cu2_e + h_e <=> cu2_c + h_c\n",
    "- r_4592: h_e + mg2_e <=> h_c + mg2_c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4772f051",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:lipGSM]",
   "language": "python",
   "name": "conda-env-lipGSM-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
